<!DOCTYPE html>



  


<html class="theme-next muse use-motion" lang>
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform">
<meta http-equiv="Cache-Control" content="no-siteapp">















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css">




  
  
  
  

  
    
    
  

  

  

  

  

  
    
    
    <link href="//fonts.googleapis.com/css?family=Lato:300,300italic,400,400italic,700,700italic&subset=latin,latin-ext" rel="stylesheet" type="text/css">
  






<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css">

<link href="/css/main.css?v=5.1.2" rel="stylesheet" type="text/css">


  <meta name="keywords" content="Hexo, NexT">








  <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico?v=5.1.2">






<meta name="description" content="软件准备在相关软件官网下载最新版本的软件    bwa-0.7.17.tar.bz2    gatk-4.1.2.0.zip   picard-2.20.0-0.tar.bz2    samtools-1.9.tar.bz2   原始数据质量控制分析数据质量的几个方面：    read各个位置的碱基质量分布 碱基的总体质量分布 read各个位置上碱基分布比例看，目的是为了分析碱基的分离程度 GC含">
<meta name="keywords" content="Hexo, NexT">
<meta property="og:type" content="article">
<meta property="og:title" content="Whole-exome analysis">
<meta property="og:url" content="http://yoursite.com/2019/06/05/Whole-exome_analysis/index.html">
<meta property="og:site_name" content="Jing">
<meta property="og:description" content="软件准备在相关软件官网下载最新版本的软件    bwa-0.7.17.tar.bz2    gatk-4.1.2.0.zip   picard-2.20.0-0.tar.bz2    samtools-1.9.tar.bz2   原始数据质量控制分析数据质量的几个方面：    read各个位置的碱基质量分布 碱基的总体质量分布 read各个位置上碱基分布比例看，目的是为了分析碱基的分离程度 GC含">
<meta property="og:locale" content="default">
<meta property="og:image" content="https://note.youdao.com/yws/public/resource/bed499f2133fd1cc1e1ced485447f8c1/xmlnote/WEBRESOURCE22c444411c67fe341deac1f544ccb7ae/6557">
<meta property="og:image" content="https://note.youdao.com/yws/public/resource/bed499f2133fd1cc1e1ced485447f8c1/xmlnote/WEBRESOURCEe3710742911832dc4bc8e26297d630c7/6567">
<meta property="og:updated_time" content="2019-06-05T09:15:31.191Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Whole-exome analysis">
<meta name="twitter:description" content="软件准备在相关软件官网下载最新版本的软件    bwa-0.7.17.tar.bz2    gatk-4.1.2.0.zip   picard-2.20.0-0.tar.bz2    samtools-1.9.tar.bz2   原始数据质量控制分析数据质量的几个方面：    read各个位置的碱基质量分布 碱基的总体质量分布 read各个位置上碱基分布比例看，目的是为了分析碱基的分离程度 GC含">
<meta name="twitter:image" content="https://note.youdao.com/yws/public/resource/bed499f2133fd1cc1e1ced485447f8c1/xmlnote/WEBRESOURCE22c444411c67fe341deac1f544ccb7ae/6557">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Muse',
    sidebar: {"position":"left","display":"post","offset":12,"offset_float":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: true,
    duoshuo: {
      userId: '0',
      author: 'Author'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://yoursite.com/2019/06/05/Whole-exome_analysis/">





  <title>Whole-exome analysis | Jing</title>
  














</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="default">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail ">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Jing</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br>
            
            Home
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br>
            
            Archives
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br>
            
            Tags
          </a>
        </li>
      

      
    </ul>
  

  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/06/05/Whole-exome_analysis/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Jing Yu">
      <meta itemprop="description" content>
      <meta itemprop="image" content="http://www.jf258.com/uploads/2013-07-13/171053918.jpg">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Jing">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">Whole-exome analysis</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              
              <time title="Post created" itemprop="dateCreated datePublished" datetime="2019-06-05T12:27:52+08:00">
                2019-06-05
              </time>
            

            

            
          </span>

          

          
            
          

          
          

          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h5 id="软件准备"><a href="#软件准备" class="headerlink" title="软件准备"></a>软件准备</h5><p>在相关软件官网下载最新版本的软件  </p>
<ul>
<li>bwa-0.7.17.tar.bz2   </li>
<li>gatk-4.1.2.0.zip  </li>
<li>picard-2.20.0-0.tar.bz2   </li>
<li>samtools-1.9.tar.bz2 </li>
</ul>
<h5 id="原始数据质量控制"><a href="#原始数据质量控制" class="headerlink" title="原始数据质量控制"></a>原始数据质量控制</h5><p>分析数据质量的几个方面：  </p>
<ul>
<li>read各个位置的碱基质量分布</li>
<li>碱基的总体质量分布</li>
<li>read各个位置上碱基分布比例看，目的是为了分析碱基的分离程度</li>
<li>GC含量分布</li>
<li>read各个位置的N含量</li>
<li>read是否还包含测序的接头序列</li>
<li>read重复率，这个是实验的扩增所引入的  </li>
</ul>
<p><strong>需要的软件</strong></p>
<ul>
<li>FastQC，一个java程序，可以很好的帮助我们理解测序数据的质量情况，唯一不足的就是图太丑了</li>
<li>MultiQC,一个基于Python的小工具，能够将测序数据的多个QC结果整合成一个HTLM网页交互式报告，同时也能导出pdf</li>
<li>Trimmomatic（cutadapt、sickle、seqtk）等可用于切除接头序列和read的低质量序列</li>
</ul>
<h5 id="全外显子组分析（WES）"><a href="#全外显子组分析（WES）" class="headerlink" title="全外显子组分析（WES）"></a>全外显子组分析（WES）</h5><p>需要的工具：  </p>
<ul>
<li>bwa</li>
<li>samtools</li>
<li>bgzip</li>
<li>GATK 4.0</li>
<li>GATK 3.x</li>
<li>Picard</li>
</ul>
<h6 id="数据预处理"><a href="#数据预处理" class="headerlink" title="数据预处理"></a>数据预处理</h6><p><strong>序列比对</strong></p>
<ul>
<li><p>为参考基因组构建索引</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bwa index hg38.fa</span><br></pre></td></tr></table></figure>
</li>
<li><p>将read比对到参考基因组</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bwa mem -t 4 -R &apos;@RG\tID:foo_lane\tPL:illumina\tLB:library\tSM:sample_name&apos; /path/to/human.fasta read_1.fq.gz read_2.fq.gz &gt; sample_name.sam</span><br></pre></td></tr></table></figure>

</li>
</ul>
<p>-t，线程数，我们在这里使用4个线程  ；<br>-R 接的是Read Group的字符串信息，这是一个非常重要的信息，以@RG开头，它是用来将比对的read进行分组的。这个信息对于我们后续对比对数据进行错误率分析和Mark duplicate时非常重要。在Read Group中，有如下几个信息非常重要：  </p>
<ul>
<li>1）ID，这是Read Group的分组ID，一般设置为测序的lane ID（不同lane之间的测序过程认为是独立的），下机数据中我们都能看到这个信息的，一般都是包含在fastq的文件名中；</li>
<li>2）PL，指的是所用的测序平台，这个信息不要随便写！特别是当我们需要使用GATK进行后续分析的时候，更是如此！这是一个很多新手都容易忽视的一个地方，在GATK中，PL只允许被设置为：ILLUMINA,SLX,SOLEXA,SOLID,454,LS454,COMPLETE,PACBIO,IONTORRENT,CAPILLARY,HELICOS或UNKNOWN这几个信息。</li>
<li>3）SM，样本ID，同样非常重要，有时候我们测序的数据比较多的时候，那么可能会分成多个不同的lane分布测出来，这个时候SM名字就是可以用于区分这些样本；</li>
<li>4）LB，测序文库的名字，这个重要性稍微低一些，主要也是为了协助区分不同的group而存在。文库名字一般可以在下机的fq文件名中找到，如果上面的lane ID足够用于区分的话，也可以不用设置LB。  </li>
</ul>
<p>为了方便后续的分析，一般在输出文件时，将其转换为bam文件</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">bwa mem -t 4 -R &apos;@RG\tID:foo_lane\tPL:illumina\tLB:library\tSM:sample_name&apos; /path/to/human.fasta read_1.fq.gz read_2.fq.gz | samtools view -S -b - &gt; sample_name.bam</span><br></pre></td></tr></table></figure>

<p><strong>排序（sort）</strong><br>第一步的比对是按照FASTQ文件的顺序把read逐一定位到参考基因组上之后，随即就输出了，它不会也不可能在这一步里面能够自动识别比对位置的先后位置重排比对结果。因此，比对后得到的结果文件中，每一条记录之间位置的先后顺序是乱的，我们后续去重复等步骤都需要在比对记录按照顺序从小到大排序下来才能进行，所以这才是需要进行排序的原因。</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">time samtools sort -@ 4 -m 4G -O bam -o sample_name.sorted.bam sample_name.bam</span><br></pre></td></tr></table></figure>

<p>-@，用于设定排序时的线程数，我们设为4；<br>-m，限制排序时最大的内存消耗，这里设为4GB；<br>-O 指定输出为bam格式；<br>-o 是输出文件的名字，</p>
<p><strong>标记重复序列（删除重复序列）</strong><br>我们使用Picard来完成这个事情</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">##标记重复序列</span><br><span class="line">java -jar picard.jar MarkDuplicates \ </span><br><span class="line">  I=sample_name.sorted.bam \</span><br><span class="line">  O=sample_name.sorted.markdup.bam \</span><br><span class="line">  M=sample_name.markdup_metrics.txt</span><br></pre></td></tr></table></figure>

<p>这里只把重复序列在输出的新结果中标记出来，但不删除。如果我们非要把这些序列完全删除的话可以这样做：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">java -jar picard.jar MarkDuplicates \ </span><br><span class="line">  REMOVE_DUPLICATES=true \</span><br><span class="line">  I=sample_name.sorted.bam \</span><br><span class="line">  O=sample_name.sorted.markdup.bam \</span><br><span class="line">  M=sample_name.markdup_metrics.txt</span><br></pre></td></tr></table></figure>

<p>在后面的分析中，在<code>Somatic SNVs+Indels</code>的过程中使用标记的bam的文件，并且使用GATK4.0的主流pipline；在<code>Detect MSI（微卫星不稳定性）</code>的过程中使用GATK3.x的过程</p>
<h6 id="Somatic-SNVs-Indels"><a href="#Somatic-SNVs-Indels" class="headerlink" title="Somatic SNVs+Indels"></a>Somatic SNVs+Indels</h6><p><strong>重新矫正碱基质量值（BQSR）</strong><br>分为两步:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"> gatk BaseRecalibrator \</span><br><span class="line">   -I my_reads.bam \</span><br><span class="line">   -R reference.fasta \</span><br><span class="line">   --known-sites sites_of_variation.vcf \</span><br><span class="line">   --known-sites another/optional/setOfSitesToMask.vcf \</span><br><span class="line">   -O recal_data.table</span><br><span class="line">## 用到的已知变异位点</span><br><span class="line">common_all_20180418_dbsnp_151_hg38.vcf</span><br><span class="line">Homo_sapiens_assembly38.known_indels.vcf</span><br><span class="line">Mills_and_1000G_gold_standard.indels.hg38.vcf</span><br><span class="line">下载地址（GATK的bundle以及1000genomic官网）：ftp://ftp.broadinstitute.org/bundle/hg38/</span><br><span class="line">https://ftp.ncbi.nlm.nih.gov/snp/organisms/</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">gatk ApplyBQSR \</span><br><span class="line">  -R reference.fasta \</span><br><span class="line">  -I input.bam \</span><br><span class="line">  --bqsr-recal-file recalibration.table \</span><br><span class="line">  -O output.bam</span><br></pre></td></tr></table></figure>

<p><strong>查看测序深度</strong></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools depth  bamfile  |  awk &apos;&#123;sum+=$3&#125; END &#123; print &quot;Average = &quot;sum/NR&#125;&apos;</span><br></pre></td></tr></table></figure>

<p><strong>使用qualimap来查看比对好的bam文件的质量</strong><br>在<a href="http://qualimap.bioinfo.cipf.es/" target="_blank" rel="noopener">官网</a>下载使用qualimap</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget https://bitbucket.org/kokonech/qualimap/downloads/qualimap_v.2.12.zip</span><br><span class="line">unzip qualimap_v.2.12.zip</span><br><span class="line">cd qualimap_v.2.12</span><br><span class="line">./qualimap -h</span><br><span class="line"># 使用qualimap时 一般需要添加内存参数,对于外显子数据，需要添加关于外显子区域的bed文件</span><br><span class="line">qualimap --java-mem-size=8G bamqc -bam in.bam -gff exome.bed</span><br></pre></td></tr></table></figure>

<h6 id="使用旧版的mutects来call-somatic-mutation（现在GATK网站已经不推荐了）"><a href="#使用旧版的mutects来call-somatic-mutation（现在GATK网站已经不推荐了）" class="headerlink" title="使用旧版的mutects来call somatic mutation（现在GATK网站已经不推荐了）"></a>使用旧版的mutects来call somatic mutation（现在GATK网站已经不推荐了）</h6><p>** CreateSomaticPanelOfNormals **
panel of normals (PoN) containing germline and artifactual sites for use with Mutect2.  </p>
<p>Step 1. Run Mutect2 in tumor-only mode for each normal sample.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">~/biosoft/GATK4.0/gatk-4.0.5.1/gatk Mutect2 \</span><br><span class="line">-R ~/reference/genome/gatk_hg38/Homo_sapiens_assembly38.fasta \</span><br><span class="line">-I HG00190.bam \</span><br><span class="line">-tumor HG00190 \</span><br><span class="line">--disable-read-filter MateOnSameContigOrNoMappedMateReadFilter \</span><br><span class="line">-L interval_list \</span><br><span class="line">--max-map-distance 0 \ ## 防止在产生的vcf文件中出现MNPS现象</span><br><span class="line">-O 3_HG00190.vcf.gz</span><br></pre></td></tr></table></figure>

<p>其中的interval_list来自于安捷伦公司的官网，使用的具体bed文件为：S07604514_Padded.bed<br>或者是可以通过CCDS文件自己制作interval_list   </p>
<ul>
<li><p>首先在CCDS官网（<a href="ftp://ftp.ncbi.nlm.nih.gov/pub/CCDS/current_human）下载文件" target="_blank" rel="noopener">ftp://ftp.ncbi.nlm.nih.gov/pub/CCDS/current_human）下载文件</a></p>
</li>
<li><p>将txt文件转换为bed形式</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cat CCDS.20160908.txt |grep -w -v Withdrawn|perl -alne &apos;&#123;/\[(.*?)\]/;next unless $1;$gene=$F[2];$exons=$1;$exons=~s/\s//g;$exons=~s/-/\t/g;print &quot;$F[0]\t$_\t$gene&quot; foreach split/,/,$exons;&#125;&apos;|sort -u |bedtools sort -i &gt;exon_probe.hg38.gene.bed</span><br></pre></td></tr></table></figure>
</li>
<li><p>制作interval_list</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">cat exon_probe.hg38.gene.bed | awk &apos;&#123;print &quot;chr&quot;$0&#125;&apos; &gt;hg38.chr.bed</span><br><span class="line"></span><br><span class="line">picard BedToIntervalList \</span><br><span class="line">       I=hg38.chr.bed \</span><br><span class="line">       O=hg38.interval_list\</span><br><span class="line">       SD=/data/reference/GRCh38/GRCh38.primary_assembly.genome.dict</span><br><span class="line">       </span><br><span class="line">gatk4 PreprocessIntervals </span><br><span class="line">      -L hg38.interval_list  </span><br><span class="line">      --sequence-dictionary /data/reference/GRCh38/GRCh38.primary_assembly.genome.dict </span><br><span class="line">      --reference /data/reference/GRCh38/GRCh38.primary_assembly.genome.fa </span><br><span class="line">      --padding 100 </span><br><span class="line">      --bin-length 0 </span><br><span class="line">      --interval-merging-rule OVERLAPPING_ONLY</span><br><span class="line">      --output hg38.Padded.interval_list</span><br></pre></td></tr></table></figure>

</li>
</ul>
<p>Step 2. Create a GenomicsDB from the normal Mutect2 calls.<br>由于在第一步产生的vcf文件中有多行对应一个位点，所以在进行Step2之前要先将这些同一位点的行合并起来</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">bcftools norm \</span><br><span class="line">    -m +any --do-not-normalize \</span><br><span class="line">    delta_af-only-gnomad_Hg19toGRCh38.vcf.gz \</span><br><span class="line">    -Oz -o zeta_af-only-gnomad_Hg19toGRCh38.vcf.gz</span><br><span class="line">## 为norm之后的新vcf文件创建索引</span><br><span class="line">gatk IndexFeatureFile \</span><br><span class="line">     -F cohort.vcf.gz</span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## step.2</span></span><br><span class="line"> gatk GenomicsDBImport -R reference.fasta -L intervals.interval_list \</span><br><span class="line">       --genomicsdb-workspace-path pon_db \</span><br><span class="line">       -V normal1.vcf.gz \</span><br><span class="line">       -V normal2.vcf.gz \</span><br><span class="line">       -V normal3.vcf.gz</span><br></pre></td></tr></table></figure>

<p>Step 3. Combine the normal calls using CreateSomaticPanelOfNormals.</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">gatk CreateSomaticPanelOfNormals -R reference.fasta -V gendb://pon_db -O pon.vcf.gz</span><br></pre></td></tr></table></figure>

<p>** 对tumor和matched normal进行calls somatic variants **</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">for sample in *C*.bam;  </span><br><span class="line">do base=$&#123;sample%-*&#125;;  </span><br><span class="line">   gatk4 Mutect2   </span><br><span class="line">   -R /data/reference/GRCh38/GRCh38.primary_assembly.genome.fa   </span><br><span class="line">   -I &quot;$base&quot;-N.sorted.markdup.recal.bam  </span><br><span class="line">   -I $sample   </span><br><span class="line">   -tumor &quot;$base&quot;-C   </span><br><span class="line">   -normal &quot;$base&quot;-N   </span><br><span class="line">   -pon ./Mutect2_interval/pon.vcf.gz  </span><br><span class="line">   --germline-resource ../../reference/somatic-hg38_af-only-gnomad.hg38.vcf.gz  </span><br><span class="line">   --af-of-alleles-not-in-resource 0.0000025  </span><br><span class="line">   -disable-read-filter MateOnSameContigOrNoMappedMateReadFilter  </span><br><span class="line">   -L ../../reference/S07604514_hs_hg38/S07604514_hs_hg38/S07604514_Padded.bed   </span><br><span class="line">   -O ./Mutect2_interval/Mutect2_output/&quot;$base&quot;_somatic_m2.vcf.gz   </span><br><span class="line">   -bamout ./Mutect2_interval/Mutect2_output/&quot;$base&quot;_tumor_normal_m2.bam;done</span><br></pre></td></tr></table></figure>

<p>** 使用<code>GetPileupSummaries</code>工具来计算肿瘤样本的污染情况 ** 
先使用Selectvariants工具生成包含commom biallelic variants的.vcf文件</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"> </span><br><span class="line">gatk SelectVariants \</span><br><span class="line">-R ~/reference/genome/gatk_hg38/Homo_sapiens_assembly38.fasta \</span><br><span class="line">-V af-only-gnomad.hg38.vcf.gz \</span><br><span class="line">--select-type-to-include SNP \</span><br><span class="line">--restrict-alleles-to BIALLELIC \</span><br><span class="line">-O af-only-gnomad.hg38.SNP_biallelic.vcf.gz</span><br></pre></td></tr></table></figure>

<p>接着用GetPileupSummaries计算resource site位点上的count数，这样并不是计算所有的位点，而是用AF参数过滤后的（大于0.01并小于0.2），同时也可以用 -L 参数指定区域，分别对Normal和Tumor样本计算</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">~/biosoft/GATK4.0/gatk-4.0.5.1/gatk GetPileupSummaries \</span><br><span class="line">-I Normal_blood.allready.bam \</span><br><span class="line">-L ../exon_150bp.list \</span><br><span class="line">-V ~/annotation/GATK/resources/bundle/hg38/af-only-gnomad.hg38.SNP_biallelic.vcf.gz \</span><br><span class="line">-O Normal_blood.pileups.table</span><br></pre></td></tr></table></figure>

<p>总之最后通过计算出的污染比例（XXX.calculatecontamination.table文件中）来过滤掉somatic variant中的一些可能是污染导致的假阳性突变</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">gatk CalculateContamination \</span><br><span class="line">-I Primary-IIIG.pileups.table \</span><br><span class="line">-matched Normal_blood.pileups.table \</span><br><span class="line">-O Primary-IIIG.calculatecontamination.table</span><br></pre></td></tr></table></figure>

<h6 id="使用Muect2新流程进行calling-somatic-mutation"><a href="#使用Muect2新流程进行calling-somatic-mutation" class="headerlink" title="使用Muect2新流程进行calling somatic mutation"></a>使用Muect2新流程进行calling somatic mutation</h6><p>首先需要创建PON，步骤与旧版的相同，接下买的操作如下</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><span class="line">## First, run Mutect2 with the --f1r2-tar-gz argument. This creates an output with raw data</span><br><span class="line">if [ 1 -eq 1 ];</span><br><span class="line">then</span><br><span class="line">	workdir=/media/data1/YJ/CRC/tag_duplicates/tag-dup_rec/;</span><br><span class="line">outdir=/media/data1/YJ/CRC/tag_duplicates/tag-dup_rec/Mutect2_interval/Mutect2_new_withnewPON_matchedmodel/;</span><br><span class="line">refdir=/media/data1/YJ/CRC/reference/;</span><br><span class="line">cd $workdir;</span><br><span class="line">for sample in *C*.bam;</span><br><span class="line">do</span><br><span class="line">	base=$&#123;sample%%-*&#125;;</span><br><span class="line">	gatk4 Mutect2 -R /data/reference/GRCh38/GRCh38.primary_assembly.genome.fa \</span><br><span class="line">		-L /media/data1/YJ/CRC/reference/S07604514_hs_hg38/S07604514_hs_hg38/S07604514_Padded.bed \</span><br><span class="line">		-I $sample \</span><br><span class="line">		-I &quot;$base&quot;-N.sorted.markdup.recal.bam \</span><br><span class="line">		-tumor &quot;$base&quot;-C \</span><br><span class="line">		-normal &quot;$base&quot;-N \</span><br><span class="line">		-germline-resource &quot;$refdir&quot;somatic-hg38_af-only-gnomad.hg38.vcf.gz \</span><br><span class="line">		-pon &quot;$workdir&quot;Mutect2_interval/pon_withgnomAD.vcf.gz \</span><br><span class="line">		--f1r2-tar-gz &quot;$outdir&quot;&quot;$base&quot;.f1r2.tar.gz \</span><br><span class="line">		-O &quot;$outdir&quot;&quot;$base&quot;.unfiltered.vcf;</span><br><span class="line">done;</span><br><span class="line">fi</span><br><span class="line"></span><br><span class="line">## Next, pass this raw data to LearnReadOrientationModel:</span><br><span class="line">if [ 1 -eq 1 ] ;</span><br><span class="line">then</span><br><span class="line">	workdir=/media/data1/YJ/CRC/tag_duplicates/tag-dup_rec/Mutect2_interval/Mutect2_new_withnewPON_matchedmodel;</span><br><span class="line">	cd $workdir;</span><br><span class="line">	for sample in *.tar.gz;</span><br><span class="line">	do</span><br><span class="line">		base=$&#123;sample%%.*&#125;;</span><br><span class="line">		gatk4 LearnReadOrientationModel -I $sample -O &quot;$base&quot;_read-orientation-model.tar.gz;</span><br><span class="line">	done;</span><br><span class="line">fi</span><br><span class="line"></span><br><span class="line">## Finally, pass the learned read orientation model to FilterMutectCallswith the -ob-priors argument:</span><br><span class="line">if [ 1 -eq 1 ];</span><br><span class="line">then</span><br><span class="line">	workdir=/media/data1/YJ/CRC/tag_duplicates/tag-dup_rec/Mutect2_interval/Mutect2_new_withnewPON_matchedmodel;</span><br><span class="line">	refdir=/data/reference/GRCh38;</span><br><span class="line">	cd $workdir</span><br><span class="line">	for sample in *.unfiltered.vcf;</span><br><span class="line">	do</span><br><span class="line">		base=$&#123;sample%%.*&#125;;</span><br><span class="line">		gatk4 FilterMutectCalls -V $sample \</span><br><span class="line">			--ob-priors &quot;$base&quot;_read-orientation-model.tar.gz \</span><br><span class="line">			-R &quot;$refdir&quot;/GRCh38.primary_assembly.genome.fa \</span><br><span class="line">			-O &quot;$base&quot;_filtered.vcf</span><br><span class="line">	done;</span><br><span class="line">fi</span><br></pre></td></tr></table></figure>

<p>最终获得利用<code>FilterMutectCalls</code>过滤过的filtered.vcf文件然后得到通过过滤的vcf文件</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##filter the .filtered.vcf and annotation it with gene_based model</span></span><br><span class="line"><span class="keyword">if</span> [ <span class="number">1</span> -eq <span class="number">1</span> ];</span><br><span class="line">then</span><br><span class="line">	<span class="keyword">for</span> sample <span class="keyword">in</span> *_filtered.vcf;</span><br><span class="line">	do</span><br><span class="line">	</span><br><span class="line">		base=$&#123;sample%%_*&#125;;</span><br><span class="line">		grep -v <span class="string">"#"</span> $sample | awk <span class="string">'$7=="PASS"'</span> &gt; <span class="string">"$base"</span>_filtered_PASS.cvf;</span><br><span class="line">	done;</span><br><span class="line">fi</span><br></pre></td></tr></table></figure>

<p>可以通过<code>VariantAnnotator</code>对vcf文件进行dbsnp注释</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">##Annotation vcf with dbsnp                                                                                                   </span><br><span class="line">if [ 1 -eq 0 ];                                                                                                                </span><br><span class="line">then                                                                                                   </span><br><span class="line">        for sample in *_filtered.vcf;                                                               </span><br><span class="line">        do</span><br><span class="line">                base=$&#123;sample%%_*&#125;;  </span><br><span class="line">                gatk4 VariantAnnotator -R /data/reference/GRCh38/GRCh38.primary_assembly.genome.fa \ </span><br><span class="line">                        -I ../../&quot;$base&quot;-C.sorted.markdup.recal.bam \  </span><br><span class="line">                        -V $sample \        </span><br><span class="line">                        --output &quot;$base&quot;_filtered_dbsnp.vcf \ </span><br><span class="line">                        -A Coverage \           </span><br><span class="line">                        --dbsnp /media/data1/YJ/CRC/reference/common_all_20180418_dbsnp_151_hg38.vcf ;</span><br><span class="line">        done                   </span><br><span class="line">fi</span><br></pre></td></tr></table></figure>

<p>注释完的vcf文件中多了一列ID的信息，里面有snp号，可以通过<a href="https://www.snpedia.com/index.php/SNPedia" target="_blank" rel="noopener">SNPedia网站</a>查看snp的具体信息</p>
<h6 id="使用不同的方法进行annotation"><a href="#使用不同的方法进行annotation" class="headerlink" title="使用不同的方法进行annotation"></a>使用不同的方法进行annotation</h6><p><strong>ANNOVAR</strong><br>在ANNOVAR官网下载最新版本的软件</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">## 下载适合的数据库(下载的数据储存在GRCH38文件夹中)</span></span><br><span class="line"> annotate_variation.pl -downdb -buildver hg19 -webfrom annovar refGene GRCH38/</span><br><span class="line"><span class="comment">## 利用 ANNOVAR进行基于gene的annotation</span></span><br><span class="line"><span class="keyword">if</span> [ <span class="number">1</span> -eq <span class="number">1</span> ];</span><br><span class="line">then</span><br><span class="line">	sofadir=/media/data1/YJ/CRC/softwave/annovar;</span><br><span class="line">	<span class="keyword">for</span> sample <span class="keyword">in</span> *filtered_PASS.vcf;</span><br><span class="line">	do</span><br><span class="line">		base=$&#123;sample%%.*&#125;;</span><br><span class="line">		</span><br><span class="line">		$sofadir/convert2annovar.pl -format vcf4 $sample -outfile <span class="string">"$base"</span>.avinput;</span><br><span class="line">		$sofadir/annotate_variation.pl -out <span class="string">"$base"</span> \</span><br><span class="line">			-build hg38 \</span><br><span class="line">			<span class="string">"$base"</span>.avinput \</span><br><span class="line">			--geneanno -dbtype refGene \</span><br><span class="line">			$sofadir/GRCH38/</span><br><span class="line">	done;</span><br><span class="line">fi</span><br></pre></td></tr></table></figure>

<p>结果会得到两个文件：</p>
<ul>
<li>在外显子位点的注释：s0020487998_filtered_PASS.exonic_variant_function<br><img src="https://note.youdao.com/yws/public/resource/bed499f2133fd1cc1e1ced485447f8c1/xmlnote/WEBRESOURCE22c444411c67fe341deac1f544ccb7ae/6557" alt="image"></li>
<li>在所有位点的注释：<br>s0020487998_filtered_PASS.variant_function<br><img src="https://note.youdao.com/yws/public/resource/bed499f2133fd1cc1e1ced485447f8c1/xmlnote/WEBRESOURCEe3710742911832dc4bc8e26297d630c7/6567" alt="image"><br>XX.variant_function文件一般关心前两列，后面几列均是变异位点的一些信息<br>第一列是变异所在基因组的位置，如：exonic、splicing、ncRNA等；其优先级：exonic = splicing &gt; ncRNA&gt; &gt; UTR5/UTR3 &gt; intron &gt; upstream/downstream &gt; intergenic<br>第二列信息则给突变位点所在的基因名称（如果突变在exonic/intronic/ncRNA），或者给出临近基因的名称</li>
</ul>
<p><strong>ClinEff</strong><br>在ClinEff的<a href="http://www.dnaminer.com/clineff.html" target="_blank" rel="noopener">官网</a>下载最新版本的软件已经相关的database</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">tar -xvf clinEff.1.0h.tgz</span><br><span class="line">tar -xvf clinEff_db38_v1.0.tgz</span><br></pre></td></tr></table></figure>

<p><strong>SnpEff</strong><br>在<a href="http://snpeff.sourceforge.net/download.html" target="_blank" rel="noopener">SnpEff的官网</a>下载最新版本的软件</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line">wget http://sourceforge.net/projects/snpeff/files/snpEff_latest_core.zip</span><br><span class="line">unzip snpEff_latest_core.zip</span><br><span class="line"><span class="comment">#查看homo_sapiens相关的数据库</span></span><br><span class="line">java -jar snpEff.jar databases | grep <span class="string">"Home_sapiens"</span> | cut -f <span class="number">1</span>,<span class="number">2</span></span><br><span class="line"></span><br><span class="line">GRCh37.75       Homo_sapiens</span><br><span class="line">GRCh38.86       Homo_sapiens</span><br><span class="line">hg19            Homo_sapiens (UCSC)</span><br><span class="line">hg19kg          Homo_sapiens (UCSC KnownGenes)</span><br><span class="line">hg38            Homo_sapiens (UCSC)</span><br><span class="line">hg38kg          Homo_sapiens (UCSC KnownGenes)</span><br><span class="line">testHg19ChrM    Homo_sapiens (UCSC)</span><br><span class="line"></span><br><span class="line"><span class="comment">#由于我们之前是使用的GRCh38的基因组，所以我们就下载GRCh38.86</span></span><br><span class="line">java -jar snpEff.jar download GRCh38.86</span><br><span class="line"><span class="comment">#注释</span></span><br><span class="line">java -Xmx4g -jar /media/data1/YJ/CRC/softwave/snpEff/snpEff.jar GRCh38.86 s0020347988_filtered_PASS.vcf &gt; s0020347988_filtered_PASS.snpEff.vcf</span><br></pre></td></tr></table></figure>

<p>输出的vcf文件，在输入vcf文件的基础上添加了一些tag：ANN、LOF、NMD<br>因此我可以将这个vcf格式文件稍微处理下，保留原来的vcf文件的前5列，再加上ANNtag形成一个新文件来查看</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">for sample in *PASS*;</span><br><span class="line">do </span><br><span class="line">    base=$&#123;sample%%_*&#125;;</span><br><span class="line">    perl -alne &apos;next if $_ =~ /^#/;$F[7] =~ /(ANN=\S+)/;print &quot;$F[1]\t$F[2]\t$F[3]\t$F[4]\t$F[5]\t$1&quot;&apos; $sample &gt; &quot;$base&quot;.anntag.vcf</span><br><span class="line">done</span><br></pre></td></tr></table></figure>

<p>vcf文件的内容如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">29275630        .       C       T       .       ANN=T|stop_gained|HIGH|PTPRU|ENSG00000060656|transcript|ENST00000345512.7|protein_coding|8/31|c.1327C&gt;T|p.Gln443*|1456/4470|1327/4341|443/1446||, </span><br><span class="line">                                                    T|stop_gained|HIGH|PTPRU|ENSG00000060656|transcript|ENST00000373779.7|protein_coding|8/30|c.1327C&gt;T|p.Gln443*|1456/5579|1327/4311|443/1436||,</span><br><span class="line">                                                    T|stop_gained|HIGH|PTPRU|ENSG00000060656|transcript|ENST00000428026.6|protein_coding|8/30|c.1327C&gt;T|p.Gln443*|1423/5550|1327/4302|443/1433||,</span><br><span class="line">                                                    T|stop_gained|HIGH|PTPRU|ENSG00000060656|transcript|ENST00000460170.2|protein_coding|8/31|c.1327C&gt;T|p.Gln443*|1333/5337|1327/4323|443/1440||,</span><br><span class="line">                                                    T|upstream_gene_variant|MODIFIER|PTPRU|ENSG00000060656|transcript|ENST00000415600.6|processed_transcript||n.-3921C&gt;T|||||3921|,</span><br><span class="line">                                                    T|non_coding_transcript_exon_variant|MODIFIER|PTPRU|ENSG00000060656|transcript|ENST00000527027.1|processed_transcript|3/3|n.564C&gt;T||||||;</span><br><span class="line">                                                    LOF=(PTPRU|ENSG00000060656|11|0.36);</span><br><span class="line">                                                    NMD=(PTPRU|ENSG00000060656|11|0.36)</span><br></pre></td></tr></table></figure>

<p>往往一个突变条目对应着多个annotation报告，主要有一下几种原因：</p>
<ul>
<li>一个突变位点可以影响多个基因，比如一个突变可能是一个基因的DOWNSTREAM同时可能是另一个基因的UPSTREAM</li>
<li>在复杂的基因组中，一个基因可能对应着多个转录组，snpeff会指出受该突变影响的每一个转录本</li>
<li>一条vcf记录可能包含不止一种突变<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">#CHROM  POS      ID    REF  ALT    QUAL FILTER    INFO</span><br><span class="line">1       889455   .     G    A,T    .    .         ANN=A|stop_gained|HIGH|NOC2L|ENSG00000188976|transcript|ENST00000327044|protein_coding|7/19|c.706C&gt;T|p.Gln236*|756/2790|706/2250|236/749||</span><br><span class="line">                                                     ,T|missense_variant|MODERATE|NOC2L|ENSG00000188976|transcript|ENST00000327044|protein_coding|7/19|c.706C&gt;A|p.Gln236Lys|756/2790|706/2250|236/749||</span><br><span class="line">                                                     ,A|downstream_gene_variant|MODIFIER|NOC2L|ENSG00000188976|transcript|ENST00000487214|processed_transcript||n.*865C&gt;T|||||351|</span><br><span class="line">                                                     ,T|downstream_gene_variant|MODIFIER|NOC2L|ENSG00000188976|transcript|ENST00000487214|processed_transcript||n.*865C&gt;A|||||351|</span><br><span class="line">                                                     ,A|downstream_gene_variant|MODIFIER|NOC2L|ENSG00000188976|transcript|ENST00000469563|retained_intron||n.*878C&gt;T|||||4171|</span><br><span class="line">                                                     ,T|downstream_gene_variant|MODIFIER|NOC2L|ENSG00000188976|transcript|ENST00000469563|retained_intron||n.*878C&gt;A|||||4171|</span><br><span class="line">                                                     ,A|non_coding_exon_variant|MODIFIER|NOC2L|ENSG00000188976|transcript|ENST00000477976|retained_intron|5/17|n.2153C&gt;T||||||</span><br><span class="line">                                                     ,T|non_coding_exon_variant|MODIFIER|NOC2L|ENSG00000188976|transcript|ENST00000477976|retained_intron|5/17|n.2153C&gt;A||||||;LOF=(NOC2L|ENSG00000188976|6|0.17);NMD=(NOC2L|ENSG00000188976|6|0.17)</span><br></pre></td></tr></table></figure>

</li>
</ul>
<p>当存在多种effect时，effect排序的依据：</p>
<ul>
<li>根据Putative impact：具有更高等级的putative impact将会排在前面</li>
<li>Effect type：有害的effect会排在前面</li>
<li>Canonical trancript before non-canonical.</li>
<li>Marker genomic coordinates (e.g. genes starting before first).　　</li>
</ul>
<p><code>ANN</code>tag将注释信息以“|”分割，每个filed有其对应的信息：</p>
<ul>
<li>Allele（or ALT）：突变的碱基</li>
<li>Annotation：表示突变的类型，当有多种类型存在时，用@连接，可以在 <a href="http://sequenceontology.org/browser/current_svn/term/SO:0001792" target="_blank" rel="noopener">The Sequence ontology官网</a>查询每一种突变类型的具体含义</li>
<li>Putative_impact：对突变的影响进行的预测，有4个程度</li>
</ul>
<p>1.HIGH：突变对蛋白的影响很大，可能导致蛋白质截短、功能丧失或引发无意义的介导衰退。eg.stop_gained, frameshift_variant<br>2.MODERATE：一种可能改变蛋白质有效性的无破坏性突变  eg.missense_variant, inframe_deletion<br>3.LOW：几乎是没害的，不会改变蛋白质的性能，eg.synonymous_variant<br>4.MODIFIER:通常是影响非编码基因的突变或非编码区的突变，对其很难预测它的影响。eg.exon_variant, downstream_gene_variant</p>
<ul>
<li>Gene ID：使用ENSEMBL id</li>
<li>Feature type：表示突变所在区域的类型，比如transcript, motif, miRNA等</li>
<li>Feature ID ：表示Feature type对应的id</li>
<li>Transcript biotype ：关于transcript的类型，根据ENSEMBL biotypes来定义，</li>
<li>Rank/total：1/1表示Exon or Intron rank / total number of exons or introns，前面的1表示这个突变是在第1个exon上（因为annatation已经给出了这个是突变是在exon上），后面的11表示这个突变所在的transcript总共有1个exon</li>
<li>HGVS.c ：n.321T&gt;C表示Variant using HGVS notation (DNA level)</li>
<li>HGVS.p：如果突变发生在ｃｏｄｉｎｇ区域，根据HGVS标记法在蛋白质水平表示突变类型。p.Gln443*表示443位的Gln氨基酸突变成了终止密码子</li>
<li>cDNA_position/cDNA_len:cDNA的位置以及，cDNA的长度</li>
<li>CDS_position / CDS_len: Position and number of coding bases (one based includes START and STOP codons).</li>
<li>Protein_position / Protein_len: Position and number of AA (one based, including START, but not STOP).</li>
</ul>
<p>至于LOF和NMD标签这是表示：Loss of funcation(LOF) and nonsense-mediated   decay(NMD) 预测  ，分析相关的effect是否可以产生LOF或者是NMD影响  </p>
<p>NMD：无义介导的mRNA降解作为真核细胞中重要RNA监控机制，识别并降解开放阅读框中含有提前终止密码子（premature termination codon，PTC）的mRNA，以避免因截短的蛋白产物积累对细胞造成毒害. NMD还调控正常生理基因的表达，暗示其在真核细胞中扮演重要角色.  </p>
<p> LOF和NMD标签的形式如下：</p>
<ul>
<li>Gene：gene的名称</li>
<li>ID：gene id，通常是ENSEMBL id</li>
<li>Num_transcripts：gene中的转录本的数量</li>
<li>percent_affected：被改突变多影响的转录本的比例</li>
</ul>
<h6 id="使用VarScan2-Somatic来call-somatic（没有写完）"><a href="#使用VarScan2-Somatic来call-somatic（没有写完）" class="headerlink" title="使用VarScan2 Somatic来call somatic（没有写完）"></a>使用VarScan2 Somatic来call somatic（没有写完）</h6><h6 id="环境搭建"><a href="#环境搭建" class="headerlink" title="环境搭建"></a>环境搭建</h6><p>1 创建文件夹</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mkdir -p input reference scripts temp</span><br></pre></td></tr></table></figure>

<p>2 安装conda并使用conda创建env</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 添加信号源</span></span><br><span class="line">conda config --add channels r</span><br><span class="line">conda config --add channels defaults</span><br><span class="line">conda config --add channels conda-forge</span><br><span class="line">conda config --add channels bioconda</span><br><span class="line"><span class="comment"># 创建env</span></span><br><span class="line">conda create --name altwes gatk R samtools trimmomatic picard bam-readcount</span><br><span class="line">varscan</span><br><span class="line"><span class="comment"># 激活env</span></span><br><span class="line">conda activate altwes</span><br></pre></td></tr></table></figure>

<p>3 安装GATK3.8</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">WES=/</span><br><span class="line">cd $WES/scripts</span><br><span class="line">wget -r -np -nd \</span><br><span class="line">-O gatk3.tar.bz2 \</span><br><span class="line">’https://software.broadinstitute.org/gatk/download/auth?package=GATK-archive&amp;version=3.8-0-ge9d806836’</span><br><span class="line">bunzip2 gatk3.tar.bz2</span><br><span class="line">tar -xf gatk3.tar</span><br></pre></td></tr></table></figure>

<h5 id="Detect-MSI"><a href="#Detect-MSI" class="headerlink" title="Detect MSI"></a>Detect MSI</h5><h6 id="数据预处理-1"><a href="#数据预处理-1" class="headerlink" title="数据预处理"></a>数据预处理</h6><p><strong>局部重比对</strong><br>接下来是局部区域重比对，通常也叫Indel局部区域重比对。有时在进行这一步骤之前还有一个merge的操作，将同个样本的所有比对结果合并成唯一一个大的BAM文件，merge的例子如下：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">samtools merge &lt;out.bam&gt; &lt;in1.bam&gt; [&lt;in2.bam&gt;...&lt;inN.bam&gt;]</span><br></pre></td></tr></table></figure>

<p>局部重比对的目的是将BWA比对过程中所发现有潜在序列插入或者序列删除（insertion和deletion，简称Indel）的区域进行重新校正。这个过程往往还会把一些已知的Indel区域一并作为重比对的区域<br>GATK4.0中没有分离出单独的局部比对脚本。只能使用3.X版本中的脚本,这里包含了两个步骤：</p>
<ul>
<li>第一步，RealignerTargetCreator ，目的是定位出所有需要进行序列重比对的目标区域</li>
<li>第二步，IndelRealigner，对所有在第一步中找到的目标区域运用算法进行序列重比对，最后得到捋顺了的新结果。<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">java -jar /path/to/GenomeAnalysisTK.jar \</span><br><span class="line"> -T RealignerTargetCreator \</span><br><span class="line"> -R /path/to/human.fasta \</span><br><span class="line"> -I sample_name.sorted.markdup.bam \</span><br><span class="line"> -known /path/to/gatk/bundle/1000G_phase1.indels.b37.vcf \</span><br><span class="line"> -known /path/to/gatk/bundle/Mills_and_1000G_gold_standard.indels.b37.vcf \</span><br><span class="line"> -o sample_name.IndelRealigner.intervals</span><br><span class="line"></span><br><span class="line">java -jar /path/to/GenomeAnalysisTK.jar \</span><br><span class="line"> -T IndelRealigner \</span><br><span class="line"> -R /path/to/human.fasta \</span><br><span class="line"> -I sample_name.sorted.markdup.bam \</span><br><span class="line"> -known /path/to/gatk/bundle/1000G_phase1.indels.b37.vcf \</span><br><span class="line"> -known /path/to/gatk/bundle/Mills_and_1000G_gold_standard.indels.b37.vcf \</span><br><span class="line"> -o sample_name.sorted.markdup.realign.bam \</span><br><span class="line"> --targetIntervals sample_name.IndelRealigner.intervals</span><br></pre></td></tr></table></figure>

</li>
</ul>
<p>这一步中我们实际使用到的known indel集为：<br> Homo_sapiens_assembly38.known_indels.vcf<br> Mills_and_1000G_gold_standard.indels.hg38.vcf<br> 这些文件可以方便的在GATK bundle里面下载（<a href="ftp://ftp.broadinstitute.org/bundle/hg38/）" target="_blank" rel="noopener">ftp://ftp.broadinstitute.org/bundle/hg38/）</a>  </p>
<p> 后面的变异检测使用GATK，而且使用GATK的HaplotypeCaller模块或者是Mutect2模块时，当这个时候才可以减少这个Indel局部重比对的步骤。 </p>
<p> <strong>重新矫正碱基质量值</strong><br> 分为两步:</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"> gatk BaseRecalibrator \</span><br><span class="line">   -I my_reads.bam \</span><br><span class="line">   -R reference.fasta \</span><br><span class="line">   --known-sites sites_of_variation.vcf \</span><br><span class="line">   --known-sites another/optional/setOfSitesToMask.vcf \</span><br><span class="line">   -O recal_data.table</span><br><span class="line">## 用到的已知编译位点</span><br><span class="line">common_all_20180418_dbsnp_151_hg38.vcf</span><br><span class="line">Homo_sapiens_assembly38.known_indels.vcf</span><br><span class="line">Mills_and_1000G_gold_standard.indels.hg38.vcf</span><br><span class="line">下载地址（GATK的bundle以及1000genomic官网）：ftp://ftp.broadinstitute.org/bundle/hg38/</span><br><span class="line">https://ftp.ncbi.nlm.nih.gov/snp/organisms/</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">gatk ApplyBQSR \</span><br><span class="line">  -R reference.fasta \</span><br><span class="line">  -I input.bam \</span><br><span class="line">  --bqsr-recal-file recalibration.table \</span><br><span class="line">  -O output.bam</span><br></pre></td></tr></table></figure>

<h6 id="MSI鉴定"><a href="#MSI鉴定" class="headerlink" title="MSI鉴定"></a>MSI鉴定</h6><p>通过以上步骤我么就获得了干净的bam文件可以用于下一步的MSI鉴定分析。<br>在此次鉴定中一共使用到了三种MSI鉴定软件用于鉴定我们的tumer-normal paired样本</p>
<p>** MSIsensor **  
MSIsensor通过分别判断每个微卫星位点的稳定性，然后以不稳定微卫星位点的比例作为MSI得分。MSIsensor需要基于配对的肿瘤-正常样本进行MSI的判定。<br> 首先, 对于在肿瘤和正常样本中测序深度均大于等于20的微卫星位点, 计算其等位基因的分布信息; 其次, 通过卡方检验比较肿瘤和正常样本的相同微卫星位点的等位基因分布, 若显著不同, 则认为该微卫星位点不稳定; 最后统计不稳定位点的比例, 若该比例超过阈值, 则判定为MSI-H,这个阈值一般是3.5%<br> <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#重参考基因组中获取微卫星位点</span></span><br><span class="line">msisensor scan -d reference.fa -o microsatellites.list</span><br><span class="line"><span class="comment">#MSI扫描</span></span><br><span class="line">msisensor msi -d microsatellites.list -n normal.bam -t tumor.bam -e bed.file -o output.prefix</span><br></pre></td></tr></table></figure></p>
<p> 结果显示3个样本都是MSS<br>** MANTIS **
<code>MANTIS</code>软件使用Python编辑，最新的版本2.7.1,但是Python3.0版本也是支持的<br>下载地址：<a href="https://github.com/OSU-SRLab/MANTIS" target="_blank" rel="noopener">https://github.com/OSU-SRLab/MANTIS</a><br>类似于MSIsensor, MANTIS也获得了肿瘤-正常配对样本在每个微卫星位点的等位基因分布信息; 与MSIsensor不同的是, 对于每个微卫星位点, MANTIS把上述两组数据看作两个向量, 定义这两个向量的 L1范数为样本中该位点的稳定程度, 对所有位点的L1范数求平均值即为样本的MSI得分</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#从参考基因组中获得微卫星位点</span></span><br><span class="line">./RepeatFinder –i /path/to/genome.fasta -o /path/to/loci.bed</span><br><span class="line"><span class="comment">#对上一步获得的微卫星位点根据exon测序实验室中的target位置信息进行过滤</span></span><br><span class="line">bedtools intersect -a MANTIS.loc.bed -b hg38_Regions.bed &gt; MANTIS_filter.bed</span><br><span class="line"><span class="comment">#进行MSI鉴定</span></span><br><span class="line">python mantis.py --bedfile /path/to/loci.bed --genome /path/to/genome.fasta -n /path/to/normal.bam -t /path/to/tumor.bam -o /path/to/output/file.txt</span><br></pre></td></tr></table></figure>

<p> 对于外显子测序，可使用以下不太严格的参数设置：<br> <figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">-mrq <span class="number">20.0</span></span><br><span class="line">-mlq <span class="number">25</span></span><br><span class="line">-mlc <span class="number">20</span></span><br><span class="line">-mrr <span class="number">1</span></span><br></pre></td></tr></table></figure></p>
<p> 结果显示3个样本都是MSS<br> ** VisualMSI **  
 通过模仿PCR来检测MSI。VisualMSI从参考基因组中提取PCR adapter，并将其map到测序read中。加入adapter可以成功的map到raed上，将计算这些adapter的插入长度。VisualMSI统计计算这些长度的分布情况。<br> <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">visualmsi -i tumor.sorted.bam -n normal.sorted.bam -r hg19.fasta -t targets/msi.tsv</span><br></pre></td></tr></table></figure></p>
<p> 关于target file<br> VisualMSI官网提供的参考文件为：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">#CHROM  POSITION  NAME</span><br><span class="line">chr4  55598216  BAT25</span><br><span class="line">chr2  47641568  BAT26</span><br><span class="line">chr14 23652365  NR-21</span><br><span class="line">chr11 102193518 NR-27</span><br><span class="line">chr2  95849372  NR-24</span><br></pre></td></tr></table></figure>

<p>但是他是根据hg19参考基因组来做的，我们需要将它转换为hg38参考基因组的,可以利用UCSC的liftover工具</p>
<ul>
<li>下载注释文件 <a href="https://hgdownload-test.gi.ucsc.edu/goldenPath/hg19/liftOver/" target="_blank" rel="noopener">https://hgdownload-test.gi.ucsc.edu/goldenPath/hg19/liftOver/</a>    中的hg19ToHg38.over.chain.gz文件</li>
<li>输入文件需要是标准的bed格式</li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">必须的字段：</span><br><span class="line">- 染色体</span><br><span class="line">- 起始位置（第一个氨基酸为0）</span><br><span class="line">- 终止位置</span><br><span class="line">chr1  213941196  213942363</span><br><span class="line">chr1  213942363  213943530</span><br><span class="line">chr1  213943530  213944697</span><br><span class="line">chr2  158364697  158365864</span><br><span class="line">chr2  158365864  158367031</span><br><span class="line">可选字段：name、score、strand....</span><br></pre></td></tr></table></figure>

<ul>
<li>进行转换<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">../softwave/liftOver CRC_cancer-Specific_marker.bed hg19ToHg38.over.chain.gz hg38_CRC_cancer-Specific_marker.tsv unmap</span><br></pre></td></tr></table></figure>

</li>
</ul>
<p>不知道为什么大多数的position都显示能通过过滤标准的reads特别的少，不能通过质量控制，这个软件不适合？还是说参数设置不对？<br>** mSINGS **
1) Catalog microsatellites present in your host genome.  There are a number of algorithms available to to this, but one we have found particularly easy to use is MISA (<a href="http://pgrc.ipk-gatersleben.de/misa/" target="_blank" rel="noopener">http://pgrc.ipk-gatersleben.de/misa/</a>).<br>2) Limit the list of microsatellites to those which are present in your capture design.  This can be done by converting the location of identified microsatellites from step #1 to BED file format, converting the coordinates of your capture design to BED file format, and using the BEDTOOLS intersect function to find where these two files overlap (<a href="http://bedtools.readthedocs.io/en/latest/content/tools/intersect.html" target="_blank" rel="noopener">http://bedtools.readthedocs.io/en/latest/content/tools/intersect.html</a>).<br>3) Generate support files and a baseline file from no fewer than 10-20 specimens, as detailed in the readme.<br>4) Examine the baseline file.  Identify any loci which have an average peak count of 1.0 / standard deviation of 0.0, as these loci may cause artifacts.  These should be removed from the baseline file and the original locus BED file.  The intervals file will need to be re-generated with the edited input BED file.<br>5) Perform mSINGS analysis of the individual files that you used to generate your baseline with the edited files.  Any remaining loci which are called as unstable in a measurable fraction (~10% or more) of known negative specimens should also be dropped due to their potential to cause artifacts.  Remove these loci and edit files as in step #4<br>6) Optional - if you have access to a number of known MSI positive specimens (10 or more), it can be helpful to analyze them at this point.  Discriminatory power of mSINGS can be improved if uninformative loci (those which are not unstable in 1 or more known MSI specimens) are removed from the panel.  You may remove uninformative loci as in step #4.  If these specimens are not available, be aware that an empirically-determined cutoff for discriminating MSS form MSI-H specimens may need to be established for your assay.  </p>
<p>mSINGs的使用流程：</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">##由于mSINGs是基于Python 2.7写的，所以首先要将shell切换到Python 2.7环境</span></span><br><span class="line"><span class="keyword">source</span> activate py2.7</span><br><span class="line">git clone https://bitbucket.org/uwlabmed/msings.git</span><br><span class="line">cd msings</span><br><span class="line">bash ./dev/bootstrap.sh</span><br><span class="line"><span class="comment">##切换到msings-env环境</span></span><br><span class="line"><span class="keyword">source</span> msings-env/bin/activate</span><br><span class="line"><span class="comment">##利用MANTIS软件中的获得的filter_bed（capture.bed的交集）文件进行下一步分析</span></span><br><span class="line">./scripts/create_intervals.sh ../../reference/mSINGS/MANTIS_filtered.bed</span><br><span class="line"><span class="comment">##创建msi_bed文件</span></span><br><span class="line">./scripts/create_baseline.sh ../../reference/mSINGS/normal_bamlist ../../reference/mSINGS/MANTIS_filtered.msi_intervals ../../reference/mSINGS/MANTIS_filtered.bed /data/reference/GRCh38/GRCh38.primary_assembly.genome.fa</span><br></pre></td></tr></table></figure>

<p>上面的到<code>MSI_BASELINE.txt</code>文件，筛选到其中<code>Standard_Deviation=0</code>的Position，这些Position被认为是没有意义的误差，然后与之前获得的<code>bed</code>文件取交集，获得有意义的bed文件<br>对normal样本进行MSI鉴定，对bed文件进行进一步的验证筛选</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">./scripts/run_msings.sh</span><br><span class="line">          ../../reference/mSINGS/cancer_bamlist ../../reference/mSINGS/MANTIS_filtered.filterSD.msi_intervals</span><br><span class="line">          ../../reference/mSINGS/MANTIS_filtered.filterSD.bed </span><br><span class="line">          /data/reference/GRCh38/GRCh38.primary_assembly.genome.fa </span><br><span class="line">          ../../reference/mSINGS/MSI_BASELINE_filterSD.txt</span><br></pre></td></tr></table></figure>

<p>根据获的<code>Combined_MSI.txt</code>结果，再从<code>MSI_BASELIEN_filterSD.txt</code>文件中删除掉在instable_num/total_num &gt; 10% 的position，同样的<code>bed</code>文件和<code>interval</code>也做相同的处理,得到最终的<code>MSI_BASELINE_filterSD_checknorm.txt</code><br>最后进行cancer样本的MSI鉴定</p>
<pre><code>./scripts/run_msings.sh
          ../../reference/mSINGS/cancer_bamlist
          ../../reference/mSINGS/MANTIS_filtered.filterSD.checknorm.msi_intervals
          ../../reference/mSINGS/MANTIS_filtered.filterSD.checknorm.bed
          /data/reference/GRCh38/GRCh38.primary_assembly.genome.fa 
          ../../reference/mSINGS/MSI_BASELINE_filterSD_checknorm.txt</code></pre><p>结果显示3个样本都是MSS的</p>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/06/05/hello-world/" rel="next" title="Hello World">
                <i class="fa fa-chevron-left"></i> Hello World
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          
  <div class="comments" id="comments">
    
  </div>


        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview">
            Overview
          </li>
        </ul>
      

      <section class="site-overview sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
          <img class="site-author-image" itemprop="image" src="http://www.jf258.com/uploads/2013-07-13/171053918.jpg" alt="Jing Yu">
          <p class="site-author-name" itemprop="name">Jing Yu</p>
           
              <p class="site-description motion-element" itemprop="description"></p>
          
        </div>
        <nav class="site-state motion-element">

          
            <div class="site-state-item site-state-posts">
              <a href="/archives/">
                <span class="site-state-item-count">2</span>
                <span class="site-state-item-name">posts</span>
              </a>
            </div>
          

          

          

        </nav>

        

        <div class="links-of-author motion-element">
          
        </div>

        
        

        
        

        


      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-5"><a class="nav-link" href="#软件准备"><span class="nav-number">1.</span> <span class="nav-text">软件准备</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#原始数据质量控制"><span class="nav-number">2.</span> <span class="nav-text">原始数据质量控制</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#全外显子组分析（WES）"><span class="nav-number">3.</span> <span class="nav-text">全外显子组分析（WES）</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#数据预处理"><span class="nav-number">3.1.</span> <span class="nav-text">数据预处理</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#Somatic-SNVs-Indels"><span class="nav-number">3.2.</span> <span class="nav-text">Somatic SNVs+Indels</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#使用旧版的mutects来call-somatic-mutation（现在GATK网站已经不推荐了）"><span class="nav-number">3.3.</span> <span class="nav-text">使用旧版的mutects来call somatic mutation（现在GATK网站已经不推荐了）</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#使用Muect2新流程进行calling-somatic-mutation"><span class="nav-number">3.4.</span> <span class="nav-text">使用Muect2新流程进行calling somatic mutation</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#使用不同的方法进行annotation"><span class="nav-number">3.5.</span> <span class="nav-text">使用不同的方法进行annotation</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#使用VarScan2-Somatic来call-somatic（没有写完）"><span class="nav-number">3.6.</span> <span class="nav-text">使用VarScan2 Somatic来call somatic（没有写完）</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#环境搭建"><span class="nav-number">3.7.</span> <span class="nav-text">环境搭建</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#Detect-MSI"><span class="nav-number">4.</span> <span class="nav-text">Detect MSI</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#数据预处理-1"><span class="nav-number">4.1.</span> <span class="nav-text">数据预处理</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#MSI鉴定"><span class="nav-number">4.2.</span> <span class="nav-text">MSI鉴定</span></a></li></ol></li></ol></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">
  
  &copy; 
  <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-heart"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Jing Yu</span>
</div>


<div class="powered-by">
  Powered by <a class="theme-link" href="https://hexo.io">Hexo</a>
</div>

<div class="theme-info">
  Theme -
  <a class="theme-link" href="https://github.com/iissnan/hexo-theme-next">
    NexT.Muse
  </a>
</div>


        

        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>

  
  <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>

  
  <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>

  
  <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.2"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.2"></script>



  
  

  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.2"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.2"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.2"></script>



  


  




	





  





  






  





  

  

  

  

  

  

</body>
</html>
