<!DOCTYPE html>



  


<html class="theme-next pisces use-motion" lang>
<head><meta name="generator" content="Hexo 3.8.0">
  <meta charset="UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=1">
<meta name="theme-color" content="#222">









<meta http-equiv="Cache-Control" content="no-transform">
<meta http-equiv="Cache-Control" content="no-siteapp">
















  
  
  <link href="/lib/fancybox/source/jquery.fancybox.css?v=2.1.5" rel="stylesheet" type="text/css">







<link href="/lib/font-awesome/css/font-awesome.min.css?v=4.6.2" rel="stylesheet" type="text/css">

<link href="/css/main.css?v=5.1.4" rel="stylesheet" type="text/css">


  <link rel="apple-touch-icon" sizes="180x180" href="/images/favicon.gif?v=5.1.4">


  <link rel="icon" type="image/png" sizes="32x32" href="/images/favicon.gif?v=5.1.4">


  <link rel="icon" type="image/png" sizes="16x16" href="/images/favicon.gif?v=5.1.4">






  <meta name="keywords" content="single cell,">










<meta name="description" content="1 Tabula Muris1.1 介绍他们将高通量但低覆盖率的10倍数据与低通量高覆盖率的数据结合起来，FACS-sortedcell+Smartseq2。 1.22 数据下载下载FACS data">
<meta name="keywords" content="single cell">
<meta property="og:type" content="article">
<meta property="og:title" content="Analysis of single cell RNA-seq data">
<meta property="og:url" content="http://yoursite.com/2019/06/06/Analysis of single cell RNA-seq data/index.html">
<meta property="og:site_name" content="Jing">
<meta property="og:description" content="1 Tabula Muris1.1 介绍他们将高通量但低覆盖率的10倍数据与低通量高覆盖率的数据结合起来，FACS-sortedcell+Smartseq2。 1.22 数据下载下载FACS data">
<meta property="og:locale" content="default">
<meta property="og:image" content="https://note.youdao.com/yws/public/resource/12beaf1b878bc1059015bf8c805c27d5/xmlnote/A5BCB784861A4B3CB16854DDB816A6AC/889">
<meta property="og:image" content="https://note.youdao.com/favicon.ico">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/14-exprs-qc_files/figure-html/mt-vs-counts-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/14-exprs-qc_files/figure-html/ercc-vs-counts-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/14-exprs-qc_files/figure-html/auto-cell-filt-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/14-exprs-qc_files/figure-html/top50-gene-expr-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/pca.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-pca-before-qc1-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-pca-before-qc2-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-pca-after-qc-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-pca-after-qc-exercise1-2-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-tsne-before-qc-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-tsne-after-qc-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-pca-before-qc-reads1-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-pca-before-qc-reads2-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-pca-after-qc-reads-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-tsne-before-qc-reads-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-tsne-after-qc-reads-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-tsne-after-qc-exercise2-1-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-tsne-after-qc-exercise2-2-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/18-confounders_files/figure-html/confound-pca-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/18-confounders_files/figure-html/confound-find-pcs-total-features-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/18-confounders_files/figure-html/confound-find-expl-vars-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-raw-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-cpm-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-cpm-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-rle-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-rle-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-uq-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-uq-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-tmm-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-tmm-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-lsf-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-scran-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-downsample-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-downsample-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-2.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-3.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-4.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-5.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-6.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-7.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-8.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-9.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-10.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-12.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-11-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-2.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-3.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-4.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-5.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-6.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-7.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-8.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-9.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-10.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-11.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-12.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-13-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-2.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-3.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-4.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-5.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-6.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-7.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-8.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-9.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-10.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-11.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-12.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-11-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-2.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-3.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-4.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-5.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-6.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-7.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-8.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-9.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-10.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-11.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-12.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-14-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/hierarchical_clustering1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/hierarchical_clustering2.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/k-means.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/graph_network.jpg">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/sc3.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-5-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-7-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-9-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-10-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-11-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-12-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-13-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-18-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/clust-pca-reduce2-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/clust-tsne-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/clust-tsne-kmeans2-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-21-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/clust-sincera-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/26-dropouts_files/figure-html/unnamed-chunk-6-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/26-dropouts_files/figure-html/unnamed-chunk-7-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/26-dropouts_files/figure-html/unnamed-chunk-10-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/26-dropouts_files/figure-html/unnamed-chunk-15-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/cannoodt_pseudotime_properties.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/cannoodt_pseudotime_methods.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/data-overview-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/pca-pseudotime-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/tscan-all-genes-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/tscan-vs-truth-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/m3d-select-genes-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/monocle-all-genes-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/monocle-vs-truth-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/destiny-deng-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/destiny-deng-2.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/slicer-analyis-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/slicer-build-graph-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/slicer-1.png">
<meta property="og:image" content="https://note.youdao.com/yws/public/resource/12beaf1b878bc1059015bf8c805c27d5/xmlnote/33E31DDE55D94CE1AEEF02639EABBC86/2019">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-plot-exprs-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-plot-switch-times-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-plot-switch-times-2.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-consistency-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-pseudotime-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-pseudotime-2.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-states-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/compare-results-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-pc1-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-tscan-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-monocle-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-diff-map-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-slicer-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-ouija-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/28-imputation_files/figure-html/unnamed-chunk-4-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/28-imputation_files/figure-html/unnamed-chunk-5-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/28-imputation_files/figure-html/unnamed-chunk-7-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/28-imputation_files/figure-html/unnamed-chunk-8-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/29-de-intro_files/figure-html/nb-plot-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/29-de-intro_files/figure-html/zero-inflation-plot-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/29-de-intro_files/figure-html/pois-beta-plot-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/KS2_Example.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/30-de-real_files/figure-html/ks-roc-plot-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/30-de-real_files/figure-html/wilcox-plot-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/30-de-real_files/figure-html/edger-plot-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/30-de-real_files/figure-html/Monocle-plot-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/30-de-real_files/figure-html/MAST-plot-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/31-projection_files/figure-html/unnamed-chunk-12-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/31-projection_files/figure-html/unnamed-chunk-13-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/31-projection_files/figure-html/unnamed-chunk-15-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/31-projection_files/figure-html/unnamed-chunk-32-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/mnnCorrectDiagramCropped.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/31-projection_files/figure-html/unnamed-chunk-38-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/muraro_seurat_hvg.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/seger_seurat_hvg.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/cca_before.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/cca_after.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/scfind.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/32-search_files/figure-html/unnamed-chunk-7-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-5-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-5-2.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-11-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-11-2.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-12-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-14-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-15-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-19-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-21-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-21-2.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-23-1.png">
<meta property="og:image" content="https://hemberg-lab.github.io/scRNA.seq.course/figures/Pipeline-batches.png">
<meta property="og:updated_time" content="2019-06-06T05:41:42.182Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:title" content="Analysis of single cell RNA-seq data">
<meta name="twitter:description" content="1 Tabula Muris1.1 介绍他们将高通量但低覆盖率的10倍数据与低通量高覆盖率的数据结合起来，FACS-sortedcell+Smartseq2。 1.22 数据下载下载FACS data">
<meta name="twitter:image" content="https://note.youdao.com/yws/public/resource/12beaf1b878bc1059015bf8c805c27d5/xmlnote/A5BCB784861A4B3CB16854DDB816A6AC/889">



<script type="text/javascript" id="hexo.configurations">
  var NexT = window.NexT || {};
  var CONFIG = {
    root: '/',
    scheme: 'Pisces',
    version: '5.1.4',
    sidebar: {"position":"left","display":"post","offset":12,"b2t":false,"scrollpercent":false,"onmobile":false},
    fancybox: true,
    tabs: true,
    motion: {"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},
    duoshuo: {
      userId: '0',
      author: 'Author'
    },
    algolia: {
      applicationID: '',
      apiKey: '',
      indexName: '',
      hits: {"per_page":10},
      labels: {"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}
    }
  };
</script>



  <link rel="canonical" href="http://yoursite.com/2019/06/06/Analysis of single cell RNA-seq data/">





  <title>Analysis of single cell RNA-seq data | Jing</title>
  








</head>

<body itemscope itemtype="http://schema.org/WebPage" lang="default">

  
  
    
  

  <div class="container sidebar-position-left page-post-detail">
    <div class="headband"></div>

    <header id="header" class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-wrapper">
  <div class="site-meta ">
    

    <div class="custom-logo-site-title">
      <a href="/" class="brand" rel="start">
        <span class="logo-line-before"><i></i></span>
        <span class="site-title">Jing</span>
        <span class="logo-line-after"><i></i></span>
      </a>
    </div>
      
        <p class="site-subtitle"></p>
      
  </div>

  <div class="site-nav-toggle">
    <button>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
      <span class="btn-bar"></span>
    </button>
  </div>
</div>

<nav class="site-nav">
  

  
    <ul id="menu" class="menu">
      
        
        <li class="menu-item menu-item-home">
          <a href="/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-home"></i> <br>
            
            Home
          </a>
        </li>
      
        
        <li class="menu-item menu-item-tags">
          <a href="/tags/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-tags"></i> <br>
            
            Tags
          </a>
        </li>
      
        
        <li class="menu-item menu-item-categories">
          <a href="/categories/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-th"></i> <br>
            
            Categories
          </a>
        </li>
      
        
        <li class="menu-item menu-item-archives">
          <a href="/archives/" rel="section">
            
              <i class="menu-item-icon fa fa-fw fa-archive"></i> <br>
            
            Archives
          </a>
        </li>
      

      
        <li class="menu-item menu-item-search">
          
            <a href="javascript:;" class="popup-trigger">
          
            
              <i class="menu-item-icon fa fa-search fa-fw"></i> <br>
            
            Search
          </a>
        </li>
      
    </ul>
  

  
    <div class="site-search">
      
  <div class="popup search-popup local-search-popup">
  <div class="local-search-header clearfix">
    <span class="search-icon">
      <i class="fa fa-search"></i>
    </span>
    <span class="popup-btn-close">
      <i class="fa fa-times-circle"></i>
    </span>
    <div class="local-search-input-wrapper">
      <input autocomplete="off" placeholder="Searching..." spellcheck="false" type="text" id="local-search-input">
    </div>
  </div>
  <div id="local-search-result"></div>
</div>



    </div>
  
</nav>



 </div>
    </header>

    <main id="main" class="main">
      <div class="main-inner">
        <div class="content-wrap">
          <div id="content" class="content">
            

  <div id="posts" class="posts-expand">
    

  

  
  
  

  <article class="post post-type-normal" itemscope itemtype="http://schema.org/Article">
  
  
  
  <div class="post-block">
    <link itemprop="mainEntityOfPage" href="http://yoursite.com/2019/06/06/Analysis of single cell RNA-seq data/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="name" content="Jing Yu">
      <meta itemprop="description" content>
      <meta itemprop="image" content="/images/avatar.png">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Jing">
    </span>

    
      <header class="post-header">

        
        
          <h1 class="post-title" itemprop="name headline">Analysis of single cell RNA-seq data</h1>
        

        <div class="post-meta">
          <span class="post-time">
            
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              
                <span class="post-meta-item-text">Posted on</span>
              
              <time title="Post created" itemprop="dateCreated datePublished" datetime="2019-06-06T13:39:55+08:00">
                2019-06-06
              </time>
            

            

            
          </span>

          
            <span class="post-category">
            
              <span class="post-meta-divider">|</span>
            
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              
                <span class="post-meta-item-text">In</span>
              
              
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/categories/single-cell/" itemprop="url" rel="index">
                    <span itemprop="name">single_cell</span>
                  </a>
                </span>

                
                
              
            </span>
          

          
            
              <span class="post-comments-count">
                <span class="post-meta-divider">|</span>
                <span class="post-meta-item-icon">
                  <i class="fa fa-comment-o"></i>
                </span>
                <a href="/2019/06/06/Analysis of single cell RNA-seq data/#comments" itemprop="discussionUrl">
                  <span class="post-comments-count valine-comment-count" data-xid="/2019/06/06/Analysis of single cell RNA-seq data/" itemprop="commentCount"></span>
                </a>
              </span>
            
          

          
          

          
            <span class="post-meta-divider">|</span>
            <span class="page-pv"><i class="fa fa-file-o"></i>
            <span class="busuanzi-value" id="busuanzi_value_page_pv"></span>
            </span>
          

          

          

        </div>
      </header>
    

    
    
    
    <div class="post-body" itemprop="articleBody">

      
      

      
        <h3 id="1-Tabula-Muris"><a href="#1-Tabula-Muris" class="headerlink" title="1 Tabula Muris"></a>1 Tabula Muris</h3><h4 id="1-1-介绍"><a href="#1-1-介绍" class="headerlink" title="1.1 介绍"></a>1.1 介绍</h4><p>他们将高通量但低覆盖率的10倍数据与低通量高覆盖率的数据结合起来，FACS-sortedcell+Smartseq2。</p>
<h4 id="1-22-数据下载"><a href="#1-22-数据下载" class="headerlink" title="1.22 数据下载"></a>1.22 数据下载</h4><p>下载FACS data</p>
<a id="more"></a>
<figure class="highlight vim"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget http<span class="variable">s:</span>//ndownloader.figshare.<span class="keyword">com</span>/<span class="keyword">files</span>/<span class="number">10038307</span></span><br><span class="line">unzip <span class="number">10038307</span></span><br><span class="line">wget http<span class="variable">s:</span>//ndownloader.figshare.<span class="keyword">com</span>/<span class="keyword">files</span>/<span class="number">10038310</span></span><br><span class="line">mv <span class="number">10038310</span> FACS_metadata.csv</span><br><span class="line">wget http<span class="variable">s:</span>//ndownloader.figshare.<span class="keyword">com</span>/<span class="keyword">files</span>/<span class="number">10039267</span></span><br><span class="line">mv <span class="number">10039267</span> FACS_annotations.csv</span><br></pre></td></tr></table></figure>

<p>下载 10X data</p>
<figure class="highlight vim"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">wget http<span class="variable">s:</span>//ndownloader.figshare.<span class="keyword">com</span>/<span class="keyword">files</span>/<span class="number">10038325</span></span><br><span class="line">unzip <span class="number">10038325</span></span><br><span class="line">wget http<span class="variable">s:</span>//ndownloader.figshare.<span class="keyword">com</span>/<span class="keyword">files</span>/<span class="number">10038328</span></span><br><span class="line">mv <span class="number">10038328</span> droplet_metadata.csv</span><br><span class="line">wget http<span class="variable">s:</span>//ndownloader.figshare.<span class="keyword">com</span>/<span class="keyword">files</span>/<span class="number">10039264</span></span><br><span class="line">mv <span class="number">10039264</span> droplet_annotation.csv</span><br></pre></td></tr></table></figure>

<p>下载好的数据如下所示：<br><img src="https://note.youdao.com/yws/public/resource/12beaf1b878bc1059015bf8c805c27d5/xmlnote/A5BCB784861A4B3CB16854DDB816A6AC/889" alt="DownLoadData"><br>现在有两个文件夹：“FACS”和“droplet”，每一个文件夹包括一个annotation和metadata文件。要检查这些文件，您可以使用head查看文本文件的前几行（按“q”键退出）：</p>
<figure class="highlight css"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="selector-tag">head</span> <span class="selector-tag">-n</span> 10 <span class="selector-tag">droplet_metadata</span><span class="selector-class">.csv</span></span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">## channel,mouse.id,tissue,subtissue,mouse.sex</span><br><span class="line">## <span class="number">10</span>X_P4_0,<span class="number">3</span>-M<span class="number">-8</span>,Tongue,NA,M</span><br><span class="line">## <span class="number">10</span>X_P4_1,<span class="number">3</span>-M<span class="number">-9</span>,Tongue,NA,M</span><br><span class="line">## <span class="number">10</span>X_P4_2,<span class="number">3</span>-M<span class="number">-8</span>/<span class="number">9</span>,Liver,hepatocytes,M</span><br><span class="line">## <span class="number">10</span>X_P4_3,<span class="number">3</span>-M<span class="number">-8</span>,Bladder,NA,M</span><br><span class="line">## <span class="number">10</span>X_P4_4,<span class="number">3</span>-M<span class="number">-9</span>,Bladder,NA,M</span><br><span class="line">## <span class="number">10</span>X_P4_5,<span class="number">3</span>-M<span class="number">-8</span>,Kidney,NA,M</span><br><span class="line">## <span class="number">10</span>X_P4_6,<span class="number">3</span>-M<span class="number">-9</span>,Kidney,NA,M</span><br><span class="line">## <span class="number">10</span>X_P4_7,<span class="number">3</span>-M<span class="number">-8</span>,Spleen,NA,M</span><br><span class="line">## <span class="number">10</span>X_P7_0,<span class="number">3</span>-F<span class="number">-56</span>,Liver,NA,F</span><br></pre></td></tr></table></figure>

<h4 id="1-3-读取这些数据集（Smartseq2）"><a href="#1-3-读取这些数据集（Smartseq2）" class="headerlink" title="1.3 读取这些数据集（Smartseq2）"></a>1.3 读取这些数据集（Smartseq2）</h4><p>现在，我们可以从逗号分隔的文件中读取相关的计数矩阵。然后检查产生的dataframe：</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">dat = read.delim(<span class="string">"FACS/Kidney-counts.csv"</span>, <span class="attribute">sep</span>=<span class="string">","</span>, <span class="attribute">header</span>=<span class="literal">TRUE</span>)</span><br><span class="line">dat[1:5,1:5]</span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">##               X A14.MAA000545<span class="number">.3</span>_8_M<span class="number">.1</span><span class="number">.1</span> E1.MAA000545<span class="number">.3</span>_8_M<span class="number">.1</span><span class="number">.1</span></span><br><span class="line">## <span class="number">1</span> <span class="number">0610005</span>C13Rik                       <span class="number">0</span>                      <span class="number">0</span></span><br><span class="line">## <span class="number">2</span> <span class="number">0610007</span>C21Rik                       <span class="number">1</span>                      <span class="number">0</span></span><br><span class="line">## <span class="number">3</span> <span class="number">0610007</span>L01Rik                       <span class="number">0</span>                      <span class="number">0</span></span><br><span class="line">## <span class="number">4</span> <span class="number">0610007</span>N19Rik                       <span class="number">0</span>                      <span class="number">0</span></span><br><span class="line">## <span class="number">5</span> <span class="number">0610007</span>P08Rik                       <span class="number">0</span>                      <span class="number">0</span></span><br><span class="line">##   M4.MAA000545<span class="number">.3</span>_8_M<span class="number">.1</span><span class="number">.1</span> O21.MAA000545<span class="number">.3</span>_8_M<span class="number">.1</span><span class="number">.1</span></span><br><span class="line">## <span class="number">1</span>                      <span class="number">0</span>                       <span class="number">0</span></span><br><span class="line">## <span class="number">2</span>                      <span class="number">0</span>                       <span class="number">0</span></span><br><span class="line">## <span class="number">3</span>                      <span class="number">0</span>                       <span class="number">0</span></span><br><span class="line">## <span class="number">4</span>                      <span class="number">0</span>                       <span class="number">0</span></span><br><span class="line">## <span class="number">5</span>                      <span class="number">0</span>                       <span class="number">0</span></span><br></pre></td></tr></table></figure>

<p>我们可以看到dataframe的第一列是基因名，所以首先我们把它们移到行名中这样我们就有了一个数字矩阵：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">dim</span><span class="params">(dat)</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">23433</span>   <span class="number">866</span></span><br></pre></td></tr></table></figure>

<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">rownames</span><span class="params">(dat)</span></span> &lt;- dat[,<span class="number">1</span>]</span><br><span class="line">dat &lt;- dat[,-<span class="number">1</span>]</span><br></pre></td></tr></table></figure>

<p>由于这是Smartseq2的数据集，所以他可能包含spike-ins：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">rownames</span><span class="params">(dat)</span></span>[grep(<span class="string">"^ERCC-"</span>, rownames(dat))]</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br></pre></td><td class="code"><pre><span class="line">##  [<span class="number">1</span>] <span class="string">"ERCC-00002"</span> <span class="string">"ERCC-00003"</span> <span class="string">"ERCC-00004"</span> <span class="string">"ERCC-00009"</span> <span class="string">"ERCC-00012"</span></span><br><span class="line">##  [<span class="number">6</span>] <span class="string">"ERCC-00013"</span> <span class="string">"ERCC-00014"</span> <span class="string">"ERCC-00016"</span> <span class="string">"ERCC-00017"</span> <span class="string">"ERCC-00019"</span></span><br><span class="line">## [<span class="number">11</span>] <span class="string">"ERCC-00022"</span> <span class="string">"ERCC-00024"</span> <span class="string">"ERCC-00025"</span> <span class="string">"ERCC-00028"</span> <span class="string">"ERCC-00031"</span></span><br><span class="line">## [<span class="number">16</span>] <span class="string">"ERCC-00033"</span> <span class="string">"ERCC-00034"</span> <span class="string">"ERCC-00035"</span> <span class="string">"ERCC-00039"</span> <span class="string">"ERCC-00040"</span></span><br><span class="line">## [<span class="number">21</span>] <span class="string">"ERCC-00041"</span> <span class="string">"ERCC-00042"</span> <span class="string">"ERCC-00043"</span> <span class="string">"ERCC-00044"</span> <span class="string">"ERCC-00046"</span></span><br><span class="line">## [<span class="number">26</span>] <span class="string">"ERCC-00048"</span> <span class="string">"ERCC-00051"</span> <span class="string">"ERCC-00053"</span> <span class="string">"ERCC-00054"</span> <span class="string">"ERCC-00057"</span></span><br><span class="line">## [<span class="number">31</span>] <span class="string">"ERCC-00058"</span> <span class="string">"ERCC-00059"</span> <span class="string">"ERCC-00060"</span> <span class="string">"ERCC-00061"</span> <span class="string">"ERCC-00062"</span></span><br><span class="line">## [<span class="number">36</span>] <span class="string">"ERCC-00067"</span> <span class="string">"ERCC-00069"</span> <span class="string">"ERCC-00071"</span> <span class="string">"ERCC-00073"</span> <span class="string">"ERCC-00074"</span></span><br><span class="line">## [<span class="number">41</span>] <span class="string">"ERCC-00075"</span> <span class="string">"ERCC-00076"</span> <span class="string">"ERCC-00077"</span> <span class="string">"ERCC-00078"</span> <span class="string">"ERCC-00079"</span></span><br><span class="line">## [<span class="number">46</span>] <span class="string">"ERCC-00081"</span> <span class="string">"ERCC-00083"</span> <span class="string">"ERCC-00084"</span> <span class="string">"ERCC-00085"</span> <span class="string">"ERCC-00086"</span></span><br><span class="line">## [<span class="number">51</span>] <span class="string">"ERCC-00092"</span> <span class="string">"ERCC-00095"</span> <span class="string">"ERCC-00096"</span> <span class="string">"ERCC-00097"</span> <span class="string">"ERCC-00098"</span></span><br><span class="line">## [<span class="number">56</span>] <span class="string">"ERCC-00099"</span> <span class="string">"ERCC-00104"</span> <span class="string">"ERCC-00108"</span> <span class="string">"ERCC-00109"</span> <span class="string">"ERCC-00111"</span></span><br><span class="line">## [<span class="number">61</span>] <span class="string">"ERCC-00112"</span> <span class="string">"ERCC-00113"</span> <span class="string">"ERCC-00116"</span> <span class="string">"ERCC-00117"</span> <span class="string">"ERCC-00120"</span></span><br><span class="line">## [<span class="number">66</span>] <span class="string">"ERCC-00123"</span> <span class="string">"ERCC-00126"</span> <span class="string">"ERCC-00130"</span> <span class="string">"ERCC-00131"</span> <span class="string">"ERCC-00134"</span></span><br><span class="line">## [<span class="number">71</span>] <span class="string">"ERCC-00136"</span> <span class="string">"ERCC-00137"</span> <span class="string">"ERCC-00138"</span> <span class="string">"ERCC-00142"</span> <span class="string">"ERCC-00143"</span></span><br><span class="line">## [<span class="number">76</span>] <span class="string">"ERCC-00144"</span> <span class="string">"ERCC-00145"</span> <span class="string">"ERCC-00147"</span> <span class="string">"ERCC-00148"</span> <span class="string">"ERCC-00150"</span></span><br><span class="line">## [<span class="number">81</span>] <span class="string">"ERCC-00154"</span> <span class="string">"ERCC-00156"</span> <span class="string">"ERCC-00157"</span> <span class="string">"ERCC-00158"</span> <span class="string">"ERCC-00160"</span></span><br><span class="line">## [<span class="number">86</span>] <span class="string">"ERCC-00162"</span> <span class="string">"ERCC-00163"</span> <span class="string">"ERCC-00164"</span> <span class="string">"ERCC-00165"</span> <span class="string">"ERCC-00168"</span></span><br><span class="line">## [<span class="number">91</span>] <span class="string">"ERCC-00170"</span> <span class="string">"ERCC-00171"</span></span><br></pre></td></tr></table></figure>

<p>现在，我们可以从列名中提取该数据的大部分元数据：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cellIDs &lt;- colnames(<span class="name">dat</span>)</span><br><span class="line">cell_info &lt;- strsplit(<span class="name">cellIDs</span>, <span class="string">"\\."</span>)</span><br><span class="line">Well &lt;- lapply(<span class="name">cell_info</span>, function(<span class="name">x</span>)&#123;x[<span class="number">1</span>]&#125;)</span><br><span class="line">Well &lt;- unlist(<span class="name">Well</span>)</span><br><span class="line">Plate &lt;- unlist(<span class="name">lapply</span>(<span class="name">cell_info</span>, function(<span class="name">x</span>)&#123;x[<span class="number">2</span>]&#125;))</span><br><span class="line">Mouse &lt;- unlist(<span class="name">lapply</span>(<span class="name">cell_info</span>, function(<span class="name">x</span>)&#123;x[<span class="number">3</span>]&#125;))</span><br></pre></td></tr></table></figure>

<p>我们可以检查每个元数据分类的分布：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">summary(<span class="name">factor</span>(<span class="name">Mouse</span>))</span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">## <span class="number">3</span>_10_M <span class="number">3</span>_11_M <span class="number">3</span>_38_F <span class="number">3</span>_39_F  <span class="number">3</span>_8_M  <span class="number">3</span>_9_M </span><br><span class="line">##    <span class="number">104</span>    <span class="number">196</span>    <span class="number">237</span>    <span class="number">169</span>     <span class="number">82</span>     <span class="number">77</span></span><br></pre></td></tr></table></figure>

<p>我们还可以检查是否有任何技术因素被混淆：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">table</span><span class="params">(Mouse, Plate)</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">##         Plate</span><br><span class="line">## Mouse    B001717 B002775 MAA000545 MAA000752 MAA000801 MAA000922</span><br><span class="line">##   <span class="number">3</span>_10_M       <span class="number">0</span>       <span class="number">0</span>         <span class="number">0</span>       <span class="number">104</span>         <span class="number">0</span>         <span class="number">0</span></span><br><span class="line">##   <span class="number">3</span>_11_M       <span class="number">0</span>       <span class="number">0</span>         <span class="number">0</span>         <span class="number">0</span>       <span class="number">196</span>         <span class="number">0</span></span><br><span class="line">##   <span class="number">3</span>_38_F     <span class="number">237</span>       <span class="number">0</span>         <span class="number">0</span>         <span class="number">0</span>         <span class="number">0</span>         <span class="number">0</span></span><br><span class="line">##   <span class="number">3</span>_39_F       <span class="number">0</span>     <span class="number">169</span>         <span class="number">0</span>         <span class="number">0</span>         <span class="number">0</span>         <span class="number">0</span></span><br><span class="line">##   <span class="number">3</span>_8_M        <span class="number">0</span>       <span class="number">0</span>        <span class="number">82</span>         <span class="number">0</span>         <span class="number">0</span>         <span class="number">0</span></span><br><span class="line">##   <span class="number">3</span>_9_M        <span class="number">0</span>       <span class="number">0</span>         <span class="number">0</span>         <span class="number">0</span>         <span class="number">0</span>        <span class="number">77</span></span><br></pre></td></tr></table></figure>

<p>最后，我们将推断的单元类型注释，并将它们与表达式矩阵中的单元相匹配：</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ann &lt;- read.table(<span class="string">"FACS_annotations.csv"</span>, <span class="attribute">sep</span>=<span class="string">","</span>, <span class="attribute">header</span>=<span class="literal">TRUE</span>)</span><br><span class="line">ann &lt;- ann[match(cellIDs, ann[,1]),]</span><br><span class="line">celltype &lt;- ann[,3]</span><br></pre></td></tr></table></figure>

<h4 id="1-4-构建一个scater-object"><a href="#1-4-构建一个scater-object" class="headerlink" title="1.4 构建一个scater object"></a>1.4 构建一个scater object</h4><p>为了创建一个SingleCellExperiment对象，我们必须把所有的单元注释放在一个dataframe中，因为实验性的批处理（pcr板）与供体鼠标完全混淆了，我们只保留其中一个。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">require</span><span class="params">(<span class="string">"SingleCellExperiment"</span>)</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">cell_anns</span> &lt;- <span class="class"><span class="keyword">data</span>.frame(<span class="title">mouse</span> = <span class="type">Mouse</span>, <span class="title">well</span>=<span class="type">Well</span>, <span class="title">type</span>=<span class="title">celltype</span>)</span></span><br><span class="line"><span class="title">rownames</span>(cell_anns) &lt;- colnames(dat)</span><br><span class="line"><span class="title">sceset</span> &lt;- <span class="type">SingleCellExperiment</span>(assays = list(counts = <span class="keyword">as</span>.matrix(dat)), colData=cell_anns)</span><br></pre></td></tr></table></figure>

<p>最后，如果数据集包含了“spike”——我们在singlecellexperobject对象中有一个隐藏的变量来跟踪它们：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">isSpike(<span class="name">sceset</span>, <span class="string">"ERCC"</span>) &lt;- grepl(<span class="string">"ERCC-"</span>, rownames(<span class="name">sceset</span>))</span><br></pre></td></tr></table></figure>

<h4 id="1-5-读取数据集（10X）"><a href="#1-5-读取数据集（10X）" class="headerlink" title="1.5 读取数据集（10X）"></a>1.5 读取数据集（10X）</h4><p>由于其大小和稀疏的10X数据（高达90%的表达式矩阵可能是0），它通常存储为稀疏矩阵。CellRanger的默认输出格式是一个.mtx文件，它将这个稀疏矩阵存储为一列行坐标、一列列坐标，以及一列大于0的表达式值。注意，如果您查看.mtx文件，您将看到两个标题行，后面是一行，详细描述了完整矩阵的行数、列数和计数。因为只有坐标被存储在.mtx文件中，所以每一行和列的名称必须分别存储在“genes.tsv”和“barcode.tsv”。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">require</span><span class="params">(<span class="string">"Matrix"</span>)</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight stata"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">cellbarcodes &lt;- <span class="keyword">read</span>.<span class="keyword">table</span>(<span class="string">"droplet/Kidney-10X_P4_5/barcodes.tsv"</span>)</span><br><span class="line">genenames &lt;- <span class="keyword">read</span>.<span class="keyword">table</span>(<span class="string">"droplet/Kidney-10X_P4_5/genes.tsv"</span>)</span><br><span class="line">molecules &lt;- <span class="keyword">Matrix</span>::readMM(<span class="string">"droplet/Kidney-10X_P4_5/matrix.mtx"</span>)</span><br></pre></td></tr></table></figure>

<p>现在我们将添加适当的行和列名。但是，如果您检查读取的cellbarcode，您将看到它们只是与每个单元相关联的条形码序列。这是一个问题，因为每批10个数据都使用相同的条形码池，因此如果我们需要将来自多个10个批次的数据组合起来，那么这些数据将不会是唯一的。因此，我们将把批号附加到每个细胞的条形码上：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">head</span><span class="params">(cellbarcodes)</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">##                   V1</span><br><span class="line">## <span class="number">1</span> AAACCTGAGATGCCAG<span class="number">-1</span></span><br><span class="line">## <span class="number">2</span> AAACCTGAGTGTCCAT<span class="number">-1</span></span><br><span class="line">## <span class="number">3</span> AAACCTGCAAGGCTCC<span class="number">-1</span></span><br><span class="line">## <span class="number">4</span> AAACCTGTCCTTGCCA<span class="number">-1</span></span><br><span class="line">## <span class="number">5</span> AAACGGGAGCTGAACG<span class="number">-1</span></span><br><span class="line">## <span class="number">6</span> AAACGGGCAGGACCCT<span class="number">-1</span></span><br></pre></td></tr></table></figure>

<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">rownames</span><span class="params">(molecules)</span></span> &lt;- genenames[,<span class="number">1</span>]</span><br><span class="line"><span class="function"><span class="title">colnames</span><span class="params">(molecules)</span></span> &lt;- paste(<span class="string">"10X_P4_5"</span>, cellbarcodes[,<span class="number">1</span>], sep=<span class="string">"_"</span>)</span><br></pre></td></tr></table></figure>

<p>现在，让我们获取这些数据的元数据和注释：</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">meta &lt;- read.delim(<span class="string">"droplet_metadata.csv"</span>, <span class="attribute">sep</span>=<span class="string">","</span>, <span class="attribute">header</span>=<span class="literal">TRUE</span>)</span><br><span class="line">head(meta)</span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">##    channel mouse.id  tissue   subtissue mouse.sex</span><br><span class="line">## <span class="number">1</span> <span class="number">10</span>X_P4_0    <span class="number">3</span>-M<span class="number">-8</span>  Tongue        &lt;NA&gt;         M</span><br><span class="line">## <span class="number">2</span> <span class="number">10</span>X_P4_1    <span class="number">3</span>-M<span class="number">-9</span>  Tongue        &lt;NA&gt;         M</span><br><span class="line">## <span class="number">3</span> <span class="number">10</span>X_P4_2  <span class="number">3</span>-M<span class="number">-8</span>/<span class="number">9</span>   Liver hepatocytes         M</span><br><span class="line">## <span class="number">4</span> <span class="number">10</span>X_P4_3    <span class="number">3</span>-M<span class="number">-8</span> Bladder        &lt;NA&gt;         M</span><br><span class="line">## <span class="number">5</span> <span class="number">10</span>X_P4_4    <span class="number">3</span>-M<span class="number">-9</span> Bladder        &lt;NA&gt;         M</span><br><span class="line">## <span class="number">6</span> <span class="number">10</span>X_P4_5    <span class="number">3</span>-M<span class="number">-8</span>  Kidney        &lt;NA&gt;         M</span><br></pre></td></tr></table></figure>

<p>在这里我们可以看到,我们需要使用“10 x_p4_5”找到这批的元数据,还要注意,老鼠ID是不同的格式在此元数据表为连字符而不是下划线和性别在中间的ID。检查附带论文的方法部分我们知道相同的8个老鼠用于droplet和plate-based技术。因此，我们需要修复小鼠的id，使其与FACS实验中使用的id一致。</p>
<figure class="highlight crmsh"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">meta</span>[<span class="keyword">meta</span>$channel == <span class="string">"10X_P4_5"</span>,]</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">##    channel mouse.id tissue subtissue mouse.sex</span><br><span class="line">## <span class="number">6</span> <span class="number">10</span>X_P4_5    <span class="number">3</span>-M<span class="number">-8</span> Kidney      &lt;NA&gt;         M</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">mouseID &lt;- <span class="string">"3_8_M"</span></span><br></pre></td></tr></table></figure>

<p>注意：根据你所分配的组织，你可能有来自混合样本的10倍数据：例如，鼠标id=3-m-5/6。您仍然应该重新格式化这些内容以保持一致，但是它们不会匹配来自FACS数据的鼠标id，这可能会影响您的下游分析。如果这些小鼠不是来自于近亲繁殖的病毒株，那么就有可能将单个细胞分配给特定的老鼠，使用的是无证明的snps，但这超出了本课程的范围。</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">ann &lt;- read.delim(<span class="string">"droplet_annotation.csv"</span>, <span class="attribute">sep</span>=<span class="string">","</span>, <span class="attribute">header</span>=<span class="literal">TRUE</span>)</span><br><span class="line">head(ann)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">##                        cell  tissue cell_ontology_class</span><br><span class="line">## <span class="number">1</span> <span class="number">10</span>X_P4_3_AAAGTAGAGATGCCAG Bladder    mesenchymal cell</span><br><span class="line">## <span class="number">2</span> <span class="number">10</span>X_P4_3_AACCGCGTCCAACCAA Bladder    mesenchymal cell</span><br><span class="line">## <span class="number">3</span> <span class="number">10</span>X_P4_3_AACTCCCGTCGGGTCT Bladder    mesenchymal cell</span><br><span class="line">## <span class="number">4</span> <span class="number">10</span>X_P4_3_AACTCTTAGTTGCAGG Bladder        bladder cell</span><br><span class="line">## <span class="number">5</span> <span class="number">10</span>X_P4_3_AACTCTTTCATAACCG Bladder    mesenchymal cell</span><br><span class="line">## <span class="number">6</span> <span class="number">10</span>X_P4_3_AAGACCTAGATCCGAG Bladder    endothelial cell</span><br><span class="line">##                      cell_ontology_term_iri cell_ontology_id</span><br><span class="line">## <span class="number">1</span> http:<span class="comment">//purl.obolibrary.org/obo/CL_0008019       CL:0008019</span></span><br><span class="line">## <span class="number">2</span> http:<span class="comment">//purl.obolibrary.org/obo/CL_0008019       CL:0008019</span></span><br><span class="line">## <span class="number">3</span> http:<span class="comment">//purl.obolibrary.org/obo/CL_0008019       CL:0008019</span></span><br><span class="line">## <span class="number">4</span> http:<span class="comment">//purl.obolibrary.org/obo/CL_1001319       CL:1001319</span></span><br><span class="line">## <span class="number">5</span> http:<span class="comment">//purl.obolibrary.org/obo/CL_0008019       CL:0008019</span></span><br><span class="line">## <span class="number">6</span> http:<span class="comment">//purl.obolibrary.org/obo/CL_0000115       CL:0000115</span></span><br></pre></td></tr></table></figure>

<p>再一次，你会发现在注释和cell条形码之间有一个细微的区别，在匹配之前，我们必须纠正它们。</p>
<figure class="highlight prolog"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">ann[,<span class="number">1</span>] &lt;- paste(ann[,<span class="number">1</span>], <span class="string">"-1"</span>, sep=<span class="string">""</span>)</span><br><span class="line">ann_subset &lt;- ann[match(colnames(molecules), ann[,<span class="number">1</span>]),]</span><br><span class="line">celltype &lt;- ann_subset[,<span class="number">3</span>]</span><br></pre></td></tr></table></figure>

<p>现在让我们构建单元元数据dataframe：</p>
<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">cell_anns</span> &lt;- <span class="class"><span class="keyword">data</span>.frame(<span class="title">mouse</span> = <span class="title">rep</span>(<span class="title">mouseID</span>, <span class="title">times</span>=<span class="title">ncol</span>(<span class="title">molecules</span>)), <span class="keyword">type</span>=celltype)</span></span><br><span class="line"><span class="title">rownames</span>(cell_anns) &lt;- colnames(molecules);</span><br></pre></td></tr></table></figure>

<h4 id="1-6-构建一个scater-object"><a href="#1-6-构建一个scater-object" class="headerlink" title="1.6 构建一个scater object"></a>1.6 构建一个scater object</h4><p>既然我们已经读入10X的数据，我们需要将它们组合成一个SingCellExperiment对象。首先，我们将检查基因名称是否相同，并且在所有批次中都是相同的：</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">identical(rownames(molecules1), rownames(molecules2))</span><br><span class="line">## [<span class="number">1</span>] TRUE</span><br><span class="line">identical(rownames(molecules1), rownames(molecules3))</span><br><span class="line">## [<span class="number">1</span>] TRUE</span><br></pre></td></tr></table></figure>

<p>现在我们来检查一下是否有重复的CellIDs：</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">sum(colnames(molecules1) %<span class="keyword">in</span>% colnames(molecules2))</span><br><span class="line">## [<span class="number">1</span>] <span class="number">0</span></span><br><span class="line">sum(colnames(molecules1) %<span class="keyword">in</span>% colnames(molecules3))</span><br><span class="line">## [<span class="number">1</span>] <span class="number">0</span></span><br><span class="line">sum(colnames(molecules2) %<span class="keyword">in</span>% colnames(molecules3))</span><br><span class="line">## [<span class="number">1</span>] <span class="number">0</span></span><br></pre></td></tr></table></figure>

<p>一切都OK，我们可以把它们组合起来：</p>
<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">all_molecules</span> &lt;- cbind(molecules1, molecules2, molecules3)</span><br><span class="line"><span class="title">all_cell_anns</span> &lt;- <span class="keyword">as</span>.<span class="class"><span class="keyword">data</span>.frame(<span class="title">rbind</span>(<span class="title">cell_anns1</span>, <span class="title">cell_anns2</span>, <span class="title">cell_anns3</span>))</span></span><br><span class="line"><span class="title">all_cell_anns</span>$batch &lt;- rep(c(<span class="string">"10X_P4_5"</span>, <span class="string">"10X_P4_6"</span>,<span class="string">"10X_P7_5"</span>), times = c(nrow(cell_anns1), nrow(cell_anns2), nrow(cell_anns3)))</span><br></pre></td></tr></table></figure>

<p>现在我们来构建一个SingleCellExperiments对象。SingleCellExperiment类的一个优点是,它能够将数据存储为正常数据矩阵或稀疏矩阵格式,以及HDF5格式其允许大型non-sparse矩阵存储以及以一种有效的方式访问磁盘,而不是整个加载到RAM中。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">require</span><span class="params">(<span class="string">"SingleCellExperiment"</span>)</span></span></span><br><span class="line"><span class="function"><span class="title">require</span><span class="params">(<span class="string">"scater"</span>)</span></span></span><br><span class="line">all_molecules &lt;- as.matrix(all_molecules)</span><br><span class="line">sceset &lt;- SingleCellExperiment(assays = list(counts = as.matrix(all_molecules)), colData=all_cell_anns)</span><br></pre></td></tr></table></figure>

<p>因为这是10倍的数据，所以它不会包含spike-in，所以我们只保存数据：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">saveRDS</span><span class="params">(sceset, <span class="string">"kidney_droplet.rds"</span>)</span></span></span><br></pre></td></tr></table></figure>

<h3 id="2-cleaning-the-Expression-Matrix"><a href="#2-cleaning-the-Expression-Matrix" class="headerlink" title="2 cleaning the Expression Matrix"></a>2 cleaning the Expression Matrix</h3><h4 id="2-1-ExpressionQC（UMI）"><a href="#2-1-ExpressionQC（UMI）" class="headerlink" title="2.1 ExpressionQC（UMI）"></a>2.1 ExpressionQC（UMI）</h4><h5 id="2-1-1-介绍"><a href="#2-1-1-介绍" class="headerlink" title="2.1.1 介绍"></a>2.1.1 介绍</h5><p>一旦基因表达被量化，它就被总结为一个表达矩阵，其中每一行对应一个基因（或转录），每一列对应一个单元格。应该检查这个矩阵，以去除那些没有被检测到的质量好的细胞，这些细胞在读取QC或映射质量控制步骤中没有被检测到。在这个阶段，如果不能去除低质量的细胞，可能会增加技术噪音，这种噪音有可能掩盖下游分析中感兴趣的生物信号。<br>由于目前还没有标准的方法来执行scRNASeq，因此在这里提出的各种质量控制措施的预期值可能会因实验而异。因此，为了执行QC，我们将寻找与数据集其余部分相关的异常值的单元，而不是与独立的质量标准进行比较。因此，在比较使用不同协议收集的数据集的质量度量时，应该小心谨慎。</p>
<h5 id="2-1-2-Tung-dataset"><a href="#2-1-2-Tung-dataset" class="headerlink" title="2.1.2 Tung dataset"></a>2.1.2 Tung dataset</h5><p>为了说明细胞的质量控制，<a href="http://jdblischak.github.io/singleCellSeq/analysis/" target="_blank" rel="noopener">我们考虑了一个由三个不同的个体产生的诱导多功能干细胞的数据集</a>（Tung等人，2017年），在芝加哥大学的Yoav Gilad实验室。这些实验是在FluidigmC1平台上进行的，并促进了对唯一的分子标识符（UMIs）和ERCC spike的量化</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">library</span><span class="params">(SingleCellExperiment)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(scater)</span></span></span><br><span class="line"><span class="function"><span class="title">options</span><span class="params">(stringsAsFactors = FALSE)</span></span></span><br></pre></td></tr></table></figure>

<p>加载数据和注释文件：</p>
<figure class="highlight lua"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">molecules &lt;- <span class="built_in">read</span>.<span class="built_in">table</span>(<span class="string">"tung/molecules.txt"</span>, sep = <span class="string">"\t"</span>)</span><br><span class="line">anno &lt;- <span class="built_in">read</span>.<span class="built_in">table</span>(<span class="string">"tung/annotation.txt"</span>, sep = <span class="string">"\t"</span>, header = TRUE)</span><br></pre></td></tr></table></figure>

<p>检查表达式矩阵的一小部分</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">head</span><span class="params">(molecules[ , <span class="number">1</span>:<span class="number">3</span>])</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">##                 NA19098.r1.A01 NA19098.r1.A02 NA19098.r1.A03</span><br><span class="line">## ENSG00000237683              <span class="number">0</span>              <span class="number">0</span>              <span class="number">0</span></span><br><span class="line">## ENSG00000187634              <span class="number">0</span>              <span class="number">0</span>              <span class="number">0</span></span><br><span class="line">## ENSG00000188976              <span class="number">3</span>              <span class="number">6</span>              <span class="number">1</span></span><br><span class="line">## ENSG00000187961              <span class="number">0</span>              <span class="number">0</span>              <span class="number">0</span></span><br><span class="line">## ENSG00000187583              <span class="number">0</span>              <span class="number">0</span>              <span class="number">0</span></span><br><span class="line">## ENSG00000187642              <span class="number">0</span>              <span class="number">0</span>              <span class="number">0</span></span><br></pre></td></tr></table></figure>

<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">head</span><span class="params">(anno)</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight gcode"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="attr">##   individual replicate well      batch      sample_id</span></span><br><span class="line"><span class="attr">## 1</span>    <span class="symbol">NA19098</span>        r<span class="number">1</span>  A<span class="number">01</span> <span class="symbol">NA19098</span>.r<span class="number">1</span> <span class="symbol">NA19098</span>.r<span class="number">1.</span>A<span class="number">01</span></span><br><span class="line"><span class="attr">## 2</span>    <span class="symbol">NA19098</span>        r<span class="number">1</span>  A<span class="number">02</span> <span class="symbol">NA19098</span>.r<span class="number">1</span> <span class="symbol">NA19098</span>.r<span class="number">1.</span>A<span class="number">02</span></span><br><span class="line"><span class="attr">## 3</span>    <span class="symbol">NA19098</span>        r<span class="number">1</span>  A<span class="number">03</span> <span class="symbol">NA19098</span>.r<span class="number">1</span> <span class="symbol">NA19098</span>.r<span class="number">1.</span>A<span class="number">03</span></span><br><span class="line"><span class="attr">## 4</span>    <span class="symbol">NA19098</span>        r<span class="number">1</span>  A<span class="number">04</span> <span class="symbol">NA19098</span>.r<span class="number">1</span> <span class="symbol">NA19098</span>.r<span class="number">1.</span>A<span class="number">04</span></span><br><span class="line"><span class="attr">## 5</span>    <span class="symbol">NA19098</span>        r<span class="number">1</span>  A<span class="number">05</span> <span class="symbol">NA19098</span>.r<span class="number">1</span> <span class="symbol">NA19098</span>.r<span class="number">1.</span>A<span class="number">05</span></span><br><span class="line"><span class="attr">## 6</span>    <span class="symbol">NA19098</span>        r<span class="number">1</span>  A<span class="number">06</span> <span class="symbol">NA19098</span>.r<span class="number">1</span> <span class="symbol">NA19098</span>.r<span class="number">1.</span>A<span class="number">06</span></span><br></pre></td></tr></table></figure>

<p>我们通过使用SingleCellExperiments（SCE）和scater包来对分析进行标准化。首先，创建SCE对象：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">umi &lt;- SingleCellExperiment(</span><br><span class="line">    assays = list(<span class="name">counts</span> = as.matrix(<span class="name">molecules</span>)), </span><br><span class="line">    colData = anno</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>移除在任何细胞中没有表达的基因：</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">keep_feature &lt;- rowSums(counts(umi) &gt; <span class="number">0</span>) &gt; <span class="number">0</span></span><br><span class="line">umi &lt;- umi[keep_feature, ]</span><br></pre></td></tr></table></figure>

<p>定义控制features（genes）-ERCC spike和线粒体基因：</p>
<figure class="highlight erlang"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">isSpike</span><span class="params">(umi, <span class="string">"ERCC"</span>)</span> &lt;- <span class="title">grepl</span><span class="params">(<span class="string">"^ERCC-"</span>, rownames(umi))</span></span></span><br><span class="line"><span class="function"><span class="title">isSpike</span><span class="params">(umi, <span class="string">"MT"</span>)</span> &lt;- <span class="title">rownames</span><span class="params">(umi)</span> %<span class="title">in</span>% </span></span><br><span class="line"><span class="function">    <span class="title">c</span><span class="params">(<span class="string">"ENSG00000198899"</span>, <span class="string">"ENSG00000198727"</span>, <span class="string">"ENSG00000198888"</span>,</span></span></span><br><span class="line"><span class="function"><span class="params">    <span class="string">"ENSG00000198886"</span>, <span class="string">"ENSG00000212907"</span>, <span class="string">"ENSG00000198786"</span>,</span></span></span><br><span class="line"><span class="function"><span class="params">    <span class="string">"ENSG00000198695"</span>, <span class="string">"ENSG00000198712"</span>, <span class="string">"ENSG00000198804"</span>,</span></span></span><br><span class="line"><span class="function"><span class="params">    <span class="string">"ENSG00000198763"</span>, <span class="string">"ENSG00000228253"</span>, <span class="string">"ENSG00000198938"</span>,</span></span></span><br><span class="line"><span class="function"><span class="params">    <span class="string">"ENSG00000198840"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p>计算质量指标:</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">umi &lt;- calculateQCMetrics(</span><br><span class="line">    umi,</span><br><span class="line">    feature_controls = list(</span><br><span class="line">        ERCC = isSpike(<span class="name">umi</span>, <span class="string">"ERCC"</span>), </span><br><span class="line">        MT = isSpike(<span class="name">umi</span>, <span class="string">"MT"</span>)</span><br><span class="line">    )</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<h5 id="2-1-3-Cell-QC"><a href="#2-1-3-Cell-QC" class="headerlink" title="2.1.3 Cell QC"></a>2.1.3 Cell QC</h5><h6 id="2-1-3-1-文库大小（Library-size）"><a href="#2-1-3-1-文库大小（Library-size）" class="headerlink" title="2.1.3.1 文库大小（Library size）"></a>2.1.3.1 文库大小（Library size）</h6><p>接下来，我们考虑每个样本检测到的RNA分子总数（如果我们使用的是读计数而不是UMI计数，这将是read的总数）。没有多少reads/molecules的cells可能会被破坏或无法捕获一个细胞，因此应该被移除。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">hist(</span><br><span class="line">    umi$total_counts,</span><br><span class="line">    breaks = <span class="number">100</span></span><br><span class="line">)</span><br><span class="line">abline(<span class="name">v</span> = <span class="number">25000</span>, col = <span class="string">"red"</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://note.youdao.com/favicon.ico" alt="histplot"></p>
<h6 id="2-1-3-2-Detected-genes"><a href="#2-1-3-2-Detected-genes" class="headerlink" title="2.1.3.2 Detected genes"></a>2.1.3.2 Detected genes</h6><p>除了为每个样本确保足够的排序深度之外，我们还希望确保read分布在transcriptome中。因此，我们计算每个样本中检测到的唯一基因的总数。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">hist(</span><br><span class="line">    umi$total_features,</span><br><span class="line">    breaks = <span class="number">100</span></span><br><span class="line">)</span><br><span class="line">abline(<span class="name">v</span> = <span class="number">7000</span>, col = <span class="string">"red"</span>)</span><br></pre></td></tr></table></figure>

<p>从图中我们得出结论，大多数细胞有7000-10000个检测到的基因，这对于high-depth的scRNA-seq来说是正常的。然而，这取决于实验流程和测序深度。例如，基于droplet的方法或具有较低测序深度的样本通常检测每个细胞的基因更少。上面的图中最显著的特征是在分布的左边的“重尾（heavy tail）”。如果在整个细胞中检测率是相等的，那么分布应该是近似正常的。因此，我们将这些细胞移出分布的尾部（少于7000个检测到的基因）。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">filter_by_total_counts &lt;- (<span class="name">umi</span>$total_counts &gt; <span class="number">25000</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">table</span><span class="params">(filter_by_total_counts)</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">## filter_by_total_counts</span><br><span class="line">## FALSE  TRUE </span><br><span class="line">##    <span class="number">46</span>   <span class="number">818</span></span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">filter_by_total_features &lt;- (<span class="name">umi</span>$total_features &gt; <span class="number">7000</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">## filter_by_expr_features</span><br><span class="line">## FALSE  TRUE </span><br><span class="line">##   <span class="number">116</span>   <span class="number">748</span></span><br></pre></td></tr></table></figure>

<h6 id="2-1-3-3-ERCCs和MTS"><a href="#2-1-3-3-ERCCs和MTS" class="headerlink" title="2.1.3.3 ERCCs和MTS"></a>2.1.3.3 ERCCs和MTS</h6><p>另一种衡量细胞质量的指标是ERCC spike-in rna和内源性rna之间的比率。这个比率可以用来估计被捕获的细胞中RNA的总量。具有高水平 spike-in RNA的细胞具有较低的RNA起始数量，这可能是由于细胞死亡或压力导致RNA被降解。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">plotPhenoData(</span><br><span class="line">    umi,</span><br><span class="line">    aes_string(</span><br><span class="line">        x = <span class="string">"total_features"</span>,</span><br><span class="line">        y = <span class="string">"pct_counts_MT"</span>,</span><br><span class="line">        colour = <span class="string">"batch"</span></span><br><span class="line">    )</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/14-exprs-qc_files/figure-html/mt-vs-counts-1.png" alt="image"></p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">plotPhenoData(</span><br><span class="line">    umi,</span><br><span class="line">    aes_string(</span><br><span class="line">        x = <span class="string">"total_features"</span>,</span><br><span class="line">        y = <span class="string">"pct_counts_ERCC"</span>,</span><br><span class="line">        colour = <span class="string">"batch"</span></span><br><span class="line">    )</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/14-exprs-qc_files/figure-html/ercc-vs-counts-1.png" alt="image"></p>
<figure class="highlight nginx"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="attribute">filter_by_MT</span> &lt;- (umi<span class="variable">$total_counts_MT</span>/umi<span class="variable">$total_counts</span>&lt;<span class="number">0</span>.<span class="number">1</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">table(filter_by_MT)</span><br><span class="line">##filter_by_MT</span><br><span class="line">##FALSE  TRUE </span><br><span class="line">##   <span class="number">31</span>   <span class="number">833</span></span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">filter_by_ERCC&lt;- (<span class="name">umi</span>$batch!=<span class="string">"NA19098.r2"</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">## filter_by_ERCC</span><br><span class="line">## FALSE  TRUE </span><br><span class="line">##    <span class="number">96</span>   <span class="number">768</span></span><br></pre></td></tr></table></figure>

<h5 id="2-1-4-Cell-filtering"><a href="#2-1-4-Cell-filtering" class="headerlink" title="2.1.4 Cell filtering"></a>2.1.4 Cell filtering</h5><h6 id="2-1-4-1-Manual"><a href="#2-1-4-1-Manual" class="headerlink" title="2.1.4.1 Manual"></a>2.1.4.1 Manual</h6><p>现在我们可以根据我们之前的分析定义一个细胞过滤：</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">umi$use &lt;- (</span><br><span class="line">    # sufficient features (genes)</span><br><span class="line">    filter_by_expr_features &amp;</span><br><span class="line">    # sufficient molecules counted</span><br><span class="line">    filter_by_total_counts &amp;</span><br><span class="line">    # sufficient endogenous RNA</span><br><span class="line">    filter_by_ERCC &amp;</span><br><span class="line">    # remove cells <span class="keyword">with</span> unusual number <span class="keyword">of</span> reads <span class="keyword">in</span> MT genes</span><br><span class="line">    filter_by_MT</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">table(umi$</span><span class="bash">use)</span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># FALSE  TRUE </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#   207   657</span></span></span><br></pre></td></tr></table></figure>

<h6 id="2-1-4-2-Automatic"><a href="#2-1-4-2-Automatic" class="headerlink" title="2.1.4.2 Automatic"></a>2.1.4.2 Automatic</h6><p>scater的另一种选择是在一组QC指标上进行PCA，然后使用自动离群检测来识别潜在的问题细胞。<br>默认情况下，以下指标用于基于PCA的异常检测:</p>
<ul>
<li>pct_counts_top_100_features</li>
<li>total_features</li>
<li>pct_counts_feature_controls</li>
<li>n_detected_feature_controls</li>
<li>log10_counts_endogenous_features</li>
<li>log10_counts_feature_controls<br>scater首先创建一个矩阵其中行表示cell，列表示不同的QC指标.在这里，PCA图提供了由其质量指标排序的细胞的2D表示。然后使用mvoutlier包中的方法检测出异常值</li>
</ul>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">umi &lt;- plotPCA(</span><br><span class="line">    umi,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>, </span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"use"</span>,</span><br><span class="line">    <span class="attr">pca_data_input</span> = <span class="string">"pdata"</span>,</span><br><span class="line">    <span class="attr">detect_outliers</span> = TRUE,</span><br><span class="line">    <span class="attr">return_SCE</span> = TRUE</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/14-exprs-qc_files/figure-html/auto-cell-filt-1.png" alt="image"></p>
<h5 id="2-1-5-Compare-filterings"><a href="#2-1-5-Compare-filterings" class="headerlink" title="2.1.5 Compare filterings"></a>2.1.5 Compare filterings</h5><p>从limma包中使用venn计数和vennDiagram函数来制作维恩图。</p>
<h5 id="2-1-6-Gene-analysis"><a href="#2-1-6-Gene-analysis" class="headerlink" title="2.1.6 Gene analysis"></a>2.1.6 Gene analysis</h5><h6 id="2-1-6-1-Gene-expression"><a href="#2-1-6-1-Gene-expression" class="headerlink" title="2.1.6.1 Gene expression"></a>2.1.6.1 Gene expression</h6><p>除了去除质量不佳的细胞外，在我们怀疑技术人工制品可能扭曲了结果的情况下，排除基因是一个好主意。此外，对基因表达谱的检查可以提供关于如何改进实验过程的见解。<br>考虑前50个表达基因所消耗的reads通常是有益的。</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotQC(umi,<span class="built_in"> type </span>= <span class="string">"highest-expression"</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/14-exprs-qc_files/figure-html/top50-gene-expr-1.png" alt="image"><br>这些分布是相对平坦的表示（但不是保证！）对这些cell的完整的转录组的良好覆盖。然而，在前15个基因中，有几只“spike-ins”的基因表明，如果实验要重复的话，对spike-ins的稀释度可能会更大。</p>
<h6 id="2-1-6-2-Gene-filtering"><a href="#2-1-6-2-Gene-filtering" class="headerlink" title="2.1.6.2 Gene filtering"></a>2.1.6.2 Gene filtering</h6><p>去除那些表达水平被认为是“不可检测”的基因是一个好主意。如果1个基因的转录本在至少两个细胞中被检测到，我们就认为该基因是可以被检测的。如果我们考虑的是reads，而不是UMI，那么一个合理的阈值就是至少需要在两个cell中至少有5个reads。然而，在这两种情况下，阈值都依赖于测序深度。重要的是要记住，基因必须在细胞过滤后进行过滤，因为有些基因只能在质量较差的细胞中被检测出来（注：colData（umi）$use过滤器应用于umi数据集）</p>
<figure class="highlight openscad"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">filter_genes &lt;- apply(</span><br><span class="line">    counts(umi[ , colData(umi)$<span class="keyword">use</span>]), </span><br><span class="line">    <span class="number">1</span>, </span><br><span class="line">    <span class="function"><span class="keyword">function</span><span class="params">(x)</span> <span class="title">length</span><span class="params">(x[x &gt; <span class="number">1</span>])</span> &gt;=</span> <span class="number">2</span></span><br><span class="line">)</span><br><span class="line">rowData(umi)$<span class="meta"><span class="meta-keyword">use</span> &lt;- filter_genes</span></span><br></pre></td></tr></table></figure>

<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">table</span><span class="params">(filter_genes)</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">## filter_genes</span><br><span class="line">## FALSE  TRUE </span><br><span class="line">##  <span class="number">4660</span> <span class="number">14066</span></span><br></pre></td></tr></table></figure>

<h5 id="2-1-7-Save-the-Data"><a href="#2-1-7-Save-the-Data" class="headerlink" title="2.1.7 Save the Data"></a>2.1.7 Save the Data</h5><p>QCed dataset数据集的维度（不要忘记我们上面定义的基因过滤参数）：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">dim(<span class="name">umi</span>[rowData(<span class="name">umi</span>)$use, colData(<span class="name">umi</span>)$use])</span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">14066</span>   <span class="number">657</span></span><br></pre></td></tr></table></figure>

<p>让我们用log-treansformed count创建一个额外的slot（我们将在下一章中需要它），并从还原的slot中删除保存的PCA结果。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">assay(<span class="name">umi</span>, <span class="string">"logcounts_raw"</span>) &lt;- log2(<span class="name">counts</span>(<span class="name">umi</span>) + <span class="number">1</span>)</span><br><span class="line">reducedDim(<span class="name">umi</span>) &lt;- NULL</span><br></pre></td></tr></table></figure>

<p>保存数据</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">saveRDS</span><span class="params">(umi, file = <span class="string">"tung/umi.rds"</span>)</span></span></span><br></pre></td></tr></table></figure>

<h4 id="2-2-Expression-QC（Reads）"><a href="#2-2-Expression-QC（Reads）" class="headerlink" title="2.2 Expression QC（Reads）"></a>2.2 Expression QC（Reads）</h4><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">library</span><span class="params">(SingleCellExperiment)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(scater)</span></span></span><br><span class="line"><span class="function"><span class="title">options</span><span class="params">(stringsAsFactors = FALSE)</span></span></span><br></pre></td></tr></table></figure>

<p>同上面的一节</p>
<h4 id="2-3-Data-visualization"><a href="#2-3-Data-visualization" class="headerlink" title="2.3 Data visualization"></a>2.3 Data visualization</h4><h5 id="2-3-1-介绍"><a href="#2-3-1-介绍" class="headerlink" title="2.3.1 介绍"></a>2.3.1 介绍</h5><p>single-cell RNA-seq 的一个重要方面是控制批处理影响。批处理效果是在处理过程中添加到样品中的由人为产生的影响。例如，如果在不同的实验室或在同一实验室的不同日子里准备了两组样本，那么我们可能会观察到在一起处理的样本之间有更大的相似性。在最坏的情况下，批处理可能被误认为是真正的生物变异。Tung数据使我们能够以可控的方式探索这些问题，因为这些样本的处理方式的一些重要方面已经被记录下来。理想情况下，我们希望看到来自同一分组的批次，以及对应于每个个体的不同组。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">library</span><span class="params">(SingleCellExperiment)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(scater)</span></span></span><br><span class="line"><span class="function"><span class="title">options</span><span class="params">(stringsAsFactors = FALSE)</span></span></span><br><span class="line">umi &lt;- readRDS(<span class="string">"tung/umi.rds"</span>)</span><br><span class="line">umi<span class="selector-class">.qc</span> &lt;- umi[rowData(umi)<span class="variable">$use</span>, colData(umi)<span class="variable">$use</span>]</span><br><span class="line">endog_genes &lt;- !rowData(umi.qc)<span class="variable">$is_feature_control</span></span><br></pre></td></tr></table></figure>

<h5 id="2-3-2-PCA-Plot"><a href="#2-3-2-PCA-Plot" class="headerlink" title="2.3.2 PCA Plot"></a>2.3.2 PCA Plot</h5><p>对数据进行概述的最简单方法是使用主成分分析转换它，然后将前两个主要组件可视化。<br>主成分分析（PCA）是一个统计过程，它使用一个转换将一组观察转换成一组称为主成分（pc）的线性无关变量的值。主成分的数量小于或等于原始变量的数量。<br>从数学上说，pc对应于协方差矩阵的特征向量。特征向量是按特征值排序的所以第一个主成分在数据中尽可能多地反映了数据的可变性，而每一个后续的组件在约束条件下都有最高的方差它与前面的组件是正交的<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/pca.png" alt="image"></p>
<h6 id="2-3-2-1-Before-QC"><a href="#2-3-2-1-Before-QC" class="headerlink" title="2.3.2.1 Before QC"></a>2.3.2.1 Before QC</h6><p>在做log转换之前：</p>
<figure class="highlight makefile"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotPCA（</span><br><span class="line">umi[endog_genens,],</span><br><span class="line">exprs_values=<span class="string">"counts"</span>,</span><br><span class="line">colour_by=<span class="string">"batch"</span>,</span><br><span class="line">size_by=<span class="string">"total_features"</span>,</span><br><span class="line">shape_by=<span class="string">"individiual"</span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-pca-before-qc1-1.png" alt="image"><br>做log转换之后：</p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    umi[endog_genes, ],</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"logcounts_raw"</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-pca-before-qc2-1.png" alt="image"><br>很明显，对数转换对我们的数据是有利的——它减少了第一个主成分的方差，并且已经分离了一些生物效应。此外，它使表达式的分布更加正常。在接下来的分析和章节中，我们将在默认情况下使用日志转换的原始计数。<br>但是，请注意，仅仅一个对数转换不足以解释单元之间的不同技术因素（例如，排序深度）。因此，请不要使用logcounts_raw来进行下游分析，而是将其作为最小的合适的数据使用，它不仅是对数转换的，而且还被库的大小（如CPM normalisation）规范化。在本课程中，我们仅为演示目的使用logcounts-raw格式！</p>
<h6 id="2-3-2-2-After-QC"><a href="#2-3-2-2-After-QC" class="headerlink" title="2.3.2.2 After QC"></a>2.3.2.2 After QC</h6><p>plotPCA(<br>    umi.qc[endog_genes, ],<br>    exprs_values = “logcounts_raw”,<br>    colour_by = “batch”,<br>    size_by = “total_features”,<br>    shape_by = “individual”<br>)
<img src="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-pca-after-qc-1.png" alt="image"><br>比较上面两图，很明显，在质量控制之后，NA19098.r2中的细胞不再形成一组异常值。<br>默认情况下，scater使用最多的500个最可变的基因来计算PCA。这可以通过改变ntop参数进行调整</p>
<figure class="highlight ceylon"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(umi.qc[endog<span class="number">_</span>genes,],exprs<span class="number">_</span>values=<span class="string">"logcounts_raw"</span>,colour<span class="number">_</span><span class="meta">by</span>=<span class="string">"batch"</span>,size<span class="number">_</span><span class="meta">by</span>=<span class="string">"total_features"</span>,shape<span class="number">_</span><span class="meta">by</span>=<span class="string">"indivaidual"</span>,ntop=<span class="number">50</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-pca-after-qc-exercise1-2-1.png" alt="image"></p>
<h5 id="2-3-3-tSNE-map"><a href="#2-3-3-tSNE-map" class="headerlink" title="2.3.3 tSNE map"></a>2.3.3 tSNE map</h5><p>替代PCA的将scRNASeq数据可视化另一种方法是tSNE图。tSNE（t-分布式随机邻居嵌入）将维度减少（例如PCA）与在邻近的网络上随机漫步，将高维数据（即我们的14,214维表达式矩阵）映射到二维空间，同时保持细胞之间的距离。与PCA相比，tSNE是一种随机算法，这意味着在同一数据集上多次运行该方法将导致不同的绘图块与PCA相比，tSNE是一种随机算法，这意味着在同一数据集上多次运行该方法将导致不同的绘图块。由于该算法的非线性和随机性，tSNE更难直观地解释tSNE。为了确保重现性，我们在下面的代码中修复了随机数字生成器的“seed”，这样我们就可以得到相同的图。</p>
<h6 id="2-3-3-1-Before-QC"><a href="#2-3-3-1-Before-QC" class="headerlink" title="2.3.3.1 Before QC"></a>2.3.3.1 Before QC</h6><figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">plotTSNE(</span><br><span class="line">    umi[endog_genes, ],</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"logcounts_raw"</span>,</span><br><span class="line">    <span class="attr">perplexity</span> = <span class="number">130</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"individual"</span>,</span><br><span class="line">    <span class="attr">rand_seed</span> = <span class="number">123456</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-tsne-before-qc-1.png" alt="image"></p>
<h6 id="2-3-3-2-After-QC"><a href="#2-3-3-2-After-QC" class="headerlink" title="2.3.3.2 After QC"></a>2.3.3.2 After QC</h6><figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">plotTSNE(</span><br><span class="line">    umi.qc[endog_genes, ],</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"logcounts_raw"</span>,</span><br><span class="line">    <span class="attr">perplexity</span> = <span class="number">130</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"individual"</span>,</span><br><span class="line">    <span class="attr">rand_seed</span> = <span class="number">123456</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/16-exprs-overview_files/figure-html/expr-overview-tsne-after-qc-1.png" alt="image"></p>
<p>解释PCA和tSNE Plot通常具有挑战性，由于它们的随机和非线性特性，他们不太直观。然而，在这种情况下，很明显，它们提供了类似的数据图像。比较上面两图，同样清楚的是来自NA19098.r2的样本在QC过滤之后不再是异常值。<br>此外，tSNE要求你提供perplexity值，它反映了用来建立近邻网络的邻居的数量;高值创造了一个密集的网络，将细胞聚集在一起，而低的值使网络更加稀疏，使得细胞群相互分离。</p>
<h4 id="2-4-Data-visualization（Reads）"><a href="#2-4-Data-visualization（Reads）" class="headerlink" title="2.4 Data visualization（Reads）"></a>2.4 Data visualization（Reads）</h4><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">library</span><span class="params">(scater)</span></span></span><br><span class="line"><span class="function"><span class="title">options</span><span class="params">(stringsAsFactors = FALSE)</span></span></span><br><span class="line">reads &lt;- readRDS(<span class="string">"tung/reads.rds"</span>)</span><br><span class="line">reads<span class="selector-class">.qc</span> &lt;- reads[rowData(reads)<span class="variable">$use</span>, colData(reads)<span class="variable">$use</span>]</span><br><span class="line">endog_genes &lt;- !rowData(reads.qc)<span class="variable">$is_feature_control</span></span><br></pre></td></tr></table></figure>

<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    reads[endog_genes, ],</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"counts"</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-pca-before-qc-reads1-1.png" alt="image"></p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    reads[endog_genes, ],</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"logcounts_raw"</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-pca-before-qc-reads2-1.png" alt="image"></p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    reads.qc[endog_genes, ],</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"logcounts_raw"</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-pca-after-qc-reads-1.png" alt="image"></p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">plotTSNE(</span><br><span class="line">    reads[endog_genes, ],</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"logcounts_raw"</span>,</span><br><span class="line">    <span class="attr">perplexity</span> = <span class="number">130</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"individual"</span>,</span><br><span class="line">    <span class="attr">rand_seed</span> = <span class="number">123456</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-tsne-before-qc-reads-1.png" alt="image"></p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">plotTSNE(</span><br><span class="line">    reads.qc[endog_genes, ],</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"logcounts_raw"</span>,</span><br><span class="line">    <span class="attr">perplexity</span> = <span class="number">130</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"individual"</span>,</span><br><span class="line">    <span class="attr">rand_seed</span> = <span class="number">123456</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-tsne-after-qc-reads-1.png" alt="image"><br>tSNE map of the tung data<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-tsne-after-qc-exercise2-1-1.png" alt="image"><br>tSNE map of the tung data (perplexity = 10)<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/17-exprs-overview-reads_files/figure-html/expr-overview-tsne-after-qc-exercise2-2-1.png" alt="image"><br>tSNE map of the tung data (perplexity = 200)</p>
<h4 id="2-5-Identifying-confounding-factors"><a href="#2-5-Identifying-confounding-factors" class="headerlink" title="2.5 Identifying confounding factors"></a>2.5 Identifying confounding factors</h4><h5 id="2-5-1-Introduction"><a href="#2-5-1-Introduction" class="headerlink" title="2.5.1 Introduction"></a>2.5.1 Introduction</h5><p>在前几章中，我们考虑了批处理效果，在这一章中，我们将继续探索如何识别和移除实验中的影响。我们将继续使用scater包，因为它提供了一组专门用于测试和解释变量的质量控制的方法。此外，我们将继续使用前一章中使用的Blischak数据。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">library</span><span class="params">(scater, quietly = TRUE)</span></span></span><br><span class="line"><span class="function"><span class="title">options</span><span class="params">(stringsAsFactors = FALSE)</span></span></span><br><span class="line">umi &lt;- readRDS(<span class="string">"tung/umi.rds"</span>)</span><br><span class="line">umi<span class="selector-class">.qc</span> &lt;- umi[rowData(umi)<span class="variable">$use</span>, colData(umi)<span class="variable">$use</span>]</span><br><span class="line">endog_genes &lt;- !rowData(umi.qc)<span class="variable">$is_feature_control</span></span><br></pre></td></tr></table></figure>

<p>umi.qc数据集包含经过细胞和基因过滤的数据。</p>
<h5 id="2-5-2-Correlations-with-PCs"><a href="#2-5-2-Correlations-with-PCs" class="headerlink" title="2.5.2 Correlations with PCs"></a>2.5.2 Correlations with PCs</h5><p>我们再回顾一下PCA 图以及QCed数据集</p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    umi.qc[endog_genes, ],</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"logcounts_raw"</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/18-confounders_files/figure-html/confound-pca-1.png" alt="image"><br>scater允许一个识别与实验和QC变量相关的主要成分<br>让我们来测试一些变量是否与任何pc相关联。</p>
<h6 id="2-5-2-1-Detected-genes"><a href="#2-5-2-1-Detected-genes" class="headerlink" title="2.5.2.1 Detected genes"></a>2.5.2.1 Detected genes</h6><figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotQC(</span><br><span class="line">    umi.qc[endog_genes, ],</span><br><span class="line">   <span class="built_in"> type </span>= <span class="string">"find-pcs"</span>,</span><br><span class="line">    exprs_values = <span class="string">"logcounts_raw"</span>,</span><br><span class="line">    variable = <span class="string">"total_features"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/18-confounders_files/figure-html/confound-find-pcs-total-features-1.png" alt="image"><br>事实上，我们可以看到PC1几乎可以完全由检测到的基因的数量来解释。事实上，在上面的PCA图上也可以看出。<br>scater也可以对每一个变量计算mariginal R2，当拟合线性模型时，每个基因的表达值都与这个变量相对应，并显示出基因边缘的密度图。</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">plotQC(</span><br><span class="line">    umi.qc[endog_genes, ],</span><br><span class="line">   <span class="built_in"> type </span>= <span class="string">"expl"</span>,</span><br><span class="line">    exprs_values = <span class="string">"logcounts_raw"</span>,</span><br><span class="line">    variables = c(</span><br><span class="line">        <span class="string">"total_features"</span>,</span><br><span class="line">        <span class="string">"total_counts"</span>,</span><br><span class="line">        <span class="string">"batch"</span>,</span><br><span class="line">        <span class="string">"individual"</span>,</span><br><span class="line">        <span class="string">"pct_counts_ERCC"</span>,</span><br><span class="line">        <span class="string">"pct_counts_MT"</span></span><br><span class="line">    )</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/18-confounders_files/figure-html/confound-find-expl-vars-1.png" alt="image"><br>这一分析表明，检测到的基因（再次）和测序深度（计数的数量）的数量对许多基因具有重要的解释力，因此，这些变量是在标准化过程或下游的统计模型中进行调整的候选目标哦。ERCCs的表达似乎也是一个重要的解释变量，上面的一个显著特征是，批处理比individual更能解释。</p>
<h4 id="2-6-Identifying-confounding-factors-Reads"><a href="#2-6-Identifying-confounding-factors-Reads" class="headerlink" title="2.6 Identifying confounding factors (Reads)"></a>2.6 Identifying confounding factors (Reads)</h4><p>处理同上</p>
<h4 id="2-7-Normalization-theory"><a href="#2-7-Normalization-theory" class="headerlink" title="2.7 Normalization theory"></a>2.7 Normalization theory</h4><h5 id="2-7-1-介绍"><a href="#2-7-1-介绍" class="headerlink" title="2.7.1 介绍"></a>2.7.1 介绍</h5><p>在前一章中，我们确定了重要的混淆因素和解释变量。scater允许在随后的统计模型中解释这些变量，或者在需要的情况下使用normaliseExprs（）对其进行条件去除。这可以通过为normaliseExprs（）提供一个设计矩阵来实现。</p>
<h5 id="2-7-2-文库大小"><a href="#2-7-2-文库大小" class="headerlink" title="2.7.2 文库大小"></a>2.7.2 文库大小</h5><p>由于scRNA-seq数据通常是在高度多路复用的平台上进行测序的，所以库的大小各不相同，从每个单元中获得的总读数可能会有很大的不同。一些(如量化方法当确定基因表达估计时（如Cufflinks、RSEM）将库的大小合并起来，因此处理这些软件输出的数据时不需要这种标准化。然而，如果使用另一种量化方法，那么必须通过将表达式矩阵的每一列乘以或除以一个“标准化因子”来修正库的大小，这是相对于其他cell的库大小的估计。许多修正库大小的方法已经被开发出来用于bulkRNA-seq，并且可以同样应用于scRNA-seq（例如。UQ，SF，CPM，RPKM，FPKM，TPM）。</p>
<h5 id="2-7-3-Normalisations"><a href="#2-7-3-Normalisations" class="headerlink" title="2.7.3 Normalisations"></a>2.7.3 Normalisations</h5><h6 id="2-7-3-1-CPM"><a href="#2-7-3-1-CPM" class="headerlink" title="2.7.3.1 CPM"></a>2.7.3.1 CPM</h6><p>使这些数据规范化的最简单方法是将每一列除以它的总数，然后乘以100 000，将它转换为每百万（CPM）计数。请注意，为了纠正总细胞RNA的含量，应该将spike排除在总表达的计算中，因此我们只使用内源性基因。</p>
<figure class="highlight actionscript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">calc_cpm &lt;-</span><br><span class="line"><span class="function"><span class="keyword">function</span> <span class="params">(expr_mat, spikes = NULL)</span> </span></span><br><span class="line"><span class="function"></span>&#123;</span><br><span class="line">    norm_factor &lt;- colSums(expr_mat[-spikes, ])</span><br><span class="line">    <span class="keyword">return</span>(t(t(expr_mat)/norm_factor)) * <span class="number">10</span>^<span class="number">6</span></span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>CPM的一个潜在缺点是，如果你的样本包含了在细胞中高度表达和不同表达的基因。在这种情况下，细胞内的总分子可能依赖于这些基因在细胞中是否开启，而由总分子正常化可能会隐藏这些基因的差异表达，或者为剩余的基因错误地创造出不同的表达。注意，RPKM、FPKM和TPM是CPM的变体，它根据各自的基因/转录本的长度进一步调整计数。</p>
<h6 id="2-7-3-2-RLE（SF）"><a href="#2-7-3-2-RLE（SF）" class="headerlink" title="2.7.3.2 RLE（SF）"></a>2.7.3.2 RLE（SF）</h6><p>size factr（SF）是由DESeq提出并推广的。首先计算出所有细胞中每个基因的几何平均值。每个细胞的大小因子是表达与基因的几何平均比率的基因的中值。这种方法的一个缺点是，由于它使用的是几何平均值，所以在所有的细胞中，只有非零表达值的基因可以被用于计算，这使得对于大型低深度的scRNASeq实验来说是不可取的。edgeR&amp;scater称该方法为“relative log expression”。</p>
<figure class="highlight lua"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">calc_sf &lt;-</span><br><span class="line"><span class="function"><span class="keyword">function</span> <span class="params">(expr_mat, spikes = NULL)</span></span> </span><br><span class="line">&#123;</span><br><span class="line">    geomeans &lt;- <span class="built_in">exp</span>(rowMeans(<span class="built_in">log</span>(expr_mat[-spikes, ])))</span><br><span class="line">    SF &lt;- <span class="function"><span class="keyword">function</span><span class="params">(cnts)</span></span> &#123;</span><br><span class="line">        median((cnts/geomeans)[(is.finite(geomeans) &amp; geomeans &gt; </span><br><span class="line">            <span class="number">0</span>)])</span><br><span class="line">    &#125;</span><br><span class="line">    norm_factor &lt;- apply(expr_mat[-spikes, ], <span class="number">2</span>, SF)</span><br><span class="line">    <span class="keyword">return</span>(t(t(expr_mat)/norm_factor))</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h6 id="2-7-3-3-UQ"><a href="#2-7-3-3-UQ" class="headerlink" title="2.7.3.3 UQ"></a>2.7.3.3 UQ</h6><p>上四分位数（upperquartile，UQ）是由（Bullard et al.2010）提出的。这里每一列都除以每个库的75%的计数。通常，计算出的分位数是由单元间的中值缩放的，以保持表达式的绝对水平相对一致。这种方法的一个缺点是，对于低深度的scRNASeq实验，大量未被发现的基因可能导致75%的量子数为零（或接近它）。这个限制可以通过概括这个想法和使用一个更高的分位数来克服（例如。99%的分位数是scater的默认值），或者在计算75%的分位数之前排除0。</p>
<figure class="highlight actionscript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">calc_uq &lt;-</span><br><span class="line"><span class="function"><span class="keyword">function</span> <span class="params">(expr_mat, spikes = NULL)</span> </span></span><br><span class="line"><span class="function"></span>&#123;</span><br><span class="line">    UQ &lt;- <span class="function"><span class="keyword">function</span><span class="params">(x)</span> </span>&#123;</span><br><span class="line">        quantile(x[x &gt; <span class="number">0</span>], <span class="number">0.75</span>)</span><br><span class="line">    &#125;</span><br><span class="line">    uq &lt;- unlist(apply(expr_mat[-spikes, ], <span class="number">2</span>, UQ))</span><br><span class="line">    norm_factor &lt;- uq/median(uq)</span><br><span class="line">    <span class="keyword">return</span>(t(t(expr_mat)/norm_factor))</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h6 id="2-7-3-4-TMM"><a href="#2-7-3-4-TMM" class="headerlink" title="2.7.3.4 TMM"></a>2.7.3.4 TMM</h6><p>另一种叫做TMM的方法是由（罗宾逊和Oshlack 2010）提出的m值（对参考）的加权平均化平均值。问题的m值是基因的log2在单个细胞之间的变化。一个cell用作参考，然后计算每个单元格的m值与这个引用相比较。然后，通过删除顶部和底部30%的值来调整这些值，并且通过加权它们来计算剩余值的平均值，以考虑对数尺度对方差的影响。每个非引用cell乘以计算因子。这种方法的两个潜在问题是在修剪后留下的非零基因，以及大多数基因没有差异表达的假设。</p>
<h6 id="2-7-3-5-scran"><a href="#2-7-3-5-scran" class="headerlink" title="2.7.3.5 scran"></a>2.7.3.5 scran</h6><p>scran包实现了针对single-cell数据（l.Lun、巴赫和Marioni 2016）的CPM的个性化。简单地说，这个方法解决了单个cell的大量零值的问题，通过将cell计算一个标准化因子（类似于CPM）来计算每个pool的总和。由于一个cell存在于许多不同的pool中，所以可以使用线性代数从特定的因子的集合中分离出特定于cell的因素。</p>
<h6 id="2-7-3-6-Downsampling"><a href="#2-7-3-6-Downsampling" class="headerlink" title="2.7.3.6 Downsampling"></a>2.7.3.6 Downsampling</h6><p>修正库大小的最后一种方法是对表达式矩阵进行Downsampling，这样每个cell的总molecules大约是相同的。这种方法的好处是，零值将被向下的采样引入，从而消除由于检测到的基因数量不同而产生的任何偏差。然而，主要的缺点是过程不是确定的，所以每次向下抽样运行时，结果表达式矩阵略有不同。因此，经常分析必须在多个向下取样上运行，以确保结果是健壮的。</p>
<figure class="highlight scilab"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">Down_Sample_Matrix &lt;-</span><br><span class="line"><span class="function"><span class="keyword">function</span> <span class="params">(expr_mat)</span> </span></span><br><span class="line">&#123;</span><br><span class="line">    min_lib_size &lt;- <span class="built_in">min</span>(colSums(expr_mat))</span><br><span class="line">    down_sample &lt;- <span class="function"><span class="keyword">function</span><span class="params">(x)</span> &#123;</span></span><br><span class="line">        prob &lt;- min_lib_size/<span class="built_in">sum</span>(x)</span><br><span class="line">        <span class="keyword">return</span>(unlist(lapply(x, <span class="function"><span class="keyword">function</span><span class="params">(y)</span> &#123;</span></span><br><span class="line">            rbinom(<span class="number">1</span>, y, prob)</span><br><span class="line">        &#125;)))</span><br><span class="line">    &#125;</span><br><span class="line">    down_sampled_mat &lt;- apply(expr_mat, <span class="number">2</span>, down_sample)</span><br><span class="line">    <span class="keyword">return</span>(down_sampled_mat)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h5 id="2-7-4-Effectiveness"><a href="#2-7-4-Effectiveness" class="headerlink" title="2.7.4 Effectiveness"></a>2.7.4 Effectiveness</h5><p>为了比较不同的标准化方法的效率，我们将使用PCA图的视觉检查，通过scater的plotRLE（）函数来计算细胞相关的相对对数表达式。也就是说，有许多（很少）的细胞比大多数基因的中位表达要高（更低），从而在细胞中产生一个正的（负的）RLE，而标准化细胞的RLE接近于零。</p>
<figure class="highlight maxima"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">calc_cell_RLE &lt;-</span><br><span class="line">function (expr_mat, spikes = NULL) </span><br><span class="line">&#123;</span><br><span class="line">    RLE_gene &lt;- function(x) &#123;</span><br><span class="line">        <span class="keyword">if</span> (<span class="built_in">median</span>(unlist(x)) &gt; <span class="number">0</span>) &#123;</span><br><span class="line">            <span class="built_in">log</span>((x + <span class="number">1</span>)/(<span class="built_in">median</span>(unlist(x)) + <span class="number">1</span>))/<span class="built_in">log</span>(<span class="number">2</span>)</span><br><span class="line">        &#125;</span><br><span class="line">        <span class="keyword">else</span> &#123;</span><br><span class="line">            rep(NA, times = <span class="built_in">length</span>(x))</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">if</span> (!<span class="built_in">is</span>.null(spikes)) &#123;</span><br><span class="line">        RLE_matrix &lt;- t(<span class="built_in">apply</span>(expr_mat[-spikes, ], <span class="number">1</span>, RLE_gene))</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">else</span> &#123;</span><br><span class="line">        RLE_matrix &lt;- t(<span class="built_in">apply</span>(expr_mat, <span class="number">1</span>, RLE_gene))</span><br><span class="line">    &#125;</span><br><span class="line">    cell_RLE &lt;- <span class="built_in">apply</span>(RLE_matrix, <span class="number">2</span>, <span class="built_in">median</span>, na.rm = T)</span><br><span class="line">    <span class="built_in">return</span>(cell_RLE)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p>请注意，RLE、TMM和UQ大小因子方法是为bulkRNA-seq数据开发的，根据实验环境的不同，可能不适合single-cell RNA-seq数据，因为它们的基本假设可能会受到问题的违反。<br>注意，scater是edgeR的calcNormFactors函数的包装器，它实现了几个库大小的标准化方法，使我们可以很容易地将这些方法应用到我们的数据中。</p>
<h4 id="2-8-Nomalization-practice（UMI）"><a href="#2-8-Nomalization-practice（UMI）" class="headerlink" title="2.8 Nomalization practice（UMI）"></a>2.8 Nomalization practice（UMI）</h4><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">install.packages(<span class="string">"devtools"</span>)</span><br><span class="line">devtools::install_github(<span class="string">"hemberg-lab/scRNA.seq.funcs"</span>)</span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(scRNA.seq.funcs)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(scater)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(scran)</span></span></span><br><span class="line"><span class="function"><span class="title">options</span><span class="params">(stringsAsFactors = FALSE)</span></span></span><br><span class="line">set.seed(<span class="number">1234567</span>)</span><br><span class="line">umi &lt;- readRDS(<span class="string">"tung/umi.rds"</span>)</span><br><span class="line">umi<span class="selector-class">.qc</span> &lt;- umi[rowData(umi)<span class="variable">$use</span>, colData(umi)<span class="variable">$use</span>]</span><br><span class="line">endog_genes &lt;- !rowData(umi.qc)<span class="variable">$is_feature_control</span></span><br></pre></td></tr></table></figure>

<h5 id="2-8-1-Raw"><a href="#2-8-1-Raw" class="headerlink" title="2.8.1 Raw"></a>2.8.1 Raw</h5><figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    umi.qc[endog_genes, ],</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"logcounts_raw"</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-raw-1.png" alt="image"></p>
<h5 id="2-8-2-CPM"><a href="#2-8-2-CPM" class="headerlink" title="2.8.2 CPM"></a>2.8.2 CPM</h5><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">logcounts</span><span class="params">(umi.qc)</span></span> &lt;- log2(calculateCPM(umi<span class="selector-class">.qc</span>, use<span class="selector-class">.size</span><span class="selector-class">.factors</span> = FALSE) + <span class="number">1</span>)</span><br><span class="line">plotPCA(</span><br><span class="line">    umi<span class="selector-class">.qc</span>[endog_genes, ],</span><br><span class="line">    colour_by = <span class="string">"batch"</span>,</span><br><span class="line">    size_by = <span class="string">"total_features"</span>,</span><br><span class="line">    shape_by = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-cpm-1.png" alt="image"></p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotRLE(</span><br><span class="line">    umi.qc[endog_genes, ], </span><br><span class="line">    exprs_mats = list(<span class="name">Raw</span> = <span class="string">"logcounts_raw"</span>, CPM = <span class="string">"logcounts"</span>),</span><br><span class="line">    exprs_logged = c(<span class="name">TRUE</span>, TRUE),</span><br><span class="line">    colour_by = <span class="string">"batch"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-cpm-1.png" alt="image"></p>
<h5 id="2-8-3-Size-factor（RLE）"><a href="#2-8-3-Size-factor（RLE）" class="headerlink" title="2.8.3 Size-factor（RLE）"></a>2.8.3 Size-factor（RLE）</h5><figure class="highlight oxygene"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">umi.qc &lt;- normaliseExprs(</span><br><span class="line">    umi.qc,</span><br><span class="line">    <span class="function"><span class="keyword">method</span> = "<span class="title">RLE</span>", </span></span><br><span class="line"><span class="function">    <span class="title">feature_set</span> = <span class="title">endog_genes</span>,</span></span><br><span class="line"><span class="function">    <span class="title">return_log</span> = <span class="title">TRUE</span>,</span></span><br><span class="line"><span class="function">    <span class="title">return_norm_as_exprs</span> = <span class="title">TRUE</span></span></span><br><span class="line"><span class="function">)</span></span><br></pre></td></tr></table></figure>

<figure class="highlight ceylon"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    umi.qc[endog<span class="number">_</span>genes, ],</span><br><span class="line">    colour<span class="number">_</span><span class="meta">by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    size<span class="number">_</span><span class="meta">by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    shape<span class="number">_</span><span class="meta">by</span> = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-rle-1.png" alt="image"></p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotRLE(</span><br><span class="line">    umi.qc[endog_genes, ], </span><br><span class="line">    exprs_mats = list(<span class="name">Raw</span> = <span class="string">"logcounts_raw"</span>, RLE = <span class="string">"logcounts"</span>),</span><br><span class="line">    exprs_logged = c(<span class="name">TRUE</span>, TRUE),</span><br><span class="line">    colour_by = <span class="string">"batch"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-rle-1.png" alt="image"></p>
<h5 id="2-8-4-Upperquantile"><a href="#2-8-4-Upperquantile" class="headerlink" title="2.8.4 Upperquantile"></a>2.8.4 Upperquantile</h5><figure class="highlight oxygene"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">umi.qc &lt;- normaliseExprs(</span><br><span class="line">    umi.qc,</span><br><span class="line">    <span class="function"><span class="keyword">method</span> = "<span class="title">upperquartile</span>", </span></span><br><span class="line"><span class="function">    <span class="title">feature_set</span> = <span class="title">endog_genes</span>,</span></span><br><span class="line"><span class="function">    <span class="title">p</span> = 0.99,</span></span><br><span class="line"><span class="function">    <span class="title">return_log</span> = <span class="title">TRUE</span>,</span></span><br><span class="line"><span class="function">    <span class="title">return_norm_as_exprs</span> = <span class="title">TRUE</span></span></span><br><span class="line"><span class="function">)</span></span><br></pre></td></tr></table></figure>

<figure class="highlight ceylon"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    umi.qc[endog<span class="number">_</span>genes, ],</span><br><span class="line">    colour<span class="number">_</span><span class="meta">by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    size<span class="number">_</span><span class="meta">by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    shape<span class="number">_</span><span class="meta">by</span> = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-uq-1.png" alt="image"></p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotRLE(</span><br><span class="line">    umi.qc[endog_genes, ], </span><br><span class="line">    exprs_mats = list(<span class="name">Raw</span> = <span class="string">"logcounts_raw"</span>, UQ = <span class="string">"logcounts"</span>),</span><br><span class="line">    exprs_logged = c(<span class="name">TRUE</span>, TRUE),</span><br><span class="line">    colour_by = <span class="string">"batch"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-uq-1.png" alt="image"></p>
<h5 id="TMM"><a href="#TMM" class="headerlink" title="TMM"></a>TMM</h5><figure class="highlight oxygene"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">umi.qc &lt;- normaliseExprs(</span><br><span class="line">    umi.qc,</span><br><span class="line">    <span class="function"><span class="keyword">method</span> = "<span class="title">TMM</span>",</span></span><br><span class="line"><span class="function">    <span class="title">feature_set</span> = <span class="title">endog_genes</span>,</span></span><br><span class="line"><span class="function">    <span class="title">return_log</span> = <span class="title">TRUE</span>,</span></span><br><span class="line"><span class="function">    <span class="title">return_norm_as_exprs</span> = <span class="title">TRUE</span></span></span><br><span class="line"><span class="function">)</span></span><br></pre></td></tr></table></figure>

<figure class="highlight ceylon"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    umi.qc[endog<span class="number">_</span>genes, ],</span><br><span class="line">    colour<span class="number">_</span><span class="meta">by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    size<span class="number">_</span><span class="meta">by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    shape<span class="number">_</span><span class="meta">by</span> = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-tmm-1.png" alt="image"></p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotRLE(</span><br><span class="line">    umi.qc[endog_genes, ], </span><br><span class="line">    exprs_mats = list(<span class="name">Raw</span> = <span class="string">"logcounts_raw"</span>, TMM = <span class="string">"logcounts"</span>),</span><br><span class="line">    exprs_logged = c(<span class="name">TRUE</span>, TRUE),</span><br><span class="line">    colour_by = <span class="string">"batch"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-tmm-1.png" alt="image"></p>
<h5 id="2-8-6-scran"><a href="#2-8-6-scran" class="headerlink" title="2.8.6 scran"></a>2.8.6 scran</h5><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">qclust &lt;- quickCluster(umi<span class="selector-class">.qc</span>, min<span class="selector-class">.size</span> = <span class="number">30</span>)</span><br><span class="line">umi<span class="selector-class">.qc</span> &lt;- computeSumFactors(umi<span class="selector-class">.qc</span>, sizes = <span class="number">15</span>, clusters = qclust)</span><br><span class="line">umi<span class="selector-class">.qc</span> &lt;- normalize(umi.qc)</span><br></pre></td></tr></table></figure>

<figure class="highlight ceylon"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    umi.qc[endog<span class="number">_</span>genes, ],</span><br><span class="line">    colour<span class="number">_</span><span class="meta">by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    size<span class="number">_</span><span class="meta">by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    shape<span class="number">_</span><span class="meta">by</span> = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-lsf-1.png" alt="image"></p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotRLE(</span><br><span class="line">    umi.qc[endog_genes, ], </span><br><span class="line">    exprs_mats = list(<span class="name">Raw</span> = <span class="string">"logcounts_raw"</span>, scran = <span class="string">"logcounts"</span>),</span><br><span class="line">    exprs_logged = c(<span class="name">TRUE</span>, TRUE),</span><br><span class="line">    colour_by = <span class="string">"batch"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-scran-1.png" alt="image"><br>scran有时会计算出负或零大小的因素。这些将完全扭曲标准化的表达式矩阵。我们可以检查scran的大小因素，如：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">summary(<span class="name">sizeFactors</span>(<span class="name">umi</span>.qc))</span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">##    Min. <span class="number">1</span>st Qu.  Median    Mean <span class="number">3</span>rd Qu.    Max. </span><br><span class="line">##  <span class="number">0.4646</span>  <span class="number">0.7768</span>  <span class="number">0.9562</span>  <span class="number">1.0000</span>  <span class="number">1.1444</span>  <span class="number">3.4348</span></span><br></pre></td></tr></table></figure>

<p>对于这个数据集，所有大小的因素都是合理的，所以我们已经完成了。如果您发现scran已经计算出了负大小的因素，那么尝试增加cluster和pool的大小，直到它们都是正的。</p>
<h5 id="2-8-7-Downsampling"><a href="#2-8-7-Downsampling" class="headerlink" title="2.8.7 Downsampling"></a>2.8.7 Downsampling</h5><figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">logcounts(<span class="name">umi</span>.qc) &lt;- log2(<span class="name">Down_Sample_Matrix</span>(<span class="name">counts</span>(<span class="name">umi</span>.qc)) + <span class="number">1</span>)</span><br><span class="line">plotPCA(</span><br><span class="line">    umi.qc[endog_genes, ],</span><br><span class="line">    colour_by = <span class="string">"batch"</span>,</span><br><span class="line">    size_by = <span class="string">"total_features"</span>,</span><br><span class="line">    shape_by = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-pca-downsample-1.png" alt="image"></p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">plotRLE(</span><br><span class="line">    umi.qc[endog_genes, ], </span><br><span class="line">    exprs_mats = list(<span class="name">Raw</span> = <span class="string">"logcounts_raw"</span>, DownSample = <span class="string">"logcounts"</span>),</span><br><span class="line">    exprs_logged = c(<span class="name">TRUE</span>, TRUE),</span><br><span class="line">    colour_by = <span class="string">"batch"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/20-exprs-norm_files/figure-html/norm-ours-rle-downsample-1.png" alt="image"></p>
<h5 id="2-8-8-Normalisation-for-gene-transcript-length"><a href="#2-8-8-Normalisation-for-gene-transcript-length" class="headerlink" title="2.8.8 Normalisation for gene/transcript length"></a>2.8.8 Normalisation for gene/transcript length</h5><p>有些方法结合了库的大小和片段/基因长度的标准化，例如：</p>
<ul>
<li>RPKM：每千碱基的read（单端测序）</li>
<li>FPKM：每千碱基的Fragment（与RPKM相同但是是用于双端测序的数据，确保映射到同一片段的成对结束不会被计数两次）</li>
<li>TPM-每千碱基的Transcript（与RPKM相同，但是，标准化的顺序是颠倒的——长度优先，顺序深度第二）<br>这些方法不适用于我们的数据集，因为包含了UMI的文字记录的末尾被优先测序。此外，一般情况下，这些数据只能使用来自对齐的BAM文件的适当的量化软件，而不是从读取计数中计算出来，因为通常只有一部分的整个遗传/转录本是被测序的，而不是整个长度。如果有疑问，检查基因/转录长度和表达水平之间的关系。<br>然而，这里我们展示了如何使用scater来计算这些normalisations。首先，我们需要在kilobase中找到有效的转录长度。然而，我们的数据集只包含了基因id，因此我们将使用基因长度而不是转录本。scater使用biomaRt，它允许通过其他属性对基因进行注释：</li>
</ul>
<figure class="highlight vala"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br></pre></td><td class="code"><pre><span class="line">umi.qc &lt;- getBMFeatureAnnos(</span><br><span class="line">    umi.qc,</span><br><span class="line">    filters = <span class="string">"ensembl_gene_id"</span>, </span><br><span class="line">    attributes = c(</span><br><span class="line">        <span class="string">"ensembl_gene_id"</span>,</span><br><span class="line">        <span class="string">"hgnc_symbol"</span>,</span><br><span class="line">        <span class="string">"chromosome_name"</span>,</span><br><span class="line">        <span class="string">"start_position"</span>,</span><br><span class="line">        <span class="string">"end_position"</span></span><br><span class="line">    ), </span><br><span class="line">    feature_symbol = <span class="string">"hgnc_symbol"</span>,</span><br><span class="line">    feature_id = <span class="string">"ensembl_gene_id"</span>,</span><br><span class="line">    biomart = <span class="string">"ENSEMBL_MART_ENSEMBL"</span>, </span><br><span class="line">    dataset = <span class="string">"hsapiens_gene_ensembl"</span>,</span><br><span class="line">    host = <span class="string">"www.ensembl.org"</span></span><br><span class="line">)</span><br><span class="line"></span><br><span class="line"><span class="meta"># If you have mouse data, change the arguments based on this example:</span></span><br><span class="line"><span class="meta"># getBMFeatureAnnos(</span></span><br><span class="line"><span class="meta">#     object,</span></span><br><span class="line"><span class="meta">#     filters = "ensembl_transcript_id",</span></span><br><span class="line"><span class="meta">#     attributes = c(</span></span><br><span class="line"><span class="meta">#         "ensembl_transcript_id",</span></span><br><span class="line"><span class="meta">#         "ensembl_gene_id", </span></span><br><span class="line"><span class="meta">#         "mgi_symbol",</span></span><br><span class="line"><span class="meta">#         "chromosome_name",</span></span><br><span class="line"><span class="meta">#         "transcript_biotype",</span></span><br><span class="line"><span class="meta">#         "transcript_start",</span></span><br><span class="line"><span class="meta">#         "transcript_end",</span></span><br><span class="line"><span class="meta">#         "transcript_count"</span></span><br><span class="line"><span class="meta">#     ),</span></span><br><span class="line"><span class="meta">#     feature_symbol = "mgi_symbol",</span></span><br><span class="line"><span class="meta">#     feature_id = "ensembl_gene_id",</span></span><br><span class="line"><span class="meta">#     biomart = "ENSEMBL_MART_ENSEMBL",</span></span><br><span class="line"><span class="meta">#     dataset = "mmusculus_gene_ensembl",</span></span><br><span class="line"><span class="meta">#     host = "www.ensembl.org"</span></span><br><span class="line"><span class="meta"># )</span></span><br></pre></td></tr></table></figure>

<p>有些基因没有被注释，因此我们把它们过滤掉：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">umi<span class="selector-class">.qc</span><span class="selector-class">.ann</span> &lt;- umi<span class="selector-class">.qc</span>[!is.na(rowData(umi.qc)<span class="variable">$ensembl_gene_id</span>), ]</span><br></pre></td></tr></table></figure>

<p>现在我们通过使用endposition和startposition来计算kilobase的总基因长度：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">eff_length &lt;- </span><br><span class="line">    abs(rowData(umi<span class="selector-class">.qc</span><span class="selector-class">.ann</span>)<span class="variable">$end_position</span> - rowData(umi<span class="selector-class">.qc</span><span class="selector-class">.ann</span>)<span class="variable">$start_position</span>) / <span class="number">1000</span></span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plot(<span class="name">eff_length</span>, rowMeans(<span class="name">counts</span>(<span class="name">umi</span>.qc.ann)))</span><br></pre></td></tr></table></figure>

<p>基因长度和平均表达之间没有关系，所以FPKM s和TPM对于这个数据集是不合适的。但无论如何我们都会证明他们的。<br>这里要注意的是计算总基因长度而不是exon总长度。许多基因会包含大量的内含子，所以它们的长度将与我们计算的结果大不相同。请把我们的计算看作是近似值。如果你想使用总exon长度，请参考<a href="https://www.biostars.org/p/83901/" target="_blank" rel="noopener">这一页</a>。
现在，我们已经准备好了进行规范：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">tpm(<span class="name">umi</span>.qc.ann) &lt;- log2(<span class="name">calculateTPM</span>(<span class="name">umi</span>.qc.ann, eff_length) + <span class="number">1</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    umi.qc.ann,</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"tpm"</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">tpm(<span class="name">umi</span>.qc.ann) &lt;- log2(<span class="name">calculateFPKM</span>(<span class="name">umi</span>.qc.ann, eff_length) + <span class="number">1</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    umi.qc.ann,</span><br><span class="line">    <span class="attr">exprs_values</span> = <span class="string">"tpm"</span>,</span><br><span class="line">    <span class="attr">colour_by</span> = <span class="string">"batch"</span>,</span><br><span class="line">    <span class="attr">size_by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">    <span class="attr">shape_by</span> = <span class="string">"individual"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>PCA寻找细胞之间的差异。每个基因的细胞的基因长度是相同的，因此FPKM几乎与CPM图相同（它只是旋转），因为它首先执行CPM然后标准化基因长度。然而，TPM是不同的，因为它在执行CPM之前根据它们的长度对基因进行加权。</p>
<h4 id="2-10-Dealing-with-confounders"><a href="#2-10-Dealing-with-confounders" class="headerlink" title="2.10 Dealing with confounders"></a>2.10 Dealing with confounders</h4><h5 id="2-10-1-Introduction"><a href="#2-10-1-Introduction" class="headerlink" title="2.10.1 Introduction"></a>2.10.1 Introduction</h5><p>在上一章中，我们对库大小进行了规范化，有效地将其作为混淆器删除。现在我们将考虑从我们的数据中删除其他不太明确的混淆器。技术混杂因素（又称批次效应）可能源于试剂，分离方法，进行实验的实验室/实验者，甚至实验的实施日期/时间的差异。特别是技术混杂因素和批次效应的会计是一个涉及实验设计原则的大课题。在这里，我们讨论了在实验设计合适时可以考虑混杂因素的方法。</p>
<p>从根本上说，技术混杂因素的解释涉及识别并且理想地去除表达数据中与感兴趣的生物信号无关（即混淆）的变异源。存在各种方法，其中一些使用加入或管家基因，其中一些使用内源基因。</p>
<h6 id="2-10-1-1-Advantages-and-disadvantages-of-using-spike-ins-to-remove-confounders"><a href="#2-10-1-1-Advantages-and-disadvantages-of-using-spike-ins-to-remove-confounders" class="headerlink" title="2.10.1.1 Advantages and disadvantages of using spike-ins to remove confounders"></a>2.10.1.1 Advantages and disadvantages of using spike-ins to remove confounders</h6><p>使用加标作为对照基因是有吸引力的，因为在我们的实验中向每个细胞添加了相同量的ERCC（或其他）spike-in。原则上，我们观察到的这些基因的所有变异都是由技术噪声引起的;而内源基因受技术噪声和生物变异性的影响。通过将模型拟合到spike-in内并从内源基因“减去”这种噪声，可以消除技术噪声。基于这个前提有几种方法可用（例如BASiCS，scLVM，RUVg）;每个使用不同的噪声模型和不同的拟合程序。或者，可以鉴定表现出超出技术噪声的显着变异的基因（例如，与中值的距离，高度可变的基因）。然而，使用加标用于标准化存在问题（特别是ERCC，源自细菌序列），包括由于各种原因，它们的变异性实际上可能高于内源基因的变异性。<br>鉴于使用spike-ins的问题，通常可以通过使用内源基因获得更好的结果。在我们有大量内源基因的情况下，平均而言，细胞之间并没有系统地变化，我们期望技术效应影响大量基因（这是一个非常常见和合理的假设），那么这些方法（例如， RUVs方法）可以很好地执行。</p>
<figure class="highlight vhdl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(scRNA.seq.funcs)</span><br><span class="line"><span class="keyword">library</span>(RUVSeq)</span><br><span class="line"><span class="keyword">library</span>(scater)</span><br><span class="line"><span class="keyword">library</span>(SingleCellExperiment)</span><br><span class="line"><span class="keyword">library</span>(scran)</span><br><span class="line"><span class="keyword">library</span>(kBET)</span><br><span class="line"><span class="keyword">library</span>(sva) # Combat</span><br><span class="line"><span class="keyword">library</span>(edgeR)</span><br><span class="line">set.seed(<span class="number">1234567</span>)</span><br><span class="line">options(stringsAsFactors = <span class="literal">FALSE</span>)</span><br><span class="line">umi &lt;- readRDS(<span class="string">"tung/umi.rds"</span>)</span><br><span class="line">umi.qc &lt;- umi[rowData(umi)$<span class="keyword">use</span>, colData(umi)$<span class="keyword">use</span>]</span><br><span class="line">endog_genes &lt;- !rowData(umi.qc)$is_feature_control</span><br><span class="line">erccs &lt;- rowData(umi.qc)$is_feature_control</span><br><span class="line"></span><br><span class="line">qclust &lt;- quickCluster(umi.qc, min.size = <span class="number">30</span>)</span><br><span class="line">umi.qc &lt;- computeSumFactors(umi.qc, sizes = <span class="number">15</span>, clusters = qclust)</span><br><span class="line">umi.qc &lt;- normalize(umi.qc)</span><br></pre></td></tr></table></figure>

<h5 id="2-10-2-Remove-Unwanted-Variation"><a href="#2-10-2-Remove-Unwanted-Variation" class="headerlink" title="2.10.2 Remove Unwanted Variation"></a>2.10.2 Remove Unwanted Variation</h5><p>导致技术噪声的因素经常表现为“批次效应”，其中在不同日期或不同技术人员处理的细胞系统地彼此不同。通常可以使用相同的工具或轻微的变体来消除技术噪音并校正批量效应。我们将考虑Remove Unwanted Variation（RUVSeq）。简而言之，RUVSeq的工作原理如下。对于ñ样品和Ĵ基因，考虑以下广义线性模型（GLM），其中RNA-Seq读数计数在已知的感兴趣的协变量和不想要的变异的未知因素上回归：logE[Y|W,X,O]=Wα+Xβ+O</p>
<p>这里，ÿ是ñ×Ĵ观察到的基因水平读数计数矩阵，w 是一个ñ×ķ对应于“不想要的变化”因素的矩阵0是ñ×Ĵ偏移矩阵可以设置为零或使用其他一些归一化过程（例如上四分位标准化）进行估计。同时评估W，α，β,k是不可行的。对于一个给定的<br>k，下面的三种方法来估计不需要的变化的因素W<br>:</p>
<ul>
<li>RUVg使用阴性对照基因（例如ERCC），假定在样品中具有恒定表达;</li>
<li>RUVs使用中心（技术）重复/阴性对照样品，其中感兴趣的协变量是恒定的;</li>
<li>RUVr使用残差，例如来自对感兴趣的协变量的计数的第一遍GLM回归。<br>我们将集中讨论前两种方法。<h6 id="2-10-1-1-RUVg"><a href="#2-10-1-1-RUVg" class="headerlink" title="2.10.1.1 RUVg"></a>2.10.1.1 RUVg</h6></li>
</ul>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">ruvg &lt;- RUVg(<span class="name">counts</span>(<span class="name">umi</span>.qc), erccs, k = <span class="number">1</span>)</span><br><span class="line">assay(<span class="name">umi</span>.qc, <span class="string">"ruvg1"</span>) &lt;- log2(</span><br><span class="line">    <span class="literal">t</span>(<span class="name">t</span>(<span class="name">ruvg</span>$normalizedCounts) / colSums(<span class="name">ruvg</span>$normalizedCounts) * 1e6) + 1</span><br><span class="line">)</span><br><span class="line">ruvg &lt;- RUVg(counts(umi.qc), erccs, k = 10)</span><br><span class="line">assay(umi.qc, "ruvg10") &lt;- log2(</span><br><span class="line">    t(t(ruvg$normalizedCounts) / colSums(ruvg$normalizedCounts) * <span class="number">1e6</span>) + <span class="number">1</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<h6 id="2-10-2-2"><a href="#2-10-2-2" class="headerlink" title="2.10.2.2"></a>2.10.2.2</h6><figure class="highlight stata"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">scIdx &lt;- <span class="built_in">matrix</span>(-1, ncol = <span class="built_in">max</span>(<span class="keyword">table</span>(umi.qc<span class="variable">$individual</span>)), nrow = 3)</span><br><span class="line">tmp &lt;- <span class="keyword">which</span>(umi.qc<span class="variable">$individual</span> == <span class="string">"NA19098"</span>)</span><br><span class="line">scIdx[1, 1:<span class="built_in">length</span>(tmp)] &lt;- tmp</span><br><span class="line">tmp &lt;- <span class="keyword">which</span>(umi.qc<span class="variable">$individual</span> == <span class="string">"NA19101"</span>)</span><br><span class="line">scIdx[2, 1:<span class="built_in">length</span>(tmp)] &lt;- tmp</span><br><span class="line">tmp &lt;- <span class="keyword">which</span>(umi.qc<span class="variable">$individual</span> == <span class="string">"NA19239"</span>)</span><br><span class="line">scIdx[3, 1:<span class="built_in">length</span>(tmp)] &lt;- tmp</span><br><span class="line">cIdx &lt;- rownames(umi.qc)</span><br><span class="line">ruvs &lt;- RUVs(counts(umi.qc), cIdx, k = 1, scIdx = scIdx, isLog = FALSE)</span><br><span class="line">assay(umi.qc, <span class="string">"ruvs1"</span>) &lt;- log2(</span><br><span class="line">    t(t(ruvs<span class="variable">$normalizedCounts</span>) / colSums(ruvs<span class="variable">$normalizedCounts</span>) * 1e6) + 1</span><br><span class="line">)</span><br><span class="line">ruvs &lt;- RUVs(counts(umi.qc), cIdx, k = 10, scIdx = scIdx, isLog = FALSE)</span><br><span class="line">assay(umi.qc, <span class="string">"ruvs10"</span>) &lt;- log2(</span><br><span class="line">    t(t(ruvs<span class="variable">$normalizedCounts</span>) / colSums(ruvs<span class="variable">$normalizedCounts</span>) * 1e6) + 1</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<h5 id="2-10-3-Combat"><a href="#2-10-3-Combat" class="headerlink" title="2.10.3 Combat"></a>2.10.3 Combat</h5><p>如果您有平衡设计的实验，Combat可用于消除批量效应，同时通过使用mod参数指定生物效应来保留生物效应。然而，Tung数据包含多个实验重复而不是平衡设计，因此使用mod1来保持生物变异性将导致错误。</p>
<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">combat_data</span> &lt;- logcounts(umi.qc)</span><br><span class="line"><span class="title">mod_data</span> &lt;- <span class="keyword">as</span>.<span class="class"><span class="keyword">data</span>.frame(<span class="title">t</span>(<span class="title">combat_data</span>))</span></span><br><span class="line"><span class="meta"># Basic batch removal</span></span><br><span class="line"><span class="title">mod0</span> = model.matrix(~ <span class="number">1</span>, <span class="class"><span class="keyword">data</span> = mod_data) </span></span><br><span class="line"><span class="meta"># Preserve biological variability</span></span><br><span class="line"><span class="title">mod1</span> = model.matrix(~ umi.qc$individual, <span class="class"><span class="keyword">data</span> = mod_data) </span></span><br><span class="line"><span class="meta"># adjust for total genes detected</span></span><br><span class="line"><span class="title">mod2</span> = model.matrix(~ umi.qc$total_features, <span class="class"><span class="keyword">data</span> = mod_data)</span></span><br><span class="line"><span class="title">assay</span>(umi.qc, <span class="string">"combat"</span>) &lt;- <span class="type">ComBat</span>(</span><br><span class="line">    dat = t(mod_data), </span><br><span class="line">    batch = factor(umi.qc$batch), </span><br><span class="line">    mod = mod0,</span><br><span class="line">    par.prior = <span class="type">TRUE</span>,</span><br><span class="line">    prior.plots = <span class="type">FALSE</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<h5 id="2-10-4-mnnCorrect"><a href="#2-10-4-mnnCorrect" class="headerlink" title="2.10.4 mnnCorrect"></a>2.10.4 mnnCorrect</h5><p>mnnCorrect（Haghverdi等人，2017）假设每批次彼此共享至少一种生物学条件。因此，它适用于各种平衡的实验设计。然而，Tung数据包含每个个体而非平衡批次的多个重复，因此我们将分别标准化每个个体。请注意，由于混淆的实验设计，这将消除同一个体中批次之间的批次效应，但不会消除不同个体中批次之间的批次效应。</p>
<p>因此，我们将合并来自每个个体的复制品以形成三个批次。</p>
<figure class="highlight lua"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><span class="line">do_mnn &lt;- <span class="function"><span class="keyword">function</span><span class="params">(data.qc)</span></span> &#123;</span><br><span class="line">    batch1 &lt;- logcounts(data.qc[, data.qc$replicate == <span class="string">"r1"</span>])</span><br><span class="line">    batch2 &lt;- logcounts(data.qc[, data.qc$replicate == <span class="string">"r2"</span>])</span><br><span class="line">    batch3 &lt;- logcounts(data.qc[, data.qc$replicate == <span class="string">"r3"</span>])</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">if</span> (ncol(batch2) &gt; <span class="number">0</span>) &#123;</span><br><span class="line">        x = mnnCorrect(</span><br><span class="line">          batch1, batch2, batch3,  </span><br><span class="line">          k = <span class="number">20</span>,</span><br><span class="line">          sigma = <span class="number">0.1</span>,</span><br><span class="line">          <span class="built_in">cos</span>.norm.<span class="keyword">in</span> = TRUE,</span><br><span class="line">          svd.dim = <span class="number">2</span></span><br><span class="line">        )</span><br><span class="line">        res1 &lt;- x$corrected<span class="string">[[1]]</span></span><br><span class="line">        res2 &lt;- x$corrected<span class="string">[[2]]</span></span><br><span class="line">        res3 &lt;- x$corrected<span class="string">[[3]]</span></span><br><span class="line">        dimnames(res1) &lt;- dimnames(batch1)</span><br><span class="line">        dimnames(res2) &lt;- dimnames(batch2)</span><br><span class="line">        dimnames(res3) &lt;- dimnames(batch3)</span><br><span class="line">        <span class="keyword">return</span>(cbind(res1, res2, res3))</span><br><span class="line">    &#125; <span class="keyword">else</span> &#123;</span><br><span class="line">        x = mnnCorrect(</span><br><span class="line">          batch1, batch3,  </span><br><span class="line">          k = <span class="number">20</span>,</span><br><span class="line">          sigma = <span class="number">0.1</span>,</span><br><span class="line">          <span class="built_in">cos</span>.norm.<span class="keyword">in</span> = TRUE,</span><br><span class="line">          svd.dim = <span class="number">2</span></span><br><span class="line">        )</span><br><span class="line">        res1 &lt;- x$corrected<span class="string">[[1]]</span></span><br><span class="line">        res3 &lt;- x$corrected<span class="string">[[2]]</span></span><br><span class="line">        dimnames(res1) &lt;- dimnames(batch1)</span><br><span class="line">        dimnames(res3) &lt;- dimnames(batch3)</span><br><span class="line">        <span class="keyword">return</span>(cbind(res1, res3))</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">indi1 &lt;- do_mnn(umi.qc[, umi.qc$individual == <span class="string">"NA19098"</span>])</span><br><span class="line">indi2 &lt;- do_mnn(umi.qc[, umi.qc$individual == <span class="string">"NA19101"</span>])</span><br><span class="line">indi3 &lt;- do_mnn(umi.qc[, umi.qc$individual == <span class="string">"NA19239"</span>])</span><br><span class="line"></span><br><span class="line">assay(umi.qc, <span class="string">"mnn"</span>) &lt;- cbind(indi1, indi2, indi3)</span><br><span class="line"></span><br><span class="line"># For a balanced design: </span><br><span class="line">#assay(umi.qc, <span class="string">"mnn"</span>) &lt;- mnnCorrect(</span><br><span class="line">#    list(B1 = logcounts(batch1), B2 = logcounts(batch2), B3 = logcounts(batch3)),  </span><br><span class="line">#    k = <span class="number">20</span>,</span><br><span class="line">#    sigma = <span class="number">0.1</span>,</span><br><span class="line">#    <span class="built_in">cos</span>.norm = TRUE,</span><br><span class="line">#    svd.dim = <span class="number">2</span></span><br><span class="line">#)</span><br></pre></td></tr></table></figure>

<h5 id="2-10-5-GLM"><a href="#2-10-5-GLM" class="headerlink" title="2.10.5 GLM"></a>2.10.5 GLM</h5><p>一般线性模型是Combat的更简单版本。如果你有一个平衡的设计，它可以纠正批量，同时保留生物效应。在混淆/复制设计中，生物效应将不适合/保留。与mnnCorrect类似，我们可以分别从每个个体中删除批量效应，以保持个体之间的生物（和技术）差异。出于演示目的，我们将天真地纠正所有共同创建的批处理效果：</p>
<figure class="highlight php"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">glm_fun &lt;- <span class="function"><span class="keyword">function</span><span class="params">(g, batch, indi)</span> </span>&#123;</span><br><span class="line">  model &lt;- glm(g ~ batch + indi)</span><br><span class="line">  model$coef[<span class="number">1</span>] &lt;- <span class="number">0</span> <span class="comment"># replace intercept with 0 to preserve reference batch.</span></span><br><span class="line">  <span class="keyword">return</span>(model$coef)</span><br><span class="line">&#125;</span><br><span class="line">effects &lt;- apply(</span><br><span class="line">    logcounts(umi.qc), </span><br><span class="line">    <span class="number">1</span>, </span><br><span class="line">    glm_fun, </span><br><span class="line">    batch = umi.qc$batch, </span><br><span class="line">    indi = umi.qc$individual</span><br><span class="line">)</span><br><span class="line">corrected &lt;- logcounts(umi.qc) - t(effects[<span class="keyword">as</span>.numeric(factor(umi.qc$batch)), ])</span><br><span class="line">assay(umi.qc, <span class="string">"glm"</span>) &lt;- corrected</span><br></pre></td></tr></table></figure>

<h5 id="2-10-6-How-to-evaluate-and-compare-confounder-removal-strategies"><a href="#2-10-6-How-to-evaluate-and-compare-confounder-removal-strategies" class="headerlink" title="2.10.6 How to evaluate and compare confounder removal strategies"></a>2.10.6 How to evaluate and compare confounder removal strategies</h5><p>在考虑去除混杂因素的不同方法时，一个关键问题是如何定量确定哪一种方法最有效。比较具有挑战性的主要原因是因为通常很难知道什么对应于技术counfounders和什么是有趣的生物变异。在这里，我们根据我们对实验设计的了解，考虑三种不同的指标。根据您希望解决的生物学问题，选择一个允许您评估可能是特定情况下最关注的混杂因素的指标非常重要。</p>
<h6 id="2-10-6-1-Effectiveness-1"><a href="#2-10-6-1-Effectiveness-1" class="headerlink" title="2.10.6.1 Effectiveness 1"></a>2.10.6.1 Effectiveness 1</h6><p>我们通过检查PCA图来评估标准化的有效性，其中颜色对应于技术重复，并且形状对应于不同的生物样品（个体）。生物样品和散布批次的分离表明技术变异已被消除。我们总是使用log2-cpm标准化数据来匹配PCA的假设。</p>
<figure class="highlight ceylon"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">for</span>(n <span class="keyword">in</span> assayNames(umi.qc)) &#123;</span><br><span class="line">    print(</span><br><span class="line">        plotPCA(</span><br><span class="line">            umi.qc[endog<span class="number">_</span>genes, ],</span><br><span class="line">            colour<span class="number">_</span><span class="meta">by</span> = <span class="string">"batch"</span>,</span><br><span class="line">            size<span class="number">_</span><span class="meta">by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">            shape<span class="number">_</span><span class="meta">by</span> = <span class="string">"individual"</span>,</span><br><span class="line">            exprs<span class="number">_</span>values = n</span><br><span class="line">        ) +</span><br><span class="line">        ggtitle(n)</span><br><span class="line">    )</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-1.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-2.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-3.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-4.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-5.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-6.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-7.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-8.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-9.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-10.png" alt="image"><br>![image]<a href="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-11.png" target="_blank" rel="noopener">https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-11.png</a>)
<img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-10-12.png" alt="image"></p>
<h6 id="2-10-6-2-Effectiveness-2"><a href="#2-10-6-2-Effectiveness-2" class="headerlink" title="2.10.6.2 Effectiveness 2"></a>2.10.6.2 Effectiveness 2</h6><p>我们还可以使用跨cell的相对对数表达式（RLE）检查校正的有效性，以确认已从数据集中删除技术噪声。注意RLE仅评估每个细胞的高于和低于平均值的基因数量是否相等 - 即系统性技术效应。 RLE可能无法检测到批次之间的随机技术噪音。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">res &lt;- list()</span><br><span class="line"><span class="function"><span class="title">for</span><span class="params">(n in assayNames(umi.qc)</span></span>) &#123;</span><br><span class="line">    res[[n]] &lt;- suppressWarnings(calc_cell_RLE(assay(umi<span class="selector-class">.qc</span>, n), erccs))</span><br><span class="line">&#125;</span><br><span class="line"><span class="function"><span class="title">par</span><span class="params">(mar=c(<span class="number">6</span>,<span class="number">4</span>,<span class="number">1</span>,<span class="number">1</span>)</span></span>)</span><br><span class="line"><span class="function"><span class="title">boxplot</span><span class="params">(res, las=<span class="number">2</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-11-1.png" alt="image"></p>
<h6 id="2-10-6-3-Effectiveness-3"><a href="#2-10-6-3-Effectiveness-3" class="headerlink" title="2.10.6.3 Effectiveness 3"></a>2.10.6.3 Effectiveness 3</h6><p>我们可以重复前面的分析，检查批处理效果是否被删除</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">for</span>(n <span class="keyword">in</span> assayNames(umi.qc)) &#123;</span><br><span class="line">    <span class="builtin-name">print</span>(</span><br><span class="line">        plotQC(</span><br><span class="line">            umi.qc[endog_genes, ],</span><br><span class="line">           <span class="built_in"> type </span>= <span class="string">"expl"</span>,</span><br><span class="line">            exprs_values = n,</span><br><span class="line">            variables = c(</span><br><span class="line">                <span class="string">"total_features"</span>,</span><br><span class="line">                <span class="string">"total_counts"</span>,</span><br><span class="line">                <span class="string">"batch"</span>,</span><br><span class="line">                <span class="string">"individual"</span>,</span><br><span class="line">                <span class="string">"pct_counts_ERCC"</span>,</span><br><span class="line">                <span class="string">"pct_counts_MT"</span></span><br><span class="line">            )</span><br><span class="line">        ) +</span><br><span class="line">        ggtitle(n)</span><br><span class="line">    )</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-1.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-2.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-3.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-4.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-5.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-6.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-7.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-8.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-9.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-10.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-11.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/confound-cpm-12.png" alt="image"></p>
<h6 id="2-10-6-4-Effectiveness-4"><a href="#2-10-6-4-Effectiveness-4" class="headerlink" title="2.10.6.4 Effectiveness 4"></a>2.10.6.4 Effectiveness 4</h6><p>检查批次效应校正功效的另一种方法是考虑在数据的局部子样本中来自不同批次的点的混合。如果没有批次效应，则任何局部区域中每批次的细胞比例应等于每批次中细胞的全局比例。</p>
<p>kBET（Buttner等人，2017）在随机细胞周围采用kNN网络，并根据二项分布测试每批次的细胞数量。这些测试的拒绝率表明数据中仍存在批次效应的严重性（高拒绝率=强批次效应）。 kBET假设每批包含相同的生物组补充，因此如果使用完美平衡的设计，它只能应用于整个数据集。但是，如果将kBET分别应用于每个生物组，也可以将其应用于复制数据。在Tung数据的情况下，我们将独立地将kBET应用于每个个体以检查残余批次效应。然而，该方法不能识别与生物条件混淆的残留批次效应。此外，kBET不确定是否保留了生物信号。</p>
<figure class="highlight php"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">compare_kBET_results &lt;- <span class="function"><span class="keyword">function</span><span class="params">(sce)</span></span>&#123;</span><br><span class="line">    indiv &lt;- unique(sce$individual)</span><br><span class="line">    norms &lt;- assayNames(sce) <span class="comment"># Get all normalizations</span></span><br><span class="line">    results &lt;- <span class="keyword">list</span>()</span><br><span class="line">    <span class="keyword">for</span> (i in indiv)&#123; </span><br><span class="line">        <span class="keyword">for</span> (j in norms)&#123;</span><br><span class="line">            tmp &lt;- kBET(</span><br><span class="line">                df = t(assay(sce[,sce$individual== i], j)), </span><br><span class="line">                batch = sce$batch[sce$individual==i], </span><br><span class="line">                heuristic = <span class="keyword">TRUE</span>, </span><br><span class="line">                verbose = <span class="keyword">FALSE</span>, </span><br><span class="line">                addTest = <span class="keyword">FALSE</span>, </span><br><span class="line">                plot = <span class="keyword">FALSE</span>)</span><br><span class="line">            results[[i]][[j]] &lt;- tmp$summary$kBET.observed[<span class="number">1</span>]</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">return</span>(<span class="keyword">as</span>.data.frame(results))</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">eff_debatching &lt;- compare_kBET_results(umi.qc)</span><br></pre></td></tr></table></figure>

<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">require(<span class="string">"reshape2"</span>)</span><br><span class="line">require(<span class="string">"RColorBrewer"</span>)</span><br><span class="line"><span class="comment"># Plot results</span></span><br><span class="line">dod &lt;- melt(as.matrix(eff_debatching),  value.<span class="attr">name</span> = <span class="string">"kBET"</span>)</span><br><span class="line">colnames(dod)[<span class="number">1</span>:<span class="number">2</span>] &lt;- c(<span class="string">"Normalisation"</span>, <span class="string">"Individual"</span>)</span><br><span class="line"></span><br><span class="line">colorset &lt;- c('gray', brewer.pal(<span class="attr">n</span> = <span class="number">9</span>, <span class="string">"RdYlBu"</span>))</span><br><span class="line"></span><br><span class="line">ggplot(dod, aes(Normalisation, Individual, <span class="attr">fill=kBET))</span> +  </span><br><span class="line">    geom_tile() +</span><br><span class="line">    scale_fill_gradient2(</span><br><span class="line">        na.<span class="attr">value</span> = <span class="string">"gray"</span>,</span><br><span class="line">        <span class="attr">low</span> = colorset[<span class="number">2</span>],</span><br><span class="line">        <span class="attr">mid=colorset[6],</span></span><br><span class="line">        <span class="attr">high</span> = colorset[<span class="number">10</span>],</span><br><span class="line">        <span class="attr">midpoint</span> = <span class="number">0.5</span>, <span class="attr">limit</span> = c(<span class="number">0</span>,<span class="number">1</span>)) +</span><br><span class="line">    scale_x_discrete(<span class="attr">expand</span> = c(<span class="number">0</span>, <span class="number">0</span>)) +</span><br><span class="line">    scale_y_discrete(<span class="attr">expand</span> = c(<span class="number">0</span>, <span class="number">0</span>)) + </span><br><span class="line">    theme(</span><br><span class="line">        axis.text.<span class="attr">x</span> = element_text(</span><br><span class="line">            <span class="attr">angle</span> = <span class="number">45</span>, </span><br><span class="line">            <span class="attr">vjust</span> = <span class="number">1</span>, </span><br><span class="line">            <span class="attr">size</span> = <span class="number">12</span>, </span><br><span class="line">            <span class="attr">hjust</span> = <span class="number">1</span></span><br><span class="line">        )</span><br><span class="line">    ) + </span><br><span class="line">    ggtitle(<span class="string">"Effect of batch regression methods per individual"</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/22-remove-conf_files/figure-html/unnamed-chunk-13-1.png" alt="image"></p>
<h5 id="2-11-Dealing-with-confounders（Reads）"><a href="#2-11-Dealing-with-confounders（Reads）" class="headerlink" title="2.11 Dealing with confounders（Reads）"></a>2.11 Dealing with confounders（Reads）</h5><figure class="highlight vhdl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">library</span>(scRNA.seq.funcs)</span><br><span class="line"><span class="keyword">library</span>(RUVSeq)</span><br><span class="line"><span class="keyword">library</span>(scater)</span><br><span class="line"><span class="keyword">library</span>(SingleCellExperiment)</span><br><span class="line"><span class="keyword">library</span>(scran)</span><br><span class="line"><span class="keyword">library</span>(kBET)</span><br><span class="line"><span class="keyword">library</span>(sva) # Combat</span><br><span class="line"><span class="keyword">library</span>(edgeR)</span><br><span class="line">set.seed(<span class="number">1234567</span>)</span><br><span class="line">options(stringsAsFactors = <span class="literal">FALSE</span>)</span><br><span class="line">reads &lt;- readRDS(<span class="string">"tung/reads.rds"</span>)</span><br><span class="line">reads.qc &lt;- reads[rowData(reads)$<span class="keyword">use</span>, colData(reads)$<span class="keyword">use</span>]</span><br><span class="line">endog_genes &lt;- !rowData(reads.qc)$is_feature_control</span><br><span class="line">erccs &lt;- rowData(reads.qc)$is_feature_control</span><br><span class="line"></span><br><span class="line">qclust &lt;- quickCluster(reads.qc, min.size = <span class="number">30</span>)</span><br><span class="line">reads.qc &lt;- computeSumFactors(reads.qc, sizes = <span class="number">15</span>, clusters = qclust)</span><br><span class="line">reads.qc &lt;- normalize(reads.qc)</span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">ruvg &lt;- RUVg(<span class="name">counts</span>(<span class="name">reads</span>.qc), erccs, k = <span class="number">1</span>)</span><br><span class="line">assay(<span class="name">reads</span>.qc, <span class="string">"ruvg1"</span>) &lt;- log2(</span><br><span class="line">    <span class="literal">t</span>(<span class="name">t</span>(<span class="name">ruvg</span>$normalizedCounts) / colSums(<span class="name">ruvg</span>$normalizedCounts) * 1e6) + 1</span><br><span class="line">)</span><br><span class="line">ruvg &lt;- RUVg(counts(reads.qc), erccs, k = 10)</span><br><span class="line">assay(reads.qc, "ruvg10") &lt;- log2(</span><br><span class="line">    t(t(ruvg$normalizedCounts) / colSums(ruvg$normalizedCounts) * <span class="number">1e6</span>) + <span class="number">1</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<figure class="highlight stata"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line">scIdx &lt;- <span class="built_in">matrix</span>(-1, ncol = <span class="built_in">max</span>(<span class="keyword">table</span>(reads.qc<span class="variable">$individual</span>)), nrow = 3)</span><br><span class="line">tmp &lt;- <span class="keyword">which</span>(reads.qc<span class="variable">$individual</span> == <span class="string">"NA19098"</span>)</span><br><span class="line">scIdx[1, 1:<span class="built_in">length</span>(tmp)] &lt;- tmp</span><br><span class="line">tmp &lt;- <span class="keyword">which</span>(reads.qc<span class="variable">$individual</span> == <span class="string">"NA19101"</span>)</span><br><span class="line">scIdx[2, 1:<span class="built_in">length</span>(tmp)] &lt;- tmp</span><br><span class="line">tmp &lt;- <span class="keyword">which</span>(reads.qc<span class="variable">$individual</span> == <span class="string">"NA19239"</span>)</span><br><span class="line">scIdx[3, 1:<span class="built_in">length</span>(tmp)] &lt;- tmp</span><br><span class="line">cIdx &lt;- rownames(reads.qc)</span><br><span class="line">ruvs &lt;- RUVs(counts(reads.qc), cIdx, k = 1, scIdx = scIdx, isLog = FALSE)</span><br><span class="line">assay(reads.qc, <span class="string">"ruvs1"</span>) &lt;- log2(</span><br><span class="line">    t(t(ruvs<span class="variable">$normalizedCounts</span>) / colSums(ruvs<span class="variable">$normalizedCounts</span>) * 1e6) + 1</span><br><span class="line">)</span><br><span class="line">ruvs &lt;- RUVs(counts(reads.qc), cIdx, k = 10, scIdx = scIdx, isLog = FALSE)</span><br><span class="line">assay(reads.qc, <span class="string">"ruvs10"</span>) &lt;- log2(</span><br><span class="line">    t(t(ruvs<span class="variable">$normalizedCounts</span>) / colSums(ruvs<span class="variable">$normalizedCounts</span>) * 1e6) + 1</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">combat_data</span> &lt;- logcounts(reads.qc)</span><br><span class="line"><span class="title">mod_data</span> &lt;- <span class="keyword">as</span>.<span class="class"><span class="keyword">data</span>.frame(<span class="title">t</span>(<span class="title">combat_data</span>))</span></span><br><span class="line"><span class="meta"># Basic batch removal</span></span><br><span class="line"><span class="title">mod0</span> = model.matrix(~ <span class="number">1</span>, <span class="class"><span class="keyword">data</span> = mod_data) </span></span><br><span class="line"><span class="meta"># Preserve biological variability</span></span><br><span class="line"><span class="title">mod1</span> = model.matrix(~ reads.qc$individual, <span class="class"><span class="keyword">data</span> = mod_data) </span></span><br><span class="line"><span class="meta"># adjust for total genes detected</span></span><br><span class="line"><span class="title">mod2</span> = model.matrix(~ reads.qc$total_features, <span class="class"><span class="keyword">data</span> = mod_data)</span></span><br><span class="line"><span class="title">assay</span>(reads.qc, <span class="string">"combat"</span>) &lt;- <span class="type">ComBat</span>(</span><br><span class="line">    dat = t(mod_data), </span><br><span class="line">    batch = factor(reads.qc$batch), </span><br><span class="line">    mod = mod0,</span><br><span class="line">    par.prior = <span class="type">TRUE</span>,</span><br><span class="line">    prior.plots = <span class="type">FALSE</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## Standardizing Data across genes</span><br></pre></td></tr></table></figure>

<figure class="highlight lua"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br></pre></td><td class="code"><pre><span class="line">do_mnn &lt;- <span class="function"><span class="keyword">function</span><span class="params">(data.qc)</span></span> &#123;</span><br><span class="line">    batch1 &lt;- logcounts(data.qc[, data.qc$replicate == <span class="string">"r1"</span>])</span><br><span class="line">    batch2 &lt;- logcounts(data.qc[, data.qc$replicate == <span class="string">"r2"</span>])</span><br><span class="line">    batch3 &lt;- logcounts(data.qc[, data.qc$replicate == <span class="string">"r3"</span>])</span><br><span class="line">    </span><br><span class="line">    <span class="keyword">if</span> (ncol(batch2) &gt; <span class="number">0</span>) &#123;</span><br><span class="line">        x = mnnCorrect(</span><br><span class="line">          batch1, batch2, batch3,  </span><br><span class="line">          k = <span class="number">20</span>,</span><br><span class="line">          sigma = <span class="number">0.1</span>,</span><br><span class="line">          <span class="built_in">cos</span>.norm.<span class="keyword">in</span> = TRUE,</span><br><span class="line">          svd.dim = <span class="number">2</span></span><br><span class="line">        )</span><br><span class="line">        res1 &lt;- x$corrected<span class="string">[[1]]</span></span><br><span class="line">        res2 &lt;- x$corrected<span class="string">[[2]]</span></span><br><span class="line">        res3 &lt;- x$corrected<span class="string">[[3]]</span></span><br><span class="line">        dimnames(res1) &lt;- dimnames(batch1)</span><br><span class="line">        dimnames(res2) &lt;- dimnames(batch2)</span><br><span class="line">        dimnames(res3) &lt;- dimnames(batch3)</span><br><span class="line">        <span class="keyword">return</span>(cbind(res1, res2, res3))</span><br><span class="line">    &#125; <span class="keyword">else</span> &#123;</span><br><span class="line">        x = mnnCorrect(</span><br><span class="line">          batch1, batch3,  </span><br><span class="line">          k = <span class="number">20</span>,</span><br><span class="line">          sigma = <span class="number">0.1</span>,</span><br><span class="line">          <span class="built_in">cos</span>.norm.<span class="keyword">in</span> = TRUE,</span><br><span class="line">          svd.dim = <span class="number">2</span></span><br><span class="line">        )</span><br><span class="line">        res1 &lt;- x$corrected<span class="string">[[1]]</span></span><br><span class="line">        res3 &lt;- x$corrected<span class="string">[[2]]</span></span><br><span class="line">        dimnames(res1) &lt;- dimnames(batch1)</span><br><span class="line">        dimnames(res3) &lt;- dimnames(batch3)</span><br><span class="line">        <span class="keyword">return</span>(cbind(res1, res3))</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">indi1 &lt;- do_mnn(reads.qc[, reads.qc$individual == <span class="string">"NA19098"</span>])</span><br><span class="line">indi2 &lt;- do_mnn(reads.qc[, reads.qc$individual == <span class="string">"NA19101"</span>])</span><br><span class="line">indi3 &lt;- do_mnn(reads.qc[, reads.qc$individual == <span class="string">"NA19239"</span>])</span><br><span class="line"></span><br><span class="line">assay(reads.qc, <span class="string">"mnn"</span>) &lt;- cbind(indi1, indi2, indi3)</span><br><span class="line"></span><br><span class="line"># For a balanced design: </span><br><span class="line">#assay(reads.qc, <span class="string">"mnn"</span>) &lt;- mnnCorrect(</span><br><span class="line">#    list(B1 = logcounts(batch1), B2 = logcounts(batch2), B3 = logcounts(batch3)),  </span><br><span class="line">#    k = <span class="number">20</span>,</span><br><span class="line">#    sigma = <span class="number">0.1</span>,</span><br><span class="line">#    <span class="built_in">cos</span>.norm = TRUE,</span><br><span class="line">#    svd.dim = <span class="number">2</span></span><br><span class="line">#)</span><br></pre></td></tr></table></figure>

<figure class="highlight php"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">glm_fun &lt;- <span class="function"><span class="keyword">function</span><span class="params">(g, batch, indi)</span> </span>&#123;</span><br><span class="line">  model &lt;- glm(g ~ batch + indi)</span><br><span class="line">  model$coef[<span class="number">1</span>] &lt;- <span class="number">0</span> <span class="comment"># replace intercept with 0 to preserve reference batch.</span></span><br><span class="line">  <span class="keyword">return</span>(model$coef)</span><br><span class="line">&#125;</span><br><span class="line">effects &lt;- apply(</span><br><span class="line">    logcounts(reads.qc), </span><br><span class="line">    <span class="number">1</span>, </span><br><span class="line">    glm_fun, </span><br><span class="line">    batch = reads.qc$batch, </span><br><span class="line">    indi = reads.qc$individual</span><br><span class="line">)</span><br><span class="line">corrected &lt;- logcounts(reads.qc) - t(effects[<span class="keyword">as</span>.numeric(factor(reads.qc$batch)), ])</span><br><span class="line">assay(reads.qc, <span class="string">"glm"</span>) &lt;- corrected</span><br></pre></td></tr></table></figure>

<figure class="highlight ceylon"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">for</span>(n <span class="keyword">in</span> assayNames(reads.qc)) &#123;</span><br><span class="line">    print(</span><br><span class="line">        plotPCA(</span><br><span class="line">            reads.qc[endog<span class="number">_</span>genes, ],</span><br><span class="line">            colour<span class="number">_</span><span class="meta">by</span> = <span class="string">"batch"</span>,</span><br><span class="line">            size<span class="number">_</span><span class="meta">by</span> = <span class="string">"total_features"</span>,</span><br><span class="line">            shape<span class="number">_</span><span class="meta">by</span> = <span class="string">"individual"</span>,</span><br><span class="line">            exprs<span class="number">_</span>values = n</span><br><span class="line">        ) +</span><br><span class="line">        ggtitle(n)</span><br><span class="line">    )</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-1.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-2.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-3.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-4.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-5.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-6.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-7.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-8.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-9.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-10.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-11.png" alt="image"><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-10-12.png" alt="image"></p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">res &lt;- list()</span><br><span class="line"><span class="function"><span class="title">for</span><span class="params">(n in assayNames(reads.qc)</span></span>) &#123;</span><br><span class="line">    res[[n]] &lt;- suppressWarnings(calc_cell_RLE(assay(reads<span class="selector-class">.qc</span>, n), erccs))</span><br><span class="line">&#125;</span><br><span class="line"><span class="function"><span class="title">par</span><span class="params">(mar=c(<span class="number">6</span>,<span class="number">4</span>,<span class="number">1</span>,<span class="number">1</span>)</span></span>)</span><br><span class="line"><span class="function"><span class="title">boxplot</span><span class="params">(res, las=<span class="number">2</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-11-1.png" alt="image"></p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">for</span>(n <span class="keyword">in</span> assayNames(reads.qc)) &#123;</span><br><span class="line">    <span class="builtin-name">print</span>(</span><br><span class="line">        plotQC(</span><br><span class="line">            reads.qc[endog_genes, ],</span><br><span class="line">           <span class="built_in"> type </span>= <span class="string">"expl"</span>,</span><br><span class="line">            exprs_values = n,</span><br><span class="line">            variables = c(</span><br><span class="line">                <span class="string">"total_features"</span>,</span><br><span class="line">                <span class="string">"total_counts"</span>,</span><br><span class="line">                <span class="string">"batch"</span>,</span><br><span class="line">                <span class="string">"individual"</span>,</span><br><span class="line">                <span class="string">"pct_counts_ERCC"</span>,</span><br><span class="line">                <span class="string">"pct_counts_MT"</span></span><br><span class="line">            )</span><br><span class="line">        ) +</span><br><span class="line">        ggtitle(n)</span><br><span class="line">    )</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-1.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-2.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-3.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-4.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-5.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-6.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-7.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-8.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-9.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-10.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-11.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-12-12.png" alt="image"></p>
<figure class="highlight php"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br></pre></td><td class="code"><pre><span class="line">compare_kBET_results &lt;- <span class="function"><span class="keyword">function</span><span class="params">(sce)</span></span>&#123;</span><br><span class="line">    indiv &lt;- unique(sce$individual)</span><br><span class="line">    norms &lt;- assayNames(sce) <span class="comment"># Get all normalizations</span></span><br><span class="line">    results &lt;- <span class="keyword">list</span>()</span><br><span class="line">    <span class="keyword">for</span> (i in indiv)&#123; </span><br><span class="line">        <span class="keyword">for</span> (j in norms)&#123;</span><br><span class="line">            tmp &lt;- kBET(</span><br><span class="line">                df = t(assay(sce[,sce$individual== i], j)), </span><br><span class="line">                batch = sce$batch[sce$individual==i], </span><br><span class="line">                heuristic = <span class="keyword">TRUE</span>, </span><br><span class="line">                verbose = <span class="keyword">FALSE</span>, </span><br><span class="line">                addTest = <span class="keyword">FALSE</span>, </span><br><span class="line">                plot = <span class="keyword">FALSE</span>)</span><br><span class="line">            results[[i]][[j]] &lt;- tmp$summary$kBET.observed[<span class="number">1</span>]</span><br><span class="line">        &#125;</span><br><span class="line">    &#125;</span><br><span class="line">    <span class="keyword">return</span>(<span class="keyword">as</span>.data.frame(results))</span><br><span class="line">&#125;</span><br><span class="line"></span><br><span class="line">eff_debatching &lt;- compare_kBET_results(reads.qc)</span><br></pre></td></tr></table></figure>

<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br></pre></td><td class="code"><pre><span class="line">require(<span class="string">"reshape2"</span>)</span><br><span class="line">require(<span class="string">"RColorBrewer"</span>)</span><br><span class="line"><span class="comment"># Plot results</span></span><br><span class="line">dod &lt;- melt(as.matrix(eff_debatching),  value.<span class="attr">name</span> = <span class="string">"kBET"</span>)</span><br><span class="line">colnames(dod)[<span class="number">1</span>:<span class="number">2</span>] &lt;- c(<span class="string">"Normalisation"</span>, <span class="string">"Individual"</span>)</span><br><span class="line"></span><br><span class="line">colorset &lt;- c('gray', brewer.pal(<span class="attr">n</span> = <span class="number">9</span>, <span class="string">"RdYlBu"</span>))</span><br><span class="line"></span><br><span class="line">ggplot(dod, aes(Normalisation, Individual, <span class="attr">fill=kBET))</span> +  </span><br><span class="line">    geom_tile() +</span><br><span class="line">    scale_fill_gradient2(</span><br><span class="line">        na.<span class="attr">value</span> = <span class="string">"gray"</span>,</span><br><span class="line">        <span class="attr">low</span> = colorset[<span class="number">2</span>],</span><br><span class="line">        <span class="attr">mid=colorset[6],</span></span><br><span class="line">        <span class="attr">high</span> = colorset[<span class="number">10</span>],</span><br><span class="line">        <span class="attr">midpoint</span> = <span class="number">0.5</span>, <span class="attr">limit</span> = c(<span class="number">0</span>,<span class="number">1</span>)) +</span><br><span class="line">    scale_x_discrete(<span class="attr">expand</span> = c(<span class="number">0</span>, <span class="number">0</span>)) +</span><br><span class="line">    scale_y_discrete(<span class="attr">expand</span> = c(<span class="number">0</span>, <span class="number">0</span>)) + </span><br><span class="line">    theme(</span><br><span class="line">        axis.text.<span class="attr">x</span> = element_text(</span><br><span class="line">            <span class="attr">angle</span> = <span class="number">45</span>, </span><br><span class="line">            <span class="attr">vjust</span> = <span class="number">1</span>, </span><br><span class="line">            <span class="attr">size</span> = <span class="number">12</span>, </span><br><span class="line">            <span class="attr">hjust</span> = <span class="number">1</span></span><br><span class="line">        )</span><br><span class="line">    ) + </span><br><span class="line">    ggtitle(<span class="string">"Effect of batch regression methods per individual"</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/23-remove-conf-reads_files/figure-html/unnamed-chunk-14-1.png" alt="image"></p>
<h3 id="3-Biological-Analysis"><a href="#3-Biological-Analysis" class="headerlink" title="3 Biological Analysis"></a>3 Biological Analysis</h3><h4 id="3-1-Clustering-Introduction"><a href="#3-1-Clustering-Introduction" class="headerlink" title="3.1 Clustering Introduction"></a>3.1 Clustering Introduction</h4><p>一旦我们对数据进行了标准化并删除了混淆因素，我们就可以进行与手头的生物学问题相关的分析。分析的确切性质取决于数据集。然而，有一些方面在广泛的背景下是有用的，我们将在接下来的几章中讨论其中的一些方面。我们将从scRNA-seq数据的聚类开始。</p>
<h5 id="3-1-1-Introduction"><a href="#3-1-1-Introduction" class="headerlink" title="3.1.1 Introduction"></a>3.1.1 Introduction</h5><p>scRNA-seq最有希望的应用之一是基于转录谱的de novo discovery和细胞类型的注释。在计算上，这是一个难题，因为它相当于无监督的聚类。也就是说，我们需要根据转录组的相似性识别细胞群，而无需事先了解标记。此外，在大多数情况下，我们甚至不知道先验集群的数量。由于高水平的噪音（技术和生物）和维度，问题变得更具挑战性。</p>
<h5 id="3-1-2-Dimensionality-reductions"><a href="#3-1-2-Dimensionality-reductions" class="headerlink" title="3.1.2 Dimensionality reductions"></a>3.1.2 Dimensionality reductions</h5><p>在处理大型数据集时，应用某种降维方法通常是有益的。通过将数据投影到较低维度的子空间上，人们通常能够显着减少噪声量。另一个好处是，在2维或3维子空间中可视化数据通常要容易得多。我们已经讨论过PCA和t-SNE。</p>
<h5 id="3-1-3-Clustering-methods"><a href="#3-1-3-Clustering-methods" class="headerlink" title="3.1.3 Clustering methods"></a>3.1.3 Clustering methods</h5><p>无监督聚类在许多不同的应用中是有用的，并且已经在机器学习中被广泛研究。一些最流行的方法是层次聚类，k均值聚类和基于图的聚类。</p>
<h6 id="3-1-3-1-Hierarchical-clustering（层次聚类）"><a href="#3-1-3-1-Hierarchical-clustering（层次聚类）" class="headerlink" title="3.1.3.1 Hierarchical clustering（层次聚类）"></a>3.1.3.1 Hierarchical clustering（层次聚类）</h6><p>在分层聚类中，可以使用自下而上或自上而下的方法。在前一种情况下，每个单元最初被分配到它自己的集群，随后合并成对的集合以创建一个hieararchy：<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/hierarchical_clustering1.png" alt="image"><br>原始的数据<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/hierarchical_clustering2.png" alt="image"><br>使用自上而下策略，首先从一个集群中的所有观察开始，然后递归地拆分每个集群以形成层次结构。该策略的一个优点是该方法是确定性的。</p>
<h6 id="3-1-3-2-k-means-k均值聚类"><a href="#3-1-3-2-k-means-k均值聚类" class="headerlink" title="3.1.3.2 k-means(k均值聚类)"></a>3.1.3.2 k-means(k均值聚类)</h6><p>在k均值聚类中，目标是将N个单元划分为k个不同的聚类。以迭代方式，分配集群中心，并将每个单元分配给最近的集群：<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/k-means.png" alt="image"><br>用于scRNA-seq分析的大多数方法在某些时候包括k均值步骤。</p>
<h6 id="3-1-3-3-Graph-based-methods（基于图的聚类）"><a href="#3-1-3-3-Graph-based-methods（基于图的聚类）" class="headerlink" title="3.1.3.3 Graph-based methods（基于图的聚类）"></a>3.1.3.3 Graph-based methods（基于图的聚类）</h6><p>在过去的二十年中，人们对分析各个领域的网络产生了很大的兴趣。一个目标是识别网络中节点的组或模块。<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/graph_network.jpg" alt="image"><br>通过构建图表，其中一些方法可以应用于scRNA-seq数据，其中每个节点代表一个细胞。请注意，构造图形并为边缘指定权重并非易事。基于图的方法的一个优点是它们中的一些非常有效并且可以应用于包含数百万个节点的网络。</p>
<h5 id="3-1-4-Challenge-in-clustering"><a href="#3-1-4-Challenge-in-clustering" class="headerlink" title="3.1.4 Challenge in clustering"></a>3.1.4 Challenge in clustering</h5><ul>
<li>What is the number of clusters k?</li>
<li>What is a cell type?</li>
<li>Scalability: in the last few years the number of cells in scRNA-seq experiments has grown by several orders of magnitude </li>
</ul>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">from ~ </span><br><span class="line">10^2to ~ 10^6</span><br></pre></td></tr></table></figure>

<ul>
<li>Tools are not user-friendly<h5 id="3-1-5-Tools-for-scRNA-seq-data"><a href="#3-1-5-Tools-for-scRNA-seq-data" class="headerlink" title="3.1.5 Tools for scRNA-seq data"></a>3.1.5 Tools for scRNA-seq data</h5><h6 id="3-1-5-1-SINCERA"><a href="#3-1-5-1-SINCERA" class="headerlink" title="3.1.5.1 SINCERA"></a>3.1.5.1 SINCERA</h6></li>
<li>SINCERA（Guo等人，2015）基于层次聚类</li>
<li>在聚类之前将数据转换为z分数</li>
<li>通过查找层次结构中的第一个单例群集来识别k</li>
</ul>
<h6 id="3-1-5-2-pcaReduce"><a href="#3-1-5-2-pcaReduce" class="headerlink" title="3.1.5.2 pcaReduce"></a>3.1.5.2 pcaReduce</h6><p>pcaReduce (žurauskienė and Yau 2016) combines PCA, k-means and “iterative” hierarchical clustering. Starting from a large number of clusters pcaReduce iteratively merges similar clusters; after each merging event it removes the principle component explaning the least variance in the data.</p>
<h6 id="3-1-5-3-SC3"><a href="#3-1-5-3-SC3" class="headerlink" title="3.1.5.3 SC3"></a>3.1.5.3 SC3</h6><p><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/sc3.png" alt="image"></p>
<ul>
<li>SC3（Kiselev等人，2017）基于PCA和光谱维数减少</li>
<li>利用k-means</li>
<li>另外执行一致性聚类<h6 id="3-1-5-4-tSNE-k-means"><a href="#3-1-5-4-tSNE-k-means" class="headerlink" title="3.1.5.4 tSNE+k-means"></a>3.1.5.4 tSNE+k-means</h6></li>
<li>基于tSEN</li>
<li>利用k-means<h6 id="3-1-5-5-SNN-Cliq"><a href="#3-1-5-5-SNN-Cliq" class="headerlink" title="3.1.5.5 SNN-Cliq"></a>3.1.5.5 SNN-Cliq</h6>SNN-Cliq（C. Xu和Su 2015）是一种基于图的方法。首先，该方法根据距离测量识别每个cell的k-最近邻居。这用于计算每对cell之间的共享最近邻居（SNN）的数量。如果它们具有至少一个SNN，则通过在两个cell之间放置边来构建图。集群被定义为使用“clique”方法在它们之间具有许多边缘的cell组。SNN-Cliq需要手动定义几个参数。<h6 id="3-1-5-6-Seurat-clustering"><a href="#3-1-5-6-Seurat-clustering" class="headerlink" title="3.1.5.6 Seurat clustering"></a>3.1.5.6 Seurat clustering</h6>Seurat聚类基于类似于SNN-Cliq的社区检测方法以及之前提出的用于分析CyTOF数据的方法（Levine等人2015）。由于Seurat变得更像是scRNA-seq数据分析的一体化工具，我们专门用一章来更详细地讨论它。<h5 id="3-1-6-Comparing-clustering"><a href="#3-1-6-Comparing-clustering" class="headerlink" title="3.1.6 Comparing clustering"></a>3.1.6 Comparing clustering</h5></li>
</ul>
<p>为了比较两组聚类标签，我们可以使用调整后的Rand索引。该指数衡量两个数据集群之间的相似性。调整后的兰德指数的值在于[0;1]间隔，在哪里1意味着两个聚类是相同的0表示偶然预期的相似程度。</p>
<h4 id="3-2-Clustering-example"><a href="#3-2-Clustering-example" class="headerlink" title="3.2 Clustering example"></a>3.2 Clustering example</h4><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">library</span><span class="params">(pcaMethods)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(pcaReduce)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(SC3)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(scater)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(SingleCellExperiment)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(pheatmap)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(mclust)</span></span></span><br><span class="line">set.seed(<span class="number">1234567</span>)</span><br></pre></td></tr></table></figure>

<p>为了说明scRNA-seq数据的聚类，我们考虑了来自发育小鼠胚胎的细胞的Deng数据集（Deng等人，2014）。我们已经预处理了数据集并提前创建了一个SingleCellExperiment对象。我们还使用原始出版物中标识的cell类型注释了cell（它是colData中的cell_type2列）。</p>
<h5 id="3-2-1-Deng-dataset"><a href="#3-2-1-Deng-dataset" class="headerlink" title="3.2.1 Deng dataset"></a>3.2.1 Deng dataset</h5><p>载入数据</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">deng &lt;- readRDS(<span class="string">"deng/deng-reads.rds"</span>)</span><br><span class="line">deng</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">## <span class="keyword">class</span>: SingleCellExperiment </span><br><span class="line">## dim: <span class="number">22431</span> <span class="number">268</span> </span><br><span class="line">## metadata(<span class="number">0</span>):</span><br><span class="line">## assays(<span class="number">2</span>): counts logcounts</span><br><span class="line">## rownames(<span class="number">22431</span>): Hvcn1 Gbp7 ... Sox5 Alg11</span><br><span class="line">## rowData names(<span class="number">10</span>): feature_symbol is_feature_control ...</span><br><span class="line">##   total_counts log10_total_counts</span><br><span class="line">## colnames(<span class="number">268</span>): <span class="number">16</span>cell <span class="number">16</span>cell<span class="number">.1</span> ... zy<span class="number">.2</span> zy<span class="number">.3</span></span><br><span class="line">## colData names(<span class="number">30</span>): cell_type2 cell_type1 ... pct_counts_ERCC</span><br><span class="line">##   is_cell_control</span><br><span class="line">## reducedDimNames(<span class="number">0</span>):</span><br><span class="line">## spikeNames(<span class="number">1</span>): ERCC</span><br></pre></td></tr></table></figure>

<p>查看细胞类型注释：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">table(<span class="name">colData</span>(<span class="name">deng</span>)$cell_type2)</span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">## </span><br><span class="line">##     <span class="number">16</span>cell      <span class="number">4</span>cell      <span class="number">8</span>cell early2cell earlyblast  late2cell </span><br><span class="line">##         <span class="number">50</span>         <span class="number">14</span>         <span class="number">37</span>          <span class="number">8</span>         <span class="number">43</span>         <span class="number">10</span> </span><br><span class="line">##  lateblast   mid2cell   midblast         zy </span><br><span class="line">##         <span class="number">30</span>         <span class="number">12</span>         <span class="number">60</span>          <span class="number">4</span></span><br></pre></td></tr></table></figure>

<p>一个简单的PCA分析已经分离了一些强大的细胞类型，并提供了数据结构的一些见解：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">plotPCA</span><span class="params">(deng, colour_by = <span class="string">"cell_type2"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-5-1.png" alt="image"><br>正如你所看到的，早期的细胞类型分离得很好，但是三个胚泡的时间点更难以区分。</p>
<h5 id="3-2-2-SC3"><a href="#3-2-2-SC3" class="headerlink" title="3.2.2 SC3"></a>3.2.2 SC3</h5><p>让我们对deng数据进行SC3聚类。SC3的优点是它可以直接摄取SingleCellExperiment对象。<br>现在让我们想象一下，我们不知道集群k（cell类型）的数量。SC3可以为您估计一些集群：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">deng &lt;- sc3_estimate_k(deng)</span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># Estimating k...</span></span></span><br><span class="line"><span class="meta">metadata(deng)$</span><span class="bash">sc3<span class="variable">$k_estimation</span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment"># [1] 6</span></span></span><br></pre></td></tr></table></figure>

<p>有趣的是，SC3预测的细胞类型数量小于原始数据注释。然而，在不同细胞类型的早期，中期和晚期阶段，我们将恰好有6种细胞类型。我们将合并的单元格类型存储在colData的cell_type1列中：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">plotPCA</span><span class="params">(deng, colour_by = <span class="string">"cell_type1"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-7-1.png" alt="image"><br>现在我们已经准备好运行SC3了（我们还要求它计算集群的生物属性）：</p>
<figure class="highlight xml"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">deng <span class="tag">&lt;<span class="name">-</span> <span class="attr">sc3</span>(<span class="attr">deng</span>, <span class="attr">ks</span> = <span class="string">10,</span> <span class="attr">biology</span> = <span class="string">TRUE)</span></span></span><br></pre></td></tr></table></figure>

<p>SC3的结果由几个不同的输出组成，这里我们展示了其中的一些：<br>共识度矩阵（Consensus matrix）:<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-9-1.png" alt="image"><br>轮廓图（silhouette plot）：Silhouette是指在数据集群内解释和验证一致性的方法。该技术提供了一个简洁的图形表示，表明每个对象在其集群中的位置。<br>轮廓值是对象与其自身群集（内聚力）相比与其他群集（分离）相似程度的度量。轮廓范围从-1到+1，其中高值表示对象与其自己的簇很好地匹配并且与相邻簇不匹配。如果大多数对象具有高值，则群集配置是合适的。如果许多点具有低值或负值，则群集配置可能具有太多或太少的群集。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">sc3_plot_silhouette</span><span class="params">(deng, k = <span class="number">10</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-10-1.png" alt="image"><br>表达矩阵的热图（Heatmap of the expression matrix:）：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">sc3_plot_expression</span><span class="params">(deng, k = <span class="number">10</span>, show_pdata = <span class="string">"cell_type2"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-11-1.png" alt="image"><br>确定marker基因：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">sc3_plot_markers</span><span class="params">(deng, k = <span class="number">10</span>, show_pdata = <span class="string">"cell_type2"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-12-1.png" alt="image"><br>PCA图，突出显示的SC3集群：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">plotPCA</span><span class="params">(deng, colour_by = <span class="string">"sc3_10_clusters"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-13-1.png" alt="image"><br>将SC3集群的结果与原始的cell类型标签进行比较：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adjustedRandIndex(<span class="name">colData</span>(<span class="name">deng</span>)$cell_type2, colData(<span class="name">deng</span>)$sc3_10_clusters)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">0.7705208</span></span><br></pre></td></tr></table></figure>

<p>兰德指数（Rand index）在统计学中，特别是在数据聚类中，是两个数据聚类之间相似性的度量。可以定义Rand索引的形式，其针对元素的机会分组进行调整，这是调整后的Rand索引。从数学的角度来看，Rand指数与准确性有关，但即使不使用类标签也适用。<br>SC3也可以在一个交互式的Shiny会话中运行：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">sc3_interactive</span><span class="params">(deng)</span></span></span><br></pre></td></tr></table></figure>

<p>由于直接计算距离，SC3在细胞数量&gt;5000时变得非常缓慢,对于包含的大型数据集达 </p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">10^5</span><br></pre></td></tr></table></figure>

<p>我们用Seurat重新评论的细胞</p>
<h5 id="3-2-3-pcaReduce"><a href="#3-2-3-pcaReduce" class="headerlink" title="3.2.3 pcaReduce"></a>3.2.3 pcaReduce</h5><p>pcaReduce直接在表达式矩阵上运行。建议在运行pcaReduce之前使用gene filter和log transfoemation。我们将使用默认的SC3基因过滤器（请注意，默认情况下，scater对象的exprs会进行对数转换）。</p>
<figure class="highlight stata"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"># <span class="keyword">use</span> the same <span class="keyword">gene</span> filter <span class="keyword">as</span> <span class="keyword">in</span> SC3</span><br><span class="line"><span class="keyword">input</span> &lt;- logcounts(deng[rowData(deng)<span class="variable">$sc3_gene_filter</span>, ])</span><br></pre></td></tr></table></figure>

<p>pcaReduce使用了几个参数：</p>
<ul>
<li>nbt定义了pcaReduce运行的次数（它是随机的，在不同的运行后可能有不同的解决方案）</li>
<li>q定义了开始聚类的维数。输出将包含所有分区<br>ķ从2到q + 1。</li>
<li>method定义了用于聚类的方法。<br>S- 执行基于sampleing的合并，<br>M- 基于最大概率执行合并。<br>我们运行pcaReduce一次：</li>
</ul>
<figure class="highlight oxygene"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"># run pcaReduce <span class="number">1</span> time creating hierarchies <span class="keyword">from</span> <span class="number">1</span> <span class="keyword">to</span> <span class="number">30</span> clusters</span><br><span class="line">pca.red &lt;- PCAreduce(t(input), nbt = <span class="number">1</span>, q = <span class="number">30</span>, <span class="function"><span class="keyword">method</span> = '<span class="title">S</span>')[[1]]</span></span><br></pre></td></tr></table></figure>

<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">colData</span><span class="params">(deng)</span></span><span class="variable">$pcaReduce</span> &lt;- as.character(pca<span class="selector-class">.red</span>[,<span class="number">32</span> - <span class="number">10</span>])</span><br><span class="line"><span class="function"><span class="title">plotPCA</span><span class="params">(deng, colour_by = <span class="string">"pcaReduce"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-18-1.png" alt="image"></p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">colData</span><span class="params">(deng)</span></span><span class="variable">$pcaReducek2</span> &lt;- as.character(pca<span class="selector-class">.red</span>[,<span class="number">32</span> - <span class="number">2</span>])</span><br><span class="line"><span class="function"><span class="title">plotPCA</span><span class="params">(deng, colour_by = <span class="string">"pcaReducek2"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/clust-pca-reduce2-1.png" alt="image"></p>
<h5 id="3-2-4-tSEN-kmeans"><a href="#3-2-4-tSEN-kmeans" class="headerlink" title="3.2.4 tSEN+kmeans"></a>3.2.4 tSEN+kmeans</h5><p>我们之前看到的tSNE图，我们使用scater是由Rtsne和ggplot2包构建的。在这里我们将做同样的事情：</p>
<figure class="highlight xml"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">deng <span class="tag">&lt;<span class="name">-</span> <span class="attr">plotTSNE</span>(<span class="attr">deng</span>, <span class="attr">rand_seed</span> = <span class="string">1,</span> <span class="attr">return_SCE</span> = <span class="string">TRUE)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/clust-tsne-1.png" alt="image"><br>请注意，上图中的所有点都是黑色的。这与我们之前看到的不同，当时细胞基于注释着色。这里我们没有任何注释，所有单元格都来自同一批次，因此所有点都是黑色的。<br>现在我们将k-means聚类算法应用于tSNE图上的点。你在中看到了多少组？</p>
<p>我们将从ķ=8开始</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">colData</span><span class="params">(deng)</span></span><span class="variable">$tSNE_kmeans</span> &lt;- as.character(kmeans(deng@reducedDims<span class="variable">$TSNE</span>, centers = <span class="number">8</span>)<span class="variable">$clust</span>)</span><br><span class="line"><span class="function"><span class="title">plotTSNE</span><span class="params">(deng, rand_seed = <span class="number">1</span>, colour_by = <span class="string">"tSNE_kmeans"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/clust-tsne-kmeans2-1.png" alt="image"><br>您可能已经注意到，pcaReduce和tSNE + kmeans都是随机的，每次运行时都会得到不同的结果。为了更好地了解解决方案，我们需要多次运行这些方法。 SC3也是随机的，但由于共识步骤，它更稳健，不太可能产生不同的结果。</p>
<h5 id="3-2-5-SNN-Clip"><a href="#3-2-5-SNN-Clip" class="headerlink" title="3.2.5 SNN-Clip"></a>3.2.5 SNN-Clip</h5><p>使用作者提供的SNN-cliq参数来运行：</p>
<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">distan</span> &lt;- <span class="string">"euclidean"</span></span><br><span class="line"><span class="title">par</span>.k &lt;- <span class="number">3</span></span><br><span class="line"><span class="title">par</span>.r &lt;- <span class="number">0.7</span></span><br><span class="line"><span class="title">par</span>.m &lt;- <span class="number">0.5</span></span><br><span class="line"><span class="meta"># construct a graph</span></span><br><span class="line"><span class="title">scRNA</span>.seq.funcs::<span class="type">SNN</span>(</span><br><span class="line">    <span class="class"><span class="keyword">data</span> = t(<span class="title">input</span>),</span></span><br><span class="line">    outfile = <span class="string">"snn-cliq.txt"</span>,</span><br><span class="line">    k = par.k,</span><br><span class="line">    distance = distan</span><br><span class="line">)</span><br><span class="line"><span class="meta"># find clusters in the graph</span></span><br><span class="line"><span class="title">snn</span>.res &lt;- </span><br><span class="line">    system(</span><br><span class="line">        paste0(</span><br><span class="line">            <span class="string">"python utils/Cliq.py "</span>, </span><br><span class="line">            <span class="string">"-i snn-cliq.txt "</span>,</span><br><span class="line">            <span class="string">"-o res-snn-cliq.txt "</span>,</span><br><span class="line">            <span class="string">"-r "</span>, par.r,</span><br><span class="line">            <span class="string">" -m "</span>, par.m</span><br><span class="line">        ),</span><br><span class="line">        intern = <span class="type">TRUE</span></span><br><span class="line">    )</span><br><span class="line"><span class="title">cat</span>(paste(snn.res, collapse = <span class="string">"\n"</span>))</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">## input file snn-cliq.txt</span><br><span class="line">## find <span class="number">66</span> quasi-cliques</span><br><span class="line">## merged into <span class="number">29</span> clusters</span><br><span class="line">## unique assign done</span><br></pre></td></tr></table></figure>

<figure class="highlight vim"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">snn.<span class="keyword">res</span> &lt;- <span class="keyword">read</span>.table(<span class="string">"res-snn-cliq.txt"</span>)</span><br><span class="line"># <span class="built_in">remove</span> <span class="keyword">files</span> that were created during the analysis</span><br><span class="line"><span class="built_in">system</span>(<span class="string">"rm snn-cliq.txt res-snn-cliq.txt"</span>)</span><br><span class="line"></span><br><span class="line">colData(deng)$SNNCliq &lt;- <span class="keyword">as</span>.character(snn.<span class="keyword">res</span>[,<span class="number">1</span>])</span><br><span class="line">plotPCA(deng, colour_by = <span class="string">"SNNCliq"</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/unnamed-chunk-21-1.png" alt="image"></p>
<h5 id="3-2-6-SINCREA"><a href="#3-2-6-SINCREA" class="headerlink" title="3.2.6 SINCREA"></a>3.2.6 SINCREA</h5><p>如前一章所述，SINCERA基于层次聚类。要记住的一件重要事情是它在进行聚类之前执行基因水平的z分数转换：</p>
<figure class="highlight oxygene"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># perform gene-<span class="keyword">by</span>-gene per-sample z-score transformation</span><br><span class="line">dat &lt;- apply(input, <span class="number">1</span>, <span class="function"><span class="keyword">function</span><span class="params">(y)</span> <span class="title">scRNA</span>.<span class="title">seq</span>.<span class="title">funcs</span>:</span>:z.transform.helper(y))</span><br><span class="line"># hierarchical clustering</span><br><span class="line">dd &lt;- <span class="keyword">as</span>.dist((<span class="number">1</span> - cor(t(dat), <span class="function"><span class="keyword">method</span> = "<span class="title">pearson</span>"))/2)</span></span><br><span class="line"><span class="function"><span class="title">hc</span> &lt;- <span class="title">hclust</span><span class="params">(dd, <span class="keyword">method</span> = "average")</span></span></span><br></pre></td></tr></table></figure>

<p>如果群集的数量未知，则SINCERA可以将k识别为生成不超过指定数量的单个群集（仅包含1个小区的群集）的分层树的最小高度</p>
<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">num</span>.singleton &lt;- <span class="number">0</span></span><br><span class="line"><span class="title">kk</span> &lt;- <span class="number">1</span></span><br><span class="line"><span class="title">for</span> (i <span class="keyword">in</span> <span class="number">2</span>:dim(dat)[<span class="number">2</span>]) &#123;</span><br><span class="line">    clusters &lt;- cutree(hc, k = i)</span><br><span class="line">    clustersizes &lt;- <span class="keyword">as</span>.<span class="class"><span class="keyword">data</span>.frame(<span class="title">table</span>(<span class="title">clusters</span>))</span></span><br><span class="line">    singleton.clusters &lt;- which(clustersizes$<span class="type">Freq</span> &lt; <span class="number">2</span>)</span><br><span class="line">    <span class="keyword">if</span> (length(singleton.clusters) &lt;= num.singleton) &#123;</span><br><span class="line">        kk &lt;- i</span><br><span class="line">    &#125; <span class="keyword">else</span> &#123;</span><br><span class="line">        break;</span><br><span class="line">    &#125;</span><br><span class="line">&#125;</span><br><span class="line"><span class="title">cat</span>(kk)</span><br></pre></td></tr></table></figure>

<p>查看SINCERA结果做的热图：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">pheatmap(</span><br><span class="line">    <span class="literal">t</span>(<span class="name">dat</span>),</span><br><span class="line">    cluster_cols = hc,</span><br><span class="line">    cutree_cols = kk,</span><br><span class="line">    kmeans_k = <span class="number">100</span>,</span><br><span class="line">    show_rownames = FALSE</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/25-clustering_files/figure-html/clust-sincera-1.png" alt="image"></p>
<h4 id="3-3-Feature-Selection"><a href="#3-3-Feature-Selection" class="headerlink" title="3.3 Feature Selection"></a>3.3 Feature Selection</h4><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">library</span><span class="params">(scRNA.seq.funcs)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(matrixStats)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(M3Drop)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(RColorBrewer)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(SingleCellExperiment)</span></span></span><br><span class="line">set.seed(<span class="number">1</span>)</span><br></pre></td></tr></table></figure>

<p>单细胞RNASeq能够测量每个细胞中数千个基因的表达。然而，在大多数情况下，只有一部分将显示对感兴趣的生物学状况的反应，例如，细胞类型的差异，分化的驱动因素，对环境刺激的反应。由于技术噪音，在scRNASeq实验中检测到的大多数基因仅在不同水平检测到。这样做的一个结果是技术噪声和批量效应可能使感兴趣的生物信号模糊不清。</p>
<p>因此，通常有利的是进行特征选择以去除那些仅表现出来自下游分析的技术噪音的基因。这通常不仅会增加数据中的信噪比;它还通过减少要处理的数据总量来降低分析的计算复杂性。</p>
<p>对于scRNASeq数据，我们将关注无需监督的特征选择方法，这些方法不需要任何先验信息，例如细胞类型标记或生物组，因为对于许多实验它们不可用，或者可能不可靠。相反，差异表达（第3.6章）可以被认为是监督特征选择的一种形式，因为它使用每个样本的已知生物标记来识别在不同组间表达的不同水平的特征（即基因）。<br>仍然使用Deng的数据：</p>
<figure class="highlight elm"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">deng</span> &lt;- readRDS(<span class="string">"deng/deng-reads.rds"</span>)</span><br><span class="line"><span class="title">cellLabels</span> &lt;- colData(deng)$cell_type2</span><br></pre></td></tr></table></figure>

<p>可以使用M3Drop对该数据进行QCed并对文库大小进行标准化，从而去除检测到的基因很少的细胞，去除未检测到的基因，并将原始计数转换为CPM。</p>
<figure class="highlight maxima"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">deng_list &lt;- M3DropCleanData(</span><br><span class="line">    counts(deng),</span><br><span class="line">    <span class="built_in">labels</span> = cellLabels,</span><br><span class="line">    min_detected_genes = <span class="number">100</span>,</span><br><span class="line">    <span class="built_in">is</span>.counts = TRUE</span><br><span class="line">)</span><br><span class="line">expr_matrix &lt;- deng_list$data # Normalized &amp; filtered expression <span class="built_in">matrix</span></span><br><span class="line">celltype_labs &lt;- <span class="built_in">factor</span>(deng_list$<span class="built_in">labels</span>) # filtered cell-type <span class="built_in">labels</span></span><br><span class="line">cell_colors &lt;- brewer.pal(<span class="built_in">max</span>(<span class="number">3</span>,<span class="built_in">length</span>(<span class="built_in">unique</span>(celltype_labs))), <span class="string">"Set3"</span>)</span><br></pre></td></tr></table></figure>

<h5 id="3-3-1-Identify-Genes-vs-a-Null-model"><a href="#3-3-1-Identify-Genes-vs-a-Null-model" class="headerlink" title="3.3.1 Identify Genes vs a Null model"></a>3.3.1 Identify Genes vs a Null model</h5><p>无监督特征选择有两种主要方法。第一个是识别与仅描述数据集中预期的技术噪声的空模型不同的基因。</p>
<p>如果数据集包含spike-in RNA，它们可用于直接模拟技术噪声。然而，加spike-in的测量可能不会像内源性转录物那样经历相同的技术噪声（Svensson等，2017）。此外，scRNASeq实验通常只包含少量的加spike-in，这降低了我们对拟合模型参数的信心。</p>
<h6 id="3-3-1-1-Highly-Variable-Genes"><a href="#3-3-1-1-Highly-Variable-Genes" class="headerlink" title="3.3.1.1 Highly Variable Genes"></a>3.3.1.1 Highly Variable Genes</h6><p>提出用于鉴定scRNASeq数据集中的特征的第一种方法是鉴定高度可变的基因（HVG）。 HVG假设如果基因在细胞中的表达差异很大，那么这些差异中的一些差异是由于细胞之间的生物学差异而不是技术噪声。然而，由于计数数据的性质，基因的平均表达与细胞间读数的变化之间存在正相关关系。必须纠正这种关系才能正确识别HVG。<br>Brennecke等人提出了一种校正方差和平均表达之间关系的流行方法。为了使用Brennecke方法，我们首先对文库大小进行标准化，然后计算平均值和平方变异系数（变化除以平方均值）表达）。对于ERCC spike-in，二次曲线拟合这两个变量之间的关系，然后使用卡方检验找到显着高于曲线的基因。此方法作为Brennecke_getVariableGenes（count，spike）函数包含在M3Drop包中。但是，此数据集不包含spike-ins，因此我们将使用整个数据集来估算技术噪音。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plot(<span class="name">rowMeans</span>(<span class="name">expr_matrix</span>),rowVars(<span class="name">expr_matrix</span>),log='xy')</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/26-dropouts_files/figure-html/unnamed-chunk-6-1.png" alt="image"><br>在下图中，红色曲线是拟合的技术噪声模型，虚线是95％CI。粉红点是多次测试校正后具有显着生物变异性的基因。</p>
<figure class="highlight armasm"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">Brennecke_HVG </span>&lt;- <span class="keyword">BrenneckeGetVariableGenes(</span></span><br><span class="line"><span class="keyword"> </span>   expr_matrix,</span><br><span class="line">    fdr = <span class="number">0</span>.<span class="number">01</span>,</span><br><span class="line">    minBiolDisp = <span class="number">0</span>.<span class="number">5</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/26-dropouts_files/figure-html/unnamed-chunk-7-1.png" alt="image"></p>
<h6 id="3-3-1-2-High-Dropout-Genes"><a href="#3-3-1-2-High-Dropout-Genes" class="headerlink" title="3.3.1.2 High Dropout Genes"></a>3.3.1.2 High Dropout Genes</h6><p>寻找HVG的另一种方法是鉴定具有意外大量零的基因。被称为“dropout rate”的零的频率与scRNASeq数据中的表达水平密切相关。零是单细胞RNASeq数据的主要特征，通常占最终表达矩阵中一半以上的条目。这些零点主要是由于mRNA未能被逆转录而失败（Andrews和Hemberg，2016）。逆转录是酶反应，因此可以使用Michaelis-Menten方程建模：</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">P_dropout=1-S/(K+S)</span><br></pre></td></tr></table></figure>

<p>S是细胞中的mRNA浓度（我们将其估计为平均表达）和ķ是Michaelis-Menten常数。<br>因为Michaelis-Menten方程是一个凸非线性函数，在我们的数据集中，两个或更多的细胞群的不同表达的基因将会被转移到Michaelis-Menten模型中.</p>
<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br></pre></td><td class="code"><pre><span class="line">K &lt;- <span class="number">49</span></span><br><span class="line">S_sim &lt;- <span class="number">10</span>^seq(from = <span class="number">-3</span>, to = <span class="number">4</span>, by = <span class="number">0.05</span>) # range of expression values</span><br><span class="line">MM &lt;- <span class="number">1</span> - S_sim / (K + S_sim)</span><br><span class="line">plot(</span><br><span class="line">    S_sim, </span><br><span class="line">    MM, </span><br><span class="line">    type = <span class="string">"l"</span>, </span><br><span class="line">    lwd = <span class="number">3</span>, </span><br><span class="line">    xlab = <span class="string">"Expression"</span>, </span><br><span class="line">    ylab = <span class="string">"Dropout Rate"</span>, </span><br><span class="line">    xlim = c(<span class="number">1</span>,<span class="number">1000</span>)</span><br><span class="line">)</span><br><span class="line">S1 &lt;- <span class="number">10</span></span><br><span class="line">P1 &lt;- <span class="number">1</span> - S1 / (K + S1) # Expression &amp; dropouts for cells in condition <span class="number">1</span></span><br><span class="line">S2 &lt;- <span class="number">750</span></span><br><span class="line">P2 &lt;- <span class="number">1</span> - S2 / (K + S2) # Expression &amp; dropouts for cells in condition <span class="number">2</span></span><br><span class="line">points(</span><br><span class="line">    c(S1, S2),</span><br><span class="line">    c(P1, P2), </span><br><span class="line">    pch = <span class="number">16</span>, </span><br><span class="line">    col = <span class="string">"grey85"</span>, </span><br><span class="line">    cex = <span class="number">3</span></span><br><span class="line">)</span><br><span class="line">mix &lt;- <span class="number">0.5</span> # proportion of cells in condition <span class="number">1</span></span><br><span class="line">points(</span><br><span class="line">    S1 * mix + S2 * (<span class="number">1</span> - mix), </span><br><span class="line">    P1 * mix + P2 * (<span class="number">1</span> - mix), </span><br><span class="line">    pch = <span class="number">16</span>, </span><br><span class="line">    col = <span class="string">"grey35"</span>, </span><br><span class="line">    cex = <span class="number">3</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>我们使用M3Drop来识别在MM曲线右侧的重要异常值。我们也应用了1%的FDR多次测试修正：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">M3Drop_genes &lt;- M3DropFeatureSelection(</span><br><span class="line">    expr_matrix,</span><br><span class="line">    mt_method = <span class="string">"fdr"</span>,</span><br><span class="line">    mt_threshold = <span class="number">0.01</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/26-dropouts_files/figure-html/unnamed-chunk-10-1.png" alt="image"><br>M3Drop包中包含一种替代方法，该方法专门针对UMI标记数据而定制，该数据通常包含由低序列覆盖率产生的许多零，这是由于逆转录不足产生的。 该模型是深度调整的负二项式（Depth-Adjusted Negative Binomial ，DANB）。 该方法将每个表达观察描述为负二项式模型，其具有与相应基因的平均表达和相应细胞的测序深度相关的平均值，以及与基因的平均表达相关的方差。</p>
<p>与Michaelis-Menten和HVG方法不同，对于该模型选择的特征没有可靠的统计检验，因此我们将考虑前1500个基因。</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">deng_int &lt;- NBumiConvertToInteger(counts(deng))</span><br><span class="line">DANB_fit &lt;- NBumiFitModel(deng_int) # DANB is fit <span class="keyword">to</span> the<span class="built_in"> raw </span>count matrix</span><br><span class="line"><span class="comment"># Perform DANB feature selection</span></span><br><span class="line">DropFS &lt;- NBumiFeatureSelectionCombinedDrop(DANB_fit)</span><br><span class="line">DANB_genes &lt;- names(DropFS[1:1500])</span><br></pre></td></tr></table></figure>

<h5 id="3-3-2-Correlated-Expression"><a href="#3-3-2-Correlated-Expression" class="headerlink" title="3.3.2 Correlated Expression"></a>3.3.2 Correlated Expression</h5><p>一种完全不同的特征选择方法是使用基因 - 基因相关性。 该方法基于多个基因将在不同细胞类型或细胞状态之间差异表达的想法。 在相同细胞群中表达的基因将彼此正相关，其中在不同细胞群中表达的基因将彼此负相关。 因此，重要的基因可以通过它们与其他基因的相关程度来鉴定。<br>这种方法的局限性在于它假设技术噪声对于每个细胞是随机且独立的，因此不应产生基因 - 基因相关性，但是这种假设被批次效应所违反，批次效应通常在不同实验批次之间是系统的并且将产生基因 - 基因相关性。 因此，通过基因 - 基因相关性排序前几千个基因比考虑相关性的显着性更合适。</p>
<figure class="highlight oxygene"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">cor_mat &lt;- cor(t(expr_matrix), <span class="function"><span class="keyword">method</span> = "<span class="title">spearman</span>") # <span class="title">Gene</span>-<span class="title">gene</span> <span class="title">correlations</span></span></span><br><span class="line"><span class="function"><span class="title">diag</span><span class="params">(cor_mat)</span> &lt;- <span class="title">rep</span><span class="params">(0, times = nrow(expr_matrix)</span>)</span></span><br><span class="line"><span class="function"><span class="title">score</span> &lt;- <span class="title">apply</span><span class="params">(cor_mat, 1, <span class="keyword">function</span>(x)</span> <span class="comment">&#123;max(abs(x))&#125;</span>) #<span class="title">Correlation</span> <span class="title">of</span> <span class="title">highest</span> <span class="title">magnitude</span></span></span><br><span class="line"><span class="function"><span class="title">names</span><span class="params">(score)</span> &lt;- <span class="title">rownames</span><span class="params">(expr_matrix)</span>;</span></span><br><span class="line">score &lt;- score[<span class="keyword">order</span>(-score)]</span><br><span class="line">Cor_genes &lt;- names(score[<span class="number">1</span>:<span class="number">1500</span>])</span><br></pre></td></tr></table></figure>

<p>最后，scRNASeq数据中另一种常用的特征选择方法是使用PCA loading。 具有高PCA负荷的基因可能是高度可变的并且与许多其他可变基因相关，因此可能与潜在的生物学相关。 然而，与基因 - 基因相关性一样，PCA载量往往易于检测由于批次效应引起的系统变异; 因此，建议绘制PCA结果以确定与生物变异相对应的那些组分而不是批次效应。</p>
<figure class="highlight stata"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"># <span class="keyword">PCA</span> is typically performed <span class="keyword">on</span> <span class="keyword">log</span>-transformed expression data</span><br><span class="line"><span class="keyword">pca</span> &lt;- prcomp(<span class="built_in">log</span>(expr_matrix + 1) / <span class="built_in">log</span>(2))</span><br><span class="line"></span><br><span class="line"># <span class="keyword">plot</span> projection</span><br><span class="line"><span class="keyword">plot</span>(</span><br><span class="line">    <span class="keyword">pca</span><span class="variable">$rotation</span>[,1], </span><br><span class="line">    <span class="keyword">pca</span><span class="variable">$rotation</span>[,2], </span><br><span class="line">    pch = 16, </span><br><span class="line">    col = cell_colors[<span class="keyword">as</span>.<span class="keyword">factor</span>(celltype_labs)]</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>![image]<a href="https://hemberg-lab.github.io/scRNA.seq.course/26-dropouts_files/figure-html/unnamed-chunk-13-1.png" target="_blank" rel="noopener">https://hemberg-lab.github.io/scRNA.seq.course/26-dropouts_files/figure-html/unnamed-chunk-13-1.png</a>)</p>
<figure class="highlight stata"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"># calculate loadings <span class="keyword">for</span> components 1 and 2</span><br><span class="line"><span class="keyword">score</span> &lt;- rowSums(<span class="built_in">abs</span>(<span class="keyword">pca</span><span class="variable">$x</span>[,c(1,2)])) </span><br><span class="line">names(<span class="keyword">score</span>) &lt;- rownames(expr_matrix)</span><br><span class="line"><span class="keyword">score</span> &lt;- <span class="keyword">score</span>[<span class="keyword">order</span>(-<span class="keyword">score</span>)]</span><br><span class="line">PCA_genes &lt;- names(<span class="keyword">score</span>[1:1500])</span><br></pre></td></tr></table></figure>

<h5 id="3-3-3-Comparing-Methods"><a href="#3-3-3-Comparing-Methods" class="headerlink" title="3.3.3 Comparing Methods"></a>3.3.3 Comparing Methods</h5><p>我们可以检查这些被识别的特征是否真的代表了在这个数据集中的细胞类型之间的不同表达。</p>
<figure class="highlight autohotkey"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">MDrop：：M3DropExpressionHeatmap(</span><br><span class="line"><span class="built_in">    M3Drop_genes,</span></span><br><span class="line"><span class="built_in">    expr_matrix,</span></span><br><span class="line">    cell_labels = celltype_labs</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/26-dropouts_files/figure-html/unnamed-chunk-15-1.png" alt="image"><br>我们还可以考虑使用Jaccard索引的查看特性选择方法的一致性：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">J &lt;- sum(<span class="name">M3Drop_genes</span> %in% HVG_genes)/length(<span class="name">unique</span>(<span class="name">c</span>(<span class="name">M3Drop_genes</span>, HVG_genes)))</span><br></pre></td></tr></table></figure>

<h4 id="3-4-Pseudotime-analysis"><a href="#3-4-Pseudotime-analysis" class="headerlink" title="3.4 Pseudotime analysis"></a>3.4 Pseudotime analysis</h4><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">library</span><span class="params">(SingleCellExperiment)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(TSCAN)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(M3Drop)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(monocle)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(destiny)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(SLICER)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(ouija)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(scater)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(ggplot2)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(ggthemes)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(ggbeeswarm)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(corrplot)</span></span></span><br><span class="line">set.seed(<span class="number">1</span>)</span><br></pre></td></tr></table></figure>

<p>在许多情况下，人们正在研究细胞不断变化的过程。这包括，例如，在开发过程中发生的许多分化过程：在刺激之后，细胞将从一种细胞类型变为另一种细胞类型。理想情况下，我们希望监测单个细胞随时间的表达水平。不幸的是，用scRNA-seq进行这种监测是不可能的，因为当提取RNA时细胞被裂解（破坏）。</p>
<p>相反，我们必须在多个时间点进行采样并获得基因表达谱的快照。由于一些细胞将在分化过程中比其他细胞更快地进行，因此每个快照可以在沿着发育进程的不同点处包含细胞。我们使用统计方法沿着一个或多个轨迹对细胞进行排序，这些轨迹代表潜在的发育轨迹，这种排序被称为“pseudotime”。</p>
<p>在本章中，我们将考虑五种不同的工具：Monocle，TSCAN，destiny，SLICER和ouija，用于根据伪时间发展对细胞进行排序。为了说明我们将使用小鼠胚胎发育数据集的方法（Deng et al.2014）。该数据集由来自早期小鼠发育的10个不同时间点的268个细胞组成。在这种情况下，不需要伪时间对准，因为单元标签提供关于发展轨迹的信息。因此，标签允许我们建立基本事实，以便我们可以评估和比较不同的方法。</p>
<p>Cannoodt等人最近的综述提供了来自单细胞转录组学的轨迹推断的各种计算方法的详细总结（Cannoodt，Saelens和Saeys 2016）。他们讨论了几种工具，但不幸的是，出于我们的目的，许多这些工具没有完整或维护良好的实现，和/或没有在R中实现。<br>下面的图表总结了各种工具的一些特性：<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/cannoodt_pseudotime_properties.png" alt="image"><br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/cannoodt_pseudotime_methods.png" alt="image"></p>
<h5 id="3-4-1-First-look-at-Deng-data"><a href="#3-4-1-First-look-at-Deng-data" class="headerlink" title="3.4.1 First look at Deng data"></a>3.4.1 First look at Deng data</h5><p>让我们先来看看Deng的数据，还没有应用复杂的pseudotime方法。 如下图所示，简单的PCA可以很好地显示这些数据中的结构。 只有在胚细胞类型（“earlyblast”，“midblat”，“latenlat”）时，PCA难以分离不同的细胞类型。</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">deng_SCE &lt;- readRDS(<span class="string">"deng/deng-reads.rds"</span>)</span><br><span class="line">deng_SCE$cell_type2 &lt;- factor(</span><br><span class="line">    deng_SCE$cell_type2,</span><br><span class="line">    levels = c(<span class="string">"zy"</span>, <span class="string">"early2cell"</span>, <span class="string">"mid2cell"</span>, <span class="string">"late2cell"</span>,</span><br><span class="line">                        <span class="string">"4cell"</span>, <span class="string">"8cell"</span>, <span class="string">"16cell"</span>, <span class="string">"earlyblast"</span>,</span><br><span class="line">                        <span class="string">"midblast"</span>, <span class="string">"lateblast"</span>)</span><br><span class="line">)</span><br><span class="line">cellLabels &lt;- deng_SCE$cell_type2</span><br><span class="line">deng &lt;- counts(deng_SCE)</span><br><span class="line">colnames(deng) &lt;- cellLabels</span><br><span class="line">deng_SCE &lt;- plotPCA(deng_SCE, colour_by = <span class="string">"cell_type2"</span>, </span><br><span class="line">                    return_SCE = TRUE)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/data-overview-1.png" alt="image"></p>
<figure class="highlight erlang"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">deng_SCE$PC1 &lt;- reducedDim(deng_SCE, <span class="string">"PCA"</span>)[,<span class="number">1</span>]</span><br><span class="line"><span class="function"><span class="title">ggplot</span><span class="params">(as.data.frame(colData(deng_SCE)), aes(x = PC1, y = cell_type2, </span></span></span><br><span class="line"><span class="function"><span class="params">                              colour = cell_type2))</span> +</span></span><br><span class="line"><span class="function">    <span class="title">geom_quasirandom</span><span class="params">(groupOnX = FALSE)</span> +</span></span><br><span class="line"><span class="function">    <span class="title">scale_color_tableau</span><span class="params">()</span> + <span class="title">theme_classic</span><span class="params">()</span> +</span></span><br><span class="line"><span class="function">    <span class="title">xlab</span><span class="params">(<span class="string">"First principal component"</span>)</span> + <span class="title">ylab</span><span class="params">(<span class="string">"Timepoint"</span>)</span> +</span></span><br><span class="line"><span class="function">    <span class="title">ggtitle</span><span class="params">(<span class="string">"Cells ordered by first principal component"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/pca-pseudotime-1.png" alt="image"><br>正如上面的图所显示的，PC1在发育时期的早期和晚期都很难正确地排列细胞，但是总的来说，在发育的时间点上对细胞进行排序是相对较好的。</p>
<h5 id="3-4-2-TSCAN"><a href="#3-4-2-TSCAN" class="headerlink" title="3.4.2 TSCAN"></a>3.4.2 TSCAN</h5><p>TSCAN将聚类与pseudotime分析相结合。 首先，它使用mclust聚类细胞，mclust基于正态分布的混合。 然后，它构建一个最小生成树来连接集群。 连接最大数量的簇的该树的分支是用于确定pseudotime的主分支。</p>
<figure class="highlight elixir"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">procdeng &lt;- TSCAN::preprocess(deng)</span><br><span class="line">colnames(procdeng) &lt;- <span class="number">1</span><span class="symbol">:ncol</span>(deng)</span><br><span class="line">dengclust &lt;- TSCAN::exprmclust(procdeng, clusternum = <span class="number">10</span>)</span><br><span class="line">TSCAN::plotmclust(dengclust)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/tscan-all-genes-1.png" alt="image"></p>
<figure class="highlight php"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">dengorderTSCAN &lt;- TSCAN::TSCANorder(dengclust, orderonly = <span class="keyword">FALSE</span>)</span><br><span class="line">pseudotime_order_tscan &lt;- <span class="keyword">as</span>.character(dengorderTSCAN$sample_name)</span><br><span class="line">deng_SCE$pseudotime_order_tscan &lt;- NA</span><br><span class="line">deng_SCE$pseudotime_order_tscan[<span class="keyword">as</span>.numeric(dengorderTSCAN$sample_name)] &lt;- </span><br><span class="line">    dengorderTSCAN$Pseudotime</span><br></pre></td></tr></table></figure>

<p>令人沮丧的是，TSCAN只提供268个细胞中的221个的pseudotime，默默地返回未分配cell的缺失值。</p>
<figure class="highlight accesslog"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cellLabels<span class="string">[dengclust$clusterid == 10]</span></span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">##  [<span class="number">1</span>] late2cell late2cell late2cell late2cell late2cell late2cell late2cell</span><br><span class="line">##  [<span class="number">8</span>] late2cell late2cell late2cell</span><br><span class="line">## <span class="number">10</span> Levels: zy early2cell mid2cell late2cell <span class="number">4</span>cell <span class="number">8</span>cell ... lateblast</span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">ggplot(<span class="name">as</span>.data.frame(<span class="name">colData</span>(<span class="name">deng_SCE</span>)), </span><br><span class="line">       aes(<span class="name">x</span> = pseudotime_order_tscan, </span><br><span class="line">           y = cell_type2, colour = cell_type2)) +</span><br><span class="line">    geom_quasirandom(<span class="name">groupOnX</span> = FALSE) +</span><br><span class="line">    scale_color_tableau() + theme_classic() +</span><br><span class="line">    xlab(<span class="string">"TSCAN pseudotime"</span>) + ylab(<span class="string">"Timepoint"</span>) +</span><br><span class="line">    ggtitle(<span class="string">"Cells ordered by TSCAN pseudotime"</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/tscan-vs-truth-1.png" alt="image"><br>TSCAN的发展轨迹是“错误的方式”，因为后来的pseudotime 值对应于早期时间点，反之亦然。 这本身并不是一个问题（很容易扭转排序以获得伪时间的直观解释），但总体而言，建议TSCAN在此数据集上的性能优于PCA将是一个延伸。 （因为这是一种基于PCA的方法，也许这并不奇怪。）</p>
<h5 id="3-4-3-monocle"><a href="#3-4-3-monocle" class="headerlink" title="3.4.3 monocle"></a>3.4.3 monocle</h5><p>生成树以连接所有细胞。 Monocle然后识别该树中最长的路径以确定伪时间。 如果数据包含发散轨迹（即一种细胞类型区分为两种不同的细胞类型），则Monacle可以识别这些。 每个得到的分叉路径被定义为单独的cell状态。</p>
<p>不幸的是，Monocle在使用所有基因时都不起作用，因此我们必须进行特征选择。 首先，我们使用M3Drop：</p>
<figure class="highlight applescript"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">m3dGenes &lt;- <span class="keyword">as</span>.<span class="built_in">character</span>(</span><br><span class="line">    M3DropFeatureSelection(deng)$Gene</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">## Warning <span class="keyword">in</span> bg__calc_variables(expr_mat): Warning: Removing <span class="number">1134</span> undetected</span><br><span class="line">## genes.</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/m3d-select-genes-1.png" alt="image"></p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">d &lt;- deng[which(rownames(deng) %<span class="keyword">in</span>% m3dGenes), ]</span><br><span class="line">d &lt;- d[!duplicated(rownames(d)), ]</span><br></pre></td></tr></table></figure>

<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">colnames</span>(d) &lt;- <span class="number">1</span>:ncol(d)</span><br><span class="line"><span class="title">geneNames</span> &lt;- rownames(d)</span><br><span class="line"><span class="title">rownames</span>(d) &lt;- <span class="number">1</span>:nrow(d)</span><br><span class="line"><span class="title">pd</span> &lt;- <span class="class"><span class="keyword">data</span>.frame(<span class="title">timepoint</span> = <span class="title">cellLabels</span>)</span></span><br><span class="line"><span class="title">pd</span> &lt;- new(<span class="string">"AnnotatedDataFrame"</span>, <span class="class"><span class="keyword">data</span>=pd)</span></span><br><span class="line"><span class="title">fd</span> &lt;- <span class="class"><span class="keyword">data</span>.frame(<span class="title">gene_short_name</span> = <span class="title">geneNames</span>)</span></span><br><span class="line"><span class="title">fd</span> &lt;- new(<span class="string">"AnnotatedDataFrame"</span>, <span class="class"><span class="keyword">data</span>=fd)</span></span><br><span class="line"></span><br><span class="line"><span class="title">dCellData</span> &lt;- newCellDataSet(d, phenoData = pd, featureData = fd, expressionFamily = tobit())</span><br><span class="line"><span class="title">dCellData</span> &lt;- setOrderingFilter(dCellData, which(geneNames %<span class="keyword">in</span>% m3dGenes))</span><br><span class="line"><span class="title">dCellData</span> &lt;- estimateSizeFactors(dCellData)</span><br><span class="line"><span class="title">dCellDataSet</span> &lt;- reduceDimension(dCellData, pseudo_expr = <span class="number">1</span>)</span><br><span class="line"><span class="title">dCellDataSet</span> &lt;- orderCells(dCellDataSet, reverse = <span class="type">FALSE</span>)</span><br><span class="line"><span class="title">plot_cell_trajectory</span>(dCellDataSet)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/monocle-all-genes-1.png" alt="image"></p>
<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta"># Store the ordering</span></span><br><span class="line"><span class="title">pseudotime_monocle</span> &lt;-</span><br><span class="line">    <span class="class"><span class="keyword">data</span>.frame(</span></span><br><span class="line"><span class="class">        <span class="type">Timepoint</span> = <span class="title">phenoData</span>(<span class="title">dCellDataSet</span>)$timepoint,</span></span><br><span class="line">        pseudotime = phenoData(dCellDataSet)$<span class="type">Pseudotime</span>,</span><br><span class="line">        <span class="type">State</span> = phenoData(dCellDataSet)$<span class="type">State</span></span><br><span class="line">    )</span><br><span class="line"><span class="title">rownames</span>(pseudotime_monocle) &lt;- <span class="number">1</span>:ncol(d)</span><br><span class="line"><span class="title">pseudotime_order_monocle</span> &lt;-</span><br><span class="line">    rownames(pseudotime_monocle[order(pseudotime_monocle$pseudotime), ])</span><br></pre></td></tr></table></figure>

<figure class="highlight erlang"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">deng_SCE$pseudotime_monocle &lt;- pseudotime_monocle$pseudotime</span><br><span class="line"><span class="function"><span class="title">ggplot</span><span class="params">(as.data.frame(colData(deng_SCE)), </span></span></span><br><span class="line"><span class="function"><span class="params">       aes(x = pseudotime_monocle, </span></span></span><br><span class="line"><span class="function"><span class="params">           y = cell_type2, colour = cell_type2))</span> +</span></span><br><span class="line"><span class="function">    <span class="title">geom_quasirandom</span><span class="params">(groupOnX = FALSE)</span> +</span></span><br><span class="line"><span class="function">    <span class="title">scale_color_tableau</span><span class="params">()</span> + <span class="title">theme_classic</span><span class="params">()</span> +</span></span><br><span class="line"><span class="function">    <span class="title">xlab</span><span class="params">(<span class="string">"monocle pseudotime"</span>)</span> + <span class="title">ylab</span><span class="params">(<span class="string">"Timepoint"</span>)</span> +</span></span><br><span class="line"><span class="function">    <span class="title">ggtitle</span><span class="params">(<span class="string">"Cells ordered by monocle pseudotime"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/monocle-vs-truth-1.png" alt="image"><br>Monocle - 至少是默认设置 - 对这些数据表现不佳。 “late2cell”组与“zy”，“early2cell”和“mid2cell”细胞完全分离（尽管这些是正确排序的），并且“4cell”，“8cell”，“16cell”或 任何blast胚细胞组之间没有分离</p>
<h5 id="3-4-4-Diffusion-maps"><a href="#3-4-4-Diffusion-maps" class="headerlink" title="3.4.4 Diffusion maps"></a>3.4.4 Diffusion maps</h5><p>扩散图由Ronald Coifman和Stephane Lafon引入，其基本思想是假设数据是来自扩散过程的样本。 该方法通过估计与数据相关的扩散算子的特征值和特征向量来推断低维流形。</p>
<p>Haghverdi等人已将扩散图概念应用于单细胞RNA-seq数据的分析，以创建称为destiny的R包。</p>
<p>我们将在这里将第一扩散映射分量中的cell的ranko prder作为“diffusion map pseudotime”。</p>
<figure class="highlight erlang"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">deng &lt;- logcounts(deng_SCE)</span><br><span class="line"><span class="function"><span class="title">colnames</span><span class="params">(deng)</span> &lt;- <span class="title">cellLabels</span></span></span><br><span class="line"><span class="function"><span class="title">dm</span> &lt;- D<span class="title">iffusionMap</span><span class="params">(t(deng))</span></span></span><br><span class="line"><span class="function"></span></span><br><span class="line"><span class="function"><span class="title">tmp</span> &lt;- <span class="title">data</span>.<span class="title">frame</span><span class="params">(DC1 = eigenvectors(dm)[,<span class="number">1</span>],</span></span></span><br><span class="line"><span class="function"><span class="params">                  DC2 = eigenvectors(dm)[,<span class="number">2</span>],</span></span></span><br><span class="line"><span class="function"><span class="params">                  Timepoint = deng_SCE$cell_type2)</span></span></span><br><span class="line"><span class="function"><span class="title">ggplot</span><span class="params">(tmp, aes(x = DC1, y = DC2, colour = Timepoint))</span> +</span></span><br><span class="line"><span class="function">    <span class="title">geom_point</span><span class="params">()</span> + <span class="title">scale_color_tableau</span><span class="params">()</span> + </span></span><br><span class="line"><span class="function">    <span class="title">xlab</span><span class="params">(<span class="string">"Diffusion component 1"</span>)</span> + </span></span><br><span class="line"><span class="function">    <span class="title">ylab</span><span class="params">(<span class="string">"Diffusion component 2"</span>)</span> +</span></span><br><span class="line"><span class="function">    <span class="title">theme_classic</span><span class="params">()</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/destiny-deng-1.png" alt="image"></p>
<figure class="highlight erlang"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">deng_SCE$pseudotime_diffusionmap &lt;- rank(eigenvectors(dm)[,<span class="number">1</span>])</span><br><span class="line"><span class="function"><span class="title">ggplot</span><span class="params">(as.data.frame(colData(deng_SCE)), </span></span></span><br><span class="line"><span class="function"><span class="params">       aes(x = pseudotime_diffusionmap, </span></span></span><br><span class="line"><span class="function"><span class="params">           y = cell_type2, colour = cell_type2))</span> +</span></span><br><span class="line"><span class="function">    <span class="title">geom_quasirandom</span><span class="params">(groupOnX = FALSE)</span> +</span></span><br><span class="line"><span class="function">    <span class="title">scale_color_tableau</span><span class="params">()</span> + <span class="title">theme_classic</span><span class="params">()</span> +</span></span><br><span class="line"><span class="function">    <span class="title">xlab</span><span class="params">(<span class="string">"Diffusion map pseudotime (first diffusion map component)"</span>)</span> +</span></span><br><span class="line"><span class="function">    <span class="title">ylab</span><span class="params">(<span class="string">"Timepoint"</span>)</span> +</span></span><br><span class="line"><span class="function">    <span class="title">ggtitle</span><span class="params">(<span class="string">"Cells ordered by diffusion map pseudotime"</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/destiny-deng-2.png" alt="image"><br>像其他方法一样，使用第一个扩散映射组件作为pseudotime，在排序早期的时间点上做得很好（如果我们在开发中采用“较早”的高值），但是它无法区分后面的时间点。</p>
<h5 id="SLICER"><a href="#SLICER" class="headerlink" title="SLICER"></a>SLICER</h5><p>SLICER方法是用于构建轨迹的算法，其描述在顺序生物过程期间基因表达变化，就像Monocle和TSCAN一样。 SLICER旨在捕获高度非线性的基因表达变化，自动选择与过程相关的基因，并检测轨迹中的多个分支和环路特征（Welch，Hartemink和Prins 2016）。 SLICER R软件包可从其GitHub存储库获得，可以使用devtools软件包从那里安装。</p>
<p>我们使用SLICER中的select_genes函数自动选择用于构建细胞轨迹的基因。 该函数使用“邻域方差”来识别在整组细胞中平滑变化而不是随机波动的基因。 在此之后，我们确定哪个“k”值（最近邻居的数量）产生最类似于轨迹的嵌入。 然后我们估计细胞的局部线性嵌入。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">library(<span class="string">"lle"</span>)</span><br><span class="line">slicer_genes &lt;- select_genes(<span class="name">t</span>(<span class="name">deng</span>))</span><br><span class="line">k &lt;- select_k(<span class="name">t</span>(<span class="name">deng</span>[slicer_genes,]), kmin = <span class="number">30</span>, kmax=60)</span><br></pre></td></tr></table></figure>

<figure class="highlight nginx"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="attribute">slicer_traj_lle</span> &lt;- lle(t(deng[slicer_genes,]), m = <span class="number">2</span>, k)<span class="variable">$Y</span></span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">reducedDim(<span class="name">deng_SCE</span>, <span class="string">"LLE"</span>) &lt;- slicer_traj_lle</span><br><span class="line">plotReducedDim(<span class="name">deng_SCE</span>, use_dimred = <span class="string">"LLE"</span>, colour_by = <span class="string">"cell_type2"</span>) +</span><br><span class="line">    xlab(<span class="string">"LLE component 1"</span>) + ylab(<span class="string">"LLE component 2"</span>) +</span><br><span class="line">    ggtitle(<span class="string">"Locally linear embedding of cells from SLICER"</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/slicer-analyis-1.png" alt="image"><br>通过计算局部线性嵌入，我们可以构造完全连接的k-最近邻图。 该图显示每个单元格的（黄色）圆圈，单元格ID号码以蓝色覆盖。 这里我们显示使用10个最近邻居计算的图表。 在这里，SLICER似乎通过一个分支检测到一个主要轨迹。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">slicer_traj_graph &lt;- conn_knn_graph(<span class="name">slicer_traj_lle</span>, <span class="number">10</span>)</span><br><span class="line">plot(<span class="name">slicer_traj_graph</span>, main = <span class="string">"Fully connected kNN graph from SLICER"</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/slicer-build-graph-1.png" alt="image"><br>从这张图中我们可以识别出“极端”细胞，它们是轨迹中的起始/结束细胞的候选细胞</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">ends &lt;- find_extreme_cells(<span class="name">slicer_traj_graph</span>, slicer_traj_lle)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/slicer-1.png" alt="image"></p>
<figure class="highlight livecodeserver"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="built_in">start</span> &lt;- <span class="keyword">ends</span>[<span class="number">1</span>]</span><br></pre></td></tr></table></figure>

<p>定义了一个开始cell后，我们可以在估计的pseudotime内对cell进行排序。</p>
<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">pseudotime_order_slicer</span> &lt;- cell_order(slicer_traj_graph, start)</span><br><span class="line"><span class="title">branches</span> &lt;- assign_branches(slicer_traj_graph, start)</span><br><span class="line"></span><br><span class="line"><span class="title">pseudotime_slicer</span> &lt;-</span><br><span class="line">    <span class="class"><span class="keyword">data</span>.frame(</span></span><br><span class="line"><span class="class">        <span class="type">Timepoint</span> = <span class="title">cellLabels</span>,</span></span><br><span class="line"><span class="class">        <span class="title">pseudotime</span> = <span class="type">NA</span>,</span></span><br><span class="line"><span class="class">        <span class="type">State</span> = <span class="title">branches</span></span></span><br><span class="line"><span class="class">    )</span></span><br><span class="line"><span class="title">pseudotime_slicer</span>$pseudotime[pseudotime_order_slicer] &lt;-</span><br><span class="line">    <span class="number">1</span>:length(pseudotime_order_slicer)</span><br><span class="line"><span class="title">deng_SCE</span>$pseudotime_slicer &lt;- pseudotime_slicer$pseudotime</span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">ggplot(<span class="name">as</span>.data.frame(<span class="name">colData</span>(<span class="name">deng_SCE</span>)), </span><br><span class="line">       aes(<span class="name">x</span> = pseudotime_slicer, </span><br><span class="line">           y = cell_type2, colour = cell_type2)) +</span><br><span class="line">    geom_quasirandom(<span class="name">groupOnX</span> = FALSE) +</span><br><span class="line">    scale_color_tableau() + theme_classic() +</span><br><span class="line">    xlab(<span class="string">"SLICER pseudotime (cell ordering)"</span>) +</span><br><span class="line">    ylab(<span class="string">"Timepoint"</span>) +</span><br><span class="line">    theme_classic()</span><br></pre></td></tr></table></figure>

<p>与前面的方法一样，SLICER为早期的时间点提供了良好的排序。它在“8cell”细胞之前放置“16cell”细胞，但比许多早期的方法提供了更好的blast细胞排序</p>
<h5 id="3-4-6-Ouija"><a href="#3-4-6-Ouija" class="headerlink" title="3.4.6 Ouija"></a>3.4.6 Ouija</h5><p>Ouija（<a href="http://kieranrcampbell.github.io/ouija/）采用了与我们迄今为止所研究的伪时间估计方法不同的方法。" target="_blank" rel="noopener">http://kieranrcampbell.github.io/ouija/）采用了与我们迄今为止所研究的伪时间估计方法不同的方法。</a> 早期的方法都是“无人监督”的，也就是说，除了可能选择信息基因之外，我们不提供任何关于我们如何期望某些基因或整个轨迹表现的先验信息的方法。</p>
<p>相比之下，Ouija是一种概率框架，它允许仅使用一小组标记基因来解释单细胞pseudotime的学习。 这个方法：</p>
<ul>
<li>从少数标记基因中推断出pseudotime，让您了解为什么pseudotime已经从这些基因中学习出来;</li>
<li>为可解释的基因调控行为（如peak time或upregulation time）提供参数估计（具有不确定性）;</li>
<li>进行了贝叶斯假设检验，以发现沿着轨迹在其他人之前受到调节的基因;</li>
<li>识别亚稳态，即沿连续轨迹的离散细胞类型。<br>我们将向Ouija提供以下标记基因（在时间点上，它们被认为是高度表达的）：</li>
<li>Early timepoints: Dazl, Rnf17, Sycp3, Nanog, Pou5f1, Fgf8, Egfr, Bmp5, Bmp15</li>
<li>Mid timepoints: Zscan4b, Foxa1, Prdm14, Sox21</li>
<li>Late timepoints: Creb3, Gpx4, Krt8, Elf5, Eomes, Cdx2, Tdgf1, Gdf3<br>使用Ouija，我们可以将基因模型表现为单调上调或下调（称为开关样行为），或基因短暂达到峰值的瞬态行为。 默认情况下，Ouija假设所有基因都表现出类似开关的行为<br>在这里，我们可以“cheat”，并检查我们选择的标记基因是否确实识别了分化过程的不同时间点。</li>
</ul>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">ouija_markers_down &lt;- c(<span class="string">"Dazl"</span>, <span class="string">"Rnf17"</span>, <span class="string">"Sycp3"</span>, <span class="string">"Fgf8"</span>, </span><br><span class="line">                        <span class="string">"Egfr"</span>, <span class="string">"Bmp5"</span>, <span class="string">"Bmp15"</span>, <span class="string">"Pou5f1"</span>)</span><br><span class="line">ouija_markers_up &lt;- c(<span class="string">"Creb3"</span>, <span class="string">"Gpx4"</span>, <span class="string">"Krt8"</span>, <span class="string">"Elf5"</span>, <span class="string">"Cdx2"</span>, </span><br><span class="line">                      <span class="string">"Tdgf1"</span>, <span class="string">"Gdf3"</span>, <span class="string">"Eomes"</span>)</span><br><span class="line">ouija_markers_transient &lt;- c(<span class="string">"Zscan4b"</span>, <span class="string">"Foxa1"</span>, <span class="string">"Prdm14"</span>, <span class="string">"Sox21"</span>)</span><br><span class="line">ouija_markers &lt;- c(<span class="name">ouija_markers_down</span>, ouija_markers_up, </span><br><span class="line">                   ouija_markers_transient)</span><br><span class="line">plotExpression(<span class="name">deng_SCE</span>, ouija_markers, x = <span class="string">"cell_type2"</span>, colour_by = <span class="string">"cell_type2"</span>) +</span><br><span class="line">    theme(<span class="name">axis</span>.text.x = element_text(<span class="name">angle</span> = <span class="number">60</span>, hjust = <span class="number">1</span>))</span><br></pre></td></tr></table></figure>

<p><img src="https://note.youdao.com/yws/public/resource/12beaf1b878bc1059015bf8c805c27d5/xmlnote/33E31DDE55D94CE1AEEF02639EABBC86/2019" alt="image"><br>为了适应pseudotime，只需调用ouija，传入预期的响应类型。 请注意，如果没有提供响应类型，则默认情况下它们都被假定为switch-like，我们将在此处执行此操作。 Ouija的输入可以是非负表达式值的cell-by-gene矩阵，也可以是ExpressionSet对象，或者，幸运的是，通过从SingleCellExperiment对象中选择logcounts值。<br>我们可以应用先前的信息，关于基因在分化过程中是否向上或向下调节，并提供关于何时表达的转换或表达的峰值可能发生的先验信息。</p>
<ul>
<li>Hamiltonian Monte Carlo（HMC） - 完整的MCMC推理，其中log-posterior的梯度信息用于“引导”随机遍历参数空间，或者</li>
<li>自动微分变分贝叶斯（ADVI或简称VI） - 近似推断，其中KL偏差到近似分布被最小化。<br>一般而言，HMC将为所有参数提供更准确的推理和近似正确的后验方差。 然而，VB比HMC快几个数量级，虽然它可能低估了后验方差，但是Ouija的作者认为，通常情况下，它通常与HMC一样发现后部pseudotime。<br>为了帮助Ouija模型，我们为它提供了关于上调和下调基因的开关强度的先验信息。 通过将下调基因的开关强度设置为-10，将上调基因的开关强度设置为10，先验强度标准差为0.5，我们告诉模型我们对这些基因在分化过程中的预期行为充满信心。</li>
</ul>
<figure class="highlight autohotkey"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="symbol">options(mc.cores = parallel::</span>detectCores())</span><br><span class="line">response_type &lt;- c(rep(<span class="string">"switch"</span>, length(ouij<span class="built_in">a_markers</span>_down) + </span><br><span class="line">                           length(ouij<span class="built_in">a_markers</span>_up)), </span><br><span class="line">                   rep(<span class="string">"transient"</span>, length(ouij<span class="built_in">a_markers</span>_transient)))</span><br><span class="line">switch_strengths &lt;- c(rep(-<span class="number">10</span>, length(ouij<span class="built_in">a_markers</span>_down)),</span><br><span class="line">                      rep(<span class="number">10</span>, length(ouij<span class="built_in">a_markers</span>_up)))</span><br><span class="line">switch_strength_sd &lt;- c(rep(<span class="number">0.5</span>, length(ouij<span class="built_in">a_markers</span>_down)),</span><br><span class="line">                      rep(<span class="number">0.5</span>, length(ouij<span class="built_in">a_markers</span>_up)))</span><br><span class="line">garbage &lt;- capture.output(</span><br><span class="line">    oui_vb &lt;- ouija(deng_SCE[ouij<span class="built_in">a_markers</span>,],</span><br><span class="line">                    single_cell_experiment_assay = <span class="string">"logcounts"</span>, </span><br><span class="line">                    response_type = response_type,</span><br><span class="line">                    switch_strengths = switch_strengths,</span><br><span class="line">                    switch_strength_sd = switch_strength_sd,</span><br><span class="line">                    inference_type = <span class="string">"vb"</span>)</span><br><span class="line">)</span><br><span class="line"></span><br><span class="line">print(oui_vb)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">## A Ouija fit <span class="keyword">with</span> <span class="number">268</span> cells and <span class="number">20</span> marker genes </span><br><span class="line">## Inference type:  Variational Bayes </span><br><span class="line">## (Gene behaviour) Switch/transient: <span class="number">16</span> / <span class="number">4</span></span><br></pre></td></tr></table></figure>

<p>我们可以使用plot_expression函数绘制伪照片上的基因表达以及平均函数（S形或高斯瞬态函数）的最大后验（MAP）估计。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">plot_expression</span><span class="params">(oui_vb)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-plot-exprs-1.png" alt="image"><br>我们还可以使用plot_switch_times和plot_transient_times函数可视化基因调控行为在轨迹中何时以切换时间或峰值时间（对于开关样或瞬态基因）的形式发生：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">plot_switch_times</span><span class="params">(oui_vb)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-plot-switch-times-1.png" alt="image"></p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">plot_peak_times</span><span class="params">(oui_vb)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-plot-switch-times-2.png" alt="image"><br>使用一致性矩阵确定亚稳态。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line">cmo &lt;- consistency_matrix(<span class="name">oui_vb</span>)</span><br><span class="line">plot_consistency(<span class="name">oui_vb</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-consistency-1.png" alt="image"></p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cell_classifications &lt;- cluster_consistency(<span class="name">cmo</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">map_pst</span> &lt;- map_pseudotime(oui_vb)</span><br><span class="line"><span class="title">ouija_pseudotime</span> &lt;- <span class="class"><span class="keyword">data</span>.frame(<span class="title">map_pst</span>, <span class="title">cell_classifications</span>)</span></span><br><span class="line"></span><br><span class="line"><span class="title">ggplot</span>(ouija_pseudotime, aes(x = map_pst, y = cell_classifications)) +</span><br><span class="line">  geom_point() +</span><br><span class="line">  xlab(<span class="string">"MAP pseudotime"</span>) +</span><br><span class="line">  ylab(<span class="string">"Cell classification"</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-pseudotime-1.png" alt="image"></p>
<figure class="highlight erlang"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">deng_SCE$pseudotime_ouija &lt;- ouija_pseudotime$map_pst</span><br><span class="line">deng_SCE$ouija_cell_class &lt;- ouija_pseudotime$cell_classifications</span><br><span class="line"></span><br><span class="line"><span class="function"><span class="title">ggplot</span><span class="params">(as.data.frame(colData(deng_SCE)), </span></span></span><br><span class="line"><span class="function"><span class="params">       aes(x = pseudotime_ouija, </span></span></span><br><span class="line"><span class="function"><span class="params">           y = cell_type2, colour = cell_type2))</span> +</span></span><br><span class="line"><span class="function">    <span class="title">geom_quasirandom</span><span class="params">(groupOnX = FALSE)</span> +</span></span><br><span class="line"><span class="function">    <span class="title">scale_color_tableau</span><span class="params">()</span> + <span class="title">theme_classic</span><span class="params">()</span> +</span></span><br><span class="line"><span class="function">    <span class="title">xlab</span><span class="params">(<span class="string">"Ouija pseudotime"</span>)</span> +</span></span><br><span class="line"><span class="function">    <span class="title">ylab</span><span class="params">(<span class="string">"Timepoint"</span>)</span> +</span></span><br><span class="line"><span class="function">    <span class="title">theme_classic</span><span class="params">()</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-pseudotime-2.png" alt="image"><br>虽然它可以对标记基因的选择和提供的先前信息敏感，但Ouija在细胞的排序方面表现相当不错。 如果选择不同的标记基因或更改先验，结果会如何变化？</p>
<p>Ouija在这里确定了四个亚稳状态，我们可以将其注释为“zygote / 2cell”，“4/8/16 cell”，“blast1”和“blast2”。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">ggplot(<span class="name">as</span>.data.frame(<span class="name">colData</span>(<span class="name">deng_SCE</span>)), </span><br><span class="line">       aes(<span class="name">x</span> = as.factor(<span class="name">ouija_cell_class</span>), </span><br><span class="line">           y = pseudotime_ouija, colour = cell_type2)) +</span><br><span class="line">    geom_boxplot() + </span><br><span class="line">    coord_flip() +</span><br><span class="line">    scale_color_tableau() + theme_classic() +</span><br><span class="line">    xlab(<span class="string">"Ouija cell classification"</span>) +</span><br><span class="line">    ylab(<span class="string">"Ouija pseudotime"</span>) +</span><br><span class="line">    theme_classic()</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/ouija-states-1.png" alt="image"></p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">gene_regs &lt;- gene_regulation(<span class="name">oui_vb</span>)</span><br><span class="line">head(<span class="name">gene_regs</span>)</span><br></pre></td></tr></table></figure>

<p>一个常见的分析是制定基因的调节顺序。 例如，基因A在基因B之前是否被上调？ 基因C在基因D下调前是否达到峰值？ Ouija根据贝叶斯假设检验来回答这些问题，该检验是关于调节时间的差异（切换时间或峰值时间）是否显着不同于0.这是使用gene_regulation函数进行整理。</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">## # A tibble: <span class="number">6</span> x <span class="number">7</span></span><br><span class="line">## # Groups:   label, gene_A [<span class="number">6</span>]</span><br><span class="line">##   label        gene_A gene_B mean_difference lower_95 upper_95 significant</span><br><span class="line">##   &lt;chr&gt;        &lt;chr&gt;  &lt;chr&gt;            &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt; &lt;lgl&gt;      </span><br><span class="line">## <span class="number">1</span> Bmp15 - Cdx2 Bmp15  Cdx2           <span class="number">-0.0434</span> <span class="number">-0.0912</span>   <span class="number">0.0110</span>  FALSE      </span><br><span class="line">## <span class="number">2</span> Bmp15 - Cre… Bmp15  Creb3           <span class="number">0.278</span>   <span class="number">0.220</span>    <span class="number">0.330</span>   TRUE       </span><br><span class="line">## <span class="number">3</span> Bmp15 - Elf5 Bmp15  Elf5           <span class="number">-0.656</span>  <span class="number">-0.688</span>   <span class="number">-0.618</span>   TRUE       </span><br><span class="line">## <span class="number">4</span> Bmp15 - Eom… Bmp15  Eomes           <span class="number">0.0766</span>  <span class="number">0.00433</span>  <span class="number">0.153</span>   TRUE       </span><br><span class="line">## <span class="number">5</span> Bmp15 - Fox… Bmp15  Foxa1          <span class="number">-0.0266</span> <span class="number">-0.0611</span>   <span class="number">0.00287</span> FALSE      </span><br><span class="line">## <span class="number">6</span> Bmp15 - Gdf3 Bmp15  Gdf3            <span class="number">0.0704</span>  <span class="number">0.00963</span>  <span class="number">0.134</span>   TRUE</span><br></pre></td></tr></table></figure>

<h5 id="3-4-7-Comparison-of-the-methods"><a href="#3-4-7-Comparison-of-the-methods" class="headerlink" title="3.4.7 Comparison of the methods"></a>3.4.7 Comparison of the methods</h5><p>TSCAN，Monocle，Diffusion Map，SLICER和Ouija推断出的轨迹哪一个更好？</p>
<p>TSCAN和Diffusion Map方法使得轨迹“错误”，因此我们将对这些比较进行调整。</p>
<figure class="highlight stata"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">df_pseudotime &lt;- <span class="keyword">as</span>.data.frame(</span><br><span class="line">    colData(deng_SCE)[, grep(<span class="string">"pseudotime"</span>, colnames(colData(deng_SCE)))]</span><br><span class="line">)</span><br><span class="line">colnames(df_pseudotime) &lt;- gsub(<span class="string">"pseudotime_"</span>, <span class="string">""</span>, </span><br><span class="line">                                colnames(df_pseudotime))</span><br><span class="line">df_pseudotime<span class="variable">$PC1</span> &lt;- deng_SCE<span class="variable">$PC1</span></span><br><span class="line">df_pseudotime<span class="variable">$order_tscan</span> &lt;- -df_pseudotime<span class="variable">$order_tscan</span></span><br><span class="line">df_pseudotime<span class="variable">$diffusionmap</span> &lt;- -df_pseudotime<span class="variable">$diffusionmap</span></span><br><span class="line"></span><br><span class="line">corrplot.mixed(<span class="keyword">cor</span>(df_pseudotime, <span class="keyword">use</span> = <span class="string">"na.or.complete"</span>), </span><br><span class="line">               <span class="keyword">order</span> = <span class="string">"hclust"</span>, tl.col = <span class="string">"black"</span>,</span><br><span class="line">               main = <span class="string">"Correlation matrix for pseudotime results"</span>,</span><br><span class="line">               mar = c(0, 0, 3.1, 0))</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/compare-results-1.png" alt="image"></p>
<h5 id="3-4-8-Expression-of-genes-through-time"><a href="#3-4-8-Expression-of-genes-through-time" class="headerlink" title="3.4.8 Expression of genes through time"></a>3.4.8 Expression of genes through time</h5><p>每个包还可以通过pseudotime显示表达式。 遵循单个基因非常有助于鉴定在分化过程中起重要作用的基因。 我们用Rhoa基因说明了这个过程。</p>
<p>我们已将使用所有方法计算的pseudotime添加到SCE对象的colData中。 完成后，scater包的完整绘图功能可用于研究基因表达，细胞群和假型的关系。 这对于不提供绘图功能的SLICER等软件包特别有用。<br>Principal components</p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">plotExpression(deng_SCE, <span class="string">"Rhoa"</span>, <span class="attr">x</span> = <span class="string">"PC1"</span>, </span><br><span class="line">               <span class="attr">colour_by</span> = <span class="string">"cell_type2"</span>, <span class="attr">show_violin</span> = FALSE,</span><br><span class="line">               <span class="attr">show_smooth</span> = TRUE)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-pc1-1.png" alt="image"><br>TSCAN</p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">plotExpression(deng_SCE, <span class="string">"Rhoa"</span>, <span class="attr">x</span> = <span class="string">"pseudotime_order_tscan"</span>, </span><br><span class="line">               <span class="attr">colour_by</span> = <span class="string">"cell_type2"</span>, <span class="attr">show_violin</span> = FALSE,</span><br><span class="line">               <span class="attr">show_smooth</span> = TRUE)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-tscan-1.png" alt="image"><br>Monocle</p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">plotExpression(deng_SCE, <span class="string">"Rhoa"</span>, <span class="attr">x</span> = <span class="string">"pseudotime_monocle"</span>, </span><br><span class="line">               <span class="attr">colour_by</span> = <span class="string">"cell_type2"</span>, <span class="attr">show_violin</span> = FALSE,</span><br><span class="line">               <span class="attr">show_smooth</span> = TRUE)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-monocle-1.png" alt="image"><br>Diffusion Map</p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">plotExpression(deng_SCE, <span class="string">"Rhoa"</span>, <span class="attr">x</span> = <span class="string">"pseudotime_diffusionmap"</span>, </span><br><span class="line">               <span class="attr">colour_by</span> = <span class="string">"cell_type2"</span>, <span class="attr">show_violin</span> = FALSE,</span><br><span class="line">               <span class="attr">show_smooth</span> = TRUE)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-diff-map-1.png" alt="image"><br>SLICER</p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">plotExpression(deng_SCE, <span class="string">"Rhoa"</span>, <span class="attr">x</span> = <span class="string">"pseudotime_slicer"</span>, </span><br><span class="line">               <span class="attr">colour_by</span> = <span class="string">"cell_type2"</span>, <span class="attr">show_violin</span> = FALSE,</span><br><span class="line">               <span class="attr">show_smooth</span> = TRUE)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-slicer-1.png" alt="image"><br>Ouija</p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">plotExpression(deng_SCE, <span class="string">"Rhoa"</span>, <span class="attr">x</span> = <span class="string">"pseudotime_ouija"</span>, </span><br><span class="line">               <span class="attr">colour_by</span> = <span class="string">"cell_type2"</span>, <span class="attr">show_violin</span> = FALSE,</span><br><span class="line">               <span class="attr">show_smooth</span> = TRUE)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/27-pseudotime_files/figure-html/Rhoa-ouija-1.png" alt="image"></p>
<h5 id="3-5-Imputation"><a href="#3-5-Imputation" class="headerlink" title="3.5 Imputation"></a>3.5 Imputation</h5><p>如前所述，分析scRNA-seq数据时的主要挑战之一是存在零或丢失。假设丢失出现了三个可能的原因：</p>
<ul>
<li>该基因不在细胞中表达，因此没有序列的转录物</li>
<li>该基因被表达，但由于某种原因，转录物在测序前丢失了某处</li>
<li>表达基因并捕获转录物并转化为cDNA，但测序深度不足以产生任何读数。<br>因此，丢失可能是实验缺陷的结果，如果是这种情况，那么我们想提供计算校正。一种可能的解决方案是将丢失量计算在表达式矩阵中。为了能够克服基因表达值，必须有一个基础模型。然而，由于我们不知道哪些丢失事件是技术人为因素，哪些本来就存在的，因此归咎是一项艰巨的挑战。</li>
</ul>
<p>据我们所知，目前有两种不同的插补方法：MAGIC（Dijk等人2017）和scImpute（W. V. Li和Li 2017）。 MAGIC仅适用于Python或Matlab，但我们将从R内部运行它。</p>
<h5 id="3-5-1-scImpute"><a href="#3-5-1-scImpute" class="headerlink" title="3.5.1 scImpute"></a>3.5.1 scImpute</h5><p>为了测试scImpute，我们使用默认参数，并将其应用到我们以前处理过的Deng数据集。scImpute将一个.csv或.txt文件作为输入：</p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">deng &lt;- readRDS(<span class="string">"deng/deng-reads.rds"</span>)</span><br><span class="line">write.csv(counts(deng), <span class="string">"deng.csv"</span>)</span><br><span class="line">scimpute(</span><br><span class="line">    <span class="attr">count_path</span> = <span class="string">"deng.csv"</span>,</span><br><span class="line">    <span class="attr">infile</span> = <span class="string">"csv"</span>,</span><br><span class="line">    <span class="attr">outfile</span> = <span class="string">"txt"</span>, </span><br><span class="line">    <span class="attr">out_dir</span> = <span class="string">"./"</span>,</span><br><span class="line">    <span class="attr">Kcluster</span> = <span class="number">10</span>,</span><br><span class="line">    <span class="attr">ncores</span> = <span class="number">2</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>现在我们可以通过考虑一个PCA图来比较结果和原始数据</p>
<figure class="highlight vim"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">res</span> &lt;- <span class="keyword">read</span>.table(<span class="string">"scimpute_count.txt"</span>)</span><br><span class="line">colnames(<span class="keyword">res</span>) &lt;- NULL</span><br><span class="line"><span class="keyword">res</span> &lt;- SingleCellExperiment(</span><br><span class="line">    assays = <span class="keyword">list</span>(logcounts = log2(<span class="keyword">as</span>.matrix(<span class="keyword">res</span>) + <span class="number">1</span>)), </span><br><span class="line">    colData = colData(deng)</span><br><span class="line">)</span><br><span class="line">rowData(<span class="keyword">res</span>)$feature_symbol &lt;- rowData(deng)$feature_symbol</span><br><span class="line">plotPCA(</span><br><span class="line">    <span class="keyword">res</span>, </span><br><span class="line">    colour_by = <span class="string">"cell_type2"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/28-imputation_files/figure-html/unnamed-chunk-4-1.png" alt="image"><br>为了评估imputation的影响，我们使用SC3来对估算矩阵进行分组</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">res &lt;- sc3_estimate_k(res)</span><br><span class="line"><span class="function"><span class="title">metadata</span><span class="params">(res)</span></span><span class="variable">$sc3</span><span class="variable">$k_estimation</span></span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">6</span></span><br></pre></td></tr></table></figure>

<figure class="highlight xml"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">res <span class="tag">&lt;<span class="name">-</span> <span class="attr">sc3</span>(<span class="attr">res</span>, <span class="attr">ks</span> = <span class="string">10,</span> <span class="attr">gene_filter</span> = <span class="string">FALSE)</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adjustedRandIndex(<span class="name">colData</span>(<span class="name">deng</span>)$cell_type2, colData(<span class="name">res</span>)$sc3_10_clusters)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">0.4687332</span></span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    res, </span><br><span class="line">    colour_by = <span class="string">"sc3_10_clusters"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/28-imputation_files/figure-html/unnamed-chunk-5-1.png" alt="image"></p>
<h5 id="3-5-2-MAGIC"><a href="#3-5-2-MAGIC" class="headerlink" title="3.5.2 MAGIC"></a>3.5.2 MAGIC</h5><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">system</span><span class="params">(<span class="string">"python3 utils/MAGIC.py -d deng.csv -o MAGIC_count.csv --cell-axis columns -l 1 --pca-non-random csv"</span>)</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight vim"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">res</span> &lt;- <span class="keyword">read</span>.csv(<span class="string">"MAGIC_count.csv"</span>, header = TRUE)</span><br><span class="line">rownames(<span class="keyword">res</span>) &lt;- <span class="keyword">res</span>[,<span class="number">1</span>]</span><br><span class="line"><span class="keyword">res</span> &lt;- <span class="keyword">res</span>[,-<span class="number">1</span>]</span><br><span class="line"><span class="keyword">res</span> &lt;- t(<span class="keyword">res</span>)</span><br><span class="line"><span class="keyword">res</span> &lt;- SingleCellExperiment(</span><br><span class="line">    assays = <span class="keyword">list</span>(logcounts = <span class="keyword">res</span>), </span><br><span class="line">    colData = colData(deng)</span><br><span class="line">)</span><br><span class="line">rowData(<span class="keyword">res</span>)$feature_symbol &lt;- rownames(<span class="keyword">res</span>)</span><br><span class="line">plotPCA(</span><br><span class="line">    <span class="keyword">res</span>, </span><br><span class="line">    colour_by = <span class="string">"cell_type2"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/28-imputation_files/figure-html/unnamed-chunk-7-1.png" alt="image"><br>为了评估imputation的影响，我们使用SC3来对估算矩阵进行分组</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">res &lt;- sc3_estimate_k(res)</span><br><span class="line"><span class="function"><span class="title">metadata</span><span class="params">(res)</span></span><span class="variable">$sc3</span><span class="variable">$k_estimation</span></span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">4</span></span><br></pre></td></tr></table></figure>

<figure class="highlight xml"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">res <span class="tag">&lt;<span class="name">-</span> <span class="attr">sc3</span>(<span class="attr">res</span>, <span class="attr">ks</span> = <span class="string">10,</span> <span class="attr">gene_filter</span> = <span class="string">FALSE)</span></span></span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">adjustedRandIndex(<span class="name">colData</span>(<span class="name">deng</span>)$cell_type2, colData(<span class="name">res</span>)$sc3_10_clusters)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">0.3752866</span></span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">plotPCA(</span><br><span class="line">    res, </span><br><span class="line">    colour_by = <span class="string">"sc3_10_clusters"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/28-imputation_files/figure-html/unnamed-chunk-8-1.png" alt="image"></p>
<h4 id="3-6-Differential-Expression-DE-analysis"><a href="#3-6-Differential-Expression-DE-analysis" class="headerlink" title="3.6 Differential Expression(DE) analysis"></a>3.6 Differential Expression(DE) analysis</h4><h5 id="3-6-1-Bulk-RNA-seq"><a href="#3-6-1-Bulk-RNA-seq" class="headerlink" title="3.6.1 Bulk RNA-seq"></a>3.6.1 Bulk RNA-seq</h5><p>在处理大量RNA-seq数据时，最常见的分析类型之一是识别不同表达的基因。通过比较两个条件之间的变化，例如突变和野生型，或者刺激和不受刺激，有可能描述这种变化背后的分子机制。<br>几种不同的方法，例如 已经开发出DESeq2和edgeR用于大量RNA-seq。 此外，还有广泛的数据集，其中RNA-seq数据已经使用RT-qPCR验证。 这些数据可用于对DE查找算法进行基准测试，现有证据表明算法表现良好。</p>
<h5 id="3-6-2-Single-cell-RNA-seq"><a href="#3-6-2-Single-cell-RNA-seq" class="headerlink" title="3.6.2 Single cell  RNA-seq"></a>3.6.2 Single cell  RNA-seq</h5><p>与大量RNA-seq相反，在scRNA-seq中，我们通常没有一组确定的实验条件。 相反，如前一章（3.2）所示，我们可以通过使用无监督的聚类方法来识别细胞群。 一旦鉴定了组，就可以通过比较各组之间的方差差异（如SC3中实施的Kruskal-Wallis检验），或通过以成对方式比较簇之间的基因表达，找到差异表达的基因。 在下一章中，我们将主要考虑为成对比较开发的工具。<br>与大量RNA-seq不同，我们通常在单细胞实验中比较的每组中有大量样品（即细胞）。因此，我们可以利用每组中表达值的完整分布来识别组之间的差异，而不是如bulk RNA-seq一样仅仅比较平均表达的估计。</p>
<p>比较分布有两种主要方法。首先，我们可以使用现有的统计模型/分布，并将相同类型的模型拟合到每个组中的表达式，然后测试每个模型的参数差异，或者测试如果允许特定参数不同，模型是否更合适据小组说。例如，在第2.10章中，我们使用edgeR来测试在不同批次中允许平均表达是否不同显着改善了数据的负二项模型的拟合。</p>
<p>或者，我们可以使用非参数测试，该测试不假设表达值遵循任何特定分布，例如， Kolmogorov-Smirnov检验（KS检验）。非参数测试通常将观察到的表达值转换为等级并测试一组的等级分布是否明显不同于另一组的等级分布。然而，一些非参数方法在存在大量绑定值时失败，例如单细胞RNA-seq表达数据中的丢失（零）的情况。此外，如果参数测试的条件成立，那么它通常比非参数测试更强大。<br>下面用一个测试数据来评价一下不同的算法的表现。处理同样的表达矩阵得到差异结果跟已知的差异结果进行比较看看overlap怎么样。评价指标主要是：</p>
<ul>
<li>召回率，True positive rate (TPR), TP/(TP + FN)</li>
<li>准确率，False positive rate (FPR), FP/(FP+TP)</li>
<li>receiver-operating-characteristic curve (ROC)</li>
<li>area under this curve (AUC)</li>
</ul>
<p>3.6.4 Models of single-cell RNAseq Data<br>RNAseq数据最常见的模型是负二项模型</p>
<figure class="highlight gams"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">set</span>.seed(1)</span><br><span class="line">hist(</span><br><span class="line">    rnbinom(</span><br><span class="line">        1000, </span><br><span class="line">        mu <span class="comment">= 10,</span> </span><br><span class="line">        size <span class="comment">= 100),</span> </span><br><span class="line">    col <span class="comment">=</span> <span class="comment">"grey50"</span><span class="comment">,</span> </span><br><span class="line">    xlab <span class="comment">=</span> <span class="comment">"Read Counts"</span><span class="comment">,</span> </span><br><span class="line">    main <span class="comment">=</span> <span class="comment">"Negative Binomial"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/29-de-intro_files/figure-html/nb-plot-1.png" alt="image"><br>Negative Binomial distribution of read counts for a single gene across 1000 cells</p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Mean：μ=mu</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Variance:σ^2=mu+mu^2/size</span><br></pre></td></tr></table></figure>

<p>它通过平均表达式（mu）和分散（size）参数化，其与方差成反比。负二项式模型很好地拟合了大量RNA-seq数据，并且它被用于为这些数据设计的大多数统计方法。此外，它已被证明适合从独特的分子标识符（UMI）标记的数据获得的分子计数的分布相当好.<br>然而，由于相对于非零读取计数的高丢失率，原始负二项式模型也不适合全长转录本数据。对于这种类型的数据，已经提出了各种zero-inflated负二项模型</p>
<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">d &lt;- <span class="number">0.5</span>;</span><br><span class="line">counts &lt;- rnbinom(</span><br><span class="line">    <span class="number">1000</span>, </span><br><span class="line">    mu = <span class="number">10</span>, </span><br><span class="line">    size = <span class="number">100</span></span><br><span class="line">)</span><br><span class="line">counts[runif(<span class="number">1000</span>) &lt; d] &lt;- <span class="number">0</span></span><br><span class="line">hist(</span><br><span class="line">    counts, </span><br><span class="line">    col = <span class="string">"grey50"</span>, </span><br><span class="line">    xlab = <span class="string">"Read Counts"</span>, </span><br><span class="line">    main = <span class="string">"Zero-inflated NB"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/29-de-intro_files/figure-html/zero-inflation-plot-1.png" alt="image"><br> Zero-inflated Negative Binomial distribution<br> <figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Mean：μ=mu*(1-d)</span><br></pre></td></tr></table></figure></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Variance:σ^2=mu(1-d)+(1+d*μ+μ/size)</span><br></pre></td></tr></table></figure>

<p>由于丢失率，这个模型在负二项模型中引入了一个新的参数d，正如我们在19章中所见到的，一个基因的丢失率与这个基因的平均表达值密切相关。不同的zero-inflated负二项模型利用mu和d之间的不同关系以及一些可能适合的μ和d独立的表达每一个基因。<br>最后，几种方法使用Poisson-Beta分布，该分部基于转录破裂的机制模型。这个模型有很强的实验支持，但是他比负二项模型更加不容易使用，</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">a &lt;- <span class="number">0.1</span></span><br><span class="line">b &lt;- <span class="number">0.1</span></span><br><span class="line">g &lt;- <span class="number">100</span></span><br><span class="line">lambdas &lt;- rbeta(<span class="number">1000</span>, a, b)</span><br><span class="line">counts &lt;- sapply(<span class="name">g*lambdas</span>, function(<span class="name">l</span>) &#123;rpois(<span class="number">1</span>, lambda = l)&#125;)</span><br><span class="line">hist(</span><br><span class="line">    counts, </span><br><span class="line">    col = <span class="string">"grey50"</span>, </span><br><span class="line">    xlab = <span class="string">"Read Counts"</span>, </span><br><span class="line">    main = <span class="string">"Poisson-Beta"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/29-de-intro_files/figure-html/pois-beta-plot-1.png" alt="image"></p>
<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Mean：μ=g*a/(a+b)</span><br></pre></td></tr></table></figure>

<figure class="highlight plain"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">Variance:σ^2=g^2*a*b/((a+b+1)*(a+b)^2)</span><br></pre></td></tr></table></figure>

<p>这个模型使用3个参数：a表示转率激活率；b表示转率抑制率，g转录产生的速率，即转录在该位点是活跃的。差异表达方法可以测试每个参数的组间差异或仅一个（通常是g）。<br>所有这些模型可以进一步扩展以明确地解释基因表达差异的其他来源，例如批次效应或库深度，这取决于特定的DE算法。</p>
<h4 id="3-7-DE-in-a-real-dataset"><a href="#3-7-DE-in-a-real-dataset" class="headerlink" title="3.7 DE in a real dataset"></a>3.7 DE in a real dataset</h4><figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">library</span><span class="params">(scRNA.seq.funcs)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(edgeR)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(monocle)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(MAST)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(ROCR)</span></span></span><br><span class="line">set.seed(<span class="number">1</span>)</span><br></pre></td></tr></table></figure>

<h6 id="3-7-1-Introdution"><a href="#3-7-1-Introdution" class="headerlink" title="3.7.1 Introdution"></a>3.7.1 Introdution</h6><p>为了测试不同单细胞差异表达方法我们将使用7-17章中的Blischak数据集。在这个实验中除了产生单细胞数据，每一个细胞系还产生了bulk RNA-seq数据。我们将使用在各自的批量数据上使用标准方法确定的不同表达的基因作为评估每个单细胞方法的准确性的基本事实。为了节省时间，我们预先计算了这些。您可以运行下面的命令来加载这些数据。</p>
<figure class="highlight stata"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">DE</span> &lt;- <span class="keyword">read</span>.<span class="keyword">table</span>(<span class="string">"tung/TPs.txt"</span>)</span><br><span class="line">notDE &lt;- <span class="keyword">read</span>.<span class="keyword">table</span>(<span class="string">"tung/TNs.txt"</span>)</span><br><span class="line">GroundTruth &lt;- <span class="keyword">list</span>(</span><br><span class="line">    <span class="keyword">DE</span> = <span class="keyword">as</span>.character(unlist(<span class="keyword">DE</span>)), </span><br><span class="line">    notDE = <span class="keyword">as</span>.character(unlist(notDE))</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>这对NA19101和NA19239的比较的结果。现在加载各自的单细胞数据：</p>
<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">molecules</span> &lt;- read.table(<span class="string">"tung/molecules.txt"</span>, sep = <span class="string">"\t"</span>)</span><br><span class="line"><span class="title">anno</span> &lt;- read.table(<span class="string">"tung/annotation.txt"</span>, sep = <span class="string">"\t"</span>, header = <span class="type">TRUE</span>)</span><br><span class="line"><span class="title">keep</span> &lt;- anno[,<span class="number">1</span>] == <span class="string">"NA19101"</span> | anno[,<span class="number">1</span>] == <span class="string">"NA19239"</span></span><br><span class="line"><span class="class"><span class="keyword">data</span> &lt;- molecules[,keep]</span></span><br><span class="line"><span class="title">group</span> &lt;- anno[keep,<span class="number">1</span>]</span><br><span class="line"><span class="title">batch</span> &lt;- anno[keep,<span class="number">4</span>]</span><br><span class="line"><span class="meta"># remove genes that aren't expressed in at least 6 cells</span></span><br><span class="line"><span class="title">gkeep</span> &lt;- rowSums(<span class="class"><span class="keyword">data</span> &gt; 0) &gt; 5;</span></span><br><span class="line"><span class="title">counts</span> &lt;- <span class="class"><span class="keyword">data</span>[gkeep,]</span></span><br><span class="line"><span class="meta"># Library size normalization</span></span><br><span class="line"><span class="title">lib_size</span> = colSums(counts)</span><br><span class="line"><span class="title">norm</span> &lt;- t(t(counts)/lib_size * median(lib_size)) </span><br><span class="line"><span class="meta"># Variant of CPM for datasets with library sizes of fewer than 1 mil molecules</span></span><br></pre></td></tr></table></figure>

<h5 id="3-7-2-Kolmogorov-Smirnov-test"><a href="#3-7-2-Kolmogorov-Smirnov-test" class="headerlink" title="3.7.2 Kolmogorov-Smirnov test"></a>3.7.2 Kolmogorov-Smirnov test</h5><p>最容易使用的测试类型是非参数化的。最常用的非参数化测试是Kolmogorov-Smirnov test（KS-test），我们可以用它来对比两个个体中每一个基因的分布。<br>KS-test量化了这两组中每个基因表达的经验累积分布之间的距离。它对平均表达的变化和方差的变化很敏感。然而，它假设数据是连续的，并且当数据包含大量相同的值时可能表现不佳（例如。0)。KS-test的另一个问题是，它对大的样本容量非常敏感，因此它可能会变得非常重要，即使差异的大小非常小。<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/KS2_Example.png" alt="image"><br>Illustration of the two-sample Kolmogorov–Smirnov statistic. Red and blue lines each correspond to an empirical distribution function, and the black arrow is the two-sample KS statistic.</p>
<figure class="highlight monkey"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">pVals &lt;- apply(</span><br><span class="line">    norm, <span class="number">1</span>, <span class="function"><span class="keyword">function</span>(</span>x) &#123;</span><br><span class="line">        ks.test(</span><br><span class="line">            x[group == <span class="string">"NA19101"</span>], </span><br><span class="line">            x[group == <span class="string">"NA19239"</span>]</span><br><span class="line">        )$p.value</span><br><span class="line">    &#125;</span><br><span class="line">)</span><br><span class="line"><span class="meta"># multiple testing correction</span></span><br><span class="line">pVals &lt;- p.adjust(pVals, <span class="function"><span class="keyword">method</span> =</span> <span class="string">"fdr"</span>)</span><br></pre></td></tr></table></figure>

<p>“applies”函数应用于表达矩阵的每一行（1）。在这个函数中我们仅仅返回；额ks。test输出中的p.value.我们现在可以考虑，KS-test检测到多少个正和负的基因：</p>
<h6 id="3-7-2-1-Evalusting-Accuracy"><a href="#3-7-2-1-Evalusting-Accuracy" class="headerlink" title="3.7.2.1 Evalusting Accuracy"></a>3.7.2.1 Evalusting Accuracy</h6><figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br></pre></td><td class="code"><pre><span class="line">sigDE &lt;- names(pVals)[pVals &lt; <span class="number">0.05</span>]</span><br><span class="line">length(sigDE) </span><br><span class="line">## [<span class="number">1</span>] <span class="number">5095</span></span><br><span class="line"># Number <span class="keyword">of</span> KS-DE genes</span><br><span class="line">sum(GroundTruth$DE %<span class="keyword">in</span>% sigDE) </span><br><span class="line">## [<span class="number">1</span>] <span class="number">792</span></span><br><span class="line"># Number <span class="keyword">of</span> KS-DE genes that are true DE genes</span><br><span class="line">sum(GroundTruth$notDE %<span class="keyword">in</span>% sigDE)</span><br><span class="line">## [<span class="number">1</span>] <span class="number">3190</span></span><br><span class="line"># Number <span class="keyword">of</span> KS-DE genes that are truly not-DE</span><br></pre></td></tr></table></figure>

<p>正如您所看到的，通过KS检验（假阳性）确定了更多假阳性基因而不是真阳性基因（真阳性），但这可能是由于大量的NOTDE基因，因此我们通常 将这些计数归一化为真阳性率（TPR），TP /（TP + FN）和假阳性率（FPR），FP /（FP + TP）。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">tp &lt;- sum(<span class="name">GroundTruth</span>$DE %in% sigDE)</span><br><span class="line">fp &lt;- sum(<span class="name">GroundTruth</span>$notDE %in% sigDE)</span><br><span class="line">tn &lt;- sum(<span class="name">GroundTruth</span>$notDE %in% names(<span class="name">pVals</span>)[pVals &gt;= <span class="number">0.05</span>])</span><br><span class="line">fn &lt;- sum(<span class="name">GroundTruth</span>$DE %in% names(<span class="name">pVals</span>)[pVals &gt;= <span class="number">0.05</span>])</span><br><span class="line">tpr &lt;- tp/(<span class="name">tp</span> + fn)</span><br><span class="line">fpr &lt;- fp/(<span class="name">fp</span> + tn)</span><br><span class="line">cat(<span class="name">c</span>(<span class="name">tpr</span>, fpr))</span><br><span class="line">## <span class="number">0.7346939</span> <span class="number">0.2944706</span></span><br></pre></td></tr></table></figure>

<p>现在我们可以看到，TPR比FPR要高得多，表明KS测试可以识别DE基因。<br>到目前为止，我们只评估了单个显着性阈值的性能。 通常情况下，改变阈值并评估一系列值的性能是有益的。 然后将其绘制为receiver-operating-characteristic curve（ROC），并且可以将一般精度统计量计算为该曲线下的面积（AUC）。 我们将使用ROCR包来促进这种绘图。<br>可以看到KS检验判断的显著差异基因实在是太多了，高达5095个。所以它能找回来792个真正的差异基因。但是却找到了3190个假阳性。所以计算得到召回率73.46%，但是准确率只有29.44%，这个表现不佳。</p>
<figure class="highlight elixir"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Only consider genes for which we know the ground truth</span></span><br><span class="line">pVals &lt;- pVals[names(pVals) %<span class="keyword">in</span>% GroundTruth<span class="variable">$DE</span> | </span><br><span class="line">               names(pVals) %<span class="keyword">in</span>% GroundTruth<span class="variable">$notDE</span>] </span><br><span class="line">truth &lt;- rep(<span class="number">1</span>, times = length(pVals));</span><br><span class="line">truth[names(pVals) %<span class="keyword">in</span>% GroundTruth<span class="variable">$DE</span>] = <span class="number">0</span>;</span><br><span class="line">pred &lt;- ROCR::prediction(pVals, truth)</span><br><span class="line">perf &lt;- ROCR::performance(pred, <span class="string">"tpr"</span>, <span class="string">"fpr"</span>)</span><br><span class="line">ROCR::plot(perf)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/30-de-real_files/figure-html/ks-roc-plot-1.png" alt="image"></p>
<figure class="highlight lua"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">aucObj &lt;- ROCR::performance(pred, <span class="string">"auc"</span>)</span><br><span class="line">aucObj@y.values<span class="string">[[1]]</span> # AUC</span><br></pre></td></tr></table></figure>

<p>Finally to facilitate the comparisons of other DE methods let’s put this code into a function so we don’t need to repeat it:</p>
<figure class="highlight lua"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">DE_Quality_AUC &lt;- <span class="function"><span class="keyword">function</span><span class="params">(pVals)</span></span> &#123;</span><br><span class="line">    pVals &lt;- pVals[names(pVals) %<span class="keyword">in</span>% GroundTruth$DE | </span><br><span class="line">                   names(pVals) %<span class="keyword">in</span>% GroundTruth$notDE]</span><br><span class="line">    truth &lt;- <span class="built_in">rep</span>(<span class="number">1</span>, times = length(pVals));</span><br><span class="line">    truth[names(pVals) %<span class="keyword">in</span>% GroundTruth$DE] = <span class="number">0</span>;</span><br><span class="line">    pred &lt;- ROCR::prediction(pVals, truth)</span><br><span class="line">    perf &lt;- ROCR::performance(pred, <span class="string">"tpr"</span>, <span class="string">"fpr"</span>)</span><br><span class="line">    ROCR::plot(perf)</span><br><span class="line">    aucObj &lt;- ROCR::performance(pred, <span class="string">"auc"</span>)</span><br><span class="line">    <span class="keyword">return</span>(aucObj@y.values<span class="string">[[1]]</span>)</span><br><span class="line">&#125;</span><br></pre></td></tr></table></figure>

<h5 id="3-7-3-Wilcox-Mann-Whitney-U-Test"><a href="#3-7-3-Wilcox-Mann-Whitney-U-Test" class="headerlink" title="3.7.3 Wilcox/Mann-Whitney-U Test"></a>3.7.3 Wilcox/Mann-Whitney-U Test</h5><p>Wilcox-rank-sum test是另一种非参数检验，但是这个test对于一组中的值是否小于、大于另一组中的值更加有针对性。因此，它通常被认为是两组之间中值差异的检验，而KS-test对于表达值分布的任何变化都很敏感。</p>
<figure class="highlight monkey"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">pVals &lt;- apply(</span><br><span class="line">    norm, <span class="number">1</span>, <span class="function"><span class="keyword">function</span>(</span>x) &#123;</span><br><span class="line">        wilcox.test(</span><br><span class="line">            x[group == <span class="string">"NA19101"</span>], </span><br><span class="line">            x[group == <span class="string">"NA19239"</span>]</span><br><span class="line">        )$p.value</span><br><span class="line">    &#125;</span><br><span class="line">)</span><br><span class="line"><span class="meta"># multiple testing correction</span></span><br><span class="line">pVals &lt;- p.adjust(pVals, <span class="function"><span class="keyword">method</span> =</span> <span class="string">"fdr"</span>)</span><br><span class="line">DE_Quality_AUC(pVals)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/30-de-real_files/figure-html/wilcox-plot-1.png" alt="image"></p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">0.8320326</span></span><br></pre></td></tr></table></figure>

<h5 id="3-7-4-edgeR"><a href="#3-7-4-edgeR" class="headerlink" title="3.7.4 edgeR"></a>3.7.4 edgeR</h5><p>我们已经在第2.10章中使用了edgeR来表示差异表达式。 edgeR基于基因表达的负二项式模型，并使用广义线性模型（GLM）框架，使我们能够将其他因素（例如批次）包含到模型中。</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">dge &lt;- DGEList(</span><br><span class="line">    counts = counts, </span><br><span class="line">    norm.factors = rep(1, length(counts[1,])), </span><br><span class="line">   <span class="built_in"> group </span>= group</span><br><span class="line">)</span><br><span class="line">group_edgeR &lt;- factor(group)</span><br><span class="line">design &lt;- model.matrix(~ group_edgeR)</span><br><span class="line">dge &lt;- estimateDisp(dge, design = design, trend.method = <span class="string">"none"</span>)</span><br><span class="line">fit &lt;- glmFit(dge, design)</span><br><span class="line">res &lt;- glmLRT(fit)</span><br><span class="line">pVals &lt;- res<span class="variable">$table</span>[,4]</span><br><span class="line">names(pVals) &lt;- rownames(res<span class="variable">$table</span>)</span><br><span class="line"></span><br><span class="line">pVals &lt;- p.adjust(pVals, method = <span class="string">"fdr"</span>)</span><br><span class="line">DE_Quality_AUC(pVals)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/30-de-real_files/figure-html/edger-plot-1.png" alt="image"></p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">0.8466764</span></span><br></pre></td></tr></table></figure>

<h5 id="3-7-5-Monocle"><a href="#3-7-5-Monocle" class="headerlink" title="3.7.5 Monocle"></a>3.7.5 Monocle</h5><p>Monocle可以使用几种不同的DE模型。 对于count data，它建议使用负二项模型（negbinomial.size）。 对于标准化数据，它建议使用正态分布（gaussianff）对其进行对数变换。 与edgeR类似，此方法使用GLM框架，因此理论上可以考虑批次，但实际上，如果包含批次，则模型将无法适用于此数据集。</p>
<figure class="highlight haskell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">pd</span> &lt;- <span class="class"><span class="keyword">data</span>.frame(<span class="title">group</span> = <span class="title">group</span>, <span class="title">batch</span> = <span class="title">batch</span>)</span></span><br><span class="line"><span class="title">rownames</span>(pd) &lt;- colnames(counts)</span><br><span class="line"><span class="title">pd</span> &lt;- new(<span class="string">"AnnotatedDataFrame"</span>, <span class="class"><span class="keyword">data</span> = pd)</span></span><br><span class="line"></span><br><span class="line"><span class="type">Obj</span> &lt;- newCellDataSet(</span><br><span class="line">    <span class="keyword">as</span>.matrix(counts), </span><br><span class="line">    phenoData = pd, </span><br><span class="line">    expressionFamily = negbinomial.size()</span><br><span class="line">)</span><br><span class="line"><span class="type">Obj</span> &lt;- estimateSizeFactors(<span class="type">Obj</span>)</span><br><span class="line"><span class="type">Obj</span> &lt;- estimateDispersions(<span class="type">Obj</span>)</span><br><span class="line"><span class="title">res</span> &lt;- differentialGeneTest(<span class="type">Obj</span>, fullModelFormulaStr = <span class="string">"~group"</span>)</span><br><span class="line"></span><br><span class="line"><span class="title">pVals</span> &lt;- res[,<span class="number">3</span>]</span><br><span class="line"><span class="title">names</span>(pVals) &lt;- rownames(res)</span><br><span class="line"><span class="title">pVals</span> &lt;- p.adjust(pVals, method = <span class="string">"fdr"</span>)</span><br><span class="line"><span class="type">DE_Quality_AUC</span>(pVals)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/30-de-real_files/figure-html/Monocle-plot-1.png" alt="image"></p>
<h5 id="3-7-6-MAST"><a href="#3-7-6-MAST" class="headerlink" title="3.7.6 MAST"></a>3.7.6 MAST</h5><p>MAST是基于zero-inflated的负二项模型。它通过一个障碍模型测试差异表达式，将离散（0 vs非零）和连续（非零值）的基因表达的测试结合起来。同样，它使用一个线性建模框架来考虑复杂的模型。</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br></pre></td><td class="code"><pre><span class="line">log_counts &lt;- log(counts + <span class="number">1</span>) / log(<span class="number">2</span>)</span><br><span class="line">fData &lt;- data.frame(names = rownames(log_counts))</span><br><span class="line">rownames(fData) &lt;- rownames(log_counts);</span><br><span class="line">cData &lt;- data.frame(cond = group)</span><br><span class="line">rownames(cData) &lt;- colnames(log_counts)</span><br><span class="line"></span><br><span class="line">obj &lt;- FromMatrix(<span class="keyword">as</span>.matrix(log_counts), cData, fData)</span><br><span class="line">colData(obj)$cngeneson &lt;- scale(colSums(assay(obj) &gt; <span class="number">0</span>))</span><br><span class="line">cond &lt;- factor(colData(obj)$cond)</span><br><span class="line"></span><br><span class="line"># Model expression <span class="keyword">as</span> function <span class="keyword">of</span> condition &amp; number <span class="keyword">of</span> detected genes</span><br><span class="line">zlmCond &lt;- zlm.SingleCellAssay(~ cond + cngeneson, obj) </span><br><span class="line">## Warning: <span class="string">'zlm.SingleCellAssay'</span> is deprecated.</span><br><span class="line">## Use <span class="string">'zlm'</span> instead.</span><br><span class="line">## See help(<span class="string">"Deprecated"</span>)</span><br><span class="line">## Warning <span class="keyword">in</span> .nextMethod(object = object, value = value): Coefficients</span><br><span class="line">## condNA19239 are never estimible and will be dropped.</span><br><span class="line">summaryCond &lt;- summary(zlmCond, doLRT = <span class="string">"condNA19101"</span>)</span><br><span class="line">summaryDt &lt;- summaryCond$datatable</span><br><span class="line"></span><br><span class="line">summaryDt &lt;- <span class="keyword">as</span>.data.frame(summaryDt)</span><br><span class="line">pVals &lt;- unlist(summaryDt[summaryDt$component == <span class="string">"H"</span>,<span class="number">4</span>]) # H = hurdle model</span><br><span class="line">names(pVals) &lt;- unlist(summaryDt[summaryDt$component == <span class="string">"H"</span>,<span class="number">1</span>])</span><br><span class="line">pVals &lt;- p.adjust(pVals, method = <span class="string">"fdr"</span>)</span><br><span class="line">DE_Quality_AUC(pVals)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/30-de-real_files/figure-html/MAST-plot-1.png" alt="image"></p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">0.8284046</span></span><br></pre></td></tr></table></figure>

<h5 id="3-7-7-Slow-Methods"><a href="#3-7-7-Slow-Methods" class="headerlink" title="3.7.7 Slow Methods"></a>3.7.7 Slow Methods</h5><p>这个方法运行时间很长</p>
<h5 id="3-7-8-BPSC"><a href="#3-7-8-BPSC" class="headerlink" title="3.7.8 BPSC"></a>3.7.8 BPSC</h5><p>BPSC使用了单细胞基因表达的Poisson-beta模型，我们在前一章中讨论过，并将它与我们在使用edgeR时已经遇到的广义线性模型结合在一起。BPSC将一个或多个组与一个参照组（“control”）进行比较，并且可以包括其他因素，比如模型中的批次。</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">library(BPSC)</span><br><span class="line">bpsc_data &lt;- norm[,<span class="attribute">batch</span>=="NA19101.r1" | <span class="attribute">batch</span>=="NA19239.r1"]</span><br><span class="line">bpsc_group = group[<span class="attribute">batch</span>=="NA19101.r1" | <span class="attribute">batch</span>=="NA19239.r1"]</span><br><span class="line"></span><br><span class="line">control_cells &lt;- which(bpsc_group == <span class="string">"NA19101"</span>)</span><br><span class="line">design &lt;- model.matrix(~bpsc_group)</span><br><span class="line"><span class="attribute">coef</span>=2 #<span class="built_in"> group </span>label</span><br><span class="line"><span class="attribute">res</span>=BPglm(data=bpsc_data, <span class="attribute">controlIds</span>=control_cells, <span class="attribute">design</span>=design, <span class="attribute">coef</span>=coef, </span><br><span class="line">                <span class="attribute">estIntPar</span>=<span class="literal">FALSE</span>, useParallel = <span class="literal">FALSE</span>)</span><br><span class="line">pVals = res<span class="variable">$PVAL</span></span><br><span class="line">pVals &lt;- p.adjust(pVals, method = <span class="string">"fdr"</span>)</span><br><span class="line">DE_Quality_AUC(pVals)</span><br></pre></td></tr></table></figure>

<h5 id="3-7-9-SCDE"><a href="#3-7-9-SCDE" class="headerlink" title="3.7.9 SCDE"></a>3.7.9 SCDE</h5><p>SCDE是第一个单细胞特异的DE方法。它适用于使用贝叶斯统计数据的零膨胀负二项模型。下面的用法测试了不同群体中单个基因的平均表达差异，但最近的版本包括了测试平均表达或基因组分散的方法，通常代表一条途径。</p>
<figure class="highlight dust"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br></pre></td><td class="code"><pre><span class="line"><span class="xml">library(scde)</span></span><br><span class="line"><span class="xml">cnts <span class="tag">&lt;<span class="name">-</span> <span class="attr">apply</span>(</span></span></span><br><span class="line"><span class="xml">    counts,</span></span><br><span class="line"><span class="xml">    2,</span></span><br><span class="line"><span class="xml">    function(x) </span><span class="template-variable">&#123;</span></span><br><span class="line"><span class="template-variable">        storage.mode(x) &lt;- 'integer'</span></span><br><span class="line"><span class="template-variable">        return(x)</span></span><br><span class="line"><span class="template-variable">    &#125;</span><span class="xml"></span></span><br><span class="line"><span class="xml">)</span></span><br><span class="line"><span class="xml">names(group) <span class="tag">&lt;<span class="name">-</span> <span class="attr">1:length</span>(<span class="attr">group</span>)</span></span></span><br><span class="line"><span class="xml">colnames(cnts) <span class="tag">&lt;<span class="name">-</span> <span class="attr">1:length</span>(<span class="attr">group</span>)</span></span></span><br><span class="line"><span class="xml">o.ifm <span class="tag">&lt;<span class="name">-</span> <span class="attr">scde::scde.error.models</span>(</span></span></span><br><span class="line"><span class="xml">    counts = cnts,</span></span><br><span class="line"><span class="xml">    groups = group,</span></span><br><span class="line"><span class="xml">    n.cores = 1,</span></span><br><span class="line"><span class="xml">    threshold.segmentation = TRUE,</span></span><br><span class="line"><span class="xml">    save.crossfit.plots = FALSE,</span></span><br><span class="line"><span class="xml">    save.model.plots = FALSE,</span></span><br><span class="line"><span class="xml">    verbose = 0,</span></span><br><span class="line"><span class="xml">    min.size.entries = 2</span></span><br><span class="line"><span class="xml">)</span></span><br><span class="line"><span class="xml">priors <span class="tag">&lt;<span class="name">-</span> <span class="attr">scde::scde.expression.prior</span>(</span></span></span><br><span class="line"><span class="xml">    models = o.ifm,</span></span><br><span class="line"><span class="xml">    counts = cnts,</span></span><br><span class="line"><span class="xml">    length.out = 400,</span></span><br><span class="line"><span class="xml">    show.plot = FALSE</span></span><br><span class="line"><span class="xml">)</span></span><br><span class="line"><span class="xml">resSCDE <span class="tag">&lt;<span class="name">-</span> <span class="attr">scde::scde.expression.difference</span>(</span></span></span><br><span class="line"><span class="xml">    o.ifm,</span></span><br><span class="line"><span class="xml">    cnts,</span></span><br><span class="line"><span class="xml">    priors,</span></span><br><span class="line"><span class="xml">    groups = group,</span></span><br><span class="line"><span class="xml">    n.randomizations = 100,</span></span><br><span class="line"><span class="xml">    n.cores = 1,</span></span><br><span class="line"><span class="xml">    verbose = 0</span></span><br><span class="line"><span class="xml">)</span></span><br><span class="line"><span class="xml"># Convert Z-scores into 2-tailed p-values</span></span><br><span class="line"><span class="xml">pVals <span class="tag">&lt;<span class="name">-</span> <span class="attr">pnorm</span>(<span class="attr">abs</span>(<span class="attr">resSCDE</span>$<span class="attr">cZ</span>), <span class="attr">lower.tail</span> = <span class="string">FALSE)</span> * <span class="attr">2</span></span></span></span><br><span class="line"><span class="xml">DE_Quality_AUC(pVals)</span></span><br></pre></td></tr></table></figure>

<h4 id="3-8-Comparing-Combining-scRNAseq-datasets"><a href="#3-8-Comparing-Combining-scRNAseq-datasets" class="headerlink" title="3.8 Comparing/Combining scRNAseq datasets"></a>3.8 Comparing/Combining scRNAseq datasets</h4><p>随着越来越多的scRNA-seq数据集变得可用，将merged-seurat进行比较是关键。比较scRNASeq数据集有两种主要方法。第一种方法是“label-centric”，它专注于通过比较单个细胞或细胞群来识别不同数据集的等效细胞类型/状态。另一种方法是“交叉数据标准化(cross-dataset normalization)”，它试图通过计算来消除特定于实验的技术/生物效应，从而使来自多个实验的数据能够被组合和联合分析。<br>以标签为中心的方法可以与具有高置信度细胞注释的数据集一起使用，例如，人类细胞图谱（HCA）（Regev等人2017）或Tabula Muris（???）一旦完成，将细胞或簇从新样本投射到该参考上以考虑组织成分和/或识别细胞小说/未知身份。从概念上讲，这种预测类似于流行的BLAST方法（Altschul等人，1990），这使得可以在数据库中快速找到最接近的匹配，用于新鉴定的核苷酸或氨基酸序列。以标签为中心的方法还可用于比较由不同实验室收集的相似生物来源的数据集，以确保注释和分析是一致的<br>cross-dataset标准化方法也可以用来比较类似的生物起源的数据集，不像label-centric，它支持多个数据集的连接分析，以方便识别罕见的细胞类型，这些类型可能在每个单独的数据集中被很稀疏地取样，以便可靠地检测到。然而，cross-dataset规范化并不适用于非常大的和多样化的引用，因为它假定每个数据集的大部分生物变量都与其他数据集重叠。</p>
<h6 id="3-8-2-Datasets"><a href="#3-8-2-Datasets" class="headerlink" title="3.8.2 Datasets"></a>3.8.2 Datasets</h6><p>我们将在两个人类胰腺数据集上运行这些方法。由于胰腺已经被广泛研究，这些数据集都有很好的注释。</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">muraro &lt;- readRDS(<span class="string">"pancreas/muraro.rds"</span>)</span><br><span class="line">segerstolpe &lt;- readRDS(<span class="string">"pancreas/segerstolpe.rds"</span>)</span><br></pre></td></tr></table></figure>

<p>此数据已针对scmap格式化。 cell类型标签必须存储在colData的cell_type1列中，并且两个数据集之间一致的基因ID必须存储在rowData的feature_symbol列中。<br>首先，让我们检查两个数据集的基因id匹配：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sum(<span class="name">rowData</span>(<span class="name">muraro</span>)$feature_symbol %in% rowData(<span class="name">segerstolpe</span>)$feature_symbol)/nrow(<span class="name">muraro</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">0.9599519</span></span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">sum(<span class="name">rowData</span>(<span class="name">segerstolpe</span>)$feature_symbol %in% rowData(<span class="name">muraro</span>)$feature_symbol)/nrow(<span class="name">segerstolpe</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">## [<span class="number">1</span>] <span class="number">0.719334</span></span><br></pre></td></tr></table></figure>

<p>在这里，我们可以看到muraro中存在的96％基因与segerstople中的基因匹配，而segerstolpe中72％的基因是muraro中的匹配基因。 这是预期的，因为segerstolpe数据集比muraro数据集更深入地排序。 然而，它强调了比较scRNASeq数据集的一些困难。<br>我们可以通过检查这两个数据集的总体大小来确认这一点。</p>
<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">dim(muraro)</span><br><span class="line">## [<span class="number">1</span>] <span class="number">19127</span>  <span class="number">2126</span></span><br><span class="line">dim(segerstolpe)</span><br><span class="line">## [<span class="number">1</span>] <span class="number">25525</span>  <span class="number">3514</span></span><br></pre></td></tr></table></figure>

<p>此外，我们还可以使用下面的命令检查每个数据集的cell类型注解：</p>
<figure class="highlight shell"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"><span class="meta">summary(factor(colData(muraro)$</span><span class="bash">cell_type1))</span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#      acinar       alpha        beta       delta      ductal endothelial </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#         219         812         448         193         245          21 </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#     epsilon       gamma mesenchymal     unclear </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#           3         101          80           4</span></span></span><br><span class="line"><span class="meta">summary(factor(colData(segerstolpe)$</span><span class="bash">cell_type1))</span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#                 acinar                  alpha                   beta </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#                    185                    886                    270 </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#          co-expression                  delta                 ductal </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#                     39                    114                    386 </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#            endothelial                epsilon                  gamma </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#                     16                      7                    197 </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#                   mast           MHC class II         not applicable </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#                      7                      5                   1305 </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#                    PSC           unclassified unclassified endocrine </span></span></span><br><span class="line"><span class="meta">#</span><span class="bash"><span class="comment">#                     54                      2                     41</span></span></span><br></pre></td></tr></table></figure>

<p>在这里我们可以看到，尽管两个数据集都将两个数据集视为相同的生物组织，但它们已经注释了一组略有不同的细胞类型。 如果您熟悉胰腺生物学，您可能会认识到segerstolpe中的胰腺星状细胞（PSCs）是一种间充质干细胞，在muraro中属于“间充质”类型。 但是，目前尚不清楚这两个注释是否应该被视为同义词。 我们可以使用以标签为中心的比较方法来确定这两个细胞类型的注释是否确实相同。<br>或者，我们可能有兴趣了解那些“未分类的内分泌”细胞的功能，或者通过利用数据集中的形成来理解每个数据集中原始聚类的质量太差（“不适用”）。 我们可以尝试使用以标签为中心的方法来推断它们最可能属于哪些现有注释，或者我们可以尝试使用跨数据集规范化来揭示其中的新颖cell类型（或现有注释中的子类型）。<br>为了简化我们的演示分析，我们将删除少量未分配的cell和质量较差的cell。 我们将保留“未分类的内分泌”，看看这些方法中是否有任何一种可以阐明它们属于哪种细胞类型。</p>
<figure class="highlight nginx"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="attribute">segerstolpe</span> &lt;- segerstolpe[,colData(segerstolpe)<span class="variable">$cell_type1</span> != <span class="string">"unclassified"</span>]</span><br><span class="line">segerstolpe &lt;- segerstolpe[,colData(segerstolpe)<span class="variable">$cell_type1</span> != <span class="string">"not applicable"</span>,]</span><br><span class="line">muraro &lt;- muraro[,colData(muraro)<span class="variable">$cell_type1</span> != <span class="string">"unclear"</span>]</span><br></pre></td></tr></table></figure>

<h5 id="3-8-3-Projecting-cells-onto-annotated-cell-types（scmap）"><a href="#3-8-3-Projecting-cells-onto-annotated-cell-types（scmap）" class="headerlink" title="3.8.3 Projecting cells onto annotated cell-types（scmap）"></a>3.8.3 Projecting cells onto annotated cell-types（scmap）</h5><p>我们最近开发了scmap（kis列夫和Hemberg，2017）——一种将细胞从scRNA-seq实验投射到其他实验中确定的细胞类型的方法。</p>
<h6 id="3-8-3-1-Feature-Selection"><a href="#3-8-3-1-Feature-Selection" class="headerlink" title="3.8.3.1 Feature Selection"></a>3.8.3.1 Feature Selection</h6><p>一旦我们有了SingleCellExperiment对象，我们就可以运行scmap了。 首先，我们必须构建参考集群的“索引”。 由于我们想知道PSC和间充质细胞是否是同义词，我们将每个数据集投射到另一个数据集，因此我们将为每个数据集构建索引。 这需要首先为参考数据集选择信息量最大的特征。</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">muraro &lt;- selectFeatures(<span class="name">muraro</span>, suppress_plot = FALSE)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/31-projection_files/figure-html/unnamed-chunk-12-1.png" alt="image"></p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">segerstolpe &lt;- selectFeatures(<span class="name">segerstolpe</span>, suppress_plot = FALSE)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/31-projection_files/figure-html/unnamed-chunk-13-1.png" alt="image"><br>从这些图的y轴我们可以看到scmap使用基于dropmerged-seserat的特征选择方法。<br>计算cell-type索引：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">muraro &lt;- indexCluster(<span class="name">muraro</span>)</span><br><span class="line">segerstolpe &lt;- indexCluster(<span class="name">segerstolpe</span>)</span><br></pre></td></tr></table></figure>

<p>可视化索引：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">heatmap(<span class="name">as</span>.matrix(<span class="name">metadata</span>(<span class="name">muraro</span>)$scmap_cluster_index))</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/31-projection_files/figure-html/unnamed-chunk-15-1.png" alt="image"><br>如果功能过于集中在少数单元类型中，您可能需要使用setFeatures函数调整功能。 在这种情况下，基于dropmerged_seurat的功能看起来很好</p>
<h6 id="3-8-3-2-Projecting"><a href="#3-8-3-2-Projecting" class="headerlink" title="3.8.3.2 Projecting"></a>3.8.3.2 Projecting</h6><p>scmap计算从每个cell到参考索引中每个cell类型的距离，然后应用经验导出的阈值来确定将哪些cell分配给最近的参考cell类型以及哪些cell不分配。为了解释测序深度的差异，使用spearman相关和余弦距离计算距离，并且仅将具有两个距离的一致分配的单元返回为指定的。<br>我们将把segerstolpe数据集投射到muraro数据集：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">seger_to_muraro &lt;- scmapCluster(</span><br><span class="line">  projection = segerstolpe,</span><br><span class="line">  index_list = list(</span><br><span class="line">    muraro = metadata(<span class="name">muraro</span>)$scmap_cluster_index</span><br><span class="line">  )</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>muraro 投射到 segerstolpe</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">muraro_to_seger &lt;- scmapCluster(</span><br><span class="line">  projection = muraro,</span><br><span class="line">  index_list = list(</span><br><span class="line">    seger = metadata(<span class="name">segerstolpe</span>)$scmap_cluster_index</span><br><span class="line">  )</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>现在，让我们将原始的细胞类型标签与投影标签进行比较：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">table</span><span class="params">(colData(muraro)</span></span><span class="variable">$cell_type1</span>, muraro_to_seger<span class="variable">$scmap_cluster_labs</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br></pre></td><td class="code"><pre><span class="line">##              </span><br><span class="line">##               acinar alpha beta co-expression delta ductal endothelial</span><br><span class="line">##   acinar         <span class="number">211</span>     <span class="number">0</span>    <span class="number">0</span>             <span class="number">0</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   alpha            <span class="number">1</span>   <span class="number">763</span>    <span class="number">0</span>            <span class="number">18</span>     <span class="number">0</span>      <span class="number">2</span>           <span class="number">0</span></span><br><span class="line">##   beta             <span class="number">2</span>     <span class="number">1</span>  <span class="number">397</span>             <span class="number">7</span>     <span class="number">2</span>      <span class="number">2</span>           <span class="number">0</span></span><br><span class="line">##   delta            <span class="number">0</span>     <span class="number">0</span>    <span class="number">2</span>             <span class="number">1</span>   <span class="number">173</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   ductal           <span class="number">7</span>     <span class="number">0</span>    <span class="number">0</span>             <span class="number">0</span>     <span class="number">0</span>    <span class="number">208</span>           <span class="number">0</span></span><br><span class="line">##   endothelial      <span class="number">0</span>     <span class="number">0</span>    <span class="number">0</span>             <span class="number">0</span>     <span class="number">0</span>      <span class="number">0</span>          <span class="number">15</span></span><br><span class="line">##   epsilon          <span class="number">0</span>     <span class="number">0</span>    <span class="number">0</span>             <span class="number">0</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   gamma            <span class="number">2</span>     <span class="number">0</span>    <span class="number">0</span>             <span class="number">0</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   mesenchymal      <span class="number">0</span>     <span class="number">0</span>    <span class="number">0</span>             <span class="number">0</span>     <span class="number">0</span>      <span class="number">1</span>           <span class="number">0</span></span><br><span class="line">##              </span><br><span class="line">##               epsilon gamma MHC class II PSC unassigned</span><br><span class="line">##   acinar            <span class="number">0</span>     <span class="number">0</span>            <span class="number">0</span>   <span class="number">0</span>          <span class="number">8</span></span><br><span class="line">##   alpha             <span class="number">0</span>     <span class="number">2</span>            <span class="number">0</span>   <span class="number">0</span>         <span class="number">26</span></span><br><span class="line">##   beta              <span class="number">0</span>     <span class="number">5</span>            <span class="number">1</span>   <span class="number">2</span>         <span class="number">29</span></span><br><span class="line">##   delta             <span class="number">0</span>     <span class="number">0</span>            <span class="number">0</span>   <span class="number">0</span>         <span class="number">17</span></span><br><span class="line">##   ductal            <span class="number">0</span>     <span class="number">0</span>            <span class="number">5</span>   <span class="number">3</span>         <span class="number">22</span></span><br><span class="line">##   endothelial       <span class="number">0</span>     <span class="number">0</span>            <span class="number">0</span>   <span class="number">1</span>          <span class="number">5</span></span><br><span class="line">##   epsilon           <span class="number">3</span>     <span class="number">0</span>            <span class="number">0</span>   <span class="number">0</span>          <span class="number">0</span></span><br><span class="line">##   gamma             <span class="number">0</span>    <span class="number">95</span>            <span class="number">0</span>   <span class="number">0</span>          <span class="number">4</span></span><br><span class="line">##   mesenchymal       <span class="number">0</span>     <span class="number">0</span>            <span class="number">0</span>  <span class="number">77</span>          <span class="number">2</span></span><br></pre></td></tr></table></figure>

<p>在这里，我们可以看到细胞类型在segerstolpe中映射到它们的等价物，而且重要的是我们看到所有的“mesenchymal”细胞都被分配到“PSC”类中。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">table</span><span class="params">(colData(segerstolpe)</span></span><span class="variable">$cell_type1</span>, seger_to_muraro<span class="variable">$scmap_cluster_labs</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br></pre></td><td class="code"><pre><span class="line">##                         </span><br><span class="line">##                          acinar alpha beta delta ductal endothelial</span><br><span class="line">##   acinar                    <span class="number">181</span>     <span class="number">0</span>    <span class="number">0</span>     <span class="number">0</span>      <span class="number">4</span>           <span class="number">0</span></span><br><span class="line">##   alpha                       <span class="number">0</span>   <span class="number">869</span>    <span class="number">1</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   beta                        <span class="number">0</span>     <span class="number">0</span>  <span class="number">260</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   co-expression               <span class="number">0</span>     <span class="number">7</span>   <span class="number">31</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   delta                       <span class="number">0</span>     <span class="number">0</span>    <span class="number">1</span>   <span class="number">111</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   ductal                      <span class="number">0</span>     <span class="number">0</span>    <span class="number">0</span>     <span class="number">0</span>    <span class="number">383</span>           <span class="number">0</span></span><br><span class="line">##   endothelial                 <span class="number">0</span>     <span class="number">0</span>    <span class="number">0</span>     <span class="number">0</span>      <span class="number">0</span>          <span class="number">14</span></span><br><span class="line">##   epsilon                     <span class="number">0</span>     <span class="number">0</span>    <span class="number">0</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   gamma                       <span class="number">0</span>     <span class="number">2</span>    <span class="number">0</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   mast                        <span class="number">0</span>     <span class="number">0</span>    <span class="number">0</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   MHC class II                <span class="number">0</span>     <span class="number">0</span>    <span class="number">0</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   PSC                         <span class="number">0</span>     <span class="number">0</span>    <span class="number">1</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##   unclassified endocrine      <span class="number">0</span>     <span class="number">0</span>    <span class="number">0</span>     <span class="number">0</span>      <span class="number">0</span>           <span class="number">0</span></span><br><span class="line">##                         </span><br><span class="line">##                          epsilon gamma mesenchymal unassigned</span><br><span class="line">##   acinar                       <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>          <span class="number">0</span></span><br><span class="line">##   alpha                        <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>         <span class="number">16</span></span><br><span class="line">##   beta                         <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>         <span class="number">10</span></span><br><span class="line">##   co-expression                <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>          <span class="number">1</span></span><br><span class="line">##   delta                        <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>          <span class="number">2</span></span><br><span class="line">##   ductal                       <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>          <span class="number">3</span></span><br><span class="line">##   endothelial                  <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>          <span class="number">2</span></span><br><span class="line">##   epsilon                      <span class="number">6</span>     <span class="number">0</span>           <span class="number">0</span>          <span class="number">1</span></span><br><span class="line">##   gamma                        <span class="number">0</span>   <span class="number">192</span>           <span class="number">0</span>          <span class="number">3</span></span><br><span class="line">##   mast                         <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>          <span class="number">7</span></span><br><span class="line">##   MHC class II                 <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>          <span class="number">5</span></span><br><span class="line">##   PSC                          <span class="number">0</span>     <span class="number">0</span>          <span class="number">53</span>          <span class="number">0</span></span><br><span class="line">##   unclassified endocrine       <span class="number">0</span>     <span class="number">0</span>           <span class="number">0</span>         <span class="number">41</span></span><br></pre></td></tr></table></figure>

<p>我们再次看到细胞类型相互匹配，其中“PSCs”匹配“mesenchymal”细胞，提供强有力的证据证明这两个注释应该被认为是同义的。<br>使用Sankey diagram进行可视化：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">plot(<span class="name">getSankey</span>(<span class="name">colData</span>(<span class="name">muraro</span>)$cell_type1,  muraro_to_seger$scmap_cluster_labs[,<span class="number">1</span>], plot_height=400))</span><br></pre></td></tr></table></figure>

<h5 id="3-8-4-Cell-to-Cell-mapping"><a href="#3-8-4-Cell-to-Cell-mapping" class="headerlink" title="3.8.4 Cell-to-Cell mapping"></a>3.8.4 Cell-to-Cell mapping</h5><p>scmap还可以将一个数据集中的每个cell投射到参考数据集中的近似最近的相邻cell中。它使用了一个高度优化的搜索算法，允许将其扩展到非常大的引用（在理论上是10万-数百万个cell）。然而，这个过程是随机的，所以我们必须修正随机的种子，以确保我们能够重现我们的结果。<br>我们已经为这个数据集执行了特性选择，这样我们就可以直接构建索引了。</p>
<figure class="highlight gams"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">set</span>.seed(193047)</span><br><span class="line">segerstolpe <span class="comment">&lt;- indexCell(segerstolpe)</span></span><br><span class="line">muraro <span class="comment">&lt;- indexCell(muraro)</span></span><br></pre></td></tr></table></figure>

<p>在这种情况下，索引是每个cell的一系列集群，使用不同的特性集，参数k和M是集群的数量和每个子集群中使用的特性的数量。在每个子集群中，将新cell分配给最近的集群，以生成独特的集群分配模式。然后我们在参考资料集中找到具有相同或最相似的集群分配模式的cell。<br>我们可以使用以下方法检查参考数据集的集群分配模式：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">metadata</span><span class="params">(muraro)</span></span><span class="variable">$scmap_cell_index</span><span class="variable">$subclusters</span>[<span class="number">1</span>:<span class="number">5</span>,<span class="number">1</span>:<span class="number">5</span>]</span><br></pre></td></tr></table></figure>

<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">##      D28<span class="number">.1</span>_1 D28<span class="number">.1</span>_13 D28<span class="number">.1</span>_15 D28<span class="number">.1</span>_17 D28<span class="number">.1</span>_2</span><br><span class="line">## [<span class="number">1</span>,]       <span class="number">4</span>       <span class="number">42</span>       <span class="number">27</span>       <span class="number">43</span>      <span class="number">10</span></span><br><span class="line">## [<span class="number">2</span>,]       <span class="number">5</span>        <span class="number">8</span>        <span class="number">2</span>       <span class="number">33</span>      <span class="number">37</span></span><br><span class="line">## [<span class="number">3</span>,]      <span class="number">11</span>       <span class="number">32</span>       <span class="number">35</span>       <span class="number">17</span>      <span class="number">26</span></span><br><span class="line">## [<span class="number">4</span>,]       <span class="number">2</span>        <span class="number">4</span>       <span class="number">32</span>        <span class="number">2</span>      <span class="number">18</span></span><br><span class="line">## [<span class="number">5</span>,]      <span class="number">31</span>       <span class="number">18</span>       <span class="number">21</span>       <span class="number">40</span>       <span class="number">1</span></span><br></pre></td></tr></table></figure>

<p>为了找到w最近的邻居，我们使用了一个与之前类似的命令：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">muraro_to_seger &lt;- scmapCell(</span><br><span class="line">  projection = muraro,</span><br><span class="line">  index_list = list(</span><br><span class="line">    seger = metadata(<span class="name">segerstolpe</span>)$scmap_cell_index</span><br><span class="line">  ),</span><br><span class="line">  w = <span class="number">5</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>查看结果：</p>
<figure class="highlight markdown"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">muraro<span class="emphasis">_to_</span>seger$seger[<span class="string">[1</span>]][,1:5]</span><br><span class="line"><span class="section">##      D28.1_1 D28.1_13 D28.1_15 D28.1_17 D28.1_2</span></span><br><span class="line"><span class="section">## [1,]    2201     1288     1117     1623    1078</span></span><br><span class="line"><span class="section">## [2,]    1229     1724     2104     1448    1593</span></span><br><span class="line"><span class="section">## [3,]    1793     1854     2201     2039    1553</span></span><br><span class="line"><span class="section">## [4,]    1882     1737     1081     1202    1890</span></span><br><span class="line"><span class="section">## [5,]    1731      976     1903     1834    1437</span></span><br></pre></td></tr></table></figure>

<p>这显示了segerstolpe中5列的最近邻居到muraro中的每个cell。然后，我们可以通过从segerstolpe数据集的colData中选择适当的数据来计算伪时间估计，分支分配或其他单元级数据。 作为演示，我们将找到每个cell的最近邻居的cell类型。</p>
<h5 id="3-8-5-Metaneighbour"><a href="#3-8-5-Metaneighbour" class="headerlink" title="3.8.5 Metaneighbour"></a>3.8.5 Metaneighbour</h5><p>Metaneighbour专门用于询问细胞类型标签是否在数据集之间保持一致。它有两个版本。首先是一种完全监督的方法，它假设所有数据集中都知道细胞类型，并计算这些细胞类型标记的“好”。 （下面将描述“好”的确切含义）。或者，metaneighbour可以估计数据集内部和跨数据集的所有细胞类型彼此之间的相似程度。我们将只使用无监督版本，因为它具有更广泛的适用性，并且更容易解释结果。</p>
<p>Metaneighbour通过构建细胞 - 细胞spearman相关网络来比较数据集中的细胞类型。然后该方法尝试通过其最近邻居的加权“投票”来预测每个cell的标签。然后将两个聚类之间的整体相似性评分为AUROC，用于基于这些加权投票将类型A的cell分配给类型B. AUROC为1表示在任何其他细胞出现之前，所有类型A的细胞都被分配到typeB，如果细胞被随机分配，则AUROC为0.5。</p>
<p>Metanighbour只是几个R函数而不是一个完整的包，所以我们必须使用source加载它们</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">source</span><span class="params">(<span class="string">"2017-08-28-runMN-US.R"</span>)</span></span></span><br></pre></td></tr></table></figure>

<h6 id="3-8-5-1-Prepare-Data"><a href="#3-8-5-1-Prepare-Data" class="headerlink" title="3.8.5.1 Prepare Data"></a>3.8.5.1 Prepare Data</h6><p>Metaneighbour要求在运行之前将所有数据集组合成single expression 矩阵：</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line">is.common &lt;- rowData(muraro)<span class="variable">$feature_symbol</span> %<span class="keyword">in</span>% rowData(segerstolpe)<span class="variable">$feature_symbol</span></span><br><span class="line">muraro &lt;- muraro[is.common,]</span><br><span class="line">segerstolpe &lt;- segerstolpe[match(rowData(muraro)<span class="variable">$feature_symbol</span>, rowData(segerstolpe)<span class="variable">$feature_symbol</span>),]</span><br><span class="line">rownames(segerstolpe) &lt;- rowData(segerstolpe)<span class="variable">$feature_symbol</span></span><br><span class="line">rownames(muraro) &lt;- rowData(muraro)<span class="variable">$feature_symbol</span></span><br><span class="line">identical(rownames(segerstolpe), rownames(muraro))</span><br><span class="line"><span class="comment">## [1] TRUE</span></span><br><span class="line">combined_logcounts &lt;- cbind(logcounts(muraro), logcounts(segerstolpe))</span><br><span class="line">dataset_labels &lt;- rep(c(<span class="string">"m"</span>, <span class="string">"s"</span>), <span class="attribute">times</span>=c(ncol(muraro), ncol(segerstolpe)))</span><br><span class="line">cell_type_labels &lt;- c(colData(muraro)<span class="variable">$cell_type1</span>, colData(segerstolpe)<span class="variable">$cell_type1</span>)</span><br><span class="line"></span><br><span class="line">pheno &lt;- data.frame(Sample_ID = colnames(combined_logcounts),</span><br><span class="line">                <span class="attribute">Study_ID</span>=dataset_labels,</span><br><span class="line">                <span class="attribute">Celltype</span>=paste(cell_type_labels, dataset_labels, <span class="attribute">sep</span>=<span class="string">"-"</span>))</span><br><span class="line">rownames(pheno) &lt;- colnames(combined_logcounts)</span><br></pre></td></tr></table></figure>

<p>Metaneighbor包含一种特征选择方法来识别高度可变的基因。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="selector-tag">var</span><span class="selector-class">.genes</span> = get_variable_genes(combined_logcounts, pheno)</span><br></pre></td></tr></table></figure>

<p>由于Metaneighbor比scmap慢得多，我们将对这些数据集进行抽样。</p>
<figure class="highlight elm"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="title">subset</span> &lt;- sample(<span class="number">1</span>:nrow(pheno), <span class="number">2000</span>)</span><br><span class="line"><span class="title">combined_logcounts</span> &lt;- combined_logcounts[,subset]</span><br><span class="line"><span class="title">pheno</span> &lt;- pheno[subset,]</span><br><span class="line"><span class="title">cell_type_labels</span> &lt;- cell_type_labels[subset]</span><br><span class="line"><span class="title">dataset_labels</span> &lt;- dataset_labels[subset]</span><br></pre></td></tr></table></figure>

<p>现在我们已经准备好运行metanbor。首先，我们将运行无人监督的版本，它将让我们看到两个数据集之间最相似的cell类型。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">unsup &lt;- run_MetaNeighbor_US(<span class="selector-tag">var</span><span class="selector-class">.genes</span>, combined_logcounts, unique(pheno<span class="variable">$Celltype</span>), pheno)</span><br><span class="line"><span class="function"><span class="title">heatmap</span><span class="params">(unsup)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/31-projection_files/figure-html/unnamed-chunk-32-1.png" alt="image"></p>
<h5 id="3-8-6-mnnCorrect"><a href="#3-8-6-mnnCorrect" class="headerlink" title="3.8.6 mnnCorrect"></a>3.8.6 mnnCorrect</h5><p>mnnCorrect纠正数据集以促进联合分析。它为了解释两个重复或两个不同实验之间的组成差异，它首先匹配实验中的单个细胞以找到重叠的生物结构。使用该重叠，它学习表达的哪个维度对应于生物状态以及哪个维度对应于批次/实验效果; mnnCorrect假设这些维度在高维表达空间中是彼此直系的。最后，它从整个表达式矩阵中删除批次/实验效果，以返回校正后的矩阵。<br>为了跨数据集匹配单个单元格，mnnCorrect使用余弦距离来避免库大小效应，然后识别跨数据集的互补最近邻居（k确定邻域大小）。只有重叠的生物群应该有相互最近的邻居（见下面的图b）。然而，这假设k被设置为大约数据集中最小生物群的大小，但是k太低将识别太少的相互最近邻居对以获得我们想要移除的批量效果的良好估计。<br>学习生物/技术效果的方法是使用奇异值分解，类似于我们在批量校正部分遇到的RUV，或者使用优化的irlba包进行主成分分析，这应该比SVD快。参数svd.dim指定应该保留多少维度来总结数据的生物结构，我们将它设置为三，因为我们发现使用上面的Metaneighbor的三个主要组。可以通过平滑（sigma）和/或方差调整（var.adj）进一步调整这些估计。</p>
<p>mnnCorrect还假设你已经将你的表达式矩阵子集化，以便它们以相同的顺序包含相同的基因，幸运的是，当我们为Metaneighbor设置数据时，我们已经为我们的数据集做了一些事情。<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/mnnCorrectDiagramCropped.png" alt="image"></p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">require(<span class="string">"scran"</span>)</span><br><span class="line"><span class="comment"># mnnCorrect will take several minutes to run</span></span><br><span class="line">corrected &lt;- mnnCorrect(logcounts(muraro), logcounts(segerstolpe), <span class="attribute">k</span>=20, <span class="attribute">sigma</span>=1, pc.<span class="attribute">approx</span>=<span class="literal">TRUE</span>, subset.<span class="attribute">row</span>=var.genes, svd.<span class="attribute">dim</span>=3)</span><br></pre></td></tr></table></figure>

<p>首先让我们检查一下我们找到了足够数量的mnn对，mnnCorrect返回一个数据帧列表，其中包含每个数据集的mnn对。</p>
<figure class="highlight lua"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">dim(corrected$<span class="built_in">pairs</span><span class="string">[[1]]</span>) # muraro -&gt; others</span><br><span class="line">## [<span class="number">1</span>] <span class="number">0</span> <span class="number">3</span></span><br><span class="line">dim(corrected$<span class="built_in">pairs</span><span class="string">[[2]]</span>) # seger -&gt; others</span><br><span class="line">## [<span class="number">1</span>] <span class="number">2533</span>    <span class="number">3</span></span><br></pre></td></tr></table></figure>

<p>第一列和第二列包含cell列ID，第三列包含指示第2列单元属于哪个数据集/批次的数字。 在我们的例子中，我们只比较两个数据集，因此所有mnn对都已分配给第二个表，第三个列只包含一个。</p>
<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br></pre></td><td class="code"><pre><span class="line">head(corrected$pairs[[<span class="number">2</span>]])</span><br><span class="line">## DataFrame with <span class="number">6</span> rows and <span class="number">3</span> columns</span><br><span class="line">##   current.cell other.cell other.batch</span><br><span class="line">##      &lt;<span class="type">integer</span>&gt;      &lt;Rle&gt;       &lt;Rle&gt;</span><br><span class="line">## <span class="number">1</span>         <span class="number">1553</span>          <span class="number">5</span>           <span class="number">1</span></span><br><span class="line">## <span class="number">2</span>         <span class="number">1078</span>          <span class="number">5</span>           <span class="number">1</span></span><br><span class="line">## <span class="number">3</span>         <span class="number">1437</span>          <span class="number">5</span>           <span class="number">1</span></span><br><span class="line">## <span class="number">4</span>         <span class="number">1890</span>          <span class="number">5</span>           <span class="number">1</span></span><br><span class="line">## <span class="number">5</span>         <span class="number">1569</span>          <span class="number">5</span>           <span class="number">1</span></span><br><span class="line">## <span class="number">6</span>          <span class="number">373</span>          <span class="number">5</span>           <span class="number">1</span></span><br><span class="line">total_pairs &lt;- nrow(corrected$pairs[[<span class="number">2</span>]])</span><br><span class="line">n_unique_seger &lt;- length(unique((corrected$pairs[[<span class="number">2</span>]][,<span class="number">1</span>])))</span><br><span class="line">n_unique_muraro &lt;- length(unique((corrected$pairs[[<span class="number">2</span>]][,<span class="number">2</span>])))</span><br></pre></td></tr></table></figure>

<p>mnnCorrect在n_unique_seger segerstolpe cell和n_unique_muraro muraro cells之间找到了2533组相互最近邻居。 这应该是足够数量的对，但每个数据集中的独特细胞数量较少表明我们可能没有捕获每个数据集中的完整生物信号。<br>现在我们可以创建一个组合数据集来共同分析这些数据。 但是，更正的数据不再是count，通常包含负表达值，因此某些分析工具可能不再合适。 为简单起见，我们只绘制一个联合TSNE。</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">require(<span class="string">"Rtsne"</span>)</span><br><span class="line">## Loading required package: Rtsne</span><br></pre></td></tr></table></figure>

<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">joint_expression_matrix &lt;- cbind(<span class="name">corrected</span>$corrected[[<span class="number">1</span>]], corrected$corrected[[<span class="number">2</span>]])</span><br><span class="line"></span><br><span class="line"># Tsne will take some time to run on the full dataset</span><br><span class="line">joint_tsne &lt;- Rtsne(<span class="name">t</span>(<span class="name">joint_expression_matrix</span>[rownames(<span class="name">joint_expression_matrix</span>) %in% var.genes,]), initial_dims=10, theta=0.<span class="number">75</span>,</span><br><span class="line">                        check_duplicates=FALSE, max_iter=200, stop_lying_iter=50, mom_switch_iter=50)</span><br><span class="line">dataset_labels &lt;- factor(<span class="name">rep</span>(<span class="name">c</span>(<span class="string">"m"</span>, <span class="string">"s"</span>), times=c(<span class="name">ncol</span>(<span class="name">muraro</span>), ncol(<span class="name">segerstolpe</span>))))</span><br><span class="line">cell_type_labels &lt;- factor(<span class="name">c</span>(<span class="name">colData</span>(<span class="name">muraro</span>)$cell_type1, colData(<span class="name">segerstolpe</span>)$cell_type1))</span><br><span class="line">plot(<span class="name">joint_tsne</span>$Y[,<span class="number">1</span>], joint_tsne$Y[,<span class="number">2</span>], pch=c(<span class="number">16</span>,<span class="number">1</span>)[dataset_labels], col=rainbow(<span class="name">length</span>(<span class="name">levels</span>(<span class="name">cell_type_labels</span>)))[cell_type_labels])</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/31-projection_files/figure-html/unnamed-chunk-38-1.png" alt="image"></p>
<h5 id="3-8-7-Cannonical-Correlation-Analysis（Seurat）"><a href="#3-8-7-Cannonical-Correlation-Analysis（Seurat）" class="headerlink" title="3.8.7 Cannonical Correlation Analysis（Seurat）"></a>3.8.7 Cannonical Correlation Analysis（Seurat）</h5><p>Seurat包中包含另一种用于组合多个数据集的校正方法，称为CCA。但是，与mnnCorrect不同，它不直接更正表达式矩阵本身。相反，Seurat为每个数据集找到一个较低维度的子空间，然后校正这些子空间。与mnnCorrect不同，Seurat一次只组合一对数据集。</p>
<p>Seurat使用基因 - 基因相关性通过称为典型相关分析（CCA）的方法识别数据集中的生物结构。 Seurat了解基因 - 基因相关性的共享结构，然后评估每个细胞适合这种结构的程度。假设必须通过数据特定维数减少方法而不是共享相关结构更好地描述的单元表示数据集特定的单元类型/状态，并且在对齐两个数据集之前被丢弃。最后，使用“扭曲”算法对两个数据集进行对齐，该算法以对人口密度差异稳健的方式对每个数据集的低维表示进行归一化。</p>
<p>请注意，因为Seurat占用了大量的库空间，您必须重新启动R会话才能加载它，并且不会在此页面上自动生成plots / merged_seuratput。<br>载入数据：</p>
<figure class="highlight xquery"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br></pre></td><td class="code"><pre><span class="line">muraro &lt;- readRDS(<span class="string">"pancreas/muraro.rds"</span>)</span><br><span class="line">segerstolpe &lt;- readRDS(<span class="string">"pancreas/segerstolpe.rds"</span>)</span><br><span class="line">segerstolpe &lt;- segerstolpe[,colData(segerstolpe)<span class="variable">$cell_type1</span> != <span class="string">"unclassified"</span>]</span><br><span class="line">segerstolpe &lt;- segerstolpe[,colData(segerstolpe)<span class="variable">$cell_type1</span> != <span class="string">"not applicable"</span>,]</span><br><span class="line">muraro &lt;- muraro[,colData(muraro)<span class="variable">$cell_type1</span> != <span class="string">"unclear"</span>]</span><br><span class="line"><span class="literal">is</span>.common &lt;- rowData(muraro)<span class="variable">$feature_symbol</span> <span class="meta">%in</span>% rowData(segerstolpe)<span class="variable">$feature_symbol</span></span><br><span class="line">muraro &lt;- muraro[<span class="literal">is</span>.common,]</span><br><span class="line">segerstolpe &lt;- segerstolpe[match(rowData(muraro)<span class="variable">$feature_symbol</span>, rowData(segerstolpe)<span class="variable">$feature_symbol</span>),]</span><br><span class="line">rownames(segerstolpe) &lt;- rowData(segerstolpe)<span class="variable">$feature_symbol</span></span><br><span class="line">rownames(muraro) &lt;- rowData(muraro)<span class="variable">$feature_symbol</span></span><br><span class="line">identical(rownames(segerstolpe), rownames(muraro))</span><br></pre></td></tr></table></figure>

<p>首先，我们将把数据重新格式化为Seurat对象：</p>
<figure class="highlight elixir"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">require</span>(<span class="string">"Seurat"</span>)</span><br><span class="line">set.seed(<span class="number">4719364</span>)</span><br><span class="line">muraro_seurat &lt;- CreateSeuratObject(raw.data=assays(muraro)[[<span class="string">"normcounts"</span>]]) <span class="comment"># raw counts aren't available for muraro</span></span><br><span class="line">muraro_seurat<span class="variable">@meta</span>.data[, <span class="string">"dataset"</span>] &lt;- <span class="number">1</span></span><br><span class="line">muraro_seurat<span class="variable">@meta</span>.data[, <span class="string">"celltype"</span>] &lt;- paste(<span class="string">"m"</span>,colData(muraro)<span class="variable">$cell_type1</span>, sep=<span class="string">"-"</span>)</span><br><span class="line"></span><br><span class="line">seger_seurat &lt;- CreateSeuratObject(raw.data=assays(segerstolpe)[[<span class="string">"counts"</span>]])</span><br><span class="line">seger_seurat<span class="variable">@meta</span>.data[, <span class="string">"dataset"</span>] &lt;- <span class="number">2</span></span><br><span class="line">seger_seurat<span class="variable">@meta</span>.data[, <span class="string">"celltype"</span>] &lt;- paste(<span class="string">"s"</span>,colData(segerstolpe)<span class="variable">$cell_type1</span>, sep=<span class="string">"-"</span>)</span><br></pre></td></tr></table></figure>

<p>接下来，我们必须对每个数据集进行标准化、缩放和识别高度可变的基因：</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">muraro_seurat &lt;- NormalizeData(<span class="attribute">object</span>=muraro_seurat)</span><br><span class="line">muraro_seurat &lt;- ScaleData(<span class="attribute">object</span>=muraro_seurat)</span><br><span class="line">muraro_seurat &lt;- FindVariableGenes(<span class="attribute">object</span>=muraro_seurat, <span class="keyword">do</span>.<span class="attribute">plot</span>=<span class="literal">TRUE</span>)</span><br><span class="line"></span><br><span class="line">seger_seurat &lt;- NormalizeData(<span class="attribute">object</span>=seger_seurat)</span><br><span class="line">seger_seurat &lt;- ScaleData(<span class="attribute">object</span>=seger_seurat)</span><br><span class="line">seger_seurat &lt;- FindVariableGenes(<span class="attribute">object</span>=seger_seurat, <span class="keyword">do</span>.<span class="attribute">plot</span>=<span class="literal">TRUE</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/muraro_seurat_hvg.png" alt="image"><br>         muraro variable genes<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/seger_seurat_hvg.png" alt="image"><br>          segerstolpe variable genes<br>虽然Seurat校正了dispersion和平均表达之间的关系，但在排序特征时它不会使用校正值。 将以下命令的结果与上图中的结果进行比较：</p>
<figure class="highlight less"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="selector-tag">head</span>(muraro_seurat<span class="variable">@hvg</span>.info, <span class="number">50</span>)</span><br><span class="line"><span class="selector-tag">head</span>(seger_seurat<span class="variable">@hvg</span>.info, <span class="number">50</span>)</span><br></pre></td></tr></table></figure>

<p>但是我们将按照他们的例子使用2000个最分散withmerged_seurat的基因来校正每个数据集的平均表达式。</p>
<figure class="highlight xml"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">gene.use <span class="tag">&lt;<span class="name">-</span> <span class="attr">union</span>(<span class="attr">rownames</span>(<span class="attr">x</span> = <span class="string">head(x</span> = <span class="string">muraro_seurat@hvg.info,</span> <span class="attr">n</span> = <span class="string">2000)),</span></span></span><br><span class="line"><span class="tag">                  <span class="attr">rownames</span>(<span class="attr">x</span> = <span class="string">head(x</span> = <span class="string">seger_seurat@hvg.info,</span> <span class="attr">n</span> = <span class="string">2000)))</span></span></span><br></pre></td></tr></table></figure>

<p>现在我们将运行CCA来查找这两个数据集的共享相关结构：</p>
<p>请注意，为了加快计算速度，我们将仅使用前5个维度，但理想情况下，您会考虑更多维度，然后使用DimHeatmap选择信息量最大的维度。</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">merged_seurat &lt;- RunCCA(<span class="attribute">object</span>=muraro_seurat, <span class="attribute">object2</span>=seger_seurat, genes.<span class="attribute">use</span>=gene.use, add.cell.<span class="attribute">id1</span>=<span class="string">"m"</span>, add.cell.<span class="attribute">id2</span>=<span class="string">"s"</span>, num.cc = 5)</span><br><span class="line">DimPlot(object = merged_seurat, reduction.use = <span class="string">"cca"</span>, group.by = <span class="string">"dataset"</span>, pt.size = 0.5) # Before correcting</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/cca_before.png" alt="image"><br>为了识别特定于数据集的细胞类型，我们比较了CCA与数据集特定主成分分析“解释”细胞的程度。</p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">merged_seurat &lt;- CalcVarExpRatio(<span class="attr">object</span> = merged_seurat, reduction.<span class="attr">type</span> = <span class="string">"pca"</span>, grouping.<span class="attr">var</span> = <span class="string">"dataset"</span>, dims.<span class="attr">use</span> = <span class="number">1</span>:<span class="number">5</span>)</span><br><span class="line">merged.all &lt;- merged_seurat</span><br><span class="line">merged_seurat &lt;- SubsetData(<span class="attr">object=merged_seurat,</span> subset.<span class="attr">name="var.ratio.pca",</span> accept.<span class="attr">low</span> = <span class="number">0.5</span>) <span class="comment"># CCA &gt; 1/2 as good as PCA</span></span><br><span class="line">merged.discard &lt;- SubsetData(<span class="attr">object=merged.all,</span> subset.<span class="attr">name="var.ratio.pca",</span> accept.<span class="attr">high</span> = <span class="number">0.5</span>)</span><br><span class="line"></span><br><span class="line">summary(factor(merged.discard@meta.data$celltype)) <span class="comment"># check the cell-type of the discarded cells.</span></span><br></pre></td></tr></table></figure>

<p>在这里我们可以看到，尽管两个数据集都包含内皮细胞，但几乎所有数据集都被丢弃为“数据集特定”。 现在我们可以对齐数据集：</p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">merged_seurat &lt;- AlignSubspace(<span class="attr">object</span> = merged_seurat, reduction.<span class="attr">type</span> = <span class="string">"cca"</span>, grouping.<span class="attr">var</span> = <span class="string">"dataset"</span>, dims.<span class="attr">align</span> = <span class="number">1</span>:<span class="number">5</span>)</span><br><span class="line">DimPlot(<span class="attr">object</span> = merged_seurat, reduction.<span class="attr">use</span> = <span class="string">"cca.aligned"</span>, group.<span class="attr">by</span> = <span class="string">"dataset"</span>, pt.<span class="attr">size</span> = <span class="number">0.5</span>) <span class="comment"># After aligning subspaces</span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/cca_after.png" alt="image"></p>
<h4 id="3-9-Search-scRNA-Seq-data"><a href="#3-9-Search-scRNA-Seq-data" class="headerlink" title="3.9 Search scRNA-Seq data"></a>3.9 Search scRNA-Seq data</h4><p>scfind是一种工具，它允许人们使用基因列表搜索单个细胞RNA-Seq集合（Atlas），例如搜索特定的基因集合的细胞和细胞类型。<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/scfind.png" alt="image"></p>
<h5 id="3-9-2-Dataset"><a href="#3-9-2-Dataset" class="headerlink" title="3.9.2 Dataset"></a>3.9.2 Dataset</h5><figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">muraro &lt;- readRDS(<span class="string">"pancreas/muraro.rds"</span>)</span><br></pre></td></tr></table></figure>

<h5 id="3-9-3-Gene-INdex"><a href="#3-9-3-Gene-INdex" class="headerlink" title="3.9.3 Gene INdex"></a>3.9.3 Gene INdex</h5><p>使用刚刚加载的数据创建索引：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cellIndex &lt;- buildCellIndex(<span class="name">muraro</span>)</span><br></pre></td></tr></table></figure>

<p>基因索引包含表达它的细胞的每个基因索引。 这类似于表达矩阵的稀疏化。 除此之外，索引也以一种可以非常快速地访问的方式进行压缩。 我们估计用这种方法可以达到5-10的压缩系数。</p>
<p>默认情况下，SingleCellExperiment对象的colData槽的cell_type1列用于定义单元类型，但也可以使用buildCellTypeIndex函数的cell_type_column参数手动定义</p>
<h5 id="3-9-4-Marker-genes"><a href="#3-9-4-Marker-genes" class="headerlink" title="3.9.4 Marker genes"></a>3.9.4 Marker genes</h5><p>现在让我们定义细胞类型特异性标记基因的列表。我们将使用在原始出版物中确定的标记基因：</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"># these genes are taken <span class="keyword">from</span> fig. <span class="number">1</span></span><br><span class="line">muraro_alpha &lt;- c(<span class="string">"GCG"</span>, <span class="string">"LOXL4"</span>, <span class="string">"PLCE1"</span>, <span class="string">"IRX2"</span>, <span class="string">"GC"</span>, <span class="string">"KLHL41"</span>, </span><br><span class="line">                  <span class="string">"CRYBA2"</span>, <span class="string">"TTR"</span>, <span class="string">"TM4SF4"</span>, <span class="string">"RGS4"</span>)</span><br><span class="line">muraro_beta &lt;- c(<span class="string">"INS"</span>, <span class="string">"IAPP"</span>, <span class="string">"MAFA"</span>, <span class="string">"NPTX2"</span>, <span class="string">"DLK1"</span>, <span class="string">"ADCYAP1"</span>, </span><br><span class="line">                 <span class="string">"PFKFB2"</span>, <span class="string">"PDX1"</span>, <span class="string">"TGFBR3"</span>, <span class="string">"SYT13"</span>)</span><br><span class="line">muraro_gamma &lt;- c(<span class="string">"PPY"</span>, <span class="string">"SERTM1"</span>, <span class="string">"CARTPT"</span>, <span class="string">"SLITRK6"</span>, <span class="string">"ETV1"</span>, </span><br><span class="line">                  <span class="string">"THSD7A"</span>, <span class="string">"AQP3"</span>, <span class="string">"ENTPD2"</span>, <span class="string">"PTGFR"</span>, <span class="string">"CHN2"</span>)</span><br><span class="line">muraro_delta &lt;- c(<span class="string">"SST"</span>, <span class="string">"PRG4"</span>, <span class="string">"LEPR"</span>, <span class="string">"RBP4"</span>, <span class="string">"BCHE"</span>, <span class="string">"HHEX"</span>, </span><br><span class="line">                  <span class="string">"FRZB"</span>, <span class="string">"PCSK1"</span>, <span class="string">"RGS2"</span>, <span class="string">"GABRG2"</span>)</span><br></pre></td></tr></table></figure>

<h5 id="3-9-5-Search-cells-by-a-gene-list"><a href="#3-9-5-Search-cells-by-a-gene-list" class="headerlink" title="3.9.5 Search cells by a gene list"></a>3.9.5 Search cells by a gene list</h5><p>findCell函数返回与给定数据集中所有cell类型对应的p值列表。 它还输出一组细胞，其中来自给定基因列表的基因共表达。 我们将在上面定义的所有标记基因列表上运行它：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">res &lt;- findCell(<span class="name">cellIndex</span>, muraro_alpha)</span><br><span class="line">barplot(<span class="name">-log10</span>(<span class="name">res</span>$p_values), ylab = <span class="string">"-log10(pval)"</span>, las = <span class="number">2</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/32-search_files/figure-html/unnamed-chunk-7-1.png" alt="image"></p>
<h3 id="4-Seurat"><a href="#4-Seurat" class="headerlink" title="4 Seurat"></a>4 Seurat</h3><p>Seurat最初是作为scRNA-seq数据的聚类工具开发的，但是在过去几年中，该软件包的重点变得不那么具体，目前Seurat是一个流行的R软件包，可以执行QC，分析和scRNA的探索 -seq数据，即本课程涵盖的许多任务。 虽然作者提供了几个教程，但在这里我们按照Seurat的作者（2,800个外周血单核细胞）创建的例子提供了一个简要的概述。 我们主要在各种函数调用中使用默认值，有关详细信息，请参阅文档和作者。 出于课程目的，将使用前面章节中描述的小邓数据集：</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">deng &lt;- readRDS(<span class="string">"deng/deng-reads.rds"</span>)</span><br></pre></td></tr></table></figure>

<h4 id="4-1-Seurat-object-class"><a href="#4-1-Seurat-object-class" class="headerlink" title="4.1 Seurat object class"></a>4.1 Seurat object class</h4><p>Seurat没有集成上面描述的SingleCellExperiment Bioconductor类，而是引入了自己的对象类 - seurat。 本章中的所有计算都是在此类的对象上执行的。 要开始分析，我们首先需要使用原始（非规范化）数据初始化对象。 我们将保留所有表达&gt;=3的基因和所有检测到至少200个基因的细胞。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">library</span><span class="params">(SingleCellExperiment)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(Seurat)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(mclust)</span></span></span><br><span class="line"><span class="function"><span class="title">library</span><span class="params">(dplyr)</span></span></span><br><span class="line">seuset &lt;- CreateSeuratObject(</span><br><span class="line">    raw<span class="selector-class">.data</span> = counts(deng),</span><br><span class="line">    min<span class="selector-class">.cells</span> = <span class="number">3</span>, </span><br><span class="line">    min<span class="selector-class">.genes</span> = <span class="number">200</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<h4 id="4-2-Expression-QC"><a href="#4-2-Expression-QC" class="headerlink" title="4.2 Expression QC"></a>4.2 Expression QC</h4><p>Seurat允许您根据任何用户定义的标准轻松地探索QC度量和过滤cell。我们可以想象基因和分子的数量并绘制它们的关系：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">VlnPlot(</span><br><span class="line">    <span class="selector-tag">object</span> = seuset, </span><br><span class="line">    features<span class="selector-class">.plot</span> = c(<span class="string">"nGene"</span>, <span class="string">"nUMI"</span>), </span><br><span class="line">    nCol = <span class="number">2</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-5-1.png" alt="image"></p>
<figure class="highlight nix"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">GenePlot(</span><br><span class="line">    <span class="attr">object</span> = seuset, </span><br><span class="line">    <span class="attr">gene1</span> = <span class="string">"nUMI"</span>, </span><br><span class="line">    <span class="attr">gene2</span> = <span class="string">"nGene"</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-5-2.png" alt="image"><br>现在，我们将排除具有明显离群值的细胞：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">seuset &lt;- FilterCells(</span><br><span class="line">    <span class="selector-tag">object</span> = seuset, </span><br><span class="line">    subset<span class="selector-class">.names</span> = c(<span class="string">"nUMI"</span>), </span><br><span class="line">    high<span class="selector-class">.thresholds</span> = c(<span class="number">2</span>e7)</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<h4 id="4-3-Normalization"><a href="#4-3-Normalization" class="headerlink" title="4.3 Normalization"></a>4.3 Normalization</h4><p>从数据集中删除不需要的单元格后，下一步是规范化数据。 默认情况下，我们采用全局缩放规范化方法LogNormalize，通过总表达式对每个单元格的基因表达式测量值进行标准化，将其乘以比例因子（默认为10,000），并对结果进行对数转换：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">seuset &lt;- NormalizeData(</span><br><span class="line">    <span class="selector-tag">object</span> = seuset, </span><br><span class="line">    normalization<span class="selector-class">.method</span> = <span class="string">"LogNormalize"</span>, </span><br><span class="line">    scale<span class="selector-class">.factor</span> = <span class="number">10000</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<h4 id="4-4-Highly-Variable-genes"><a href="#4-4-Highly-Variable-genes" class="headerlink" title="4.4 Highly Variable genes"></a>4.4 Highly Variable genes</h4><p>Seurat计算高度可变的基因，并将重点放在这些基因上进行下游分析。findvariable基因计算每个基因的平均表达和色散，将这些基因放入bin中，然后计算每个bin内的分散度。这有助于控制可变性和平均表达之间的关系：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">seuset &lt;- FindVariableGenes(</span><br><span class="line">    <span class="selector-tag">object</span> = seuset,</span><br><span class="line">    mean<span class="selector-class">.function</span> = ExpMean, </span><br><span class="line">    dispersion<span class="selector-class">.function</span> = LogVMR, </span><br><span class="line">    x<span class="selector-class">.low</span><span class="selector-class">.cutoff</span> = <span class="number">0.0125</span>, </span><br><span class="line">    x<span class="selector-class">.high</span><span class="selector-class">.cutoff</span> = <span class="number">3</span>, </span><br><span class="line">    y<span class="selector-class">.cutoff</span> = <span class="number">0.5</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<figure class="highlight less"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="selector-tag">length</span>(x = seuset<span class="variable">@var</span>.genes)</span><br><span class="line">## <span class="selector-attr">[1]</span> <span class="selector-tag">6127</span></span><br></pre></td></tr></table></figure>

<h4 id="4-5-Dealing-with-confounders"><a href="#4-5-Dealing-with-confounders" class="headerlink" title="4.5 Dealing with confounders"></a>4.5 Dealing with confounders</h4><p>为了减轻混杂因素的影响，Seurat构建线性模型以基于用户定义的变量预测基因表达。 这些模型的缩放z得分残差存储在scale.data槽中，用于降维和聚类。</p>
<p>Seurat可以消除由批次，细胞排列率（由Drop-seq数据的Drop-seq工具提供），检测到的分子数量，线粒体基因表达和细胞周期驱动的基因表达中的细胞 - 细胞变异。 在这里，我们回归每个细胞检测到的分子数量。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">seuset &lt;- ScaleData(</span><br><span class="line">    <span class="selector-tag">object</span> = seuset, </span><br><span class="line">    vars<span class="selector-class">.to</span><span class="selector-class">.regress</span> = c(<span class="string">"nUMI"</span>)</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<h4 id="4-6-Linear-dimensionality-reduction"><a href="#4-6-Linear-dimensionality-reduction" class="headerlink" title="4.6 Linear dimensionality reduction"></a>4.6 Linear dimensionality reduction</h4><p>接下来，我们对缩放数据执行PCA。 默认情况下，<a href="mailto:object@var.genes" target="_blank" rel="noopener">object@var.genes</a>中的基因用作输入，但也可以使用pc.genes定义。 在高度可变的基因上运行降维可以提高性能。 然而，对于某些类型的数据（UMI） - 特别是在回归技术变量之后，当在更大的基因子集（包括整个转录组）上运行时，PCA返回相似（尽管更慢）的结果。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">seuset &lt;- RunPCA(</span><br><span class="line">    <span class="selector-tag">object</span> = seuset, </span><br><span class="line">    pc<span class="selector-class">.genes</span> = seuset@<span class="selector-tag">var</span><span class="selector-class">.genes</span>, </span><br><span class="line">    do<span class="selector-class">.print</span> = TRUE, </span><br><span class="line">    pcs<span class="selector-class">.print</span> = <span class="number">1</span>:<span class="number">5</span>, </span><br><span class="line">    genes<span class="selector-class">.print</span> = <span class="number">5</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>Seurat提供了几种有用的方法来可视化定义PCA的细胞和基因：</p>
<figure class="highlight routeros"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">PrintPCA(object = seuset, pcs.<span class="builtin-name">print</span> = 1:5, genes.<span class="builtin-name">print</span> = 5, use.full = <span class="literal">FALSE</span>)</span><br></pre></td></tr></table></figure>

<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">VizPCA</span><span class="params">(object = seuset, pcs.use = <span class="number">1</span>:<span class="number">2</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-11-1.png" alt="image"></p>
<figure class="highlight lsl"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">PCAPlot(object = seuset, dim<span class="number">.1</span> = <span class="number">1</span>, dim<span class="number">.2</span> = <span class="number">2</span>)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-11-2.png" alt="image"><br>特别是，PCHeatmap允许轻松探索数据集中异质性的主要来源，并且在尝试确定要包含哪些PC以进行进一步的下游分析时非常有用。 细胞和基因都根据其PCA分数排序。 将cells.use设置为数字可以绘制光谱两端的极端单元格，从而大大加快了大型数据集的绘制速度：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">PCHeatmap(</span><br><span class="line">    <span class="selector-tag">object</span> = seuset, </span><br><span class="line">    pc<span class="selector-class">.use</span> = <span class="number">1</span>:<span class="number">6</span>, </span><br><span class="line">    cells<span class="selector-class">.use</span> = <span class="number">500</span>, </span><br><span class="line">    do<span class="selector-class">.balanced</span> = TRUE, </span><br><span class="line">    <span class="selector-tag">label</span><span class="selector-class">.columns</span> = FALSE,</span><br><span class="line">    use<span class="selector-class">.full</span> = FALSE</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-12-1.png" alt="image"></p>
<h4 id="4-7-Significant-PCs"><a href="#4-7-Significant-PCs" class="headerlink" title="4.7 Significant PCs"></a>4.7 Significant PCs</h4><p>为了克服任何单个基因中scRNA-seq数据的广泛技术噪音，Seurat根据其PCA评分对细胞进行聚类，每个PC基本上代表一种结合了相关基因集的信息的metagene。 因此，确定下游包含多少PC是重要的一步。 Seurat随机置换数据的一个子集（默认为1％）并重新运行PCA，通过重复此过程构建基因分数的空分布。 我们将重要的PC识别为具有低p值基因强烈富集的PC：</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line">seuset &lt;- JackStraw(</span><br><span class="line">    <span class="selector-tag">object</span> = seuset, </span><br><span class="line">    num<span class="selector-class">.replicate</span> = <span class="number">100</span>, </span><br><span class="line">    do<span class="selector-class">.print</span> = FALSE</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>JackStrawPlot函数提供了一个可视化工具，用于比较每个具有均匀分布（虚线）的PC的p值分布。 重要的PC将显示具有低p值的基因的强烈富集（虚线上方的实线）。 在这种情况下，PC 1-8似乎很重要。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">JackStrawPlot</span><span class="params">(object = seuset, PCs = <span class="number">1</span>:<span class="number">9</span>)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-14-1.png" alt="image"><br>一种更特别的方法来确定要使用哪些PCs，看一看主成分的标准偏差的图，并在图中有一个清晰的弯头的地方画出你的截止点。这可以用PCElbowPlot来完成。在这个例子中，看起来拐点会落在PC 5周围。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line"><span class="function"><span class="title">PCElbowPlot</span><span class="params">(object = seuset)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-15-1.png" alt="image"></p>
<h4 id="4-8-Clustering-cells"><a href="#4-8-Clustering-cells" class="headerlink" title="4.8 Clustering cells"></a>4.8 Clustering cells</h4><p>Seurat实现了基于图形的聚类方法。基于先前识别的PC计算cell之间的距离。 Seurat方法受到最近manuscript的启发，这些manuscript将基于图形的聚类方法应用于scRNA-seq数据–SNN-Cliq（（C。Xu和Su 2015））和CyTOF数据 - PhenoGraph（（Levine等人，2015））。简而言之，这些方法将细胞嵌入图结构中 - 例如K-最近邻（KNN）图，在具有相似基因表达模式的细胞之间绘制边缘，然后尝试将该图划分为高度互连的准集团或群落。与PhenoGraph一样，我们首先根据PCA空间中的欧氏距离构建KNN图，并根据其局部邻域中的共享重叠（Jaccard距离）细化任意两个cell之间的边权重。为了聚类细胞，我们应用模块化优化技术 - SLM（（Blondel等人，2008）），将细胞迭代分组在一起，目的是优化标准模块化功能。</p>
<p>FindClusters函数实现该过程，并包含一个分辨率参数，用于设置下游群集的粒度，增加的值会导致更多的群集。我们发现在这之间设置此参数<br>0.6- 1.2通常会为大约包含3000细胞的单细胞数据集返回良好的结果。较大的数据集通常会增加最佳分辨率。群集保存在object@ ident中。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br></pre></td><td class="code"><pre><span class="line">seuset &lt;- FindClusters(</span><br><span class="line">    <span class="selector-tag">object</span> = seuset, </span><br><span class="line">    reduction<span class="selector-class">.type</span> = <span class="string">"pca"</span>, </span><br><span class="line">    dims<span class="selector-class">.use</span> = <span class="number">1</span>:<span class="number">8</span>, </span><br><span class="line">    resolution = <span class="number">1.0</span>, </span><br><span class="line">    print<span class="selector-class">.output</span> = <span class="number">0</span>, </span><br><span class="line">    save<span class="selector-class">.SNN</span> = TRUE</span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>Seurat的一个有用特性是能够回忆起在最新的函数调用中使用的参数。对于findcluster来说，有一个函数PrintFindClustersParams来打印一个格式良好的参数总结，这些参数是选择的：</p>
<figure class="highlight clean"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line">PrintFindClustersParams(object = seuset)</span><br><span class="line">## Parameters used <span class="keyword">in</span> latest FindClusters calculation run on: <span class="number">2018</span><span class="number">-05</span><span class="number">-29</span> <span class="number">15</span>:<span class="number">27</span>:<span class="number">42</span></span><br><span class="line">## =============================================================================</span><br><span class="line">## Resolution: <span class="number">1</span></span><br><span class="line">## -----------------------------------------------------------------------------</span><br><span class="line">## Modularity Function    Algorithm         n.start         n.iter</span><br><span class="line">##      <span class="number">1</span>                   <span class="number">1</span>                 <span class="number">100</span>             <span class="number">10</span></span><br><span class="line">## -----------------------------------------------------------------------------</span><br><span class="line">## Reduction used          k.param          k.scale          prune.SNN</span><br><span class="line">##      pca                 <span class="number">30</span>                <span class="number">25</span>              <span class="number">0.0667</span></span><br><span class="line">## -----------------------------------------------------------------------------</span><br><span class="line">## Dims used <span class="keyword">in</span> calculation</span><br><span class="line">## =============================================================================</span><br><span class="line">## <span class="number">1</span> <span class="number">2</span> <span class="number">3</span> <span class="number">4</span> <span class="number">5</span> <span class="number">6</span> <span class="number">7</span> <span class="number">8</span></span><br></pre></td></tr></table></figure>

<p>我们可以查看集群结果，并将它们与原始的细胞标签进行比较：</p>
<figure class="highlight less"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="selector-tag">table</span>(seuset<span class="variable">@ident</span>)</span><br><span class="line">## </span><br><span class="line">##  <span class="selector-tag">0</span>  <span class="selector-tag">1</span>  <span class="selector-tag">2</span>  <span class="selector-tag">3</span> </span><br><span class="line">## <span class="selector-tag">85</span> <span class="selector-tag">75</span> <span class="selector-tag">59</span> <span class="selector-tag">34</span></span><br><span class="line"><span class="selector-tag">adjustedRandIndex</span>(colData(deng)[seuset<span class="variable">@cell</span>.names, ]$cell_type2, seuset<span class="variable">@ident</span>)</span><br><span class="line">## <span class="selector-attr">[1]</span> <span class="selector-tag">0</span><span class="selector-class">.3981315</span></span><br></pre></td></tr></table></figure>

<p>Seurat还利用tSNE图来识别聚类结果。 作为对tSNE的输入，我们建议使用作为聚类分析的输入的相同PC，尽管使用genes.use参数也支持基于缩放基因表达计算tSNE。</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">seuset &lt;- RunTSNE(</span><br><span class="line">    <span class="selector-tag">object</span> = seuset,</span><br><span class="line">    dims<span class="selector-class">.use</span> = <span class="number">1</span>:<span class="number">8</span>,</span><br><span class="line">    do<span class="selector-class">.fast</span> = TRUE</span><br><span class="line">)</span><br><span class="line"><span class="function"><span class="title">TSNEPlot</span><span class="params">(object = seuset)</span></span></span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-19-1.png" alt="image"></p>
<h4 id="4-9-Marker-genes"><a href="#4-9-Marker-genes" class="headerlink" title="4.9 Marker genes"></a>4.9 Marker genes</h4><p>Seurat可以帮助您找到通过差异表达定义聚类的标记。 默认情况下，它标识单个群集的正负标记，与所有其他单元格进行比较。 您可以测试群集彼此之间或所有cell。 例如，要找到第2组的标记基因，我们可以运行：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">markers2 &lt;- FindMarkers(<span class="name">seuset</span>, <span class="number">2</span>)</span><br></pre></td></tr></table></figure>

<p>然后，标记基因可以被可视化：</p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">VlnPlot(<span class="name">object</span> = seuset, features.plot = rownames(<span class="name">markers2</span>)[<span class="number">1</span>:<span class="number">2</span>])</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-21-1.png" alt="image"></p>
<figure class="highlight lisp"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">FeaturePlot(</span><br><span class="line">    seuset, </span><br><span class="line">    head(<span class="name">rownames</span>(<span class="name">markers2</span>)), </span><br><span class="line">    cols.use = c(<span class="string">"lightgrey"</span>, <span class="string">"blue"</span>), </span><br><span class="line">    nCol = <span class="number">3</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-21-2.png" alt="image"><br>FindAllMarkers自动执行此过程并查找所有群集的标记</p>
<figure class="highlight stylus"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line">markers &lt;- FindAllMarkers(</span><br><span class="line">    <span class="selector-tag">object</span> = seuset, </span><br><span class="line">    only<span class="selector-class">.pos</span> = TRUE, </span><br><span class="line">    min<span class="selector-class">.pct</span> = <span class="number">0.25</span>, </span><br><span class="line">    thresh<span class="selector-class">.use</span> = <span class="number">0.25</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p>DoHeatmap为给定的细胞和基因生成表达式热图。 在这种情况下，我们为每个群集绘制前10个标记（或所有标记，如果小于20）：</p>
<figure class="highlight n1ql"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">top10 &lt;- markers %&gt;% group_by(cluster) %&gt;% top_n(10, avg_logFC)</span><br><span class="line">DoHeatmap(</span><br><span class="line">    <span class="keyword">object</span> = seuset, </span><br><span class="line">    genes.<span class="keyword">use</span> = top10$gene, </span><br><span class="line">    slim.col.label = <span class="literal">TRUE</span>, </span><br><span class="line">    remove.<span class="keyword">key</span> = <span class="literal">TRUE</span></span><br><span class="line">)</span><br></pre></td></tr></table></figure>

<p><img src="https://hemberg-lab.github.io/scRNA.seq.course/33-seurat_files/figure-html/unnamed-chunk-23-1.png" alt="image"></p>
<h3 id="5-“Ideal”-scRNA-seq-pipline"><a href="#5-“Ideal”-scRNA-seq-pipline" class="headerlink" title="5 “Ideal” scRNA-seq pipline"></a>5 “Ideal” scRNA-seq pipline</h3><h4 id="5-1-Experimental-Design"><a href="#5-1-Experimental-Design" class="headerlink" title="5.1 Experimental Design"></a>5.1 Experimental Design</h4><ul>
<li>避免混淆生物和批处理效果<ul>
<li>如果可能的话，应该在相同的芯片上捕获多个条件</li>
<li>对每个条件执行多个复制，如果可能的话，应该在一起执行不同条件的复制</li>
<li>统计数字不能纠正一个完全令人困惑的实验！</li>
</ul>
</li>
<li>独特的分子标识符<ul>
<li>大大减少数据中的噪声</li>
<li>可以降低基因检测率（不清楚是UMIs还是其他操作差异）</li>
<li>失去连接信息</li>
<li>使用更长的umi(~ 10 bp)</li>
<li>使用UMI-tools对UMIs中的排序错误进行校正</li>
</ul>
</li>
<li>Spike-ins<ul>
<li>有利于质量控制</li>
<li>可能对标准化读取计数很有用</li>
<li>可用于近似细胞大小/ RNA含量（如果与生物学问题相关）</li>
<li>通常表现出比内源基因更高的噪音（移液错误，混合物质量）</li>
<li>需要更多的测序才能获得每个细胞足够的内源性读数</li>
</ul>
</li>
<li>细胞数量 vs read depth<ul>
<li>基因检测平台从每个细胞1百万个读数开始</li>
<li>转录因子检测（调节网络）需要高读取深度和最敏感的方案（即Fluidigm C1）</li>
<li>细胞聚类和细胞类型鉴定受益于大量细胞，并且不需要高测序深度（每个细胞约100,000个读数）。<br><img src="https://hemberg-lab.github.io/scRNA.seq.course/figures/Pipeline-batches.png" alt="image"><h4 id="5-2-Processing-Reads"><a href="#5-2-Processing-Reads" class="headerlink" title="5.2 Processing Reads"></a>5.2 Processing Reads</h4></li>
</ul>
</li>
<li>Read QC &amp; Trimming<ul>
<li>FASTQC，cutadapt</li>
</ul>
</li>
<li>Mapping<ul>
<li>Small datasets or UMI datasets：align to genome/transcriptome using STAR</li>
<li>Large datasets :pseudo-alignment with Salmon or kallisto</li>
</ul>
</li>
<li>Quantification<ul>
<li>Small dataset,no UMI:featureCounts</li>
<li>Large dataset,no UMI:Salmon,kallisto</li>
<li>UMI dataset:UMI-tools+featureCounts<h4 id="5-3-Preparing-Expression-Matrix"><a href="#5-3-Preparing-Expression-Matrix" class="headerlink" title="5.3 Preparing Expression Matrix"></a>5.3 Preparing Expression Matrix</h4></li>
</ul>
</li>
<li>Cell QC<ul>
<li>scater</li>
<li>consider:mtRNA,rRNA,spikr-ins(if available),number of detected genes per cell,total reads/molecules per cell</li>
</ul>
</li>
<li>Library Size Normalization<ul>
<li>scran</li>
</ul>
</li>
<li>Batch correction(if appropriate)<ul>
<li>Replicates/Confounded RUVs</li>
<li>Unknown or unbalanced biological groups mnnCorrect</li>
<li>Balanced sedign Combat<h4 id="5-4-Biological-Interpretation"><a href="#5-4-Biological-Interpretation" class="headerlink" title="5.4 Biological Interpretation"></a>5.4 Biological Interpretation</h4></li>
</ul>
</li>
<li>Feature Selection<ul>
<li>M3Drop</li>
</ul>
</li>
<li>Clustering and Marker Gene Identification<ul>
<li>&lt;=5000 cells:SC3</li>
<li><blockquote>
<p>5000 cells:Seurat</p>
</blockquote>
</li>
</ul>
</li>
<li>Pseudotime <ul>
<li>distinct timepoints:TSCAN</li>
<li>small dataset/unknown number of branches:Monocle2</li>
<li>large continuous dataset:destiny</li>
</ul>
</li>
<li>Differential Expression<ul>
<li>Small number of cells and few group:scde</li>
<li>Replicates with bath effects:mixture/linear models</li>
<li>Balanced batchs:edgeR or MAST</li>
<li>Large dataset:Kruskal-Wallix test(all groups a once),or Wilcox-test(compare 2-groups at a time)</li>
</ul>
</li>
</ul>

      
    </div>
    
    
    

    

    

    

    <footer class="post-footer">
      
        <div class="post-tags">
          
            <a href="/tags/single-cell/" rel="tag"># single cell</a>
          
        </div>
      

      
      
      

      
        <div class="post-nav">
          <div class="post-nav-next post-nav-item">
            
              <a href="/2019/06/06/热图太丑怎么办(pheatmap的使用)/" rel="next" title="热图太丑怎么办(pheatmap的使用)">
                <i class="fa fa-chevron-left"></i> 热图太丑怎么办(pheatmap的使用)
              </a>
            
          </div>

          <span class="post-nav-divider"></span>

          <div class="post-nav-prev post-nav-item">
            
          </div>
        </div>
      

      
      
    </footer>
  </div>
  
  
  
  </article>



    <div class="post-spread">
      
    </div>
  </div>


          </div>
          


          

  
    <div class="comments" id="comments">
    </div>
  



        </div>
        
          
  
  <div class="sidebar-toggle">
    <div class="sidebar-toggle-line-wrap">
      <span class="sidebar-toggle-line sidebar-toggle-line-first"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-middle"></span>
      <span class="sidebar-toggle-line sidebar-toggle-line-last"></span>
    </div>
  </div>

  <aside id="sidebar" class="sidebar">
    
    <div class="sidebar-inner">

      

      
        <ul class="sidebar-nav motion-element">
          <li class="sidebar-nav-toc sidebar-nav-active" data-target="post-toc-wrap">
            Table of Contents
          </li>
          <li class="sidebar-nav-overview" data-target="site-overview-wrap">
            Overview
          </li>
        </ul>
      

      <section class="site-overview-wrap sidebar-panel">
        <div class="site-overview">
          <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
            
              <img class="site-author-image" itemprop="image" src="/images/avatar.png" alt="Jing Yu">
            
              <p class="site-author-name" itemprop="name">Jing Yu</p>
              <p class="site-description motion-element" itemprop="description"></p>
          </div>

          <nav class="site-state motion-element">

            
              <div class="site-state-item site-state-posts">
              
                <a href="/archives/">
              
                  <span class="site-state-item-count">7</span>
                  <span class="site-state-item-name">posts</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-categories">
                <a href="/categories/index.html">
                  <span class="site-state-item-count">4</span>
                  <span class="site-state-item-name">categories</span>
                </a>
              </div>
            

            
              
              
              <div class="site-state-item site-state-tags">
                <a href="/tags/index.html">
                  <span class="site-state-item-count">9</span>
                  <span class="site-state-item-name">tags</span>
                </a>
              </div>
            

          </nav>

          

          
            <div class="links-of-author motion-element">
                
                  <span class="links-of-author-item">
                    <a href="https://github.com/jingyu9603" target="_blank" title="GitHub">
                      
                        <i class="fa fa-fw fa-github"></i>GitHub</a>
                  </span>
                
                  <span class="links-of-author-item">
                    <a href="mailto:11yj3312@gmail.com" target="_blank" title="E-Mail">
                      
                        <i class="fa fa-fw fa-envelope"></i>E-Mail</a>
                  </span>
                
            </div>
          

          
          

          
          

          

        </div>
      </section>

      
      <!--noindex-->
        <section class="post-toc-wrap motion-element sidebar-panel sidebar-panel-active">
          <div class="post-toc">

            
              
            

            
              <div class="post-toc-content"><ol class="nav"><li class="nav-item nav-level-3"><a class="nav-link" href="#1-Tabula-Muris"><span class="nav-number">1.</span> <span class="nav-text">1 Tabula Muris</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#1-1-介绍"><span class="nav-number">1.1.</span> <span class="nav-text">1.1 介绍</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#1-22-数据下载"><span class="nav-number">1.2.</span> <span class="nav-text">1.22 数据下载</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#1-3-读取这些数据集（Smartseq2）"><span class="nav-number">1.3.</span> <span class="nav-text">1.3 读取这些数据集（Smartseq2）</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#1-4-构建一个scater-object"><span class="nav-number">1.4.</span> <span class="nav-text">1.4 构建一个scater object</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#1-5-读取数据集（10X）"><span class="nav-number">1.5.</span> <span class="nav-text">1.5 读取数据集（10X）</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#1-6-构建一个scater-object"><span class="nav-number">1.6.</span> <span class="nav-text">1.6 构建一个scater object</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-cleaning-the-Expression-Matrix"><span class="nav-number">2.</span> <span class="nav-text">2 cleaning the Expression Matrix</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#2-1-ExpressionQC（UMI）"><span class="nav-number">2.1.</span> <span class="nav-text">2.1 ExpressionQC（UMI）</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#2-1-1-介绍"><span class="nav-number">2.1.1.</span> <span class="nav-text">2.1.1 介绍</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-1-2-Tung-dataset"><span class="nav-number">2.1.2.</span> <span class="nav-text">2.1.2 Tung dataset</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-1-3-Cell-QC"><span class="nav-number">2.1.3.</span> <span class="nav-text">2.1.3 Cell QC</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#2-1-3-1-文库大小（Library-size）"><span class="nav-number">2.1.3.1.</span> <span class="nav-text">2.1.3.1 文库大小（Library size）</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-1-3-2-Detected-genes"><span class="nav-number">2.1.3.2.</span> <span class="nav-text">2.1.3.2 Detected genes</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-1-3-3-ERCCs和MTS"><span class="nav-number">2.1.3.3.</span> <span class="nav-text">2.1.3.3 ERCCs和MTS</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-1-4-Cell-filtering"><span class="nav-number">2.1.4.</span> <span class="nav-text">2.1.4 Cell filtering</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#2-1-4-1-Manual"><span class="nav-number">2.1.4.1.</span> <span class="nav-text">2.1.4.1 Manual</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-1-4-2-Automatic"><span class="nav-number">2.1.4.2.</span> <span class="nav-text">2.1.4.2 Automatic</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-1-5-Compare-filterings"><span class="nav-number">2.1.5.</span> <span class="nav-text">2.1.5 Compare filterings</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-1-6-Gene-analysis"><span class="nav-number">2.1.6.</span> <span class="nav-text">2.1.6 Gene analysis</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#2-1-6-1-Gene-expression"><span class="nav-number">2.1.6.1.</span> <span class="nav-text">2.1.6.1 Gene expression</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-1-6-2-Gene-filtering"><span class="nav-number">2.1.6.2.</span> <span class="nav-text">2.1.6.2 Gene filtering</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-1-7-Save-the-Data"><span class="nav-number">2.1.7.</span> <span class="nav-text">2.1.7 Save the Data</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-2-Expression-QC（Reads）"><span class="nav-number">2.2.</span> <span class="nav-text">2.2 Expression QC（Reads）</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-3-Data-visualization"><span class="nav-number">2.3.</span> <span class="nav-text">2.3 Data visualization</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#2-3-1-介绍"><span class="nav-number">2.3.1.</span> <span class="nav-text">2.3.1 介绍</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-3-2-PCA-Plot"><span class="nav-number">2.3.2.</span> <span class="nav-text">2.3.2 PCA Plot</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#2-3-2-1-Before-QC"><span class="nav-number">2.3.2.1.</span> <span class="nav-text">2.3.2.1 Before QC</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-3-2-2-After-QC"><span class="nav-number">2.3.2.2.</span> <span class="nav-text">2.3.2.2 After QC</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-3-3-tSNE-map"><span class="nav-number">2.3.3.</span> <span class="nav-text">2.3.3 tSNE map</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#2-3-3-1-Before-QC"><span class="nav-number">2.3.3.1.</span> <span class="nav-text">2.3.3.1 Before QC</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-3-3-2-After-QC"><span class="nav-number">2.3.3.2.</span> <span class="nav-text">2.3.3.2 After QC</span></a></li></ol></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-4-Data-visualization（Reads）"><span class="nav-number">2.4.</span> <span class="nav-text">2.4 Data visualization（Reads）</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-5-Identifying-confounding-factors"><span class="nav-number">2.5.</span> <span class="nav-text">2.5 Identifying confounding factors</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#2-5-1-Introduction"><span class="nav-number">2.5.1.</span> <span class="nav-text">2.5.1 Introduction</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-5-2-Correlations-with-PCs"><span class="nav-number">2.5.2.</span> <span class="nav-text">2.5.2 Correlations with PCs</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#2-5-2-1-Detected-genes"><span class="nav-number">2.5.2.1.</span> <span class="nav-text">2.5.2.1 Detected genes</span></a></li></ol></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-6-Identifying-confounding-factors-Reads"><span class="nav-number">2.6.</span> <span class="nav-text">2.6 Identifying confounding factors (Reads)</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-7-Normalization-theory"><span class="nav-number">2.7.</span> <span class="nav-text">2.7 Normalization theory</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#2-7-1-介绍"><span class="nav-number">2.7.1.</span> <span class="nav-text">2.7.1 介绍</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-7-2-文库大小"><span class="nav-number">2.7.2.</span> <span class="nav-text">2.7.2 文库大小</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-7-3-Normalisations"><span class="nav-number">2.7.3.</span> <span class="nav-text">2.7.3 Normalisations</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#2-7-3-1-CPM"><span class="nav-number">2.7.3.1.</span> <span class="nav-text">2.7.3.1 CPM</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-7-3-2-RLE（SF）"><span class="nav-number">2.7.3.2.</span> <span class="nav-text">2.7.3.2 RLE（SF）</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-7-3-3-UQ"><span class="nav-number">2.7.3.3.</span> <span class="nav-text">2.7.3.3 UQ</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-7-3-4-TMM"><span class="nav-number">2.7.3.4.</span> <span class="nav-text">2.7.3.4 TMM</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-7-3-5-scran"><span class="nav-number">2.7.3.5.</span> <span class="nav-text">2.7.3.5 scran</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-7-3-6-Downsampling"><span class="nav-number">2.7.3.6.</span> <span class="nav-text">2.7.3.6 Downsampling</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-7-4-Effectiveness"><span class="nav-number">2.7.4.</span> <span class="nav-text">2.7.4 Effectiveness</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-8-Nomalization-practice（UMI）"><span class="nav-number">2.8.</span> <span class="nav-text">2.8 Nomalization practice（UMI）</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#2-8-1-Raw"><span class="nav-number">2.8.1.</span> <span class="nav-text">2.8.1 Raw</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-8-2-CPM"><span class="nav-number">2.8.2.</span> <span class="nav-text">2.8.2 CPM</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-8-3-Size-factor（RLE）"><span class="nav-number">2.8.3.</span> <span class="nav-text">2.8.3 Size-factor（RLE）</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-8-4-Upperquantile"><span class="nav-number">2.8.4.</span> <span class="nav-text">2.8.4 Upperquantile</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#TMM"><span class="nav-number">2.8.5.</span> <span class="nav-text">TMM</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-8-6-scran"><span class="nav-number">2.8.6.</span> <span class="nav-text">2.8.6 scran</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-8-7-Downsampling"><span class="nav-number">2.8.7.</span> <span class="nav-text">2.8.7 Downsampling</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-8-8-Normalisation-for-gene-transcript-length"><span class="nav-number">2.8.8.</span> <span class="nav-text">2.8.8 Normalisation for gene/transcript length</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-10-Dealing-with-confounders"><span class="nav-number">2.9.</span> <span class="nav-text">2.10 Dealing with confounders</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#2-10-1-Introduction"><span class="nav-number">2.9.1.</span> <span class="nav-text">2.10.1 Introduction</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#2-10-1-1-Advantages-and-disadvantages-of-using-spike-ins-to-remove-confounders"><span class="nav-number">2.9.1.1.</span> <span class="nav-text">2.10.1.1 Advantages and disadvantages of using spike-ins to remove confounders</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-10-2-Remove-Unwanted-Variation"><span class="nav-number">2.9.2.</span> <span class="nav-text">2.10.2 Remove Unwanted Variation</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#2-10-1-1-RUVg"><span class="nav-number">2.9.2.1.</span> <span class="nav-text">2.10.1.1 RUVg</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-10-2-2"><span class="nav-number">2.9.2.2.</span> <span class="nav-text">2.10.2.2</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-10-3-Combat"><span class="nav-number">2.9.3.</span> <span class="nav-text">2.10.3 Combat</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-10-4-mnnCorrect"><span class="nav-number">2.9.4.</span> <span class="nav-text">2.10.4 mnnCorrect</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-10-5-GLM"><span class="nav-number">2.9.5.</span> <span class="nav-text">2.10.5 GLM</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-10-6-How-to-evaluate-and-compare-confounder-removal-strategies"><span class="nav-number">2.9.6.</span> <span class="nav-text">2.10.6 How to evaluate and compare confounder removal strategies</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#2-10-6-1-Effectiveness-1"><span class="nav-number">2.9.6.1.</span> <span class="nav-text">2.10.6.1 Effectiveness 1</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-10-6-2-Effectiveness-2"><span class="nav-number">2.9.6.2.</span> <span class="nav-text">2.10.6.2 Effectiveness 2</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-10-6-3-Effectiveness-3"><span class="nav-number">2.9.6.3.</span> <span class="nav-text">2.10.6.3 Effectiveness 3</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#2-10-6-4-Effectiveness-4"><span class="nav-number">2.9.6.4.</span> <span class="nav-text">2.10.6.4 Effectiveness 4</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#2-11-Dealing-with-confounders（Reads）"><span class="nav-number">2.9.7.</span> <span class="nav-text">2.11 Dealing with confounders（Reads）</span></a></li></ol></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-Biological-Analysis"><span class="nav-number">3.</span> <span class="nav-text">3 Biological Analysis</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#3-1-Clustering-Introduction"><span class="nav-number">3.1.</span> <span class="nav-text">3.1 Clustering Introduction</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#3-1-1-Introduction"><span class="nav-number">3.1.1.</span> <span class="nav-text">3.1.1 Introduction</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-1-2-Dimensionality-reductions"><span class="nav-number">3.1.2.</span> <span class="nav-text">3.1.2 Dimensionality reductions</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-1-3-Clustering-methods"><span class="nav-number">3.1.3.</span> <span class="nav-text">3.1.3 Clustering methods</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#3-1-3-1-Hierarchical-clustering（层次聚类）"><span class="nav-number">3.1.3.1.</span> <span class="nav-text">3.1.3.1 Hierarchical clustering（层次聚类）</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#3-1-3-2-k-means-k均值聚类"><span class="nav-number">3.1.3.2.</span> <span class="nav-text">3.1.3.2 k-means(k均值聚类)</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#3-1-3-3-Graph-based-methods（基于图的聚类）"><span class="nav-number">3.1.3.3.</span> <span class="nav-text">3.1.3.3 Graph-based methods（基于图的聚类）</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-1-4-Challenge-in-clustering"><span class="nav-number">3.1.4.</span> <span class="nav-text">3.1.4 Challenge in clustering</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-1-5-Tools-for-scRNA-seq-data"><span class="nav-number">3.1.5.</span> <span class="nav-text">3.1.5 Tools for scRNA-seq data</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#3-1-5-1-SINCERA"><span class="nav-number">3.1.5.1.</span> <span class="nav-text">3.1.5.1 SINCERA</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#3-1-5-2-pcaReduce"><span class="nav-number">3.1.5.2.</span> <span class="nav-text">3.1.5.2 pcaReduce</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#3-1-5-3-SC3"><span class="nav-number">3.1.5.3.</span> <span class="nav-text">3.1.5.3 SC3</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#3-1-5-4-tSNE-k-means"><span class="nav-number">3.1.5.4.</span> <span class="nav-text">3.1.5.4 tSNE+k-means</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#3-1-5-5-SNN-Cliq"><span class="nav-number">3.1.5.5.</span> <span class="nav-text">3.1.5.5 SNN-Cliq</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#3-1-5-6-Seurat-clustering"><span class="nav-number">3.1.5.6.</span> <span class="nav-text">3.1.5.6 Seurat clustering</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-1-6-Comparing-clustering"><span class="nav-number">3.1.6.</span> <span class="nav-text">3.1.6 Comparing clustering</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#3-2-Clustering-example"><span class="nav-number">3.2.</span> <span class="nav-text">3.2 Clustering example</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#3-2-1-Deng-dataset"><span class="nav-number">3.2.1.</span> <span class="nav-text">3.2.1 Deng dataset</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-2-2-SC3"><span class="nav-number">3.2.2.</span> <span class="nav-text">3.2.2 SC3</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-2-3-pcaReduce"><span class="nav-number">3.2.3.</span> <span class="nav-text">3.2.3 pcaReduce</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-2-4-tSEN-kmeans"><span class="nav-number">3.2.4.</span> <span class="nav-text">3.2.4 tSEN+kmeans</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-2-5-SNN-Clip"><span class="nav-number">3.2.5.</span> <span class="nav-text">3.2.5 SNN-Clip</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-2-6-SINCREA"><span class="nav-number">3.2.6.</span> <span class="nav-text">3.2.6 SINCREA</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#3-3-Feature-Selection"><span class="nav-number">3.3.</span> <span class="nav-text">3.3 Feature Selection</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#3-3-1-Identify-Genes-vs-a-Null-model"><span class="nav-number">3.3.1.</span> <span class="nav-text">3.3.1 Identify Genes vs a Null model</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#3-3-1-1-Highly-Variable-Genes"><span class="nav-number">3.3.1.1.</span> <span class="nav-text">3.3.1.1 Highly Variable Genes</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#3-3-1-2-High-Dropout-Genes"><span class="nav-number">3.3.1.2.</span> <span class="nav-text">3.3.1.2 High Dropout Genes</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-3-2-Correlated-Expression"><span class="nav-number">3.3.2.</span> <span class="nav-text">3.3.2 Correlated Expression</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-3-3-Comparing-Methods"><span class="nav-number">3.3.3.</span> <span class="nav-text">3.3.3 Comparing Methods</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#3-4-Pseudotime-analysis"><span class="nav-number">3.4.</span> <span class="nav-text">3.4 Pseudotime analysis</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#3-4-1-First-look-at-Deng-data"><span class="nav-number">3.4.1.</span> <span class="nav-text">3.4.1 First look at Deng data</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-4-2-TSCAN"><span class="nav-number">3.4.2.</span> <span class="nav-text">3.4.2 TSCAN</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-4-3-monocle"><span class="nav-number">3.4.3.</span> <span class="nav-text">3.4.3 monocle</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-4-4-Diffusion-maps"><span class="nav-number">3.4.4.</span> <span class="nav-text">3.4.4 Diffusion maps</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#SLICER"><span class="nav-number">3.4.5.</span> <span class="nav-text">SLICER</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-4-6-Ouija"><span class="nav-number">3.4.6.</span> <span class="nav-text">3.4.6 Ouija</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-4-7-Comparison-of-the-methods"><span class="nav-number">3.4.7.</span> <span class="nav-text">3.4.7 Comparison of the methods</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-4-8-Expression-of-genes-through-time"><span class="nav-number">3.4.8.</span> <span class="nav-text">3.4.8 Expression of genes through time</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-5-Imputation"><span class="nav-number">3.4.9.</span> <span class="nav-text">3.5 Imputation</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-5-1-scImpute"><span class="nav-number">3.4.10.</span> <span class="nav-text">3.5.1 scImpute</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-5-2-MAGIC"><span class="nav-number">3.4.11.</span> <span class="nav-text">3.5.2 MAGIC</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#3-6-Differential-Expression-DE-analysis"><span class="nav-number">3.5.</span> <span class="nav-text">3.6 Differential Expression(DE) analysis</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#3-6-1-Bulk-RNA-seq"><span class="nav-number">3.5.1.</span> <span class="nav-text">3.6.1 Bulk RNA-seq</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-6-2-Single-cell-RNA-seq"><span class="nav-number">3.5.2.</span> <span class="nav-text">3.6.2 Single cell  RNA-seq</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#3-7-DE-in-a-real-dataset"><span class="nav-number">3.6.</span> <span class="nav-text">3.7 DE in a real dataset</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#3-7-1-Introdution"><span class="nav-number">3.6.0.1.</span> <span class="nav-text">3.7.1 Introdution</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-7-2-Kolmogorov-Smirnov-test"><span class="nav-number">3.6.1.</span> <span class="nav-text">3.7.2 Kolmogorov-Smirnov test</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#3-7-2-1-Evalusting-Accuracy"><span class="nav-number">3.6.1.1.</span> <span class="nav-text">3.7.2.1 Evalusting Accuracy</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-7-3-Wilcox-Mann-Whitney-U-Test"><span class="nav-number">3.6.2.</span> <span class="nav-text">3.7.3 Wilcox/Mann-Whitney-U Test</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-7-4-edgeR"><span class="nav-number">3.6.3.</span> <span class="nav-text">3.7.4 edgeR</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-7-5-Monocle"><span class="nav-number">3.6.4.</span> <span class="nav-text">3.7.5 Monocle</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-7-6-MAST"><span class="nav-number">3.6.5.</span> <span class="nav-text">3.7.6 MAST</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-7-7-Slow-Methods"><span class="nav-number">3.6.6.</span> <span class="nav-text">3.7.7 Slow Methods</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-7-8-BPSC"><span class="nav-number">3.6.7.</span> <span class="nav-text">3.7.8 BPSC</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-7-9-SCDE"><span class="nav-number">3.6.8.</span> <span class="nav-text">3.7.9 SCDE</span></a></li></ol></li><li class="nav-item nav-level-4"><a class="nav-link" href="#3-8-Comparing-Combining-scRNAseq-datasets"><span class="nav-number">3.7.</span> <span class="nav-text">3.8 Comparing/Combining scRNAseq datasets</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#3-8-2-Datasets"><span class="nav-number">3.7.0.1.</span> <span class="nav-text">3.8.2 Datasets</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-8-3-Projecting-cells-onto-annotated-cell-types（scmap）"><span class="nav-number">3.7.1.</span> <span class="nav-text">3.8.3 Projecting cells onto annotated cell-types（scmap）</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#3-8-3-1-Feature-Selection"><span class="nav-number">3.7.1.1.</span> <span class="nav-text">3.8.3.1 Feature Selection</span></a></li><li class="nav-item nav-level-6"><a class="nav-link" href="#3-8-3-2-Projecting"><span class="nav-number">3.7.1.2.</span> <span class="nav-text">3.8.3.2 Projecting</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-8-4-Cell-to-Cell-mapping"><span class="nav-number">3.7.2.</span> <span class="nav-text">3.8.4 Cell-to-Cell mapping</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-8-5-Metaneighbour"><span class="nav-number">3.7.3.</span> <span class="nav-text">3.8.5 Metaneighbour</span></a><ol class="nav-child"><li class="nav-item nav-level-6"><a class="nav-link" href="#3-8-5-1-Prepare-Data"><span class="nav-number">3.7.3.1.</span> <span class="nav-text">3.8.5.1 Prepare Data</span></a></li></ol></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-8-6-mnnCorrect"><span class="nav-number">3.7.4.</span> <span class="nav-text">3.8.6 mnnCorrect</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-8-7-Cannonical-Correlation-Analysis（Seurat）"><span class="nav-number">3.7.5.</span> <span class="nav-text">3.8.7 Cannonical Correlation Analysis（Seurat）</span></a></li></ol><li class="nav-item nav-level-4"><a class="nav-link" href="#3-9-Search-scRNA-Seq-data"><span class="nav-number">3.8.</span> <span class="nav-text">3.9 Search scRNA-Seq data</span></a><ol class="nav-child"><li class="nav-item nav-level-5"><a class="nav-link" href="#3-9-2-Dataset"><span class="nav-number">3.8.1.</span> <span class="nav-text">3.9.2 Dataset</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-9-3-Gene-INdex"><span class="nav-number">3.8.2.</span> <span class="nav-text">3.9.3 Gene INdex</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-9-4-Marker-genes"><span class="nav-number">3.8.3.</span> <span class="nav-text">3.9.4 Marker genes</span></a></li><li class="nav-item nav-level-5"><a class="nav-link" href="#3-9-5-Search-cells-by-a-gene-list"><span class="nav-number">3.8.4.</span> <span class="nav-text">3.9.5 Search cells by a gene list</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#4-Seurat"><span class="nav-number">4.</span> <span class="nav-text">4 Seurat</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#4-1-Seurat-object-class"><span class="nav-number">4.1.</span> <span class="nav-text">4.1 Seurat object class</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#4-2-Expression-QC"><span class="nav-number">4.2.</span> <span class="nav-text">4.2 Expression QC</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#4-3-Normalization"><span class="nav-number">4.3.</span> <span class="nav-text">4.3 Normalization</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#4-4-Highly-Variable-genes"><span class="nav-number">4.4.</span> <span class="nav-text">4.4 Highly Variable genes</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#4-5-Dealing-with-confounders"><span class="nav-number">4.5.</span> <span class="nav-text">4.5 Dealing with confounders</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#4-6-Linear-dimensionality-reduction"><span class="nav-number">4.6.</span> <span class="nav-text">4.6 Linear dimensionality reduction</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#4-7-Significant-PCs"><span class="nav-number">4.7.</span> <span class="nav-text">4.7 Significant PCs</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#4-8-Clustering-cells"><span class="nav-number">4.8.</span> <span class="nav-text">4.8 Clustering cells</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#4-9-Marker-genes"><span class="nav-number">4.9.</span> <span class="nav-text">4.9 Marker genes</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#5-“Ideal”-scRNA-seq-pipline"><span class="nav-number">5.</span> <span class="nav-text">5 “Ideal” scRNA-seq pipline</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#5-1-Experimental-Design"><span class="nav-number">5.1.</span> <span class="nav-text">5.1 Experimental Design</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#5-2-Processing-Reads"><span class="nav-number">5.2.</span> <span class="nav-text">5.2 Processing Reads</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#5-3-Preparing-Expression-Matrix"><span class="nav-number">5.3.</span> <span class="nav-text">5.3 Preparing Expression Matrix</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#5-4-Biological-Interpretation"><span class="nav-number">5.4.</span> <span class="nav-text">5.4 Biological Interpretation</span></a></li></ol></li></div>
            

          </div>
        </section>
      <!--/noindex-->
      

      

    </div>
  </aside>


        
      </div>
    </main>

    <footer id="footer" class="footer">
      <div class="footer-inner">
        <div class="copyright">&copy; <span itemprop="copyrightYear">2019</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">Jing Yu</span>

  
</div>


  <div class="powered-by">Powered by <a class="theme-link" target="_blank" href="https://hexo.io">Hexo</a></div>



  <span class="post-meta-divider">|</span>



  <div class="theme-info">Theme &mdash; <a class="theme-link" target="_blank" href="https://github.com/iissnan/hexo-theme-next">NexT.Pisces</a> v5.1.4</div>




        
<div class="busuanzi-count">
  <script async src="https://dn-lbstatics.qbox.me/busuanzi/2.3/busuanzi.pure.mini.js"></script>

  
    <span class="site-uv">
      <i class="fa fa-user"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_uv"></span>
      
    </span>
  

  
    <span class="site-pv">
      <i class="fa fa-eye"></i>
      <span class="busuanzi-value" id="busuanzi_value_site_pv"></span>
      
    </span>
  
</div>








        
      </div>
    </footer>

    
      <div class="back-to-top">
        <i class="fa fa-arrow-up"></i>
        
      </div>
    

    

  </div>

  

<script type="text/javascript">
  if (Object.prototype.toString.call(window.Promise) !== '[object Function]') {
    window.Promise = null;
  }
</script>









  












  
  
    <script type="text/javascript" src="/lib/jquery/index.js?v=2.1.3"></script>
  

  
  
    <script type="text/javascript" src="/lib/fastclick/lib/fastclick.min.js?v=1.0.6"></script>
  

  
  
    <script type="text/javascript" src="/lib/jquery_lazyload/jquery.lazyload.js?v=1.9.7"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/velocity/velocity.ui.min.js?v=1.2.1"></script>
  

  
  
    <script type="text/javascript" src="/lib/fancybox/source/jquery.fancybox.pack.js?v=2.1.5"></script>
  


  


  <script type="text/javascript" src="/js/src/utils.js?v=5.1.4"></script>

  <script type="text/javascript" src="/js/src/motion.js?v=5.1.4"></script>



  
  


  <script type="text/javascript" src="/js/src/affix.js?v=5.1.4"></script>

  <script type="text/javascript" src="/js/src/schemes/pisces.js?v=5.1.4"></script>



  
  <script type="text/javascript" src="/js/src/scrollspy.js?v=5.1.4"></script>
<script type="text/javascript" src="/js/src/post-details.js?v=5.1.4"></script>



  


  <script type="text/javascript" src="/js/src/bootstrap.js?v=5.1.4"></script>



  


  




	





  





  










  <script src="//cdn1.lncld.net/static/js/3.0.4/av-min.js"></script>
  <script src="//unpkg.com/valine/dist/Valine.min.js"></script>
  
  <script type="text/javascript">
    var GUEST = ['nick','mail','link'];
    var guest = 'nick,mail,link';
    guest = guest.split(',').filter(item=>{
      return GUEST.indexOf(item)>-1;
    });
    new Valine({
        el: '#comments' ,
        verify: false,
        notify: false,
        appId: 'qf3QmPxySWNGSWi82McRwPS2-gzGzoHsz',
        appKey: 'jeBjNYIDyqLMmB5KgBnI0ITW',
        placeholder: 'Just go go',
        avatar:'monsterid',
        guest_info:guest,
        pageSize:'10' || 10,
    });
  </script>



  

  <script type="text/javascript">
    // Popup Window;
    var isfetched = false;
    var isXml = true;
    // Search DB path;
    var search_path = "search.xml";
    if (search_path.length === 0) {
      search_path = "search.xml";
    } else if (/json$/i.test(search_path)) {
      isXml = false;
    }
    var path = "/" + search_path;
    // monitor main search box;

    var onPopupClose = function (e) {
      $('.popup').hide();
      $('#local-search-input').val('');
      $('.search-result-list').remove();
      $('#no-result').remove();
      $(".local-search-pop-overlay").remove();
      $('body').css('overflow', '');
    }

    function proceedsearch() {
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay"></div>')
        .css('overflow', 'hidden');
      $('.search-popup-overlay').click(onPopupClose);
      $('.popup').toggle();
      var $localSearchInput = $('#local-search-input');
      $localSearchInput.attr("autocapitalize", "none");
      $localSearchInput.attr("autocorrect", "off");
      $localSearchInput.focus();
    }

    // search function;
    var searchFunc = function(path, search_id, content_id) {
      'use strict';

      // start loading animation
      $("body")
        .append('<div class="search-popup-overlay local-search-pop-overlay">' +
          '<div id="search-loading-icon">' +
          '<i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>' +
          '</div>' +
          '</div>')
        .css('overflow', 'hidden');
      $("#search-loading-icon").css('margin', '20% auto 0 auto').css('text-align', 'center');

      $.ajax({
        url: path,
        dataType: isXml ? "xml" : "json",
        async: true,
        success: function(res) {
          // get the contents from search data
          isfetched = true;
          $('.popup').detach().appendTo('.header-inner');
          var datas = isXml ? $("entry", res).map(function() {
            return {
              title: $("title", this).text(),
              content: $("content",this).text(),
              url: $("url" , this).text()
            };
          }).get() : res;
          var input = document.getElementById(search_id);
          var resultContent = document.getElementById(content_id);
          var inputEventFunction = function() {
            var searchText = input.value.trim().toLowerCase();
            var keywords = searchText.split(/[\s\-]+/);
            if (keywords.length > 1) {
              keywords.push(searchText);
            }
            var resultItems = [];
            if (searchText.length > 0) {
              // perform local searching
              datas.forEach(function(data) {
                var isMatch = false;
                var hitCount = 0;
                var searchTextCount = 0;
                var title = data.title.trim();
                var titleInLowerCase = title.toLowerCase();
                var content = data.content.trim().replace(/<[^>]+>/g,"");
                var contentInLowerCase = content.toLowerCase();
                var articleUrl = decodeURIComponent(data.url);
                var indexOfTitle = [];
                var indexOfContent = [];
                // only match articles with not empty titles
                if(title != '') {
                  keywords.forEach(function(keyword) {
                    function getIndexByWord(word, text, caseSensitive) {
                      var wordLen = word.length;
                      if (wordLen === 0) {
                        return [];
                      }
                      var startPosition = 0, position = [], index = [];
                      if (!caseSensitive) {
                        text = text.toLowerCase();
                        word = word.toLowerCase();
                      }
                      while ((position = text.indexOf(word, startPosition)) > -1) {
                        index.push({position: position, word: word});
                        startPosition = position + wordLen;
                      }
                      return index;
                    }

                    indexOfTitle = indexOfTitle.concat(getIndexByWord(keyword, titleInLowerCase, false));
                    indexOfContent = indexOfContent.concat(getIndexByWord(keyword, contentInLowerCase, false));
                  });
                  if (indexOfTitle.length > 0 || indexOfContent.length > 0) {
                    isMatch = true;
                    hitCount = indexOfTitle.length + indexOfContent.length;
                  }
                }

                // show search results

                if (isMatch) {
                  // sort index by position of keyword

                  [indexOfTitle, indexOfContent].forEach(function (index) {
                    index.sort(function (itemLeft, itemRight) {
                      if (itemRight.position !== itemLeft.position) {
                        return itemRight.position - itemLeft.position;
                      } else {
                        return itemLeft.word.length - itemRight.word.length;
                      }
                    });
                  });

                  // merge hits into slices

                  function mergeIntoSlice(text, start, end, index) {
                    var item = index[index.length - 1];
                    var position = item.position;
                    var word = item.word;
                    var hits = [];
                    var searchTextCountInSlice = 0;
                    while (position + word.length <= end && index.length != 0) {
                      if (word === searchText) {
                        searchTextCountInSlice++;
                      }
                      hits.push({position: position, length: word.length});
                      var wordEnd = position + word.length;

                      // move to next position of hit

                      index.pop();
                      while (index.length != 0) {
                        item = index[index.length - 1];
                        position = item.position;
                        word = item.word;
                        if (wordEnd > position) {
                          index.pop();
                        } else {
                          break;
                        }
                      }
                    }
                    searchTextCount += searchTextCountInSlice;
                    return {
                      hits: hits,
                      start: start,
                      end: end,
                      searchTextCount: searchTextCountInSlice
                    };
                  }

                  var slicesOfTitle = [];
                  if (indexOfTitle.length != 0) {
                    slicesOfTitle.push(mergeIntoSlice(title, 0, title.length, indexOfTitle));
                  }

                  var slicesOfContent = [];
                  while (indexOfContent.length != 0) {
                    var item = indexOfContent[indexOfContent.length - 1];
                    var position = item.position;
                    var word = item.word;
                    // cut out 100 characters
                    var start = position - 20;
                    var end = position + 80;
                    if(start < 0){
                      start = 0;
                    }
                    if (end < position + word.length) {
                      end = position + word.length;
                    }
                    if(end > content.length){
                      end = content.length;
                    }
                    slicesOfContent.push(mergeIntoSlice(content, start, end, indexOfContent));
                  }

                  // sort slices in content by search text's count and hits' count

                  slicesOfContent.sort(function (sliceLeft, sliceRight) {
                    if (sliceLeft.searchTextCount !== sliceRight.searchTextCount) {
                      return sliceRight.searchTextCount - sliceLeft.searchTextCount;
                    } else if (sliceLeft.hits.length !== sliceRight.hits.length) {
                      return sliceRight.hits.length - sliceLeft.hits.length;
                    } else {
                      return sliceLeft.start - sliceRight.start;
                    }
                  });

                  // select top N slices in content

                  var upperBound = parseInt('1');
                  if (upperBound >= 0) {
                    slicesOfContent = slicesOfContent.slice(0, upperBound);
                  }

                  // highlight title and content

                  function highlightKeyword(text, slice) {
                    var result = '';
                    var prevEnd = slice.start;
                    slice.hits.forEach(function (hit) {
                      result += text.substring(prevEnd, hit.position);
                      var end = hit.position + hit.length;
                      result += '<b class="search-keyword">' + text.substring(hit.position, end) + '</b>';
                      prevEnd = end;
                    });
                    result += text.substring(prevEnd, slice.end);
                    return result;
                  }

                  var resultItem = '';

                  if (slicesOfTitle.length != 0) {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + highlightKeyword(title, slicesOfTitle[0]) + "</a>";
                  } else {
                    resultItem += "<li><a href='" + articleUrl + "' class='search-result-title'>" + title + "</a>";
                  }

                  slicesOfContent.forEach(function (slice) {
                    resultItem += "<a href='" + articleUrl + "'>" +
                      "<p class=\"search-result\">" + highlightKeyword(content, slice) +
                      "...</p>" + "</a>";
                  });

                  resultItem += "</li>";
                  resultItems.push({
                    item: resultItem,
                    searchTextCount: searchTextCount,
                    hitCount: hitCount,
                    id: resultItems.length
                  });
                }
              })
            };
            if (keywords.length === 1 && keywords[0] === "") {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-search fa-5x" /></div>'
            } else if (resultItems.length === 0) {
              resultContent.innerHTML = '<div id="no-result"><i class="fa fa-frown-o fa-5x" /></div>'
            } else {
              resultItems.sort(function (resultLeft, resultRight) {
                if (resultLeft.searchTextCount !== resultRight.searchTextCount) {
                  return resultRight.searchTextCount - resultLeft.searchTextCount;
                } else if (resultLeft.hitCount !== resultRight.hitCount) {
                  return resultRight.hitCount - resultLeft.hitCount;
                } else {
                  return resultRight.id - resultLeft.id;
                }
              });
              var searchResultList = '<ul class=\"search-result-list\">';
              resultItems.forEach(function (result) {
                searchResultList += result.item;
              })
              searchResultList += "</ul>";
              resultContent.innerHTML = searchResultList;
            }
          }

          if ('auto' === 'auto') {
            input.addEventListener('input', inputEventFunction);
          } else {
            $('.search-icon').click(inputEventFunction);
            input.addEventListener('keypress', function (event) {
              if (event.keyCode === 13) {
                inputEventFunction();
              }
            });
          }

          // remove loading animation
          $(".local-search-pop-overlay").remove();
          $('body').css('overflow', '');

          proceedsearch();
        }
      });
    }

    // handle and trigger popup window;
    $('.popup-trigger').click(function(e) {
      e.stopPropagation();
      if (isfetched === false) {
        searchFunc(path, 'local-search-input', 'local-search-result');
      } else {
        proceedsearch();
      };
    });

    $('.popup-btn-close').click(onPopupClose);
    $('.popup').click(function(e){
      e.stopPropagation();
    });
    $(document).on('keyup', function (event) {
      var shouldDismissSearchPopup = event.which === 27 &&
        $('.search-popup').is(':visible');
      if (shouldDismissSearchPopup) {
        onPopupClose();
      }
    });
  </script>





  

  

  

  
  

  

  

  

</body>
</html>
